// MB-support-plugin/lib/utils/maybe.js
'use strict';

const _ = require('lodash');
/**
 * Are you sick of being burned by unexpected nulls or `undefined is not a
 * function'? Are you sick of null checking everytime you want to do anything
 * of any significance? Are you sick of the uncertainty? Well do I have the
 * thing for you! Hi, I'm Billy Mays, here to introduce you to Maybe.js!
 *
 * The justification for these helper functions is to abstract away
 * null-checking. It's important to deal with errors, but it rather than
 * deal with them interspersed throughout business logic (and likely forget)
 * it's better to deal with them all at once at the end, allowing you to
 * write more natural code.
 *
 * These helpers are based on Haskell monads and functors. The vocabulary
 * isn't really important. I'm just pointing it out in case you want to
 * google more information about the patterns.
 *
 * @example
 * if(printer) {
 *   const status = printer.getStatus();
 *   if(status) {
 *     const currentProcess = status.current_process;
 *     if(currentProcess) {
 *       if(currentProcess.percent_done) {
 *         // do something with the percent done
 *       }
 *     }
 *   }
 * }
 *
 * // You can probably see how this would (and does) quickly get out of hand.
 * // THERE'S GOT TO BE A BETTER WAY!
 *
 * // THERE IS!
 * maybe(this.props.printer.getStatus())
 *   .flatMap(s => maybe(s.current_process))
 *   .flatMap(cp => maybe(cp.percent_done))
 *   .map(percent => {
 *     // do something with percent done
 *   });
 *
 * // OR HOW ABOUT THIS?!
 * either(status.state, status.current_process)
 *   .left(state => i18n.t(`${state}`))
 *   .right(cp => i18n.t(`${cp.name}.${cp.step}`))
 *   .lift()
 *
 * @author Jon Laing
 */

/**
 * Maybe represents a value that might exist, or might not. This is helpful
 * for dealing with values coming from the printer, which may or may not
 * be set. Using this interface, you can write your code naturally, without
 * having to null-check on every operation.
 *
 * @interface Maybe
 */

/**
 * Apply a function to the contained value. If the value doesn't exist
 * (nothing), then this is a noop.
 *
 * @function
 * @name Maybe#map
 * @param {function} f - a function to apply to the inner value. It should
 * return anything except a Maybe. If your `f` returns a Maybe, use `flatMap`
 * instead. It takes one parameter, which is the contained value, if one
 * exists. If no value exists, then the function is not called. You don't
 * have to null-check within `f`.
 * @returns {Maybe}
 */

/**
 * Apply a function that can return a maybe to a contained value. Use this
 * instead of `map` if the closure will also return a maybe. Otherwise,
 * you will have a Maybe nested inside a Maybe, and that kind of defeats
 * the purpose.
 *
 * @function
 * @name Maybe#flatMap
 * @param {function} f - a function to apply to the inner value. `f` should
 * return a Maybe to allow the chain to continue.
 * @returns {Maybe}
 */

/**
 * Lift the value out of the Maybe. If the value exists, then this will
 * return the value. If not, this will return `null`.
 *
 * @function
 * @name Maybe#lift
 * @returns {any}
 */

/**
 * A value exists
 * @implements {Maybe}
 */


const just = val => ({
  map: f => just(f(val)),
  flatMap: f => f(val),
  lift: () => val
});
/**
 * A value doesn't exist
 * @implements {Maybe}
 */


const nothing = () => ({
  map: nothing,
  flatMap: nothing,
  lift: () => null
});
/**
 * A helper function that will abstract out null-checking.
 * If `val` is not `null` or `undefined` it will return `just(val)`
 * otherwise it will return `nothing`.
 *
 * @param {any} val - any value
 * @returns {Maybe}
 */


const maybe = val => val != null ? just(val) : nothing();

const isNothing = maybeVal => !maybeVal.map(() => true).lift();
/**
 * Either abstracts null-checking in a slightly different way from Maybe.
 * Instead of returning an empty `nothing` when its primary value is null,
 * it returns a fallback value.
 *
 * You'll notice that the interface has `left` and `right`. Use the
 * mneumonic "right" = "correct" to determine which is the primary
 * value, and which is the fallback. ie: `right` is the primary value, and
 * `left` is the fallback.
 *
 * @interface Either
 */

/**
 * Left allows you to apply transformations to the left side (fallback or
 * error side) of the Either switch. This is a no-op on the right
 * side.
 *
 * @function
 * @name Either#left
 * @param {function} f - a function to apply to the contained value
 * on the left side.
 * @returns {Either}
 */

/**
 * Right allows you to apply transformations to the right side (primary or
 * correct side) of the Either switch. This is a no-op on the left side.
 *
 * @function
 * @name Either#right
 * @param {function} f - a function to apply to the contained value
 * on the right side.
 * @returns {Either}
 */

/**
 * Lifts the contained value, be it left or right, out of the Either.
 *
 * @function
 * @name Either#lift
 * @returns {any}
 */

/** @implements {Either} */


const left = val => ({
  left: f => left(f(val)),
  map: f => left(f(val)),
  right: () => left(val),
  lift: () => val
});
/** @implements {Either} */


const right = val => ({
  left: () => right(val),
  map: f => right(f(val)),
  right: f => right(f(val)),
  lift: () => val
});
/**
 * A helper function that will abstract null-checking and create the
 * Either. If `val` is not `null` or `undefined`, this will return a `right`
 * containing `val`. Otherwise, it will return a `left` containing `fallback`.
 *
 * @param {any} fallback - A default, error, or fallback value
 * @param {any} val - the primary value
 * @returns {Either}
 */


const either = (fallback, val) => val == null ? left(fallback) : right(val);

const getProp = _.curry((propName, monad) => monad.map(val => val[propName]));

const getPropLift = _.curry((propName, monad) => monad.map(val => val[propName]).lift()); //TODO add nested property checking capability to maybe.js


module.exports = {
  just,
  nothing,
  maybe,
  isNothing,
  left,
  right,
  either,
  getProp,
  getPropLift
};
