// MB-support-plugin/lib/utils/printerDetails.js
'use strict';
/** @author Jon Laing */

const PrinterStateEnum = require('../constants').PrinterStateEnum;

const Errors = require('../errors');

const BirdWingError = require('../croissant').BirdWingError;

const {
  maybe,
  either
} = require('./maybe');
/**
 * A helper function to get the printer store.
 * @param {object} flux
 * @returns {Flux.store}
 */


const getPrinterStore = flux => flux.store('PrinterStore');
/**
 * A helper function to get a printer out of the store
 * @param {object} flux
 * @param {string} id - id of the desired printer
 * @returns {MakerbotPrinter}
 */


const getPrinter = (flux, id) => getPrinterStore(flux).getPrinter(id);
/**
 * A helper function to get the selected printer out of the store
 * @param {object} flux
 * @returns {MakerbotPrinter}
 */


const getSelectedPrinter = flux => getPrinterStore(flux).getSelectedPrinter();
/**
 * A helper object to abstract null-checking on the printer status
 *
 * @param {MakerbotPrinter} printer
 * @param {string} override - some panels need to override the printer status
 * to make the UI look the way we want.
 * @returns {object}
 */


const printerStatus = (printer, override) => ({
  check: f => {
    const s = override || printer && printer.getStatus(); // status or null

    return either(false, s).right(f).lift(); // return s != null && f(s);
  },
  apply: f => {
    const s = override || printer && printer.getStatus(); // status or null

    return maybe(s).map(f).lift(); // return s != null ? f(s) : null;
  }
});
/**
 * A helper object to abstract null-checking on the printer's current process
 *
 * @param {MakerbotPrinter} printer
 * @returns {object}
 */


const currentPrinterProcess = printer => ({
  check: f => printerStatus(printer).check(s => {
    const cp = s.current_process;
    return either(false, cp).right(f).lift(); // return cp != null && f(cp);
  }),
  apply: f => printerStatus(printer).apply(s => {
    const cp = s.current_process;
    return maybe(cp).map(f).lift(); // return cp != null ? f(cp) : null;
  })
}); // Not entirely sure if this is needed, or if there's a more succinct way.
// I took it from the old implementation of PrinterDetailsPanel.


const isAcknowledgeableErr = printer => {
  const err = printer.getLastError();
  if (!err) return false;
  const errInfo = Errors.getErrInfo(printer, err); // this comment is to offset Shirley's bus-factor
  // https://github.com/makerbot/MBCoreUtils/blob/develop/birdwing_codegen/contexts/errors.json#L645

  const acknowledgeable = [BirdWingError.Action.ACKNOWLEDGE, BirdWingError.Action.ACKNOWLEDGE_OR_CONTINUE];
  return acknowledgeable.indexOf(errInfo.action) > -1;
}; // check if the step of the current process is in a particular list
// of steps. This is basically meant to flatten current process steps into
// a handful of "states" for rendering.


const isStep = (printer, override) => stepList => override ? stepList.indexOf(override) !== -1 : currentPrinterProcess(printer).check(cp => stepList.indexOf(cp.step.toLowerCase()) != -1); // all the steps that constitute a printer being "busy"


const busySteps = ['initializing', 'initial_heating', 'final_heating', 'cooling', 'homing', 'position_found', 'preheating', 'calibrating', 'end_sequence', 'cancelling', 'preheating_loading', 'preheating_resuming', 'preheating_unloading', 'stopping_filament', 'cleaning_up', 'loading_print_tool', 'waiting_for_file', 'running', 'extrusion', 'loading_filament', 'unloading_filament']; // Steps that consitute unloading filament.

const unloadingSteps = ['unloading_filament', 'preheating_unloading']; // all the steps that constitute a printer "transferring"

const transferSteps = ['transfer', 'downloading', 'verify_firmware']; // all the steps that constitute a printer "updating its firmware"

const updateFirmwareSteps = ['writing']; // all the steps that constitute a printer "printing"

const printingSteps = ['printing']; // all the steps that constitute a printer "paused"

const pausedSteps = ['suspending', 'suspended', 'unsuspending']; // all the steps that constitute a printer having "started its print"

const startPrintSteps = ['clear_build_plate']; // all the steps that constitute a printer "clearing filament"

const clearFilamentSteps = ['clear_filament']; // all the steps that constitute a printer "removing filament
// for z calibration"

const zCalFilamentSteps = ['remove_filament']; // all the steps that constitute a printer having an "error"

const errorSteps = ['failed', 'error_step']; // all the steps that constitute a printer action having "completed"

const completedSteps = ['completed'];
/**
 * This function returns a really useful object. Basically, the printer
 * details panel can have so many different states based on printer states
 * that it got incredibly unwieldy to not have a common interface for all
 * of it. So, this function returns an object that distills a bunch of
 * printer states into booleans, to make it easier to check.
 *
 * @param {MakerbotPrinter} printer
 * @returns {object}
 */

const printerDetailsStatus = (printer, override) => {
  const methodExists = printer.processMethodExists.bind(printer);
  const status = printerStatus(printer, override).check;
  const currentProcess = currentPrinterProcess(printer).check;
  const step = isStep(printer, override); // We don't have any guarantee that `current_process` has the types
  // we expect because it comes from kaiten which is written in python
  // so we have to coerce some of the process properties into booleans,
  // namely `cancelled` and `cancellable`.

  return {
    has: {
      state: status(s => s.state != null),
      process: currentProcess(() => true),
      extruder: !printer.isMissingExtruder(),
      error: !!printer.getLastError(),
      acknowledgableError: isAcknowledgeableErr(printer),
      dualExtruders: printer.isSixthGen(),
      chamber: printer.getChamberInfo().length > 0,
      platform: printer.getPlatformInfo && printer.getPlatformInfo().length > 0
    },
    is: {
      authenticated: printer.isAuthenticated(),
      idle: status(s => s.state === PrinterStateEnum.Idle),
      pending: status(s => s.state === PrinterStateEnum.Pending),
      cancelled: currentProcess(cp => !!cp.cancelled),
      offline: status(s => s.state === PrinterStateEnum.Offline),
      reconnecting: status(s => s.state === PrinterStateEnum.Reconnecting),
      reauthenticating: status(s => s.state === PrinterStateEnum.Reauthenticating),
      authenticating: status(s => s.state === PrinterStateEnum.Authenticating),
      unauthenticated: status(s => s.state === PrinterStateEnum.Unauthenticated)
    },
    can: {
      suspend: methodExists('suspend'),
      acknowledgeError: methodExists('acknowledge_error'),
      acknowledgeFailure: methodExists('acknowledge_failure'),
      acknowledgeSuccess: methodExists('acknowledge_success'),
      stopFilament: methodExists('stop_filament'),
      printAgain: currentProcess(cp => !!cp.can_print_again),
      resume: methodExists('resume'),
      loadFilament: methodExists('load_filament'),
      unloadFilament: methodExists('unload_filament'),
      cancel: status(s => s.state !== PrinterStateEnum.Pending) && currentProcess(cp => !!cp.cancellable && !cp.cancelled)
    },
    step: {
      failed: currentProcess(cp => cp.step === 'failed'),
      busy: step(busySteps),
      transferring: step(transferSteps),
      updatingFirmware: step(updateFirmwareSteps),
      printing: step(printingSteps),
      paused: step(pausedSteps),
      suspending: currentProcess(cp => cp.step === 'suspending'),
      startPrint: step(startPrintSteps),
      clearFilament: step(clearFilamentSteps),
      unloadingFilament: step(unloadingSteps),
      zCalFilament: step(zCalFilamentSteps),
      error: step(errorSteps),
      completed: step(completedSteps),
      wifiSetUp: currentProcess(cp => cp.name === 'WifiSetupProcess'),
      reconnecting: currentProcess(cp => cp.step === 'reconnecting')
    }
  };
};

module.exports = {
  getPrinterStore,
  getPrinter,
  getSelectedPrinter,
  printerStatus,
  currentPrinterProcess,
  printerDetailsStatus
};
