// MB-support-plugin/lib/ui/views/input/increment.js
'use strict';

const React = require('react');

const ce = React.createElement;
const DOM = React.DOM;

const R = require('ramda');

const TextInput = require('eagle-print/views/ui/input_text');

const Button = require('eagle-print/views/ui/button');

const FluxComponent = require('eagle-print/views/flux_component');

const {
  rangeStep,
  roundPrec
} = require('../../../util');

class Increment extends FluxComponent {
  constructor(props) {
    super(props);
  }

  _range() {
    return R.map(x => this._normalize(this.props.step, x), rangeStep(this.props.step, this.props.min, this.props.max + this.props.step));
  }

  _precision(step) {
    const log = -Math.log10(step);
    return log > 0 ? Math.ceil(log) : Math.floor(log);
  }

  _normalize(step, val) {
    return roundPrec(this._precision(step), val);
  }

  round(val) {
    // padding with a 0 in case use types '.2' intead of '0.2'
    if (isNaN(parseInt(`0${val}`))) {
      return this.props.value;
    }

    const rounded = this._normalize(this.props.step, val);

    if (rounded > this.props.max) {
      return this.props.max;
    }

    const gteList = R.filter(x => x >= rounded, this._range());

    if (gteList.length > 0) {
      return R.reduce(R.min, R.head(gteList), gteList);
    }

    return R.pipe(R.filter(x => x <= rounded), R.reduce(R.max, 0))(this._range());
  }

  _renderButton(positive) {
    return ce(Button, {
      tabIndex: -1,
      svgSrc: positive ? 'icons/plus-button.svg' : 'icons/minus-button.svg',
      disabled: this.props.disabled,
      onClick: () => {
        const oldVal = Number(this.props.value);
        let newVal;

        if (positive) {
          newVal = oldVal + this.props.step;
        } else {
          newVal = oldVal - this.props.step;
        }

        return this.props.onChange(this.round(newVal));
      }
    });
  }

  onBlur(val) {
    this.props.onChange(this.round(val));
  }

  render() {
    return DOM.div({
      className: 'increment-row'
    }, // DOM.span({}, this.props.title),
    ce(TextInput, {
      id: this.props.name,
      name: this.props.name,
      clearOnFocus: false,
      placeholder: `${this.props.format(this.props.value)}`,
      disabled: this.props.disabled,
      // onChange: (val) => this.props.onChange(this.round(val)),
      onFocus: (data, element) => {
        element.target.placeholder = '';
      },
      onBlur: (data, element) => {
        element.target.placeholder = `${this.props.format(this.props.value)}`;
      },
      onEnterKeyPress: (val, element) => {
        element.target.blur();
        const v = this.props.sanitize(val);
        this.props.onChange(this.round(v));
      },
      value: this.props.value // validate: (val) => !isNaN(val)

    }), this._renderButton(false), // - button
    this._renderButton(true) // + button
    );
  }

}

Increment.displayName = 'Increment';
Increment.propTypes = {
  name: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  value: React.PropTypes.number.isRequired,
  sanitize: React.PropTypes.func.isRequired,
  format: React.PropTypes.func.isRequired,
  step: React.PropTypes.number.isRequired,
  type: React.PropTypes.string.isRequired,
  min: React.PropTypes.number,
  max: React.PropTypes.number,
  units: React.PropTypes.string,
  onChange: React.PropTypes.func
};
Increment.defaultProps = {
  onChange: () => {}
};
module.exports = Increment;
