// MB-support-plugin/lib/ui/views/input/listbox.js
"use strict";

const R = require('ramda');

const React = require('react');

const DOM = React.DOM; // HELPERS

const disableClass = R.curry((disabled, className) => className + (disabled ? ' disabled' : ''));
const errorClass = R.curry((hasError, className) => className + (hasError ? ' error' : ''));

const listBoxClassName = ({
  disabled,
  error
}, className) => R.pipe(disableClass(disabled), errorClass(error))(className);

const listboxButton = ({
  label,
  onClick,
  disabled
}) => DOM.button({
  className: disableClass(disabled, 'list-box-button'),
  onClick: disabled ? () => {} : onClick
}, label);

const listboxItem = (onSelect, disabled) => ({
  label,
  value,
  selected,
  disabled: itemDisabled
}) => DOM.a({
  key: value,
  className: disableClass(itemDisabled, `list-box-item${selected ? ' list-box-item-selected' : ''}`),
  onClick: disabled || itemDisabled ? () => {} : () => onSelect(value)
}, label);

const listboxItems = (onSelect, items, disabled, emptyLabel) => !items || items.length > 0 ? R.map(listboxItem(onSelect, disabled), items) : DOM.em({}, emptyLabel); // PUBLIC COMPONENT


const Listbox = ({
  items,
  newItem,
  onAdd,
  onRemove,
  onSelect,
  onChange,
  placeholder,
  disabled,
  addLabel,
  removeLabel,
  className,
  emptyLabel,
  error
}) => DOM.div({
  className: `list-box ${className || ''}`
}, DOM.div({
  className: 'list-box-items'
}, listboxItems(onSelect, items, disabled, emptyLabel)), DOM.div({
  className: 'list-box-add-remove'
}, DOM.input({
  className: listBoxClassName({
    disabled,
    error
  }, 'list-box-input'),
  type: 'text',
  onChange: e => onChange(e.target.value),
  value: newItem,
  placeholder,
  disabled
}), listboxButton({
  label: addLabel || 'Add',
  onClick: onAdd,
  disabled
}), listboxButton({
  label: removeLabel || 'Remove',
  onClick: onRemove,
  disabled
})));

Listbox.propTypes = {
  items: React.PropTypes.array.isRequired,
  newItem: React.PropTypes.string,
  onAdd: React.PropTypes.func,
  onRemove: React.PropTypes.func,
  onSelect: React.PropTypes.func
};
Listbox.defaultProps = {
  onAdd: () => {},
  onRemove: () => {},
  onSelect: () => {}
};
module.exports = Listbox;
