// MB-support-plugin/lib/composers/mb_quick_composer.js
'use strict';

const q = require('q');

const _ = require('lodash');

const ComposeProvider = require('eagle-print/services/providers/compose_provider');

const Transform = require('eagle-print/model/transform'); /// Rect object that will rep the bounding boxes of our objects to plate


const Rect = function (x, y, w, h) {
  this.x = _.round(x, 6);
  this.y = _.round(y, 6);
  this.width = _.round(w, 6);
  this.height = _.round(h, 6);
};

Rect.prototype = {
  constructor: Rect,

  /// does this rectangle fully contain the rectangle @rect
  /// @returns {bool}
  contains(rect) {
    return this.x <= rect.x && this.y <= rect.y && this.x + this.width >= rect.x + rect.width && this.y + this.height >= rect.y + rect.height;
  },

  /// does this rectangle overlap with the recatangle @rect
  /// @returns {bool}
  overlaps(rect) {
    // setup the left, right, bottom and top for this rect
    const thisLeft = this.x;
    const thisRight = this.x + this.width;
    const thisBottom = this.y;
    const thisTop = this.y + this.height; // setup the same for the rect

    const thatLeft = rect.x;
    const thatRight = rect.x + rect.width;
    const thatBottom = rect.y;
    const thatTop = rect.y + rect.height; // fun geometry trick - http://stackoverflow.com/a/306332

    return thisLeft < thatRight && thisRight > thatLeft && thisBottom < thatTop && thisTop > thatBottom;
  },

  /// union this rectangle with @rect and @return a new rect that is their
  /// overlap.
  union(rect) {
    const thisLeft = this.x;
    const thisRight = this.x + this.width;
    const thisBottom = this.y;
    const thisTop = this.y + this.height; // setup the same for the rect

    const thatLeft = rect.x;
    const thatRight = rect.x + rect.width;
    const thatBottom = rect.y;
    const thatTop = rect.y + rect.height;
    const left = thisLeft > thatLeft ? thisLeft : thatLeft;
    const width = thisRight < thatRight ? thisRight - left : thatRight - left;
    const bottom = thisBottom > thatBottom ? thisBottom : thatBottom;
    const height = thisTop < thatTop ? thisTop - bottom : thatTop - bottom;
    return new Rect(left, bottom, width, height);
  },

  /// break this rectange into an array of smaller ones, missing the area bound
  /// by @rect. the rectangles in that array may overlap.
  /// @returns {Array of Rect}
  getFreeRects(rect) {
    // if the rect doesn't overlap, just return this rectangle
    if (!this.overlaps(rect)) {
      return [this];
    }

    const freeRects = []; // setup the left, right, bottom and top for this rect

    const boundsLeft = this.x;
    const boundsRight = this.x + this.width;
    const boundsBottom = this.y;
    const boundsTop = this.y + this.height; // setup the same for the rect

    const cutoutLeft = rect.x;
    const cutoutRight = rect.x + rect.width;
    const cutoutBottom = rect.y;
    const cutoutTop = rect.y + rect.height; // bottom

    if (boundsBottom < cutoutBottom) {
      freeRects.push(new Rect(boundsLeft, boundsBottom, boundsRight - boundsLeft, cutoutBottom - boundsBottom));
    } // right


    if (boundsRight > cutoutRight) {
      freeRects.push(new Rect(cutoutRight, boundsBottom, boundsRight - cutoutRight, boundsTop - boundsBottom));
    } // top


    if (boundsTop > cutoutTop) {
      freeRects.push(new Rect(boundsLeft, cutoutTop, boundsRight - boundsLeft, boundsTop - cutoutTop));
    } // left


    if (boundsLeft < cutoutLeft) {
      freeRects.push(new Rect(boundsLeft, boundsBottom, cutoutLeft - boundsLeft, boundsTop - boundsBottom));
    }

    return freeRects;
  },

  /// can this rectangle fit @rect inside of it
  /// @returns {bool}
  canFit(rect) {
    return this.width >= rect.width && this.height >= rect.height;
  },

  /// get the distance from the orgin (0, 0) to the closest point of this
  /// rectangle.
  ///
  /// @returns {float?}
  distanceToOrigin() {
    const point = this.closestPointToOrigin();
    return Math.sqrt(point.x * point.x + point.y * point.y);
  },

  /// get the point closest to the origin (0, 0) on this rectangle
  /// @returns {x, y}
  closestPointToOrigin() {
    // http://stackoverflow.com/a/18157551
    return {
      x: Math.max(this.x, 0, -(this.x + this.width)),
      y: Math.max(this.y, 0, -(this.y + this.height))
    };
  }

}; /// On object representing a build area that we can place rectangles into

const Buildplate = function (buildArea, allowOutsideOfBounds) {
  this.allowOutsideOfBounds = allowOutsideOfBounds; // create spaces - an array of rectangles representing the build plate,
  // initialize with one in each quadrent

  const width = buildArea.x / 2;
  const height = buildArea.y / 2;
  this.buildRect = new Rect(-width, -height, 2 * width, 2 * height);

  if (this.allowOutsideOfBounds) {
    // a solid km x km build plate should be sufficiently large to avoid
    // problems
    const huge = 1000;
    this.spaces = [new Rect(-huge / 2, -huge / 2, huge, huge)];
  } else {
    this.spaces = [this.buildRect];
  }
};

Buildplate.prototype = {
  constructor: Buildplate,

  /// add a bounding box @rect to this buildplate.
  /// @return {bool} - was the rect able to be packed in this buildplate
  pack(rect) {
    // derp
    const self = this; // we'll store the first space outside the buildplate that can fit our rect
    // if we're allowed to place bodies there.

    let firstFitSpace = null;

    for (const i in self.spaces) {
      const space = self.spaces[i]; // check to see if this space can hold rect

      if (space.canFit(rect)) {
        if (self.allowOutsideOfBounds) {
          // if we're allowed to place things outside of the bounds, then some
          // the spaces extend beyond the build plate. check to see if the union
          // of the buildplate and this space can contain this rect. if it can
          // place it there, if it can't store the first space visited in case
          // none of our spaces can hold this rect.
          const spaceOnPlate = space.union(self.buildRect);

          if (spaceOnPlate.canFit(rect)) {
            self._placeRectInSpace(space, rect);

            return true;
          } else if (!firstFitSpace) {
            firstFitSpace = space;
          }
        } else {
          // if we aren't allowed to place things outside of the bounds, then
          // the rect fitting in this space is sufficient. place and exit.
          self._placeRectInSpace(space, rect);

          return true;
        }
      }
    }

    if (self._isEmpty()) {
      // if this buildplate is empty, and the model doesnt fit in the sole
      // space, then place it in the sole space anyway, since it won't fit on
      // any tray we create.
      self._placeRectInSpace(self.spaces[0], rect);

      return true;
    } else if (this.allowOutsideOfBounds) {
      if (!firstFitSpace) {
        throw new Error('Unable to find space for model on current tray.');
      } // if we' are allowed to place outside of the bounds, firstFitSpace is the
      // first space we visited that can contain rect. place the rect there.


      self._placeRectInSpace(firstFitSpace, rect);

      return true;
    } else {
      // if we aren't allowed to place outside of the bounds, return false since
      // we didn't place the rect.
      return false;
    }
  },

  /// @return {bool} - is this buildplate empty?
  _isEmpty() {
    const self = this;
    return self.spaces.length === 1 && self.spaces[0].contains(self.buildRect);
  },

  /// place the rectangle @rect inside of @space such that the center of @rect
  /// is as close to the origin as possible. then place @rect in this.spaces.
  _placeRectInSpace(space, rect) {
    // get the point in the space closest to the origin
    const point = space.closestPointToOrigin(); // set rect.x so that the center of the rect is is closest to point
    //
    // there are three places that we could want our bounding box to end up
    // on the x axis, the extreme right and left of the current space, or on
    // the y axis. we can determine which one based on the closest point of
    // our space to the origin. the same is going to be true for the y axis
    // in the next block of code.
    //
    // LEGEND
    // + - (0 , 0) origin
    // * - point in the current space closest to origin
    // # - right edge of a space
    // @ - left edge of a space
    // $ - x = 0
    //
    // EXAMPLE WHERE WE WANT X = 0 - rect.width / 2
    // point = (0, -2)
    // rect = { width:8, height:3 }
    // space = { x:-34 , y:-7, width:68, height:5 }
    //
    // ---------------------------------------------------------------------
    // |                                                                   |
    // |                                                                   |
    // |                                                                   |
    // |                           ------------                            |
    // |                           |          |                            |
    // |  object already here ->   |     +    |                            |
    // |                           |          |                            |
    // |---------------------------------*---------------------------------|
    // |                             ~~~~~~~~~                             |
    // |                             ~       ~                             |
    // #                             $~~~~~~~~                     @       |
    // |                                                                   |
    // ---------------------------------------------------------------------
    //        space we're going to fill ^
    // EXAMPLE WHERE WE WANT X = space.x
    // point = (5, 0)
    // rect = { width: 8, height: 3 }
    // space = { x:5, y:-7, width:30, height:14 }
    // ---------------------------------------------------------------------
    // |                                      |                            |
    // |        space we're going to fill ->  |                            |
    // |                                      |                            |
    // |                           ------------                            |
    // |                           |          |                            |
    // |  object already here ->   |     +    *                            |
    // |                           |          |                            |
    // |                           ------------                            |
    // |                                      |~~~~~~~                     |
    // |                                      |      ~                     |
    // |                             $        #~~~~~~~             @       |
    // |                                      |                            |
    // ---------------------------------------------------------------------
    // EXAMPLE WHERE WE WANT x = space.x + space.width - rect.width
    // point = (-5, 0)
    // rect = { width: 8, height: 3 }
    // space = { x:-34, y:-7, width:30, height:14 }
    // ---------------------------------------------------------------------
    // |                           |                                       |
    // |                           | <- space we're going to fill          |
    // |                           |                                       |
    // |                           ------------                            |
    // |                           |          |                            |
    // |                           *     +    | <- object already here     |
    // |                           |          |                            |
    // |                           ------------                            |
    // |                    ~~~~~~~|                                       |
    // |                    ~      |                                       |
    // #                    @~~~~~~| $                                     |
    // |                           |                                       |
    // ---------------------------------------------------------------------

    if (point.x === space.x) {
      rect.x = space.x;
    } else if (point.x === 0) {
      rect.x = point.x - rect.width / 2;
    } else {
      rect.x = space.x + space.width - rect.width;
    } // set rect.y so that the center of the rect is closest to the point


    if (point.y === space.y) {
      rect.y = point.y;
    } else if (point.y === 0) {
      rect.y = point.y - rect.height / 2;
    } else {
      rect.y = space.y + space.height - rect.height;
    }

    this.place(rect);
    return true;
  },

  /// update spaces by placing rectangle @rect and dividing up the space(s) it
  /// falls in
  place(rect) {
    // break up any spaces that rect falls into, concatting them into spaces
    const newSpaces = [];

    for (const i in this.spaces) {
      const newbies = this.spaces[i].getFreeRects(rect);

      for (const j in newbies) {
        newSpaces.push(newbies[j]);
      }
    } // consolidate spaces by omitting any space that is fully contained by
    // another space


    this.spaces = newSpaces.filter(function (rect, index) {
      for (let i = 0; i < newSpaces.length; i++) {
        const doesItContain = newSpaces[i].contains(rect); // compare against index to not compare rect against itself

        if (index !== i && doesItContain) {
          return false;
        }
      }

      return true;
    }); // sort the spaces by their distance to the origin. this is the order the
    // next object to be packed will visit the spaces

    this.spaces.sort(function (rectA, rectB) {
      return rectA.distanceToOrigin() - rectB.distanceToOrigin();
    });
  }

}; /// Packer object will take in recangles and assign them coordinates
/// @geometryService - service used to get bounding boxes on tray bodies
/// @tessellationMap - map of tessellations used in current tray
/// @buildArea - width and height of the build area
/// @innerGutter - how much padding each object gets when placed
/// @currentTray - the tray we're packing bodies into
/// @standIns - the objects we're adding to the build plate
/// @restrictToSingleTray - create additional trays as we need them
/// @getNextTray - function that returns the next tray we have to package
/// @projectState - the state of the current project

const Packer = function (geometryService, tessellationMap, buildArea, innerGutter, currentTray, standIns, restrictToSingleTray, getNextTray, projectState) {
  // save the geom service, tess map, inner gutter, and tray
  // TODO - inner gutter should come from tray settings?
  this.geometryService = geometryService;
  this.tessellationMap = tessellationMap;
  this.innerGutter = innerGutter; // store the build area argument and generate the first buildplate

  this.restrictToSingleTray = restrictToSingleTray;
  this.buildArea = buildArea;
  this.buildplate = new Buildplate(this.buildArea, this.restrictToSingleTray);
  this.tray = currentTray;
  this.getNextTray = getNextTray;
  this.projectState = projectState; // sort the stand ins and store that at this.standIns
  // TODO - think about different ways to sort

  this.standIns = standIns.sort(function (standInA, standInB) {
    const sizeA = standInA.getSize();
    const areaA = sizeA.x * sizeA.y;
    const sizeB = standInB.getSize();
    const areaB = sizeB.x * sizeB.y;

    if (areaA > areaB) {
      return -1;
    } else if (areaA === areaB) {
      return 0;
    } else {
      return 1;
    }
  });
};

Packer.prototype = {
  constructor: Packer,

  /// itterate through our tray, and place it into our "buildplate"
  _placeTrayBodies() {
    const self = this;
    self.tray.getTrayBodies().forEach(function (tb) {
      // get the tessellation for this object and grab its bounding box
      const tessellation = self.tessellationMap.get(tb.getOccurrence().getTessellationKey());
      const trayBodyBoundingBox = self.geometryService.getBounds(tb.getOccurrence(), tb.getTransform(), tessellation);
      const min = trayBodyBoundingBox.min;
      const max = trayBodyBoundingBox.max; // create a new rect with the same size as the bounding box, and place it

      self.buildplate.place(new Rect(min.x - self.innerGutter, min.y - self.innerGutter, max.x - min.x + 2 * self.innerGutter, max.y - min.y + 2 * self.innerGutter));
    });
  },

  /// Create a new (empty) buildplate object and store it in this.buildplate.
  /// Use the getNextTray function and reassign this.tray.
  _incrementTray() {
    const self = this;
    self.buildplate = new Buildplate(self.buildArea, self.restrictToSingleTray);
    const trayState = self.getNextTray(self.tray, self.projectState);
    self.tray = trayState.nextTray;
    self.projectState = trayState.newProjectState;
  },

  /// pack the stand ins and place them in spaces. once they are packed,
  /// translate the stand ins and add them to the tray.
  ///
  /// @returns {this.tray} - modified filled tray
  arrangeTray() {
    const self = this;
    return q.Promise(function (resolve, reject) {
      try {
        // create an object to store stand ins that won't fit onto the current tray
        let leftoverStandIns; // while we have stand ins that haven't been placed yet

        while (self.standIns.length > 0) {
          // reset leftoverStandIns to an empty array
          leftoverStandIns = []; // place all of the bodies in the current tray in the current
          // buildplate

          self._placeTrayBodies(); // try to place each stand in in the current buildplate


          for (const i in self.standIns) {
            // get the stand ins bounding box and create a Rect for it, adding in
            // the inner gutter to the height and width.
            const standIn = self.standIns[i];
            const size = standIn.getSize();
            const rect = new Rect(0, 0, size.x + self.innerGutter * 2, size.y + self.innerGutter * 2); // try to pack the bounding rect in the corresponding buildplate.
            // if successful, set the transformation for this standin, and add
            // it to the current tray

            if (self.buildplate.pack(rect)) {
              // extract the shift in (x, y) from the rect, apply it to the stand in
              // and add the stand in to the tray.
              standIn.setTransform(Transform.createTranslationXform(rect.x + self.innerGutter, rect.y + self.innerGutter, 0));
              self.tray = standIn.addBodiesToTray(self.tray).newTray;
            } else {
              // if the stand in doesn't fit on the current buildplate, store
              // it for the next one
              leftoverStandIns.push(standIn);
            }
          } // reset standins to the an array of only the ones that didnt fit


          self.standIns = leftoverStandIns;

          if (self.standIns.length > 0) {
            // if we still have stand ins, create another buildplate and tray to
            // add them to.
            self._incrementTray();
          }
        }

        resolve(self.projectState.setTray(self.tray));
      } catch (err) {
        reject(err);
      }
    });
  }

};

class MBComposer extends ComposeProvider {
  constructor(geometryService) {
    super(geometryService);
  }

  arrangeStandIns(standIns, projectState, tessellationMap, currentTray, restrictToSingleTray, getNextTray) {
    // get the envelope size and subtract the outer gutter from x and y,
    // resulting in the total build area
    const envelopeSize = projectState.getBuildEnvelope().getDimensions();
    const outerGutter = 0.00635;
    const buildArea = {
      x: envelopeSize.x - outerGutter * 2,
      y: envelopeSize.y - outerGutter * 2
    }; // set the inner gutter

    const innerGutter = 0.002; // create a promise with a resolve and reject (maybe an update soon?)

    const self = this; // create compose helper that we're using

    const packer = new Packer(self.geometryService, tessellationMap, buildArea, innerGutter, currentTray, standIns, restrictToSingleTray, getNextTray, projectState);
    return packer.arrangeTray();
  }

  accept(composeData) {
    return composeData.projectState.getGender().getVendor() === 'MakerBot';
  }

}

module.exports = MBComposer;
