// MB-support-plugin/lib/constants/networkPreferences.js
'use strict';

const Actions = Object.freeze({
  Waiting: 'MB_NEWTORK_PREFERENCES_WAITING',
  Connecting: 'MB_NEWTORK_PREFERENCES_CONNECTING',
  WifiConnected: 'MB_NETWORK_PREFERENCES_WIFI_CONNECTED',
  EthernetConnected: 'MB_NETWORK_PREFERENCES_ETHERNET_CONNECTED',
  WifiReady: 'MB_NETWORK_PREFERENCES_WIFI_READY',
  Unconnected: 'MB_NETWORK_PREFERENCES_UNCONNECTED',
  SelectNetwork: 'MB_NETWORK_PREFERENCES_SELECT_NETWORK',
  ChangePasswordAttempt: 'MB_NETWORK_PREFERENCES_CHANGE_PASSWORD_ATTEMPT',
  TogglePasswordVisibility: 'MB_NETWORK_PREFERENCES_TOGGLE_PASSWORD',
  WifiForget: 'MB_NETWORK_PREFERENCES_WIFI_FORGET',
  SetStaticIP: 'MB_NETWORK_PREFERENCES_SET_STATIC_IP',
  PrepopulateStaticIPFields: 'MB_NETWORK_PREFERENCES_PREPOPULATE_STATIC_IP_FIELDS',
  ChangeStaticIPAddress: 'MB_NETWORK_PREFERENCES_CHANGE_STATIC_IP_ADDRESS',
  ChangeNetmask: 'MB_NETWORK_PREFERENCES_CHANGE_NETMASK',
  ChangeGateway: 'MB_NETWORK_PREFERENCES_CHANGE_GATEWAY',
  ChangeDNSServers: 'MB_NETWORK_PREFERENCES_CHANGE_DNS_SERVERS',
  SelectDNSItem: 'MB_NETWORK_PREFERENCES_SELECT_DNS_ITEM',
  NewDNSItem: 'MB_NETWORK_PREFERENCES_NEW_DNS_ITEM',
  Error: 'MB_NETWORK_PREFERENCES_ERROR',
  ClearError: 'MB_NETWORK_PREFERENCES_CLEAR_ERROR'
});
const NetworkStateEnum = Object.freeze({
  state: {
    WIFI: 'wifi',
    ETHERNET: 'ethernet',
    OFFLINE: 'offline'
  },
  wifi: {
    ENABLED: 'enabled',
    DISABLED: 'disabled'
  }
}); // Error code for invalid passwords

const InvalidPassword = 52;
const CHANGE_EVENT = 'change';
module.exports = {
  Actions,
  NetworkStateEnum,
  InvalidPassword,
  CHANGE_EVENT
};
