// MB-support-plugin/lib/constants/print_settings.js
'use strict';

const {
  rangeStep
} = require('../util');

const format = require('../utils/printSettings/format');

const Actions = Object.freeze({
  ToggleProp: 'MB_TOGGLE_PROP',
  BulkToggleProp: 'MB_BULK_TOGGLE_PROPS',
  ResetToDefault: 'MB_RESET_TO_DEFAULT',
  GetPrinterSchema: 'MB_GET_PRINTER_SCHEMA',
  UpdateProperty: 'UPDATE_PROPERTY',
  SelectPrinter: 'SELECT_PRINTER_SUCCESS',
  GetExtruderTypes: 'GET_EXTRUDER_TYPES',
  GetSupportedConfigurations: 'GET_SUPPORTED_CONFIGURATIONS',
  InitalizeValues: 'COMPOSE_SUCCEEDED',
  AuthenticationWaiting: 'AUTHENTICATION_WAITING',
  ResetAllToDefault: 'MB_RESET_ALL_TO_DEFAULT',
  RemoveAllSettings: 'MB_REMOVE_ALL_SETTINGS',
  CompareSettings: 'OPEN_PROJECT_SUCCEEDED',
  CreatePreset: 'MB_CREATE_PRINT_PRESET',
  UpdatePreset: 'MB_UPDATE_PRINT_PRESET',
  WritePreset: 'MB_WRITE_PRINT_PRESET',
  DestroyPreset: 'MB_DESTROY_PRINT_PRESET',
  ReadPresets: 'MB_READ_PRINT_PRESETS',
  ApplyPreset: 'MB_APPLY_PRINT_PRESET',
  UnapplyPreset: 'MB_UNAPPLY_PRINT_PRESET',
  RenamePreset: 'MB_RENAME_PRINT_PRESET',
  PresetError: 'MB_PRINT_PRESET_ERROR',
  CloseModal: 'MB_CLOSE_SETTINGS_MODAL',
  RevertChanges: 'MB_PRINT_SETTINGS_REVERT_CHANGES',
  UpdateExtruder: 'MB_UPDATE_EXTRUDER',
  UpdateExtruderPair: 'MB_UPDATE_EXTRUDER_PAIR',
  UpdateMaterial: 'MB_UPDATE_MATERIAL',
  UpdateMaterials: 'MB_UPDATE_MATERIALS',
  UpdateSupports: 'MB_UPDATE_SUPPORTS',
  UpdatePrintMode: 'MB_UPDATE_PRINT_MODE',
  UpdatePurgeWallCenter: 'MB_UPDATE_PURGE_WALL_CENTER'
}); //  ************************************************************************
//  This object is the handoff from Slice Config getFullSchema().
//  NOTE: If the keys are not mapped correctly here, setting will not appear!
//  ************************************************************************

const SettingsKeys = Object.freeze({
  // ( Patterned after sliceconfig/blog/develop/slice_library/schema_pruner.py )
  // Special - Basic Params
  Extruders: 'extruders',
  Materials: 'materials',
  PrintModes: 'printMode',
  // Printer
  TravelSpeed: 'travelSpeedXY',
  ChamberTemp: 'chamberTemp',
  PlatformTemp: 'platformTemp',
  // BuildPlaneTemp: 'buildPlaneTemp', // Applies only to six-gen printers
  PurgeWall: 'doPurgeWall',
  PurgeEarlyEnd: 'doPurgeEarlyEnd',
  IdleTemp0: 'extruderProfiles>0>idleTemperature',
  IdleTemp1: 'extruderProfiles>1>idleTemperature',
  // Extruder
  ExtruderTemp0: 'extruderProfiles>0>temperature',
  ExtruderTemp1: 'extruderProfiles>1>temperature',
  ExtraToggleDelay0: 'extruderProfiles>0>extraToggleDelay',
  ExtraToggleDelay1: 'extruderProfiles>1>extraToggleDelay',
  RetractDistance0: 'extruderProfiles>0>retractDistance',
  RetractDistance1: 'extruderProfiles>1>retractDistance',
  // Roofs
  RoofSolidThickness: 'roofSolidThickness',
  RoofSurfaceThickness: 'roofSurfaceThickness',
  RoofSurfaceFanSpeed0: 'extruderProfiles>0>extrusionProfiles>roof_surface>fanSpeed',
  RoofSurfaceFanSpeed1: 'extruderProfiles>1>extrusionProfiles>roof_surface>fanSpeed',
  RoofSurfaceFeedRate0: 'extruderProfiles>0>extrusionProfiles>roof_surface>feedrate',
  RoofSurfaceFeedRate1: 'extruderProfiles>1>extrusionProfiles>roof_surface>feedrate',
  // Shells
  NumShells: 'modelShellProfiles>extent>numberOfShells',
  LayerHeight: 'layerHeight',
  OutlineFeedRate0: 'extruderProfiles>0>extrusionProfiles>outlines>feedrate',
  OutlineFeedRate1: 'extruderProfiles>1>extrusionProfiles>outlines>feedrate',
  OutlineFanSpeed0: 'extruderProfiles>0>extrusionProfiles>outlines>fanSpeed',
  OutlineFanSpeed1: 'extruderProfiles>1>extrusionProfiles>outlines>fanSpeed',
  FixedShellStart: 'doFixedShellStart',
  ShellStartDirection: 'fixedShellStartDirection',
  InsetsFanSpeed0: 'extruderProfiles>0>extrusionProfiles>insets>fanSpeed',
  InsetsFanSpeed1: 'extruderProfiles>1>extrusionProfiles>insets>fanSpeed',
  InsetsFeedRate0: 'extruderProfiles>0>extrusionProfiles>insets>feedrate',
  InsetsFeedRate1: 'extruderProfiles>1>extrusionProfiles>insets>feedrate',
  SmartZipper: 'doSmartZipper',
  // Infill
  InfillPattern: 'modelFillProfiles>sparse>pattern',
  InfillDensity: 'modelFillProfiles>sparse>density',
  InfillFillThickness: 'maxSparseFillThickness',
  SparseFeedRate0: 'extruderProfiles>0>extrusionProfiles>sparse>feedrate',
  SparseFeedRate1: 'extruderProfiles>1>extrusionProfiles>sparse>feedrate',
  SparseFanSpeed0: 'extruderProfiles>0>extrusionProfiles>sparse>fanSpeed',
  SparseFanSpeed1: 'extruderProfiles>1>extrusionProfiles>sparse>fanSpeed',
  SolidFeedRate0: 'extruderProfiles>0>extrusionProfiles>solid>feedrate',
  SolidFeedRate1: 'extruderProfiles>1>extrusionProfiles>solid>feedrate',
  SolidFanSpeed0: 'extruderProfiles>0>extrusionProfiles>solid>fanSpeed',
  SolidFanSpeed1: 'extruderProfiles>1>extrusionProfiles>solid>fanSpeed',
  // Floors
  FloorSolidThickness: 'floorSolidThickness',
  FloorSurfaceThickness: 'floorSurfaceThickness',
  FloorSurfaceFanSpeed0: 'extruderProfiles>0>extrusionProfiles>floor_surface>fanSpeed',
  FloorSurfaceFanSpeed1: 'extruderProfiles>1>extrusionProfiles>floor_surface>fanSpeed',
  FloorSurfaceSpeed0: 'extruderProfiles>0>extrusionProfiles>floor_surface>feedrate',
  FloorSurfaceSpeed1: 'extruderProfiles>1>extrusionProfiles>floor_surface>feedrate',
  // Support
  Supports: 'supportType',
  SupportDensity: 'supportFillProfiles>sparse>density',
  SupportedAngle: 'supportAngle',
  SupportModelSpacing: 'supportModelSpacing',
  DoSupportUnderBridges: 'doSupportUnderBridges',
  SupportExtraDistance: 'supportExtraDistance',
  // == Base Layer == :: may collapse all this into Base Layer options.
  // Rafts
  BaseLayer: 'baseLayer',
  RaftToModelSpacing: 'raftModelSpacing',
  RaftModelShellsSpacing: 'raftModelShellsSpacing',
  RaftBaseOutset: 'raft_baseOutset',
  RaftBaseOutsetFromModelEdge: 'raftProfiles>base>outsetFromModelEdge',
  RaftBaseFeedRate0: 'extruderProfiles>0>extrusionProfiles>raft_base>feedrate',
  RaftBaseFeedRate1: 'extruderProfiles>1>extrusionProfiles>raft_base>feedrate',
  RaftBaseFanSpeed0: 'extruderProfiles>0>extrusionProfiles>raft_base>fanSpeed',
  RaftBaseFanSpeed1: 'extruderProfiles>1>extrusionProfiles>raft_base>fanSpeed',
  BaseLayerSurfaceFeedRate0: 'extruderProfiles>0>extrusionProfiles>base_layer_surface>feedrate',
  BaseLayerSurfaceFeedRate1: 'extruderProfiles>1>extrusionProfiles>base_layer_surface>feedrate',
  BaseLayerSurfaceFanSpeed0: 'extruderProfiles>0>extrusionProfiles>base_layer_surface>fanSpeed',
  BaseLayerSurfaceFanSpeed1: 'extruderProfiles>1>extrusionProfiles>base_layer_surface>fanSpeed',
  BaseLayerSurfaceTemp0: 'extruderProfiles>0>extrusionProfiles>base_layer_surface>temperature',
  BaseLayerSurfaceTemp1: 'extruderProfiles>1>extrusionProfiles>base_layer_surface>temperature',
  // Brims
  NumberOfBrims: 'numberOfBrims',
  NumberOfInternalBrims: 'numberOfInternalBrims',
  BrimsModelOffset: 'brimsModelOffset',
  //
  // 5th-gen Experimental and 6th-gen LABS extruders
  //
  ExtruderRestartExtraDistance0: 'extruderProfiles>0>restartExtraDistance',
  ExtruderRetractRate0: 'extruderProfiles>0>retractRate',
  //
  // 5th-gen Experimental Extruder Only.
  //
  BridgeFanSpeed: 'extruderProfiles>0>extrusionProfiles>bridges>fanSpeed',
  BridgeFeedRate: 'extruderProfiles>0>extrusionProfiles>bridges>feedrate',
  DefaultFanSpeed: 'fanDefaultSpeed',
  ExtruderRestartExtraRate: 'extruderProfiles>0>restartExtraRate',
  ExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionVolumeMultiplier',
  FanLayer: 'fanLayer',
  FloorFanSpeed: 'extruderProfiles>0>extrusionProfiles>floorSurfaceFills>fanSpeed',
  FloorFeedRate: 'extruderProfiles>0>extrusionProfiles>floorSurfaceFills>feedrate',
  InfillFanSpeed: 'extruderProfiles>0>extrusionProfiles>infill>fanSpeed',
  NozzleDiameter: 'extruderProfiles>0>nozzleDiameter',
  RaftFeedRate: 'extruderProfiles>0>extrusionProfiles>raft>feedrate',
  RaftFanSpeed: 'extruderProfiles>0>extrusionProfiles>raft>fanSpeed',
  RoofSurfaceFillsFanSpeed: 'extruderProfiles>0>extrusionProfiles>roofSurfaceFills>fanSpeed',
  RoofSurfaceFillsFeedRate: 'extruderProfiles>0>extrusionProfiles>roofSurfaceFills>feedrate',
  SparseRoofSurfaceFanSpeed: 'extruderProfiles>0>extrusionProfiles>sparseRoofSurfaceFills>fanSpeed',
  SparseRoofSurfaceFeedRate: 'extruderProfiles>0>extrusionProfiles>sparseRoofSurfaceFills>feedrate',
  SpursFanSpeed: 'extruderProfiles>0>extrusionProfiles>spurs>fanSpeed',
  SpursFeedRate: 'extruderProfiles>0>extrusionProfiles>spurs>feedrate',
  //
  // 6th-gen LABS Extruder Only.
  //
  ExtruderRestartExtraDistance1: 'extruderProfiles>1>restartExtraDistance',
  ExtruderRetractRate1: 'extruderProfiles>1>retractRate',
  ToolchangeRestartDistance0: 'extruderProfiles>0>toolchangeRestartDistance',
  ToolchangeRestartDistance1: 'extruderProfiles>1>toolchangeRestartDistance',
  ToolchangeRestartRate0: 'extruderProfiles>0>toolchangeRestartRate',
  ToolchangeRestartRate1: 'extruderProfiles>1>toolchangeRestartRate',
  ToolchangeRetractDistance0: 'extruderProfiles>0>toolchangeRetractDistance',
  ToolchangeRetractDistance1: 'extruderProfiles>1>toolchangeRetractDistance',
  ToolchangeRetractRate0: 'extruderProfiles>0>toolchangeRetractRate',
  ToolchangeRetractRate1: 'extruderProfiles>1>toolchangeRetractRate',
  RaftBaseTemp0: 'extruderProfiles>0>extrusionProfiles>raft_base>temperature',
  RaftBaseTemp1: 'extruderProfiles>1>extrusionProfiles>raft_base>temperature',
  RaftBaseSpeed0: 'extruderProfiles>0>extrusionProfiles>raft_base>feedrate',
  RaftBaseSpeed1: 'extruderProfiles>1>extrusionProfiles>raft_base>feedrate',
  RaftInterfaceTemp0: 'extruderProfiles>0>extrusionProfiles>raft_interface>temperature',
  RaftInterfaceTemp1: 'extruderProfiles>1>extrusionProfiles>raft_interface>temperature',
  RaftInterfaceSpeed0: 'extruderProfiles>0>extrusionProfiles>raft_interface>feedrate',
  RaftInterfaceSpeed1: 'extruderProfiles>1>extrusionProfiles>raft_interface>feedrate',
  RaftSurfaceTemp0: 'extruderProfiles>0>extrusionProfiles>raft_surface>temperature',
  RaftSurfaceTemp1: 'extruderProfiles>1>extrusionProfiles>raft_surface>temperature',
  RaftSurfaceSpeed0: 'extruderProfiles>0>extrusionProfiles>raft_surface>feedrate',
  RaftSurfaceSpeed1: 'extruderProfiles>1>extrusionProfiles>raft_surface>feedrate',
  RestartRate0: 'extruderProfiles>0>restartRate',
  RestartRate1: 'extruderProfiles>1>restartRate',
  //baselayer
  XScale: 'xScale',
  YScale: 'yScale',
  ZScale: 'zScale',
  // Brims: 'doBrims',
  // BrimsBaseWidth: 'brimsBaseWidth',
  // BrimsOverlapWidth: 'brimsOverlapWidth',
  // PaddedBase: 'doPaddedBase',
  // PaddedBaseWidth: 'paddedBaseWidth',
  // PaddedBaseOutlineOffset: 'paddedBaseOutlineOffset',
  // RaftBrimsSpacing: 'raftBrimsSpacing',
  RaftBaseZOffset: 'raftProfiles>base>ZOffset',
  RaftBaseExtruderID: 'raftProfiles>base>extruderID',
  RaftBaseFirstSurfaceZOffset: 'raftProfiles>base>firstSurfaceZOffset',
  RaftBaseLayerThickness: 'raftProfiles>base>layerThickness',
  RaftBaseLayerWidth: 'raftProfiles>base>layerWidth',
  RaftBaseNumberOfLayers: 'raftProfiles>base>numberOfLayers',
  RaftInterfaceZOffset: 'raftProfiles>interface>ZOffset',
  RaftInterfaceExtruderID: 'raftProfiles>interface>extruderID',
  RaftInterfaceFirstSurfaceZOffset: 'raftProfiles>interface>firstSurfaceZOffset',
  RaftInterfaceLayerThickness: 'raftProfiles>interface>layerThickness',
  RaftInterfaceLayerWidth: 'raftProfiles>interface>layerWidth',
  RaftInterfaceNumberOfLayers: 'raftProfiles>interface>numberOfLayers',
  RaftInterfaceOutset: 'raftProfiles>interface>outsetFromModelEdge',
  RaftSurfaceZOffset: 'raftProfiles>surface>ZOffset',
  RaftSurfaceExtruderID: 'raftProfiles>surface>extruderID',
  RaftSurfaceFirstSurfaceZOffset: 'raftProfiles>surface>firstSurfaceZOffset',
  RaftSurfaceLayerThickness: 'raftProfiles>surface>layerThickness',
  RaftSurfaceLayerWidth: 'raftProfiles>surface>layerWidth',
  RaftSurfaceNumberOfLayers: 'raftProfiles>surface>numberOfLayers',
  RaftSurfaceOutset: 'raftProfiles>surface>outsetFromModelEdge',
  BaseLayerSurfaceExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>base_layer_surface>extrusionVolumeMultiplier',
  // BrimsExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>brims>extrusionVolumeMultiplier',
  //floor
  FloorSurfaceExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>floor_surface>extrusionVolumeMultiplier',
  //shells
  InsetsExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>insets>extrusionVolumeMultiplier',
  OutlinesExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>outlines>extrusionVolumeMultiplier',
  //roofs
  RoofSurfaceExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>roof_surface>extrusionVolumeMultiplier',
  //extruder
  ExtrusionVolumeMultiplierLABS: 'extruderProfiles>0>extrusionVolumeMultiplier',
  SolidExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>solid>extrusionVolumeMultiplier',
  SparseExtrusionVolumeMultiplier: 'extruderProfiles>0>extrusionProfiles>sparse>extrusionVolumeMultiplier',
  //infill
  ConvertSparsePulseValue: 'convertSparsePulseValue' // end 6th-gen LABS Extruder Only

}); //  ************************************************************************
//  Used by various UI as ENUMS
//  ************************************************************************

const GroupKeys = Object.freeze({
  BasicSettings: 'basic_settings',
  QuickSettings: 'quick_settings',
  Printer: 'printer',
  Extruder: 'extruder',
  Roofs: 'roofs',
  Shells: 'shells',
  Infill: 'infill',
  Floors: 'floors',
  SupportsBridging: 'supports_bridging',
  BaseLayer: 'base_layer'
}); //  ************************************************************************
//  Used by advancedSettingsModal.js to group and display possible settings
//  ************************************************************************

const GroupDict = Object.freeze({
  [GroupKeys.QuickSettings]: [SettingsKeys.Supports, SettingsKeys.BaseLayer, SettingsKeys.InfillDensity, SettingsKeys.NumShells, SettingsKeys.LayerHeight, SettingsKeys.SupportedAngles],
  [GroupKeys.Printer]: [SettingsKeys.TravelSpeed, SettingsKeys.ChamberTemp, SettingsKeys.BuildPlaneTemp, SettingsKeys.PlatformTemp, // SettingsKeys.BuildPlaneTemp,
  SettingsKeys.PurgeWall, SettingsKeys.PurgeEarlyEnd],
  [GroupKeys.Extruder]: [SettingsKeys.ExtruderTemp0, SettingsKeys.ExtruderTemp1, SettingsKeys.ExtraToggleDelay0, SettingsKeys.ExtraToggleDelay1, SettingsKeys.IdleTemp0, SettingsKeys.IdleTemp1, SettingsKeys.RetractDistance0, SettingsKeys.RetractDistance1, SettingsKeys.NozzleDiameter, SettingsKeys.ExtruderRetractRate0, SettingsKeys.ExtruderRetractRate1, SettingsKeys.ExtruderRestartExtraDistance0, SettingsKeys.ExtruderRestartExtraDistance1, SettingsKeys.ExtruderRestartExtraRate, SettingsKeys.ExtrusionVolumeMultiplier, SettingsKeys.DefaultFanSpeed, SettingsKeys.FanLayer, SettingsKeys.RestartRate0, SettingsKeys.RestartRate1, SettingsKeys.ToolchangeRestartDistance0, SettingsKeys.ToolchangeRestartDistance1, SettingsKeys.ToolchangeRestartRate0, SettingsKeys.ToolchangeRestartRate1, SettingsKeys.ToolchangeRetractDistance0, SettingsKeys.ToolchangeRetractDistance1, SettingsKeys.ToolchangeRetractRate0, SettingsKeys.ToolchangeRetractRate1, SettingsKeys.ExtrusionVolumeMultiplierLABS, SettingsKeys.SolidExtrusionVolumeMultiplier, SettingsKeys.SparseExtrusionVolumeMultiplier],
  [GroupKeys.Roofs]: [SettingsKeys.RoofSolidThickness, SettingsKeys.RoofSurfaceThickness, SettingsKeys.RoofSurfaceFanSpeed0, SettingsKeys.RoofSurfaceFanSpeed1, SettingsKeys.RoofSurfaceFeedRate0, SettingsKeys.RoofSurfaceFeedRate1, SettingsKeys.RoofSurfaceFillsFanSpeed, SettingsKeys.RoofSurfaceFillsFeedRate, SettingsKeys.SparseRoofSurfaceFanSpeed, SettingsKeys.SparseRoofSurfaceFeedRate, SettingsKeys.RoofSurfaceExtrusionVolumeMultiplier],
  [GroupKeys.Shells]: [SettingsKeys.NumShells, SettingsKeys.LayerHeight, SettingsKeys.OutlineFeedRate0, SettingsKeys.OutlineFeedRate1, SettingsKeys.OutlineFanSpeed0, SettingsKeys.OutlineFanSpeed1, SettingsKeys.FixedShellStart, SettingsKeys.ShellStartDirection, SettingsKeys.InsetsFeedRate0, SettingsKeys.InsetsFeedRate1, SettingsKeys.InsetsFanSpeed0, SettingsKeys.InsetsFanSpeed1, SettingsKeys.SmartZipper, SettingsKeys.InsetsExtrusionVolumeMultiplier, SettingsKeys.OutlinesExtrusionVolumeMultiplier],
  [GroupKeys.Infill]: [SettingsKeys.InfillPattern, SettingsKeys.InfillDensity, SettingsKeys.InfillFillThickness, SettingsKeys.SparseFanSpeed0, SettingsKeys.SparseFanSpeed1, SettingsKeys.SparseFeedRate0, SettingsKeys.SparseFeedRate1, SettingsKeys.SolidFeedRate0, SettingsKeys.SolidFeedRate1, SettingsKeys.InfillFanSpeed, SettingsKeys.SolidFanSpeed0, SettingsKeys.SolidFanSpeed1, SettingsKeys.SpursFanSpeed, SettingsKeys.SpursFeedRate, SettingsKeys.ConvertSparsePulseValue],
  [GroupKeys.Floors]: [SettingsKeys.FloorSolidThickness, SettingsKeys.FloorSurfaceThickness, SettingsKeys.FloorFanSpeed, SettingsKeys.FloorFeedRate, SettingsKeys.FloorSurfaceFanSpeed0, SettingsKeys.FloorSurfaceFanSpeed1, SettingsKeys.FloorSurfaceSpeed0, SettingsKeys.FloorSurfaceSpeed1, SettingsKeys.FloorSurfaceExtrusionVolumeMultiplier],
  [GroupKeys.SupportsBridging]: [SettingsKeys.Supports, SettingsKeys.SupportDensity, SettingsKeys.SupportedAngle, SettingsKeys.SupportModelSpacing, SettingsKeys.DoSupportUnderBridges, SettingsKeys.BridgeFanSpeed, SettingsKeys.BridgeFeedRate, SettingsKeys.SupportExtraDistance],
  [GroupKeys.BaseLayer]: [SettingsKeys.BaseLayer, SettingsKeys.BaseLayerSurfaceFeedRate0, SettingsKeys.BaseLayerSurfaceFeedRate1, SettingsKeys.BaseLayerSurfaceFanSpeed0, SettingsKeys.BaseLayerSurfaceFanSpeed1, SettingsKeys.BaseLayerSurfaceTemp0, SettingsKeys.BaseLayerSurfaceTemp1, SettingsKeys.RaftToModelSpacing, SettingsKeys.RaftBaseOutset, SettingsKeys.RaftBaseOutsetFromModelEdge, SettingsKeys.RaftBaseFeedRate0, SettingsKeys.RaftBaseFeedRate1, SettingsKeys.RaftBaseFanSpeed0, SettingsKeys.RaftBaseFanSpeed1, SettingsKeys.RaftBaseTemp0, SettingsKeys.RaftBaseTemp1, SettingsKeys.RaftInterfaceSpeed0, SettingsKeys.RaftInterfaceSpeed1, SettingsKeys.RaftInterfaceTemp0, SettingsKeys.RaftInterfaceTemp1, SettingsKeys.RaftModelShellsSpacing, SettingsKeys.RaftFeedRate, SettingsKeys.RaftFanSpeed, SettingsKeys.RaftSurfaceSpeed0, SettingsKeys.RaftSurfaceSpeed1, SettingsKeys.RaftSurfaceTemp0, SettingsKeys.RaftSurfaceTemp1, SettingsKeys.NumberOfBrims, SettingsKeys.NumberOfInternalBrims, //
  // 6th-gen LABS Extruder Only.
  //
  SettingsKeys.XScale, SettingsKeys.YScale, SettingsKeys.ZScale, SettingsKeys.Brims, SettingsKeys.BrimsBaseWidth, SettingsKeys.BrimsModelOffset, SettingsKeys.BrimsOverlapWidth, SettingsKeys.BrimsExtrusionVolumeMultiplier, SettingsKeys.PaddedBase, SettingsKeys.PaddedBaseWidth, SettingsKeys.PaddedBaseOutlineOffset, SettingsKeys.BaseLayerSurfaceExtrusionVolumeMultiplier, SettingsKeys.RaftBrimsSpacing, SettingsKeys.RaftBaseZOffset, SettingsKeys.RaftBaseExtruderID, SettingsKeys.RaftBaseFirstSurfaceZOffset, SettingsKeys.RaftBaseLayerThickness, SettingsKeys.RaftBaseLayerWidth, SettingsKeys.RaftBaseNumberOfLayers, SettingsKeys.RaftInterfaceZOffset, SettingsKeys.RaftInterfaceExtruderID, SettingsKeys.RaftInterfaceFirstSurfaceZOffset, SettingsKeys.RaftInterfaceLayerThickness, SettingsKeys.RaftInterfaceLayerWidth, SettingsKeys.RaftInterfaceNumberOfLayers, SettingsKeys.RaftInterfaceOutset, SettingsKeys.RaftSurfaceZOffset, SettingsKeys.RaftSurfaceExtruderID, SettingsKeys.RaftSurfaceFirstSurfaceZOffset, SettingsKeys.RaftSurfaceLayerThickness, SettingsKeys.RaftSurfaceLayerWidth, SettingsKeys.RaftSurfaceNumberOfLayers, SettingsKeys.RaftSurfaceOutset]
}); //  ************************************************************************
//  Pretty strings for display.
//  ************************************************************************

const DisplayKeys = Object.freeze({
  basic_settings: 'Basic Settings',
  quick_settings: 'Quick Settings',
  printer: 'Printer',
  extruder: 'Extruder',
  extruders: 'Extruders',
  roofs: 'Roofs',
  shells: 'Shells',
  infill: 'Infill',
  floors: 'Floors',
  supports_bridging: 'Supports + Bridging',
  support: 'Support',
  base_layer: 'Base Layer'
});
const PrintModes = Object.freeze({
  Balanced: 'balanced',
  Draft: 'draft',
  Minfill: 'minfill',
  Skeleton: 'skeleton',
  BestFinish: 'best-finish',
  HighQuality: 'high-quality',
  HighStrength: 'high-strength',
  Solid: 'solid'
});
const Types = Object.freeze({
  Array: 'array',
  Boolean: 'boolean',
  Increment: 'increment',
  String: 'string',
  Scalar: 'scalar',
  UnsignedInt: 'unsigned_integer',
  Percent: 'percent'
}); //  ************************************************************************
//  Types for validator (?): mb-project-validator-provider.js
//  NOTE:  Can we lose all of this?
//  ************************************************************************

/**
 * Add more entries to this copy Immutable map instead of using the
 * one from SettingsTypes because the types didn't match
 * for rendering UI settings panel vs. project meta.json value.
 * Rather than screwing up the UI view, copy the map.
 *
 * layerHeight/numberOfShell are numbers in meta.json vs strings in
 * UI
 *
 * If a new print settings option's type was not listed in
 * SettingsTypes, then getOptionalPluginRef will default to
 * 'GeneralElemNumber' which there is a high chance that is correct.
 */

/** @type {Immutable.Map<string, string>} */

const SettingsTypes = Object.freeze({
  [SettingsKeys.Extruders]: Types.Array,
  [SettingsKeys.Materials]: Types.Array,
  [SettingsKeys.NumShells]: Types.String,
  [SettingsKeys.FixedShellStart]: Types.Boolean,
  [SettingsKeys.InfillPattern]: Types.String,
  [SettingsKeys.Supports]: Types.String,
  [SettingsKeys.BaseLayer]: Types.String,
  [SettingsKeys.PrintModes]: Types.String
});
const ValidatorSettingsTypes = Object.freeze({
  // Special - Basic Params
  [SettingsKeys.Extruders]: Types.Array,
  [SettingsKeys.Materials]: Types.Array,
  [SettingsKeys.PrintModes]: Types.String,
  // Printer
  [SettingsKeys.TravelSpeed]: Types.UnsignedInt,
  [SettingsKeys.ChamberTemp]: Types.UnsignedInt,
  // [SettingsKeys.BuildPlaneTemp]: Types.UnsignedInt,
  [SettingsKeys.PlatformTemp]: Types.UnsignedInt,
  [SettingsKeys.PurgeWall]: Types.Boolean,
  [SettingsKeys.PurgeEarlyEnd]: Types.Boolean,
  // Extruder
  [SettingsKeys.ExtruderTemp0]: Types.UnsignedInt,
  [SettingsKeys.ExtruderTemp1]: Types.UnsignedInt,
  [SettingsKeys.ExtraToggleDelay0]: Types.UnsignedInt,
  [SettingsKeys.ExtraToggleDelay1]: Types.UnsignedInt,
  [SettingsKeys.IdleTemp0]: Types.UnsignedInt,
  [SettingsKeys.IdleTemp1]: Types.UnsignedInt,
  [SettingsKeys.RetractDistance0]: Types.UnsignedInt,
  [SettingsKeys.RetractDistance1]: Types.UnsignedInt,
  [SettingsKeys.NozzleDiameter]: Types.UnsignedInt,
  [SettingsKeys.ExtruderRetractRate0]: Types.UnsignedInt,
  [SettingsKeys.ExtruderRetractRate1]: Types.UnsignedInt,
  [SettingsKeys.ExtruderRestartExtraDistance0]: Types.UnsignedInt,
  [SettingsKeys.ExtruderRestartExtraDistance1]: Types.UnsignedInt,
  [SettingsKeys.ExtruderRestartExtraRate]: Types.UnsignedInt,
  [SettingsKeys.ExtrusionVolumeMultiplier]: Types.UnsignedInt,
  [SettingsKeys.DefaultFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.FanLayer]: Types.UnsignedInt,
  [SettingsKeys.RestartRate0]: Types.UnsignedInt,
  [SettingsKeys.RestartRate1]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRestartDistance0]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRestartDistance1]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRestartRate0]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRestartRate1]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRetractDistance0]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRetractDistance1]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRetractRate0]: Types.UnsignedInt,
  [SettingsKeys.ToolchangeRetractRate1]: Types.UnsignedInt,
  [SettingsKeys.ExtrusionVolumeMultiplierLABS]: Types.UnsignedInt,
  [SettingsKeys.SolidExtrusionVolumeMultiplier]: Types.UnsignedInt,
  [SettingsKeys.SparseExtrusionVolumeMultiplier]: Types.UnsignedInt,
  // Roofs
  [SettingsKeys.RoofSolidThickness]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceThickness]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFillsFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceFillsFeedRate]: Types.UnsignedInt,
  [SettingsKeys.SparseRoofSurfaceFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.SparseRoofSurfaceFeedRate]: Types.UnsignedInt,
  [SettingsKeys.RoofSurfaceExtrusionVolumeMultiplier]: Types.UnsignedInt,
  // Shells
  [SettingsKeys.NumShells]: Types.UnsignedInt,
  [SettingsKeys.LayerHeight]: Types.UnsignedInt,
  [SettingsKeys.OutlineFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.OutlineFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.OutlineFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.OutlineFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.FixedShellStart]: Types.Boolean,
  [SettingsKeys.ShellStartDirection]: Types.UnsignedInt,
  [SettingsKeys.InsetsFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.InsetsFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.InsetsFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.InsetsFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.SmartZipper]: Types.Boolean,
  [SettingsKeys.InsetsExtrusionVolumeMultiplier]: Types.UnsignedInt,
  [SettingsKeys.OutlinesExtrusionVolumeMultiplier]: Types.UnsignedInt,
  // Infill
  [SettingsKeys.InfillPattern]: Types.String,
  [SettingsKeys.InfillDensity]: Types.UnsignedInt,
  [SettingsKeys.InfillFillThickness]: Types.UnsignedInt,
  [SettingsKeys.SparseFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.SparseFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.SolidFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.SolidFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.SolidFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.SolidFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.InfillFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.SparseFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.SparseFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.SpursFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.SpursFeedRate]: Types.UnsignedInt,
  [SettingsKeys.ConvertSparsePulseValue]: Types.UnsignedInt,
  // Floors
  [SettingsKeys.FloorSolidThickness]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceThickness]: Types.UnsignedInt,
  [SettingsKeys.FloorFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.FloorFeedRate]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceSpeed0]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceSpeed1]: Types.UnsignedInt,
  [SettingsKeys.FloorSurfaceExtrusionVolumeMultiplier]: Types.UnsignedInt,
  // Support
  [SettingsKeys.Supports]: Types.String,
  [SettingsKeys.SupportDensity]: Types.UnsignedInt,
  [SettingsKeys.SupportedAngle]: Types.UnsignedInt,
  [SettingsKeys.SupportModelSpacing]: Types.UnsignedInt,
  [SettingsKeys.DoSupportUnderBridges]: Types.Boolean,
  [SettingsKeys.BridgeFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.BridgeFeedRate]: Types.UnsignedInt,
  [SettingsKeys.SupportExtraDistance]: Types.UnsignedInt,
  // Base Layer
  [SettingsKeys.BaseLayer]: Types.String,
  [SettingsKeys.BaseLayerSurfaceFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceTemp0]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceTemp1]: Types.UnsignedInt,
  [SettingsKeys.BaseLayerSurfaceExtrusionVolumeMultiplier]: Types.UnsignedInt,
  [SettingsKeys.RaftToModelSpacing]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseOutset]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseOutsetFromModelEdge]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseTemp0]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseTemp1]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceSpeed0]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceSpeed0]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceSpeed1]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceTemp0]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceTemp1]: Types.UnsignedInt,
  [SettingsKeys.RaftModelShellsSpacing]: Types.UnsignedInt,
  [SettingsKeys.RaftFeedRate]: Types.UnsignedInt,
  [SettingsKeys.RaftFanSpeed]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceSpeed0]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceSpeed1]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceTemp0]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceTemp1]: Types.UnsignedInt,
  [SettingsKeys.RaftBrimsSpacing]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseFirstSurfaceZOffset]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseLayerThickness]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseLayerWidth]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseNumberOfLayers]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseZOffset]: Types.UnsignedInt,
  [SettingsKeys.RaftBaseExtruderID]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceNumberOfLayers]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceOutset]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceExtruderID]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceZOffset]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceLayerThickness]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceLayerWidth]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceExtruderID]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceFirstSurfaceZOffset]: Types.UnsignedInt,
  [SettingsKeys.RaftInterfaceFirstSurfaceZOffset]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceLayerThickness]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceLayerWidth]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceNumberOfLayers]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceOutset]: Types.UnsignedInt,
  [SettingsKeys.RaftSurfaceZOffset]: Types.UnsignedInt,
  [SettingsKeys.XScale]: Types.UnsignedInt,
  [SettingsKeys.YScale]: Types.UnsignedInt,
  [SettingsKeys.ZScale]: Types.UnsignedInt,
  [SettingsKeys.PaddedBase]: Types.Boolean,
  [SettingsKeys.PaddedBaseOutlineOffset]: Types.UnsignedInt,
  [SettingsKeys.PaddedBaseWidth]: Types.UnsignedInt,
  // Brims
  [SettingsKeys.Brims]: Types.Boolean,
  [SettingsKeys.NumberOfBrims]: Types.UnsignedInt,
  [SettingsKeys.NumberOfInternalBrims]: Types.UnsignedInt,
  [SettingsKeys.BrimsBaseWidth]: Types.UnsignedInt,
  [SettingsKeys.BrimsOverlapWidth]: Types.UnsignedInt,
  [SettingsKeys.BrimsModelOffset]: Types.UnsignedInt,
  [SettingsKeys.BrimsFeedRate0]: Types.UnsignedInt,
  [SettingsKeys.BrimsFeedRate1]: Types.UnsignedInt,
  [SettingsKeys.BrimsFanSpeed0]: Types.UnsignedInt,
  [SettingsKeys.BrimsFanSpeed1]: Types.UnsignedInt,
  [SettingsKeys.BrimsExtrusionVolumeMultiplier]: Types.UnsignedInt
}); // UI formatting for display.

/** @type {Immutable.Map<string, format.formatHigherOrderFunc>} */

const Format = Object.freeze({
  [SettingsKeys.Extruders]: format.extruders,
  [SettingsKeys.PrintModes]: format.printMode,
  // Printer
  [SettingsKeys.TravelSpeed]: format.millimeterPerSecond,
  [SettingsKeys.ChamberTemp]: format.celsius,
  // [SettingsKeys.BuildPlaneTemp]: format.celsius,
  [SettingsKeys.PlatformTemp]: format.celsius,
  [SettingsKeys.PurgeWall]: format.noop,
  [SettingsKeys.PurgeEarlyEnd]: format.noop,
  [SettingsKeys.SmartZipper]: format.noop,
  // Extruder
  [SettingsKeys.ExtruderTemp0]: format.celsius,
  [SettingsKeys.ExtruderTemp1]: format.celsius,
  [SettingsKeys.ExtraToggleDelay0]: format.seconds,
  [SettingsKeys.ExtraToggleDelay1]: format.seconds,
  [SettingsKeys.IdleTemp0]: format.celsius,
  [SettingsKeys.IdleTemp1]: format.celsius,
  [SettingsKeys.RetractDistance0]: format.millimeter,
  [SettingsKeys.RetractDistance1]: format.millimeter,
  [SettingsKeys.NozzleDiameter]: format.millimeter,
  [SettingsKeys.ExtruderRetractRate0]: format.millimeterPerSecond,
  [SettingsKeys.ExtruderRetractRate1]: format.millimeterPerSecond,
  [SettingsKeys.ExtruderRestartExtraDistance0]: format.millimeter,
  [SettingsKeys.ExtruderRestartExtraDistance1]: format.millimeter,
  [SettingsKeys.ExtruderRestartExtraRate]: format.millimeterPerSecond,
  [SettingsKeys.ExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.DefaultFanSpeed]: format.percent,
  [SettingsKeys.FanLayer]: format.layers,
  [SettingsKeys.RestartRate0]: format.millimeterPerSecond,
  [SettingsKeys.RestartRate1]: format.millimeterPerSecond,
  [SettingsKeys.ToolchangeRestartDistance0]: format.millimeter,
  [SettingsKeys.ToolchangeRestartDistance1]: format.millimeter,
  [SettingsKeys.ToolchangeRestartRate0]: format.millimeterPerSecond,
  [SettingsKeys.ToolchangeRestartRate1]: format.millimeterPerSecond,
  [SettingsKeys.ToolchangeRetractDistance0]: format.millimeter,
  [SettingsKeys.ToolchangeRetractDistance1]: format.millimeter,
  [SettingsKeys.ToolchangeRetractRate0]: format.millimeterPerSecond,
  [SettingsKeys.ToolchangeRetractRate1]: format.millimeterPerSecond,
  [SettingsKeys.ExtrusionVolumeMultiplierLABS]: format.percent,
  // Roofs
  [SettingsKeys.RoofSolidThickness]: format.millimeter,
  [SettingsKeys.RoofSurfaceFanSpeed0]: format.percent,
  [SettingsKeys.RoofSurfaceFanSpeed1]: format.percent,
  [SettingsKeys.RoofSurfaceFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.RoofSurfaceFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.RoofSurfaceThickness]: format.millimeter,
  [SettingsKeys.RoofFillsSurfaceFanSpeed]: format.percent,
  [SettingsKeys.RoofSurfaceFillsFeedRate]: format.millimeterPerSecond,
  [SettingsKeys.SparseRoofSurfaceFanSpeed]: format.percent,
  [SettingsKeys.SparseRoofSurfaceFeedRate]: format.millimeterPerSecond,
  // Shells
  [SettingsKeys.NumShells]: format.shells,
  [SettingsKeys.LayerHeight]: format.layerHeight,
  [SettingsKeys.OutlineFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.OutlineFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.OutlineFanSpeed0]: format.percent,
  [SettingsKeys.OutlineFanSpeed1]: format.percent,
  [SettingsKeys.FixedShellStart]: format.noop,
  [SettingsKeys.ShellStartDirection]: format.degrees,
  [SettingsKeys.InsetsFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.InsetsFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.InsetsFanSpeed0]: format.percent,
  [SettingsKeys.InsetsFanSpeed1]: format.percent,
  // Infill
  [SettingsKeys.InfillDensity]: format.percent,
  [SettingsKeys.InfillFillThickness]: format.millimeter,
  [SettingsKeys.InfillPattern]: format.validString,
  [SettingsKeys.SparseFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.SparseFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.SolidFanSpeed0]: format.percent,
  [SettingsKeys.SolidFanSpeed1]: format.percent,
  [SettingsKeys.SolidFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.SolidFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.SparseFanSpeed0]: format.percent,
  [SettingsKeys.SparseFanSpeed1]: format.percent,
  [SettingsKeys.InfillFanSpeed]: format.percent,
  [SettingsKeys.SpursFanSpeed]: format.percent,
  [SettingsKeys.SpursFeedRate]: format.millimeterPerSecond,
  // Floors
  [SettingsKeys.FloorSolidThickness]: format.roundMillimeter,
  [SettingsKeys.FloorSurfaceThickness]: format.roundMillimeter,
  [SettingsKeys.FloorFanSpeed]: format.percent,
  [SettingsKeys.FloorFeedRate]: format.millimeterPerSecond,
  [SettingsKeys.FloorSurfaceFanSpeed0]: format.percent,
  [SettingsKeys.FloorSurfaceFanSpeed1]: format.percent,
  [SettingsKeys.FloorSurfaceSpeed0]: format.millimeterPerSecond,
  [SettingsKeys.FloorSurfaceSpeed1]: format.millimeterPerSecond,
  // Supports
  [SettingsKeys.Supports]: format.validString,
  [SettingsKeys.SupportDensity]: format.percent,
  [SettingsKeys.SupportModelSpacing]: format.millimeter,
  [SettingsKeys.SupportedAngle]: format.degrees,
  [SettingsKeys.DoSupportUnderBridges]: format.noop,
  [SettingsKeys.BridgeFanSpeed]: format.percent,
  [SettingsKeys.BridgeFeedRate]: format.millimeterPerSecond,
  [SettingsKeys.SupportExtraDistance]: format.millimeter,
  // Base Layer
  [SettingsKeys.BaseLayer]: format.validString,
  [SettingsKeys.RaftToModelSpacing]: format.millimeter,
  [SettingsKeys.RaftBaseOutset]: format.millimeter,
  [SettingsKeys.RaftBaseOutsetFromModelEdge]: format.millimeter,
  [SettingsKeys.RaftBaseFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.RaftBaseFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.RaftBaseFanSpeed0]: format.percent,
  [SettingsKeys.RaftBaseFanSpeed1]: format.percent,
  [SettingsKeys.RaftBaseTemp0]: format.celsius,
  [SettingsKeys.RaftBaseTemp1]: format.celsius,
  [SettingsKeys.RaftInterfaceSpeed0]: format.millimeterPerSecond,
  [SettingsKeys.RaftInterfaceSpeed1]: format.millimeterPerSecond,
  [SettingsKeys.RaftInterfaceTemp0]: format.celsius,
  [SettingsKeys.RaftInterfaceTemp1]: format.celsius,
  [SettingsKeys.BaseLayerSurfaceFeedRate0]: format.millimeterPerSecond,
  [SettingsKeys.BaseLayerSurfaceFeedRate1]: format.millimeterPerSecond,
  [SettingsKeys.BaseLayerSurfaceFanSpeed0]: format.percent,
  [SettingsKeys.BaseLayerSurfaceFanSpeed1]: format.percent,
  [SettingsKeys.BaseLayerSurfaceTemp0]: format.celsius,
  [SettingsKeys.BaseLayerSurfaceTemp1]: format.celsius,
  [SettingsKeys.RaftModelShellsSpacing]: format.millimeter,
  [SettingsKeys.RaftFeedRate]: format.millimeterPerSecond,
  [SettingsKeys.RaftFanSpeed]: format.percent,
  [SettingsKeys.RaftSurfaceSpeed0]: format.millimeterPerSecond,
  [SettingsKeys.RaftSurfaceSpeed1]: format.millimeterPerSecond,
  [SettingsKeys.RaftSurfaceTemp0]: format.celsius,
  [SettingsKeys.RaftSurfaceTemp1]: format.celsius,
  [SettingsKeys.BaseLayerSurfaceExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.PaddedBaseOutlineOffset]: format.millimeter,
  [SettingsKeys.PaddedBase]: format.noop,
  [SettingsKeys.PaddedBaseWidth]: format.millimeter,
  [SettingsKeys.RaftBrimsSpacing]: format.millimeter,
  [SettingsKeys.RaftInterfaceFirstSurfaceZOffset]: format.millimeter,
  [SettingsKeys.RaftBaseFirstSurfaceZOffset]: format.millimeter,
  [SettingsKeys.RaftBaseLayerThickness]: format.millimeter,
  [SettingsKeys.RaftBaseLayerWidth]: format.millimeter,
  [SettingsKeys.RaftBaseNumberOfLayers]: format.identity,
  [SettingsKeys.RaftBaseZOffset]: format.millimeter,
  [SettingsKeys.RaftBaseExtruderID]: format.identity,
  [SettingsKeys.RaftInterfaceNumberOfLayers]: format.identity,
  [SettingsKeys.RaftInterfaceOutset]: format.millimeter,
  [SettingsKeys.RaftInterfaceZOffset]: format.millimeter,
  [SettingsKeys.RaftInterfaceLayerThickness]: format.millimeter,
  [SettingsKeys.RaftInterfaceLayerWidth]: format.millimeter,
  [SettingsKeys.RaftInterfaceExtruderID]: format.identity,
  [SettingsKeys.RaftSurfaceExtruderID]: format.identity,
  [SettingsKeys.RaftSurfaceFirstSurfaceZOffset]: format.millimeter,
  [SettingsKeys.RaftSurfaceLayerThickness]: format.millimeter,
  [SettingsKeys.RaftSurfaceLayerWidth]: format.millimeter,
  [SettingsKeys.RaftSurfaceNumberOfLayers]: format.identity,
  [SettingsKeys.RaftSurfaceOutset]: format.millimeter,
  [SettingsKeys.RaftSurfaceZOffset]: format.millimeter,
  [SettingsKeys.XScale]: format.identity,
  [SettingsKeys.YScale]: format.identity,
  [SettingsKeys.ZScale]: format.identity,
  [SettingsKeys.SolidExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.SparseExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.FloorSurfaceExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.ConvertSparsePulseValue]: format.millimeter,
  [SettingsKeys.RoofSurfaceExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.InsetsExtrusionVolumeMultiplier]: format.percent,
  [SettingsKeys.OutlinesExtrusionVolumeMultiplier]: format.percent,
  // Brims
  [SettingsKeys.Brims]: format.noop,
  [SettingsKeys.NumberOfBrims]: format.identity,
  [SettingsKeys.NumberOfInternalBrims]: format.identity,
  [SettingsKeys.BrimsBaseWidth]: format.millimeter,
  [SettingsKeys.BrimsOverlapWidth]: format.millimeter,
  [SettingsKeys.BrimsModelOffset]: format.millimeter,
  [SettingsKeys.BrimsFeedRate]: format.millimeterPerSecond,
  [SettingsKeys.BrimsFanSpeed]: format.millimeterPerSecond,
  [SettingsKeys.BrimsExtrusionVolumeMultiplier]: format.percent
});
/**
 * Transforms a min/max/step relationship into an array of strings, intended to be used for
 * dropdowns and the like.
 * @callback optionFunc
 * @param {number} min - minimum value
 * @param {number} max - maximum value
 * @param {number} step - steps between values
 * @returns {number[]} all the numbers between min and max that are step apart
 */

/** @type {Immutable.Map<string, optionFunc>} */

const Options = Object.freeze({
  [SettingsKeys.NumShells]: (min, max, step) => rangeStep(step, min, 10)
});
/**
 * Use for transforming values into acceptable values for the settings
 * @callback sanitizeFunc
 * @param {any} v - value of the setting
 */

/** @type {Immutable.Map<string, sanitizeFunc}
 * As of right now, Sanitization only has one type of lamda
 * function assigned to multiple settings.
 **/

const Sanitization = Object.freeze([[SettingsKeys.ExtrusionVolumeMultiplier], [SettingsKeys.DefaultFanSpeed], [SettingsKeys.RoofSurfaceFanSpeed0], [SettingsKeys.RoofSurfaceFanSpeed1], [SettingsKeys.RoofSurfaceFillsFanSpeed], [SettingsKeys.SolidFanSpeed0], [SettingsKeys.SolidFanSpeed1], [SettingsKeys.SparseRoofSurfaceFanSpeed], [SettingsKeys.OutlineFanSpeed0], [SettingsKeys.OutlineFanSpeed1], [SettingsKeys.InsetsFanSpeed0], [SettingsKeys.InsetsFanSpeed1], [SettingsKeys.InfillDensity], [SettingsKeys.InfillFanSpeed], [SettingsKeys.SparseFanSpeed0], [SettingsKeys.SparseFanSpeed1], [SettingsKeys.SpursFanSpeed], [SettingsKeys.FloorFanSpeed], [SettingsKeys.FloorSurfaceFanSpeed0], [SettingsKeys.FloorSurfaceFanSpeed1], [SettingsKeys.SupportDensity], [SettingsKeys.BridgeFanSpeed], [SettingsKeys.RaftBaseFanSpeed0], [SettingsKeys.RaftBaseFanSpeed1], [SettingsKeys.BaseLayerSurfaceFanSpeed0], [SettingsKeys.BaseLayerSurfaceFanSpeed1], [SettingsKeys.RaftFanSpeed], [SettingsKeys.BaseLayerSurfaceExtrusionVolumeMultiplier], [SettingsKeys.BrimsExtrusionVolumeMultiplier], [SettingsKeys.ExtrusionVolumeMultiplierLABS], [SettingsKeys.SolidExtrusionVolumeMultiplier], [SettingsKeys.SparseExtrusionVolumeMultiplier], [SettingsKeys.FloorSurfaceExtrusionVolumeMultiplier], [SettingsKeys.RoofSurfaceExtrusionVolumeMultiplier], [SettingsKeys.InsetsExtrusionVolumeMultiplier], [SettingsKeys.OutlinesExtrusionVolumeMultiplier]].reduce((obj, setting) => {
  obj[setting] = v => v / 100;

  return obj;
}, {})); //  ************************************************************************
//  Leave keys out of `overrides` object before passing to Slice Config (?)
//  ************************************************************************

const PresetKeyBlacklist = [SettingsKeys.Extruders, SettingsKeys.Materials, 'applyToAll', 'rotationLock', SettingsKeys.PrintModes];
const PRESET_EXT = 'printmode'; //
// Experimental Extruder Only.
//

const EESettingsKeys = [SettingsKeys.NozzleDiameter, SettingsKeys.ExtruderRetractRate0, SettingsKeys.ExtruderRetractRate1, SettingsKeys.ExtruderRestartExtraDistance0, SettingsKeys.ExtruderRestartExtraDistance1, SettingsKeys.ExtruderRestartExtraRate, SettingsKeys.ExtrusionVolumeMultiplier, SettingsKeys.DefaultFanSpeed, SettingsKeys.FanLayer, SettingsKeys.RaftBaseTemp0, SettingsKeys.RaftBaseTemp1, SettingsKeys.RaftInterfaceSpeed0, SettingsKeys.RaftInterfaceSpeed1, SettingsKeys.RaftInterfaceTemp0, SettingsKeys.RaftInterfaceTemp1, SettingsKeys.RoofSurfaceFillsFanSpeed, SettingsKeys.RoofSurfaceFillsFeedRate, SettingsKeys.SparseRoofSurfaceFanSpeed, SettingsKeys.SparseRoofSurfaceFeedRate, SettingsKeys.InfillFanSpeed, SettingsKeys.SpursFanSpeed, SettingsKeys.SpursFeedRate, SettingsKeys.FloorFanSpeed, SettingsKeys.FloorFeedRate, SettingsKeys.BridgeFanSpeed, SettingsKeys.BridgeFeedRate, SettingsKeys.RaftFeedRate, SettingsKeys.RaftFanSpeed, SettingsKeys.RaftSurfaceSpeed0, SettingsKeys.RaftSurfaceSpeed1, SettingsKeys.RaftSurfaceTemp0, SettingsKeys.RaftSurfaceTemp1, SettingsKeys.RestartRate0, SettingsKeys.RestartRate1, SettingsKeys.ToolchangeRestartDistance0, SettingsKeys.ToolchangeRestartDistance1, SettingsKeys.ToolchangeRestartRate0, SettingsKeys.ToolchangeRestartRate1, SettingsKeys.ToolchangeRetractDistance0, SettingsKeys.ToolchangeRetractDistance1, SettingsKeys.ToolchangeRetractRate0, SettingsKeys.ToolchangeRetractRate1, SettingsKeys.BrimsBaseWidth, // SettingsKeys.BrimsModelOffset,
SettingsKeys.BrimsOverlapWidth, SettingsKeys.Brims, SettingsKeys.PaddedBase, SettingsKeys.BaseLayerSurfaceExtrusionVolumeMultiplier, SettingsKeys.BrimsExtrusionVolumeMultiplier, // SettingsKeys.NumberOfBrims,
// SettingsKeys.NumberOfInternalBrims,
SettingsKeys.PaddedBaseOutlineOffset, SettingsKeys.PaddedBaseWidth, SettingsKeys.RaftBrimsSpacing, SettingsKeys.RaftBaseExtruderID, SettingsKeys.RaftBaseFirstSurfaceZOffset, SettingsKeys.RaftBaseLayerThickness, SettingsKeys.RaftBaseLayerWidth, SettingsKeys.RaftBaseNumberOfLayers, SettingsKeys.RaftBaseOutsetFromModelEdge, SettingsKeys.RaftBaseZOffset, SettingsKeys.RaftInterfaceExtruderID, SettingsKeys.RaftInterfaceFirstSurfaceZOffset, SettingsKeys.RaftInterfaceLayerThickness, SettingsKeys.RaftInterfaceLayerWidth, SettingsKeys.RaftInterfaceNumberOfLayers, SettingsKeys.RaftInterfaceOutset, SettingsKeys.RaftInterfaceZOffset, SettingsKeys.RaftSurfaceExtruderID, SettingsKeys.RaftSurfaceFirstSurfaceZOffset, SettingsKeys.RaftSurfaceLayerThickness, SettingsKeys.RaftSurfaceLayerWidth, SettingsKeys.RaftSurfaceNumberOfLayers, SettingsKeys.RaftSurfaceOutset, SettingsKeys.RaftSurfaceZOffset, SettingsKeys.XScale, SettingsKeys.YScale, SettingsKeys.ZScale, SettingsKeys.SolidExtrusionVolumeMultiplier, SettingsKeys.SparseExtrusionVolumeMultiplier, SettingsKeys.ExtrusionVolumeMultiplierLABS, SettingsKeys.FloorSurfaceExtrusionVolumeMultiplier, SettingsKeys.ConvertSparsePulseValue, SettingsKeys.RoofSurfaceExtrusionVolumeMultiplier, SettingsKeys.InsetsExtrusionVolumeMultiplier, SettingsKeys.OutlinesExtrusionVolumeMultiplier];
module.exports = {
  Actions,
  DisplayKeys,
  SettingsKeys,
  PrintModes,
  GroupKeys,
  GroupDict,
  Types,
  SettingsTypes,
  ValidatorSettingsTypes,
  Format,
  Options,
  Sanitization,
  PresetKeyBlacklist,
  PRESET_EXT,
  EESettingsKeys
};
