// MB-support-plugin/lib/constants/printerDetails.js
'use strict';

const Immutable = require('immutable');

const Dispatch = Object.freeze({
  AcknowledgeCompleted: 'MB_PRINTER_DETAILS_ACKNOWLEDGE_COMPLETED',
  AcknowledgeError: 'MB_PRINTER_DETAILS_ACKNOWLEDGE_ERROR',
  AcknowledgeResumeError: 'MB_PRINTER_DETAILS_ACKNOWLEDGE_RESUME_ERROR',
  AcknowledgeFailure: 'MB_PRINTER_DETAILS_ACKNOWLEDGE_FAILURE',
  AcknowledgeSuccess: 'MB_PRINTER_DETAILS_ACKNOWLEDGE_SUCCESS',
  Suspend: 'MB_PRINTER_DETAILS_SUSPEND',
  Resume: 'MB_PRINTER_DETAILS_RESUME',
  Cancel: 'MB_PRINTER_DETAILS_CANCEL',
  StopFilament: 'MB_PRINTER_DETAILS_STOP_FILAMENT',
  PrintAgain: 'MB_PRINTER_DETAILS_PRINT_AGAIN',
  ShowPanelDetails: 'MB_PRINTER_DETAILS_SHOW_PANEL_DETAILS',
  ShowPanelFilament: 'MB_PRINTER_DETAILS_SHOW_PANEL_FILAMENT',
  ShowPanelUtilities: 'MB_PRINTER_DETAILS_SHOW_PANEL_UTILITIES',
  System: 'MB_PRINTER_DETAILS_SYSTEM_NOTIFICATION',
  State: 'MB_PRINTER_DETAILS_STATE_NOTIFICATION',
  Status: 'MB_PRINTER_DETAILS_STATUS_NOTIFICATION',
  ButtonWait: 'MB_PRINTER_DETAILS_BUTTON_WAIT',
  ExtruderChange: 'MB_PRINTER_DETAILS_EXTRUDER_CHANGE',
  NetworkChange: 'MB_PRINTER_DETAILS_NETWORK_CHANGE'
});
const PanelStates = Object.freeze({
  Details: 'Details',
  // The Default
  Utilities: 'Utilities',
  Filament: 'Filament'
}); // NOTE: the `status` in `enabled` and `shouldShow` is referring to the object
// returned by `printerDetailsStatus` in mb-plugin/lib/utils/printerDetails.js
//
// - `enabled` refers to whether the button can be clicked or not
// - `shouldShow` refers to whether the button is rendered or not
// - `default` will show if no other button is shown for the given printer state

const ActionButtons = Immutable.Map({
  pause: {
    label: i18n => i18n.t('printer_details.pause'),
    icon: () => 'icons/pause.svg',
    order: 0,
    enabled: status => status.can.suspend,
    shouldShow: status => status.can.suspend || !status.has.process,
    onClick: (actions, id) => () => actions.resume(id),
    default: true
  },
  continue: {
    label: i18n => i18n.t('mb:printerStatus.state.continue'),
    icon: () => 'icons/done_button.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => status.can.acknowledgeError,
    onClick: (actions, id) => () => actions.acknowledgeResumeError(id)
  },
  acknowledgeError: {
    label: i18n => i18n.t('mb:errors.button_text.OK'),
    icon: () => 'icons/done_button.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => status.has.acknowledgableError && !status.can.acknowledgeFailure,
    onClick: (actions, id) => () => actions.acknowledgeError(id)
  },
  acknowledgeFailure: {
    label: i18n => i18n.t('mb:errors.button_text.OK'),
    icon: () => 'icons/done_button.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => status.can.acknowledgeFailure && status.has.extruder,
    onClick: (actions, id) => () => actions.acknowledgeFailure(id)
  },
  acknowledgeSuccess: {
    label: i18n => i18n.t('mb:errors.button_text.OK'),
    icon: () => 'icons/done_button.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => status.can.acknowledgeSuccess,
    onClick: (actions, id) => () => actions.acknowledgeSuccess(id)
  },
  stopFilament: {
    label: i18n => i18n.t('mb:errors.button_text.OK'),
    icon: () => 'icons/done_button.svg',
    order: 0,
    enabled: status => !status.is.pending && !status.step.unloadingFilament,
    shouldShow: status => status.can.stopFilament,
    onClick: (actions, id) => () => actions.stopFilament(id)
  },
  printAgain: {
    label: i18n => i18n.t('mb:printer_details.print_again'),
    icon: () => 'icons/reprint_button.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => (status.step.failed || status.step.completed) && status.can.printAgain && !status.can.acknowledgeFailure,
    onClick: (actions, id) => () => actions.printAgain(id)
  },
  resume: {
    label: i18n => i18n.t('printer_details.resume'),
    icon: () => 'icons/resume-start.svg',
    order: 0,
    enabled: status => !status.is.pending && !status.step.suspending,
    shouldShow: status => status.can.resume && status.has.extruder,
    onClick: (actions, id) => () => actions.resume(id)
  },
  attachExtruder: {
    label: i18n => i18n.t('mb:printer_details.attach_extruder'),
    icon: () => 'icons/attach_extruder.svg',
    order: 0,
    enabled: status => !status.is.pending,
    shouldShow: status => !status.has.extruder && !status.has.dualExtruders && !status.has.error,
    onClick: (actions, id) => () => actions.attachExtruder(id)
  },
  cancel: {
    label: i18n => i18n.t('printer_details.cancel'),
    icon: () => 'icons/cancel-print.svg',
    order: 1,
    enabled: status => status.can.cancel,
    shouldShow: status => !status.step.completed,
    onClick: (actions, id) => () => actions.cancel(id),
    default: true
  },
  done: {
    label: i18n => i18n.t('mb:printer_details.done'),
    icon: () => 'icons/done_button.svg',
    order: 1,
    enabled: status => !status.is.pending,
    shouldShow: status => status.step.completed,
    onClick: (actions, id) => () => actions.done(id)
  },
  filament: {
    label: i18n => i18n.t('mb:filament.button'),
    icon: status => {
      return status.has.dualExtruders ? 'icons/filament-dual.svg' : 'icons/filament.svg';
    },
    order: 2,
    enabled: status => !!(status.is.authenticated && status.has.extruder && (status.is.idle || status.can.loadFilament || status.can.unloadFilament)),
    shouldShow: () => true,
    onClick: actions => () => actions.filament(),
    default: true
  },
  utilities: {
    label: i18n => i18n.t('printer_details.utilities'),
    icon: () => 'icons/utilities.svg',
    order: 3,
    enabled: status => !!(status.is.authenticated && status.has.state),
    shouldShow: () => true,
    onClick: actions => () => actions.utilities()
  }
});

const ProgressButtons = id => Immutable.Map({
  clear_build_plate: (i18n, actions) => ({
    label: i18n.t('mb:start_print'),
    primary: true,
    onClick: () => actions.acknowledgeBuildPlateCleared(id)
  }),
  clear_filament: (i18n, actions) => ({
    label: i18n.t('mb:errors.button_text.unload_filament'),
    onClick: () => actions.clearFilament(id)
  }),
  remove_filament: (i18n, actions) => ({
    label: i18n.t('mb:errors.button_text.runZCal'),
    onClick: () => actions.zCalFilament(id)
  })
});

module.exports = {
  Dispatch,
  PanelStates,
  ActionButtons,
  ProgressButtons
};
