// MB-support-plugin/lib/constants/urls.js
'use strict';

const url = Object.freeze({
  error: {
    1038: 'https://support.makerbot.com/troubleshooting/makerbot-replicatorplus/error-codes/error-1038_12288',
    54: 'https://support.makerbot.com/troubleshooting/makerbot-replicatorplus/error-codes/error-54_12283',
    16: 'https://support.makerbot.com/troubleshooting/makerbot-replicator/error-codes/error-16_12238',
    1014: 'https://support.makerbot.com/troubleshooting/makerbot-replicatorplus/error-codes/error-1014_12286',
    65: 'https://support.makerbot.com/troubleshooting/makerbot-replicatorplus/error-codes/error-65_12285'
  },
  zCal: 'https://support.makerbot.com/learn/makerbot-print-software/using-makerbot-print/z-calibration-and-z-axis-offset_13621',
  zOffset: 'https://support.makerbot.com/learn/makerbot-print-software/using-makerbot-print/z-calibration-and-z-axis-offset_13621',
  defaultPrintModes: 'https://support.makerbot.com/learn/13638',
  customPrintModes: 'https://support.makerbot.com/learn/13640',
  arrangeBuildPlate: 'https://support.makerbot.com/learn/makerbot-print-software/using-makerbot-print/arranging-objects-on-the-build-plate_13590'
});
module.exports = {
  url
};
