// MB-support-plugin/lib/ui/views/utilities/calibration.js
'use strict';

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const React = require('react');

const FluxComponent = require('eagle-print/views/flux_component');

const _ = require('lodash');

const path = require('path');

const UtilityTabSet = require('./utilityTabSet');

const UtilityTab = require('./utilityTab');

const ZCal = require('./zcal');

const ZOffset = require('./zOffset');

const PrinterStateEnum = require('../../../constants').PrinterStateEnum;

const PanelState = Object.freeze({
  Default: 'default',
  ZCal: 'zcal',
  ZOffset: 'zoffset'
});

class Calibration extends FluxComponent {
  constructor(props, context) {
    super(props, context);

    _.assign(this.state, {
      canCalibrate: false
    });
  }

  componentDidMount() {
    this.props.printer.invoke('HasZCalibrationRoutine').then(res => {
      this.setState({
        canCalibrate: res === true
      });
    });
  }

  _renderZCalComponent(i18n, selectedPrinter) {
    return /*#__PURE__*/React.createElement(UtilityTab, _extends({}, this.props, {
      tabKey: PanelState.ZCal,
      tabName: i18n.t('mb:utility.category.calibration.zcal'),
      tabIcon: path.resolve(__dirname, '../images/calibration-utilities-manual.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Unauthenticated || !this.state.canCalibrate
    }), /*#__PURE__*/React.createElement(ZCal, _extends({}, this.props, {
      printer: selectedPrinter,
      onClose: this.props.onClose,
      onBack: () => this.panel.back(),
      onFinish: this.props.onFinish,
      flux: this.props.flux
    })));
  }

  _renderZOffset(i18n, selectedPrinter) {
    const status = selectedPrinter.getStatus() || {};
    return /*#__PURE__*/React.createElement(UtilityTab, _extends({}, this.props, {
      tabKey: PanelState.ZOffset,
      tabName: i18n.t('mb:utility.category.calibration.zoffset'),
      tabIcon: path.resolve(__dirname, '../images/calibration-utilities.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Unauthenticated
    }), /*#__PURE__*/React.createElement(ZOffset, {
      printer: selectedPrinter,
      onClose: this.props.onClose,
      onBack: () => this.panel.back(),
      onFinish: this.props.onFinish,
      flux: this.props.flux
    }));
  }

  render() {
    const selectedPrinter = this.props.printer;
    const i18n = this.props.flux.getService('TranslationService');
    return /*#__PURE__*/React.createElement(UtilityTabSet, _extends({}, this.props, {
      title: i18n.t('mb:utility.category.calibration.header'),
      ref: c => this.panel = c,
      flux: this.props.flux,
      printer: this.props.printer,
      onBack: this.props.onBack
    }), /*#__PURE__*/React.createElement("div", {
      tabKey: "default"
    }), this._renderZCalComponent(i18n, selectedPrinter), this._renderZOffset(i18n, selectedPrinter));
  }

}

Calibration.displayName = 'Calibration';
Calibration.propTypes = {
  onBack: React.PropTypes.func,
  onClose: React.PropTypes.func,
  onFinish: React.PropTypes.func,
  printer: React.PropTypes.object.isRequired,
  flux: React.PropTypes.object.isRequired
};
Calibration.defaultProps = {
  onBack: () => {},
  onClose: () => {},
  onFinish: () => {}
};
module.exports = Calibration;
