// MB-support-plugin/lib/ui/views/utilities/copyBotLogs.js
'use strict';

const React = require('react');

const ce = React.createElement;
const {
  DOM: {
    div,
    strong
  }
} = React;

const R = require('ramda');

const FluxComponent = require('eagle-print/views/flux_component');

const PanelHeader = require('./panelHeader');

const {
  CHANGE_EVENT
} = require('../../../constants/copyBotLogs');

const {
  PrinterStateEnum
} = require('../../../constants');

const {
  pSplit
} = require('../../../util');

const WaitableButton = require('../WaitableButton');

class CopyBotLogs extends FluxComponent {
  constructor(props) {
    super(props);
    this._i18n = this.props.flux.getService('TranslationService');
    this._trackingService = this.props.flux.getService('TrackingService');
    this._loggingService = this.props.flux.getService('LoggingService');
    this._store = this.props.flux.store('CopyBotLogs');
    this._actions = this.props.flux.actions.CopyBotLogs;
    this.onChange = this.onChange.bind(this);
    this.state = this._store.getState(this.props.printer.getId());
  }

  componentDidMount() {
    this._store.addListener(CHANGE_EVENT, this.onChange);
  }

  componentWillUnmount() {
    this._store.removeListener(CHANGE_EVENT, this.onChange);
  }

  onChange() {
    this.setState(this._store.getState(this.props.printer.getId()));
  }

  renderButtons() {
    const {
      printer
    } = this.props;
    const id = printer.getId();
    const {
      tmpPath
    } = this.state;
    const disabled = printer.getStatus().state === PrinterStateEnum.Busy || printer.getStatus().state === PrinterStateEnum.Unauthenticated;
    if (this.state.error) return div({}, div({}, strong({}, this._i18n.t('mb:copyBotLogs.error'))), div({
      style: {
        marginTop: 10
      }
    }, ce(WaitableButton, {
      waiting: false,
      buttonProps: {
        disabled,
        type: 'secondary',
        onClick: () => this._actions.copyLogs(id),
        label: this._i18n.t('mb:copyBotLogs.tryAgain')
      }
    })));
    if (this.state.isCopying) return div({}, ce(WaitableButton, {
      waiting: true,
      buttonProps: {
        type: 'secondary',
        label: this._i18n.t('mb:copyBotLogs.copying'),
        disabled: true
      }
    }));
    if (tmpPath) return div({}, div({}, strong({}, this._i18n.t('mb:copyBotLogs.copied'))), div({
      style: {
        marginTop: 10
      }
    }, ce(WaitableButton, {
      waiting: false,
      buttonProps: {
        type: 'secondary',
        label: this._i18n.t('mb:copyBotLogs.send'),
        onClick: () => this._actions.sendToSupport(id)
      }
    })), div({
      style: {
        marginTop: 10
      }
    }, ce(WaitableButton, {
      waiting: false,
      buttonProps: {
        type: 'secondary',
        label: this._i18n.t('mb:copyBotLogs.save'),
        onClick: () => this._actions.saveLogs(id)
      }
    })));
    return div({}, ce(WaitableButton, {
      waiting: false,
      buttonProps: {
        disabled,
        type: 'secondary',
        label: this._i18n.t('mb:copyBotLogs.copy'),
        onClick: () => this._actions.copyLogs(id)
      }
    }));
  }

  render() {
    return div(R.merge(this.props, {
      id: 'copybotlogs',
      className: `${this.props.className || 'copybotlogs'} printerPanel`,
      style: {
        position: 'relative',
        height: '100%'
      }
    }), // PANEL HEADER
    ce(PanelHeader, {
      title: this._i18n.t('mb:copyBotLogs.title'),
      flux: this.props.flux,
      printer: this.props.printer,
      onBack: this.props.onBack,
      onClose: this.props.onClose
    }), // PANEL BODY
    div({
      className: 'panel-body'
    }, pSplit(this._i18n.t('mb:copyBotLogs.info')), this.renderButtons()), // PANEL FOOTER PILL
    div({
      className: 'printerPanel'
    }, div({
      style: {
        marginBottom: 50
      }
    })));
  }

}

CopyBotLogs.propTypes = {
  flux: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  className: React.PropTypes.string
};
module.exports = CopyBotLogs;
