// MB-support-plugin/lib/ui/views/utilities/firmwareUpdatePanel.js
'use strict';

const React = require('react');

const FluxComponent = require('eagle-print/views/flux_component');

const DOM = React.DOM;
const ce = React.createElement;

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const dialog = require('electron').remote.dialog;

const Button = require('eagle-print/views/ui/button.js');

const firmwareUtils = require('../../../firmwareUtils');

const HeaderSelectButton = require('../headerSelectButton');

class FirmwareUpdatePanel extends FluxComponent {
  /** @constructor **/
  constructor(props, context) {
    super(props, context);
    this.i18n = this.getFlux().getService('TranslationService');
    this.tracking = this.getFlux().getService('TrackingService');
    this.logger = this.getFlux().getService('LoggingService');
    this.message = this.getFlux().actions.message;

    _.assign(this.state, {
      currentVersion: this.i18n.t('mb:firmware.checking'),
      availableVersion: this.i18n.t('mb:firmware.checking'),
      availableDescription: '',
      customFilename: null,
      customManifest: null,
      canUpdate: false,
      browsing: false
    });
  }

  componentDidMount() {
    const fw = this.props.printer.firmware.availableUpdate;
    this.setState({
      currentVersion: this.props.printer.getInfo().firmware_version
    });
    this.props.printer.firmware.listen(this._handleNotif.bind(this));

    if (this.props.printer.firmware.hasUpdate()) {
      this._handleNotif(fw);
    } else {
      this.props.printer.firmware.check();
    }
  }

  componentWillUnmount() {
    this.props.printer.firmware.remove(this._handleNotif.bind(this));
  }
  /**
   * Updates the printer with either the latest version the printer knows about, or with
   * a custom firmware file that the user chooses.
   * @returns {q.Promise}
   */


  handleUpdate() {
    const flux = this.getFlux();
    this.props.onUpdate(); // if the user hasn't chosen a specific firmware file to use, just tell the printer
    // to grab one from the web (or the one it already has)

    if (this.state.customFilename == null) {
      this.tracking.addEvent({
        name: 'firmware_update_initialized'
      });
      return firmwareUtils.updateFirmware(this.props.printer, flux);
    }

    this.tracking.addEvent({
      name: 'firmware_upload_initialized'
    });
    return this.props.printer.firmware.upload(this.state.customFilename).then(() => this.logger.info('uploading firmware')).catch(err => {
      this.message.createErrorMessage({
        id: 'firmware-update-error',
        message: err
      });
    });
  }
  /**
   * Opens a "Choose File" dialog to pick a custom firmware file that
   * can be uploaded to the printer. Those files should be .zip
   **/


  handleBrowse() {
    if (this.state.browsing) {
      return;
    }

    this.setState({
      browsing: true
    });
    dialog.showOpenDialog({}, f => {
      if (f) {
        return this.props.printer.firmware.verify(f[0]).then(manifest => {
          this.setState({
            customFilename: f[0],
            customManifest: manifest,
            canUpdate: true,
            browsing: false
          });
        }).catch(err => {
          this.setState({
            browsing: false
          });
          this.message.createErrorMessage({
            id: 'firmware-update-error',
            message: err
          });
        });
      }

      this.setState({
        browsing: false
      });
    });
  }
  /**
   * @private
   * If the provided firmware is a newer version than the current firmware, update the
   * component state and allow the user to initiate the update.
   * @param {Object} firmware - the most recent version of the firmware we found
   */


  _handleNotif(firmware) {
    if (this.props.printer.firmware.hasUpdate()) {
      this.setState({
        availableVersion: firmware.manifest.version.join('.'),
        availableDescription: firmware.manifest.description,
        canUpdate: true
      });
    } else {
      this.setState({
        availableVersion: this.i18n.t('mb:firmware.none'),
        availableDescription: this.i18n.t('mb:firmware.none'),
        canUpdate: false
      });
    }
  }

  _parseCustomFirmwareName() {
    if (this.state.customManifest) {
      const product = this.state.customManifest.product;
      const version = this.state.customManifest.version.join('.');
      return `${product} ${version}`;
    }

    return null;
  }

  renderHeader() {
    return DOM.div({
      className: 'printerPanelHeader'
    }, DOM.div({
      className: 'printerPanelBackButton',
      onClick: this.props.onBack
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })), DOM.div({
      className: 'printerPanelHeaderLabel'
    }, DOM.h1({}, this.i18n.t('mb:firmware.header'))), ce(HeaderSelectButton, {
      flux: this.getFlux(),
      printer: this.props.printer
    }), DOM.div({
      className: 'printerPanelArrowDown',
      onClick: this.props.onClose
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })));
  }

  renderCurrent() {
    return DOM.div({}, DOM.label({}, this.i18n.t('mb:firmware.current')), DOM.p({}, this.state.currentVersion));
  }

  renderUpdate() {
    const version = this._parseCustomFirmwareName() || this.state.availableVersion;
    return DOM.div({}, DOM.label({}, this.i18n.t('mb:firmware.update_to')), DOM.div({
      style: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginRight: '10px'
      }
    }, DOM.input({
      readOnly: true,
      value: version,
      className: 'input-text'
    }), ce(Button, {
      style: {
        margin: '0 12px',
        padding: '8px 12px'
      },
      type: `secondary ${!this.state.canUpdate ? 'disabled' : ''}`,
      disabled: !this.state.canUpdate,
      onClick: () => this.handleUpdate()
    }, DOM.span({}, this.i18n.t('mb:firmware.update')))), DOM.a({
      style: {
        color: 'red',
        margin: '10px 0',
        display: 'block'
      },
      onClick: () => this.handleBrowse()
    }, this.i18n.t('mb:firmware.browse')));
  }

  renderDescription() {
    return DOM.div({}, DOM.label({}, this.i18n.t('mb:firmware.description')), DOM.p({}, this.state.customManifest != null ? this.state.customManifest.description : this.state.availableDescription));
  }

  render() {
    return DOM.div(_.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    }), this.renderHeader(), DOM.div({
      className: 'printerPanel'
    }, DOM.div({
      style: {
        marginBottom: 50
      }
    }, DOM.div({
      className: 'panel-body'
    }, this.renderCurrent(), this.renderUpdate(), this.renderDescription()))));
  }

}

FirmwareUpdatePanel.propTypes = _.assign({}, FluxComponent.propTypes, {
  onClose: React.PropTypes.func,
  onBack: React.PropTypes.func,
  printer: React.PropTypes.object.isRequired
});
FirmwareUpdatePanel.defaultProps = {
  onClose: () => {},
  onBack: () => {},
  onUpdate: () => {}
};
module.exports = FirmwareUpdatePanel;
