// MB-support-plugin/lib/ui/views/utilities/networkPreferences.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const path = require('path');

const classNames = require('classnames');

const Checkbox = require('eagle-print/views/ui/input_checkbox');

const Dropdown = require('eagle-print/views/ui/input_dropdown');

const FluxComponent = require('eagle-print/views/flux_component');

const {
  getPropLift
} = require('../../../utils/maybe');

const WaitableButton = require('../WaitableButton');

const HeaderSelectButton = require('../headerSelectButton');

const StaticIP = require('./staticIP');

const {
  CHANGE_EVENT,
  NetworkStateEnum
} = require('../../../constants/networkPreferences');

class NetworkPreferences extends FluxComponent {
  constructor(props) {
    super(props);
    this._i18n = this.props.flux.getService('TranslationService');
    this._trackingService = this.props.flux.getService('TrackingService');
    this._loggingService = this.props.flux.getService('LoggingService');
    this._store = this.props.flux.store('NetworkPreferencesStore');
    this._actions = this.props.flux.actions.NetworkPreferencesActions;
    this.onChange = this.onChange.bind(this);
    this.handleNetworkChange = this.handleNetworkChange.bind(this);
    this.state = this._store.getState();
  }

  componentDidMount() {
    this.handleNetworkChange();
    this.props.printer.updateNotifications.addListener('network_state_change', this.handleNetworkChange);
    this.props.printer.updateNotifications.addListener('conn_type_change', this.handleNetworkChange);

    this._store.addListener(CHANGE_EVENT, this.onChange);

    this._actions.clearError();
  }

  componentWillUnmount() {
    this.props.printer.updateNotifications.removeListener('network_state_change', this.handleNetworkChange);
    this.props.printer.updateNotifications.removeListener('conn_type_change', this.handleNetworkChange);

    this._store.removeListener(CHANGE_EVENT, this.onChange);
  }

  onChange() {
    this.setState(this._store.getState());
  }

  handleNetworkChange() {
    this._actions.networkChange(this.props.printer.getId());
  }

  handleDisconnect() {
    this._actions.disconnectFromNetwork(this.props.printer.getId(), this.state.currentPath);
  }

  handleDropdownChange({
    value: path
  }) {
    this._actions.selectNetwork(path);

    this._actions.prepopulateStaticIP(this.props.printer.getId(), path);
  }

  handlePasswordChange(event) {
    this._actions.changePasswordAttempt(event.target.value);
  }

  handleHidePasswordChange() {
    this._actions.togglePasswordVisibility();
  }

  toggleWifi() {
    const id = this.props.printer.getId();
    if (this.state.enableWifi) this._actions.disableWifi(id);else this._actions.enableWifi(id);
  }

  _shouldDisableView() {
    return !this.state.canEditWifi(this.props.printer);
  }

  _currentNetwork() {
    if (this.state.selectedNetwork.lift()) return this.state.selectedNetwork;
    return this.state.networkState;
  }

  _currentNetworkPath() {
    const connectedByEthernet = getPropLift('state', this.state.networkState) === NetworkStateEnum.state.ETHERNET;
    if (connectedByEthernet) return 'ethernet';
    return getPath(this.state.selectedNetwork) || getPath(this.state.networkState);
  }

  renderMessageView() {
    return DOM.div({
      className: 'network-preferences__status__message message'
    }, DOM.p({
      style: {
        lineHeight: 1.7,
        margin: '7px 0'
      }
    }, this.state.currentMessage.map(this._i18n.t).lift()));
  }

  renderErrorView() {
    const width = 25;
    const height = 25;
    const viewBox = `0 0 ${width} ${height}`;
    return DOM.div({
      style: {
        margin: '15px 0'
      },
      className: 'network-preferences__status__error error'
    }, this.state.currentError.map(err => {
      const key = `mb:network.errors.${_.camelCase(err)}`;

      const fallback = _.startCase(err);

      return [ce('svg', {
        ref: 'errorIcon',
        className: 'error-icon',
        width,
        height,
        viewBox
      }, ce('image', {
        height,
        width,
        xlinkHref: path.resolve(__dirname, '../images/alert-icon-white.svg')
      })), DOM.span({
        className: 'error-text'
      }, this._i18n.exists(key) ? this._i18n.t(key) : fallback)];
    }).lift());
  }

  renderStatusView() {
    return DOM.div({
      style: {
        paddingBottom: 0
      },
      className: 'network-preferences__status panel-body'
    }, DOM.div({
      style: {
        fontWeight: 'bold',
        marginBottom: '0px'
      },
      className: 'header'
    }, this._i18n.t('mb:network.header.status'), ': ', this.state.statusMsg.map(this._i18n.t).lift()), this.renderMessageView(), ce(Checkbox, {
      ref: 'wifiCheckbox',
      checked: this.state.enableWifi,
      label: this._i18n.t('mb:network.checkbox.enableWifi'),
      classNames: 'wifi-checkbox',
      labelClassNames: 'status-checkbox-label',
      onChange: () => this.toggleWifi(),
      disabled: this._shouldDisableView()
    }), this.renderErrorView());
  }

  renderRefreshButton() {
    const id = this.props.printer.getId();
    return ce(WaitableButton, {
      waiting: this.state.waiting,
      buttonProps: {
        style: {
          margin: '0 15px',
          padding: '8px 12px'
        },
        type: 'secondary',
        className: 'refresh-button',
        label: this._i18n.t('mb:network.button.refresh'),
        onClick: () => this._actions.networkChange(id)
      }
    });
  }

  renderDisconnectButton() {
    return ce(WaitableButton, {
      waiting: this.state.waiting,
      buttonProps: {
        style: {
          margin: '0 15px',
          padding: '8px 12px'
        },
        type: 'secondary',
        className: 'disconnect-button',
        label: this._i18n.t('mb:network.button.disconnect'),
        onClick: () => this.handleDisconnect()
      }
    });
  }

  renderNetworksView() {
    const networkIsSelected = this.state.selectedNetwork.map(() => true).lift() || false;
    const networkViewBodyClasses = classNames('network-preferences__networks__body', {
      'view-disabled': this._shouldDisableView()
    });
    const placeholder = this.state.availableNetworks.length > 0 ? getPropLift('name', this.state.networkState) || this._i18n.t('mb:network.dropdown.chooseANetwork') : 'No Networks Available';
    const options = this.state.availableNetworks.reduce((acc, v) => acc.concat({
      label: v.name,
      value: v.path
    }), [{
      label: placeholder,
      value: ''
    }]);
    const value = getPath(this.state.selectedNetwork);
    return DOM.div({
      className: 'network-preferences__networks'
    }, DOM.div({
      style: {
        paddingBottom: 0
      },
      className: 'panel-body'
    }, DOM.div({
      style: {
        fontWeight: 'bold',
        marginTop: '0px'
      },
      className: 'header'
    }, this._i18n.t('mb:network.header.network'), ': '), DOM.div({
      style: {
        display: 'flex',
        margin: '5px 0'
      },
      className: networkViewBodyClasses
    }, ce(Dropdown, {
      ref: 'networksDropdown',
      style: {
        flex: 2,
        width: '100%'
      },
      options,
      onChange: event => this.handleDropdownChange(event),
      disableExpandOutOfPanel: true,
      placeholder,
      value
    }), getPropLift('state', this.state.networkState) === 'wifi' ? this.renderDisconnectButton() : this.renderRefreshButton())), networkIsSelected ? this.renderConnectView() : null);
  }

  renderConnectView() {
    const connectViewBodyClasses = classNames('network-preferences__networks__connect form-group auth-input panel-body', {
      'view-disabled': this._shouldDisableView()
    });
    const pwRequired = this.state.selectedNetwork.map(network => network.password === 'required' && getPropLift('state', this.state.networkState) !== 'wifi').lift();
    return DOM.div({
      className: connectViewBodyClasses
    }, pwRequired ? this.renderPasswordInput() : null, ce(WaitableButton, {
      waiting: this.state.waiting,
      buttonProps: {
        style: {
          margin: '12px 0',
          padding: '8px 12px'
        },
        type: 'primary',
        className: 'network-button',
        label: this._i18n.t('mb:network.button.connect'),
        onClick: () => this.handleConnectButtonClicked()
      }
    }));
  }

  handleConnectButtonClicked() {
    const id = this.props.printer.getId();

    this._actions.connectToNetwork(id, this._currentNetworkPath(), this.state.passwordAttempt);
  }

  renderPasswordInput() {
    const passwordInputBodyClasses = classNames('', {
      'view-disabled': this._shouldDisableView()
    });
    return DOM.div({
      className: 'network-preferences__networks__connect__password-input'
    }, DOM.span({
      style: {
        fontWeight: 'bold'
      },
      className: 'header'
    }, this._i18n.t('mb:network.header.wifiPassword'), ': '), DOM.div({
      className: passwordInputBodyClasses
    }, DOM.input({
      ref: 'passwordInput',
      style: {
        margin: '5px 0'
      },
      className: 'input-text network-password-input',
      type: this.state.showPassword ? 'text' : 'password',
      value: this.state.passwordAttempt,
      onChange: event => this.handlePasswordChange(event),
      onKeyDown: event => event.key === 'Enter' ? this.handleConnectButtonClicked() : null
    }), ce(Checkbox, {
      ref: 'hidePasswordCheckbox',
      checked: this.state.showPassword,
      label: this._i18n.t('mb:network.checkbox.showPassword'),
      labelClassNames: 'hide-password-checkbox-label',
      classNames: 'hide-password-checkbox',
      onChange: () => this.handleHidePasswordChange()
    })));
  }

  render() {
    const staticIPprops = {
      flux: this.props.flux,
      printer: this.props.printer,
      waiting: this.state.waiting,
      enableStaticIP: this.state.enableStaticIP,
      staticIPEdited: this.state.staticIPEdited,
      staticIPAddress: this.state.staticIPAddress,
      netmask: this.state.netmask,
      gateway: this.state.gateway,
      dnsServers: this.state.dnsServers,
      editStaticIPAddress: this.state.editStaticIPAddress,
      editNetmask: this.state.editNetmask,
      editGateway: this.state.editGateway,
      newDNSItem: this.state.newDNSItem,
      canEdit: this.state.canEditStaticIP(this.props.printer),
      path: this._currentNetworkPath()
    };
    return /*#__PURE__*/React.createElement("div", {
      id: "network-preferences",
      className: `${this.props.className || 'network-preferences'} printerPanel`,
      style: {
        position: 'relative',
        height: '100%'
      }
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeader"
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelBackButton",
      onClick: this.props.onBack
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeaderLabel"
    }, /*#__PURE__*/React.createElement("h1", null, this._i18n.t('mb:network.header.main'))), /*#__PURE__*/React.createElement(HeaderSelectButton, {
      flux: this.props.flux,
      printer: this.props.printer
    }), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelArrowDown",
      onClick: this.props.onClose
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    }))), /*#__PURE__*/React.createElement("div", {
      style: {
        overflowY: 'auto',
        position: 'absolute',
        top: 60,
        left: 0,
        right: 0,
        bottom: 0
      }
    }, !_.isEmpty(this.state.networkState) ? this.renderStatusView() : null, this.state.enableWifi ? this.renderNetworksView() : null, /*#__PURE__*/React.createElement(StaticIP, staticIPprops)), /*#__PURE__*/React.createElement("div", {
      className: 'printerPanel'
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 50
      }
    })));
  }

}

NetworkPreferences.displayName = 'NetworkPreferences';
NetworkPreferences.propTypes = {
  flux: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  className: React.PropTypes.string
};
const getPath = getPropLift('path');
module.exports = NetworkPreferences;
