// MB-support-plugin/lib/ui/views/utilities/notifications.js
'use strict';

const React = require('react');

const FluxComponent = require('eagle-print/views/flux_component');

const Checkbox = require('eagle-print/views/ui/input_checkbox');

const UtilityTabSet = require('./utilityTabSet');

class Notifications extends FluxComponent {
  constructor(props) {
    super(props);
    this.i18n = this.props.flux.getService('TranslationService');
    this.logger = this.props.flux.getService('LoggingService');
    this.tracking = this.props.flux.getService('TrackingService');
    this.state = {
      detectionEnabled: true,
      soundEnabled: true,
      analyticsEnabled: true
    };
  }

  componentDidMount() {
    const detectionDisabled = this.props.printer.getDisabledErrors().includes('filament_slip');
    const soundEnabled = this.props.printer.soundState;
    const analyticsEnabled = this.props.printer.analyticsState;
    this.setState({
      detectionDisabled,
      soundEnabled,
      analyticsEnabled
    });
  }

  handleEnableDetectionToggle() {
    this.setState({
      detectionDisabled: !this.state.detectionDisabled
    }, () => {
      this.props.printer.invoke('SetToolheadErrorVisibility', ['filament_slip', this.state.detectionDisabled]);
    });
  }

  handleSoundToggle() {
    this.setState({
      soundEnabled: !this.state.soundEnabled
    }, () => {
      this.props.printer.invoke('ToggleSound', [this.state.soundEnabled]);
    });
  }

  handleAnalyticsEnabled() {
    this.setState({
      analyticsEnabled: !this.state.analyticsEnabled
    }, () => {
      this.props.printer.invoke('SetAnalyticsEnabled', [this.state.analyticsEnabled]);
    });
  }

  renderFilamentErrorDetectionOption() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement("label", null, this.i18n.t('mb:notifications.filamentErrorDetection.title')), /*#__PURE__*/React.createElement("p", null, this.i18n.t('mb:notifications.filamentErrorDetection.paragraph'))), /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement(Checkbox, {
      ref: "enableJamDetectionCheckbox",
      label: this.i18n.t('mb:notifications.filamentErrorDetection.checkbox'),
      checked: !this.state.detectionDisabled,
      onChange: () => this.handleEnableDetectionToggle()
    })));
  }

  renderSoundOption() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement("label", null, this.i18n.t('mb:notifications.sound.title')), /*#__PURE__*/React.createElement("p", null, this.i18n.t('mb:notifications.sound.paragraph'))), /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement(Checkbox, {
      ref: "enableSoundCheckbox",
      label: this.i18n.t('mb:notifications.sound.checkbox'),
      checked: this.state.soundEnabled,
      onChange: () => this.handleSoundToggle()
    })));
  }

  renderAnalyticsOption() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement("label", null, this.i18n.t('mb:notifications.analytics.title')), /*#__PURE__*/React.createElement("p", null, this.i18n.t('mb:notifications.analytics.paragraph'))), /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 20
      }
    }, /*#__PURE__*/React.createElement(Checkbox, {
      ref: "enableAnalyticsCheckbox",
      label: this.i18n.t('mb:notifications.analytics.checkbox'),
      checked: this.state.analyticsEnabled,
      onChange: () => this.handleAnalyticsEnabled()
    })));
  }

  render() {
    const isSixthGen = this.props.printer.isSixthGen();
    return /*#__PURE__*/React.createElement(UtilityTabSet, {
      title: this.i18n.t('mb:notifications.header'),
      ref: c => this.panel = c,
      flux: this.props.flux,
      printer: this.props.printer,
      onBack: this.props.onBack,
      onClose: this.props.onClose
    }, /*#__PURE__*/React.createElement("div", {
      className: "panel-body"
    }, !isSixthGen ? this.renderFilamentErrorDetectionOption() : null, !isSixthGen ? this.renderSoundOption() : null, this.renderAnalyticsOption()));
  }

}

module.exports = Notifications;
