// MB-support-plugin/lib/ui/views/utilities/printerInfo.js
'use strict';

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const React = require('react');

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const q = require('q');

const classNames = require('classnames');

const Button = require('eagle-print/views/ui/button');

const util = require('../../../util');

const NetworkStateEnum = require('../../../constants').NetworkStateEnum;

const HeaderSelectButton = require('../headerSelectButton');

const PrinterInfoRow = require('./printerInfoComponents/PrinterInfoRow');

const PrinterInfoTitle = require('./printerInfoComponents/PrinterInfoTitle');

class PrinterInfo extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      newPrinterName: '',
      printerName: this.props.printer.getName(),
      disableUpdate: true,
      network: {},
      extruders: {}
    };
    this._i18n = this.props.flux.getService('TranslationService');
    this._trackingService = this.props.flux.getService('TrackingService');
  }

  componentWillMount() {
    this.getPrinterInfo();
  }

  getPrinterInfo() {
    const networkState = () => {
      if (this.props.printer.isSketchPrinter()) return this.props.printer.getNetworkState();
      return this.props.printer.invoke('NetworkState').then(networkState => {
        const networkInfo = {};

        if (networkState) {
          networkInfo.subnet = networkState.netmask;
          networkInfo.gateway = networkState.gateway;
          networkInfo.dns = networkState.dns;
          networkInfo.state = networkState.state;
          if (networkState.state === NetworkStateEnum.state.WIFI) networkInfo.wifi = networkState.name;else networkInfo.wifi = this._i18n.t('mb:printerInfo.n_a');
        }

        return networkInfo;
      });
    };

    const extruderStats = () => {
      return this.props.printer.getAllExtruderInfo().then(extrudersInfo => {
        if (_.isEmpty(extrudersInfo)) return;
        const extruderStats = [];
        extrudersInfo.map((extruderInfo, index) => {
          extruderStats[index] = {
            temp: extruderInfo.current_temperature,
            extruder_version: extruderInfo.extruder_type,
            print_time: extruderInfo.extruder_print_time,
            filamentExtruded: extruderInfo.extrusion_distance_mm / 1000,
            mass: extruderInfo.extrusion_mass_g / 1000,
            serial: extruderInfo.serial ? this.serialNoFormat(extruderInfo.serial) : this._i18n.t('mb:printerInfo.n_a')
          };
        });
        return extruderStats;
      });
    };

    const systemInfo = () => {
      return this.props.printer.getSystemInfo();
    };

    const spoolInfo = () => {
      return this.props.printer.getSpoolInfo();
    };

    q.all([networkState(), extruderStats(), systemInfo(), spoolInfo()]).then(([network, extruders, systemInfo, spoolInfo]) => {
      const stateChanges = {
        network,
        extruders,
        systemInfo,
        spoolInfo
      };
      this.setState(stateChanges);
      return systemInfo;
    }).catch(console.error);
  }

  getMaterial(extruderIndex) {
    const printSettings = this.props.flux.store('PrintSettingsStore').getState();
    const filamentBay = this.state.systemInfo && this.state.systemInfo.filamentbays ? this.state.systemInfo.filamentbays[extruderIndex] : null;

    if (!this.state.systemInfo) {
      return this._i18n.t('mb:printerInfo.placeHolder');
    }

    if (filamentBay) {
      if (filamentBay.tag_present === 1) {
        const spoolInfo = this.state.spoolInfo;

        if (spoolInfo && spoolInfo[extruderIndex] && spoolInfo[extruderIndex].material_name) {
          return util.getMaterialPrettyStringFromCode(spoolInfo[extruderIndex].material_name);
        } else {
          // If there is a spool tag, but we don't have spool information, we assume
          // it's because we're still waiting for it, so we use '...'
          return this._i18n.t('mb:printerInfo.placeHolder');
        }
      } else {
        // If there is no spool tag, then there is no info to detect, so
        // we use 'Unknown Material' as the placeholder
        return this._i18n.t('mb:printerInfo.unknown');
      } // If there is no filament bay, get material info from the print settings store

    } else {
      const material = printSettings.materials.selected[extruderIndex];

      if (material !== undefined) {
        const materialTranslation = printSettings.materials.translations[material].name;
        return materialTranslation;
      }
    }

    return this._i18n.t('mb:printerInfo.unknown');
  }
  /**
   * Converts the recieved serial number in '12345' format to '0000-0001-2345' format
   * @param {number} serial - extruders serial number recieved
   **/


  serialNoFormat(serial) {
    let serialRecieved = serial.toString();

    while (serialRecieved.length < 12) {
      serialRecieved = `0${serialRecieved}`;
    }

    serialRecieved = serialRecieved.match(/.{4}/g).join('-');
    return serialRecieved;
  }
  /**
   * Adds event to tracking service
   * @param {string} name - message to track
   */


  addTrackingEvent(name) {
    const merged = _.merge({
      name
    }, this.props.printer.getTrackingInfo());

    this._trackingService.addEvent(merged);
  }

  updatePrinterName() {
    if (!this.state.disableUpdate) {
      this.setState({
        disableUpdate: true
      });
      this.props.printer.invoke('ChangeMachineName', [this.state.newPrinterName]).then(() => {
        this.addTrackingEvent('Printer Name Update Successful');
        this.setState({
          printerName: this.state.newPrinterName
        });
        this.refs.printerNameInput.value = '';
      }).catch(err => {
        this.addTrackingEvent('Printer Name Update Failed');
      });
    } else {
      this.addTrackingEvent('Printer Name Update Failed - invalid name');
    }
  }

  handleInputChange(e) {
    const newName = e.target.value;
    let disableUpdate = true;
    if (newName.length > 0 && newName.length < 150) disableUpdate = false;
    if (this.props.printer.isSketchPrinter()) disableUpdate = true;
    this.setState({
      newPrinterName: newName,
      disableUpdate
    });
  }

  formatLoadDate(secsFromEpochToLoad) {
    /**
     * To derive the date when a spool was loaded in bot, FW provides a couple of options:
     * 1. first_load_date = # of days since Jan 1st, 2000 (arbitrary, requires more JS date finesse - should be depricated)
     * 2. first_load_seconds = # of seconds since Jan 1st, 1970 (standard Epoch timestamp - all clients should use this)
     * Note: JS Date constructor accepts ms so (sec * 1000) is passed in -> then reformatted for the UI
     **/
    const loadDate = new Date(secsFromEpochToLoad * 1000);
    return loadDate.toLocaleDateString().replace(/[/]/g, '.');
  }

  formatMaterialAmounts(remainingMaterial, originalAmount, linearDensity) {
    const remainingMaterialMass = Math.floor(util.filamentLengthToMass(remainingMaterial, linearDensity));
    const originalAmountMass = Math.floor(util.filamentLengthToMass(originalAmount, linearDensity));

    const materialText = this._i18n.t('mb:printerInfo.filamentBay.material_amount_text', {
      remainingMaterialMass,
      originalAmountMass
    });

    return materialText;
  }

  renderUpdatePrinterName() {
    const buttonClass = classNames('update-button', {
      disabled: this.state.disableUpdate
    });

    const UpdateButton = () => {
      return /*#__PURE__*/React.createElement(Button, {
        style: {
          margin: '0 15px',
          padding: '8px 12px'
        },
        type: "secondary",
        classNames: buttonClass,
        label: this._i18n.t('mb:printerInfo.update'),
        onClick: () => this.updatePrinterName()
      });
    };

    return /*#__PURE__*/React.createElement("div", {
      className: "printerInfoPanel"
    }, /*#__PURE__*/React.createElement(PrinterInfoRow, {
      label: this._i18n.t('mb:printerInfo.printerName'),
      value: this.state.printerName
    }), /*#__PURE__*/React.createElement("div", {
      className: "printerUpdateName"
    }, /*#__PURE__*/React.createElement("input", {
      className: "truncate",
      id: "inputPrinterName",
      ref: "printerNameInput",
      type: "text",
      placeholder: this.state.printerName,
      onChange: event => this.handleInputChange(event),
      onKeyDown: event => event.key === 'Enter' ? this.updatePrinterName() : null,
      autoFocus: true
    }), /*#__PURE__*/React.createElement(UpdateButton, null)), /*#__PURE__*/React.createElement(PrinterInfoRow, {
      label: this._i18n.t('mb:printerInfo.firmware'),
      value: this.props.printer.getInfo().firmware_version
    }));
  }

  renderDNS() {
    if (!!this.state.network.dns && this.state.network.dns.length > 0) {
      const dns = this.state.network.dns.map(eachDNS => /*#__PURE__*/React.createElement("div", {
        className: "dns",
        key: eachDNS
      }, eachDNS));
      return dns;
    } else {
      return this._i18n.t('mb:printerInfo.placeHolder');
    }
  }

  renderNetwork() {
    const printerMAC = this.props.printer.getId().substring(18).match(/.{2}/g).join(':');

    const valuePlaceholder = this._i18n.t('mb:printerInfo.placeHolder');

    const networkInfoRows = [{
      label: this._i18n.t('mb:printerInfo.network.connectionType'),
      value: this.state.network.state || valuePlaceholder
    }, {
      label: this._i18n.t('mb:printerInfo.network.ip'),
      value: this.props.printer.getIpAddress() || valuePlaceholder
    }, {
      label: this._i18n.t('mb:printerInfo.network.subnet'),
      value: this.state.network.subnet || valuePlaceholder
    }, {
      label: this._i18n.t('mb:printerInfo.network.gateway'),
      value: this.state.network.gateway || valuePlaceholder
    }, {
      label: this._i18n.t('mb:printerInfo.network.dns'),
      value: this.renderDNS()
    }, {
      label: this._i18n.t('mb:printerInfo.network.wifi'),
      value: this.state.network.wifi || valuePlaceholder
    }, {
      label: this._i18n.t('mb:printerInfo.network.mac'),
      value: printerMAC || valuePlaceholder
    }];
    return /*#__PURE__*/React.createElement("div", {
      className: "printerInfoPanel"
    }, /*#__PURE__*/React.createElement(PrinterInfoTitle, {
      title: this._i18n.t('mb:printerInfo.network.title')
    }), networkInfoRows.map(props => {
      return /*#__PURE__*/React.createElement(PrinterInfoRow, _extends({
        key: `network ${props.label}`
      }, props));
    }));
  }

  renderMechanical() {
    const valuePlaceholder = this._i18n.t('mb:printerInfo.placeHolder');

    if (!(this.state && this.state.systemInfo && this.state.systemInfo.toolheads && this.state.systemInfo.toolheads.chamber)) {
      return;
    }

    const mechanicalInfoRows = [{
      label: this._i18n.t('mb:printerInfo.mechanical.chamber_temp'),
      value: this.state.systemInfo.toolheads.chamber[0].current_temperature + this._i18n.t('mb:printerInfo.extruder.temp_unit') || valuePlaceholder
    }];
    return /*#__PURE__*/React.createElement("div", {
      className: "printerInfoPanel"
    }, /*#__PURE__*/React.createElement(PrinterInfoTitle, {
      title: `${this._i18n.t('mb:printerInfo.mechanical.title')}`
    }), mechanicalInfoRows.map(props => {
      return /*#__PURE__*/React.createElement(PrinterInfoRow, _extends({
        key: `mechanical ${props.label}`
      }, props));
    }));
  }

  renderExtruder() {
    const valuePlaceholder = this._i18n.t('mb:printerInfo.placeHolder');

    const extruders = this.props.printer.getAttachedExtruders();
    const isDual = extruders.length > 1;
    const extruderComponents = extruders.map((attachedExtruder, extruderIndex) => {
      const extruder = this.state.extruders ? this.state.extruders[extruderIndex] : null; // at Method launch, Kaiten isn't returning print_time (aka extrusion_time_s)
      // so only showing on single extrusion bots until we can display something

      const extruderInfoRows = [{
        label: this._i18n.t('mb:printerInfo.extruder.temp'),
        value: extruder && extruder.temp ? extruder.temp + this._i18n.t('mb:printerInfo.extruder.temp_unit') : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.extruder.version'),
        value: extruder && extruder.extruder_version ? extruder.extruder_version : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.extruder.material'),
        value: this.getMaterial(extruderIndex)
      }, !isDual ? {
        label: this._i18n.t('mb:printerInfo.extruder.print_time'),
        value: extruder && extruder.print_time ? extruder.print_time : valuePlaceholder
      } : {
        label: '',
        value: ''
      }, {
        label: this._i18n.t('mb:printerInfo.extruder.filament'),
        value: extruder && extruder.filamentExtruded ? JSON.stringify(extruder.filamentExtruded) + this._i18n.t('mb:printerInfo.extruder.filament_unit') : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.extruder.mass'),
        value: extruder && extruder.mass ? JSON.stringify(extruder.mass) + this._i18n.t('mb:printerInfo.extruder.mass_unit') : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.extruder.serial'),
        value: extruder && extruder.serial ? extruder.serial : valuePlaceholder
      }];
      return /*#__PURE__*/React.createElement("div", {
        className: "printerInfoPanel",
        key: extruderIndex + 1
      }, /*#__PURE__*/React.createElement(PrinterInfoTitle, {
        title: `${this._i18n.t('mb:printerInfo.extruder.title')} ${isDual ? extruderIndex + 1 : ''}`
      }), extruderInfoRows.map(props => {
        return /*#__PURE__*/React.createElement(PrinterInfoRow, _extends({
          key: `extruder${extruderIndex} ${props.label}`
        }, props));
      }));
    });
    return extruderComponents;
  }

  renderMaterialBays() {
    if (!this.state.systemInfo || !this.state.systemInfo.filamentbays || !this.state.spoolInfo) {
      return;
    }

    const spoolInfo = this.state.spoolInfo;
    return this.state.systemInfo.filamentbays.map((filamentBay, bayIndex) => {
      /*
       * If there is a spool tag, but we don't have spool information, we assume
       * it's because we're still waiting for it, so we use '...'
       * If there is no spool tag, then there is no info to detect, so
       * we use 'Not Detected' as the placeholder
       */
      const valuePlaceholder = filamentBay.tag_present === 1 ? this._i18n.t('mb:printerInfo.placeHolder') : this._i18n.t('mb:printerInfo.not_detected');
      const spoolInfoRows = [{
        label: this._i18n.t('mb:printerInfo.filamentBay.material_type'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].material_type !== undefined ? util.getMaterialPrettyStringFromCode(spoolInfo[bayIndex].material_name) : this._i18n.t('mb:printerInfo.unknown')
      }, {
        label: this._i18n.t('mb:printerInfo.filamentBay.material_color'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].material_color_name ? spoolInfo[bayIndex].material_color_name : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.filamentBay.material_amount'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].amount_remaining !== undefined && spoolInfo[bayIndex].original_amount !== undefined ? this.formatMaterialAmounts(spoolInfo[bayIndex].amount_remaining, spoolInfo[bayIndex].original_amount, spoolInfo[bayIndex].linear_density) : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.filamentBay.first_used'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].first_load_date !== undefined ? this.formatLoadDate(spoolInfo[bayIndex].first_load_secs) : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.filamentBay.highest_humidity_level'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].max_humidity !== undefined ? `${spoolInfo[bayIndex].max_humidity}%` : valuePlaceholder
      }, {
        label: this._i18n.t('mb:printerInfo.filamentBay.tag_uid'),
        value: spoolInfo[bayIndex] && spoolInfo[bayIndex].tag_uid !== undefined && // tag UID can come through as null from the spool info
        spoolInfo[bayIndex].tag_uid !== null ? spoolInfo[bayIndex].tag_uid : valuePlaceholder
      }];
      return /*#__PURE__*/React.createElement("div", {
        className: "printerInfoPanel",
        key: bayIndex + 1
      }, /*#__PURE__*/React.createElement(PrinterInfoTitle, {
        title: `${this._i18n.t('mb:printerInfo.filamentBay.title')} ${bayIndex + 1}`
      }), spoolInfoRows.map(props => {
        return /*#__PURE__*/React.createElement(PrinterInfoRow, _extends({
          key: `${props.label} bay${bayIndex}`
        }, props));
      }));
    });
  }

  render() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", _extends({
      className: `${this.props.className || 'printerInfo'} printerPanel`
    }, this.props), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeader"
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelBackButton",
      onClick: this.props.onBack
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeaderLabel"
    }, /*#__PURE__*/React.createElement("h1", null, this._i18n.t('mb:printerInfo.header'))), /*#__PURE__*/React.createElement(HeaderSelectButton, {
      flux: this.props.flux,
      printer: this.props.printer
    }), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelArrowDown",
      onClick: this.props.onClose
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })))), /*#__PURE__*/React.createElement("div", {
      className: "printerInfoContainer"
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        marginBottom: 50
      }
    }, this.renderUpdatePrinterName(), this.renderNetwork(), this.renderMechanical(), this.renderExtruder(), this.renderMaterialBays())));
  }

}

PrinterInfo.displayName = 'PrinterInfo';
PrinterInfo.propTypes = {
  flux: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  className: React.PropTypes.string
};
module.exports = PrinterInfo;
