// MB-support-plugin/lib/ui/views/utilities/PrinterUtilityPanel.js
'use strict';

const React = require('react');

const FluxComponent = require('eagle-print/views/flux_component');

const R = require('ramda');

const path = require('path');

const {
  PrinterStateEnum,
  PrinterConnTypeEnum
} = require('../../../constants');

const UtilityTab = require('./utilityTab');

const UtilityHeader = require('./utilityHeader');

const UtilityTabSet = require('./utilityTabSet');

const PrinterInfo = require('./printerInfo');

const NetworkPreferences = require('./networkPreferences');

const FirmwareUpdatePanel = require('./firmwareUpdatePanel');

const CopyBotLogs = require('./copyBotLogs');

const Calibration = require('./calibration');

const ResetToFactory = require('./resetToFactory');

const Notifications = require('./notifications');

const FirmwareUpdateBanner = require('../firmwareUpdateBanner');

const PanelState = Object.freeze({
  MainMenu: 'default',
  // DEFAULT
  PrinterInfo: 'PrinterInfo',
  NetworkPreferences: 'NetworkPreferences',
  FirmwareUpdate: 'FirmwareUpdate',
  Calibration: 'Calibration',
  ResetToFactory: 'ResetToFactory',
  CopyBotLogs: 'CopyBotLogs',
  Notifications: 'Notifications'
});

class PrinterUtilityPanel extends FluxComponent {
  constructor(props) {
    super(props);
  }

  renderHeader() {
    const selectedPrinter = this.props.printer;
    return /*#__PURE__*/React.createElement("div", {
      tabKey: PanelState.MainMenu,
      style: {
        position: 'relative'
      }
    }, /*#__PURE__*/React.createElement(FirmwareUpdateBanner, {
      flux: this.getFlux(),
      printer: selectedPrinter,
      onBack: this.props.onBack
    }), /*#__PURE__*/React.createElement(UtilityHeader, {
      printer: this.props.printer,
      flux: this.props.flux
    }));
  }

  render() {
    const i18n = this.props.flux.getService('TranslationService');
    const selectedPrinter = this.props.printer;
    const status = selectedPrinter.getStatus() || {};
    const isSixthGen = selectedPrinter.isSixthGen();
    const isSketchPrinter = selectedPrinter.isSketchPrinter();
    return /*#__PURE__*/React.createElement(UtilityTabSet, R.merge(this.props, {
      title: i18n.t('mb:utility.header.utilities'),
      ref: c => this.panel = c,
      flux: this.props.flux,
      printer: this.props.printer,
      onBack: this.props.onBack
    }), this.renderHeader(), /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.PrinterInfo,
      tabName: i18n.t('mb:utility.category.printerinfo'),
      tabIcon: path.resolve(__dirname, '../images/Utilities_PrinterInfo.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Unauthenticated
    }), /*#__PURE__*/React.createElement(PrinterInfo, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      flux: this.props.flux
    })), /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.NetworkPreferences,
      tabName: i18n.t('mb:utility.category.network'),
      tabIcon: path.resolve(__dirname, '../images/Utilities_Network.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Unauthenticated || isSketchPrinter
    }), /*#__PURE__*/React.createElement(NetworkPreferences, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      flux: this.props.flux
    })), /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.FirmwareUpdate,
      tabName: i18n.t('mb:utility.category.firmware'),
      tabIcon: path.resolve(__dirname, '../images/firmware-update.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Busy || status.state === PrinterStateEnum.Unauthenticated || isSketchPrinter
    }), /*#__PURE__*/React.createElement(FirmwareUpdatePanel, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      onUpdate: () => this.props.onBack(),
      flux: this.props.flux
    })), /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.CopyBotLogs,
      tabName: i18n.t('mb:utility.category.copyBotLogs'),
      tabIcon: path.resolve(__dirname, '../images/PBL_1_v3.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: selectedPrinter.getConnectionType() === PrinterConnTypeEnum.USB || isSketchPrinter
    }), /*#__PURE__*/React.createElement(CopyBotLogs, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      onUpdate: () => this.props.onBack(),
      flux: this.props.flux
    })), !isSixthGen ? /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.Calibration,
      tabName: i18n.t('mb:utility.category.calibration.header'),
      tabIcon: path.resolve(__dirname, '../images/calibration-utilities.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: isSketchPrinter
    }), /*#__PURE__*/React.createElement(Calibration, R.merge(this.props, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      onFinish: this.props.onBack,
      flux: this.props.flux
    }))) : null, /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.Notifications,
      tabName: i18n.t('mb:utility.category.notifications'),
      tabIcon: path.resolve(__dirname, '../images/notifications.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Busy || status.state === PrinterStateEnum.Unauthenticated || isSketchPrinter
    }), /*#__PURE__*/React.createElement(Notifications, R.merge(this.props, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      onFinish: this.props.onBack,
      flux: this.props.flux
    }))), /*#__PURE__*/React.createElement(UtilityTab, R.merge(this.props, {
      tabKey: PanelState.ResetToFactory,
      tabName: i18n.t('mb:utility.category.resetToFactory'),
      tabIcon: path.resolve(__dirname, '../images/Utilities_Reset.svg'),
      onClick: tab => this.panel.select(tab),
      disabled: status.state === PrinterStateEnum.Busy || status.state === PrinterStateEnum.Unauthenticated || isSketchPrinter
    }), /*#__PURE__*/React.createElement(ResetToFactory, R.merge(this.props, {
      printer: selectedPrinter,
      onBack: () => this.panel.back(),
      onClose: this.props.onClose,
      onFinish: this.props.onBack,
      onResetSuccess: this.props.onResetSuccess,
      flux: this.props.flux
    }))));
  }

}

PrinterUtilityPanel.propTypes = {
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  onResetSuccess: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired
};
module.exports = PrinterUtilityPanel;
