// MB-support-plugin/lib/ui/views/utilities/resetToFactory.js
'use strict';

const React = require('react');

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const Button = require('eagle-print/views/ui/button');

const Checkbox = require('eagle-print/views/ui/input_checkbox');

const HeaderSelectButton = require('../headerSelectButton');

class ResetToFactory extends React.Component {
  constructor(props) {
    super(props);
    this._i18n = this.props.flux.getService('TranslationService');
    this._logger = this.props.flux.getService('LoggingService');
    this._trackingService = this.props.flux.getService('TrackingService');
    this.state = {
      clearCalibration: false,
      canCalibrate: false
    };
  }
  /**
   * Adds event to tracking service
   * @param {string} name - message to track
   */


  addTrackingEvent(name) {
    const merged = _.merge({
      name
    }, this.props.printer.getTrackingInfo());

    this._trackingService.addEvent(merged);
  }

  componentDidMount() {
    this.props.printer.invoke('HasZCalibrationRoutine').then(results => {
      this.setState({
        canCalibrate: results === true
      });
    });
  }

  handleEnableClearCalibrationToggle() {
    this.setState({
      clearCalibration: !this.state.clearCalibration
    });
  }

  onConfirm() {
    this.addTrackingEvent('Confirm Reset To Factory');

    this._logger.info(`Reset to factory setting with clearCalibration ` + `flag set to ${this.state.clearCalibration}`);

    this.props.printer.invoke('ResetToFactory', [this.state.clearCalibration]).then(() => {
      this.props.onResetSuccess();
      this.addTrackingEvent('Reset To Factory Success');
    }).catch(err => {
      this._logger.log(err);

      this.addTrackingEvent('Reset To Factory Failed');
    }).done();
  }

  onReset() {
    this.addTrackingEvent('Reset To Factory');
    const modalService = this.props.flux.getService('ModalService');
    modalService.confirm({
      title: this._i18n.t('mb:resetToFactory.modal.title'),
      body: this._i18n.t('mb:resetToFactory.modal.body', {
        printerName: this.props.printer.getName()
      }),
      cancelText: this._i18n.t('modal.cancel'),
      confirmText: this._i18n.t('mb:resetToFactory.modal.reset'),
      modalClass: 'w400 warningModal',
      onConfirm: () => this.onConfirm()
    });
  }

  renderResetButtonElm() {
    const props = {
      style: {
        margin: '0 25px',
        padding: '8px 12px'
      },
      type: 'secondary',
      label: this._i18n.t('mb:resetToFactory.button'),
      onClick: () => this.onReset()
    };
    return /*#__PURE__*/React.createElement(Button, props);
  }

  renderPrinterPanelHeader() {
    const headerSelectButtonProps = {
      flux: this.props.flux,
      printer: this.props.printer
    };
    return /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeader"
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelBackButton",
      onClick: this.props.onBack
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeaderLabel"
    }, /*#__PURE__*/React.createElement("h1", null, this._i18n.t('mb:resetToFactory.header'))), /*#__PURE__*/React.createElement(HeaderSelectButton, headerSelectButtonProps), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelArrowDown",
      onClick: this.props.onClose
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })));
  }

  renderClearCalibrationCheckbox() {
    const checkboxProps = {
      ref: 'clearCalibrationCheckbox',
      label: this._i18n.t('mb:resetToFactory.checkbox.clearCalibration'),
      classNames: 'clear-calibration-checkbox',
      labelClassNames: 'status-checkbox-label',
      onChange: () => this.handleEnableClearCalibrationToggle()
    };
    return this.state.canCalibrate ? /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 26
      }
    }, /*#__PURE__*/React.createElement(Checkbox, checkboxProps)) : null;
  }

  renderPrinterPanel() {
    return /*#__PURE__*/React.createElement("div", {
      className: "printerPanel"
    }, /*#__PURE__*/React.createElement("div", {
      style: {
        margin: 26,
        display: 'block'
      }
    }, this._i18n.t('mb:resetToFactory.warningMessage')), this.renderClearCalibrationCheckbox(), this.renderResetButtonElm());
  }

  render() {
    const props = _.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    });

    return /*#__PURE__*/React.createElement("div", props, this.renderPrinterPanelHeader(), this.renderPrinterPanel());
  }

}

ResetToFactory.propTypes = {
  flux: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  onResetSuccess: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  className: React.PropTypes.string
};
module.exports = ResetToFactory;
