// MB-support-plugin/lib/ui/views/utilities/staticIP.js
"use strict";

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const FluxComponent = require('eagle-print/views/flux_component');

const Checkbox = require('eagle-print/views/ui/input_checkbox');

const WaitableButton = require('../WaitableButton');

const Listbox = require('../input/listbox');

class StaticIP extends FluxComponent {
  constructor(props) {
    super(props);
    this._i18n = this.props.flux.getService('TranslationService');
    this._trackingService = this.props.flux.getService('TrackingService');
    this._loggingService = this.props.flux.getService('LoggingService');
    this._actions = this.props.flux.actions.NetworkPreferencesActions;
    this.state = {
      enableStaticIP: false
    };
  }

  handleStaticIPAddressChange(event) {
    this._actions.changeStaticIPAddress(event.target.value);
  }

  handleNetmaskChange(event) {
    this._actions.changeNetmask(event.target.value);
  }

  handleGatewayChange(event) {
    this._actions.changeGateway(event.target.value);
  }

  handleEnableStaticIPToggle() {
    if (this.state.enableStaticIP) {
      this.handleApplyStaticIP(false);

      this._trackingService.addEvent({
        name: 'Static IP Disabled'
      });
    } else {
      this._trackingService.addEvent({
        name: 'Static IP Enabled'
      });
    }

    this.setState({
      enableStaticIP: !this.state.enableStaticIP
    });
  }

  handleApplyStaticIP(enable = true) {
    const id = this.props.printer.getId();
    const dns = enable ? this.props.dnsServers.map(s => s.value) : [];

    this._actions.setStaticIP(id, this.props.path, this.props.editStaticIPAddress || this.props.staticIPAddress, this.props.editNetmask || this.props.netmask, this.props.editGateway || this.props.gateway, dns, enable);

    if (enable) this._trackingService.addEvent({
      name: 'Static IP Settings Applied'
    });
  }

  handleDNSAdd() {
    const items = this.props.dnsServers.concat({
      value: this.props.newDNSItem
    }).map(v => v.value);

    this._actions.changeDNSServers(items);
  }

  handleDNSRemove() {
    const items = this.props.dnsServers.reduce((acc, i) => !i.selected ? acc.concat(i.value) : acc, []);

    this._actions.changeDNSServers(items);
  }

  handleDNSSelect(val) {
    this._actions.selectDNSItem(val);
  }

  handleDNSChange(val) {
    this._actions.newDNSItem(val);
  }

  _label(text) {
    return DOM.div({
      style: {
        fontWeight: 'bold',
        marginTop: '0px'
      },
      className: 'header'
    }, text, ': ');
  }

  renderInputs() {
    return DOM.div({
      className: 'staticip'
    }, this._label(this._i18n.t('mb:network.staticIP.ip')), DOM.input({
      style: {
        margin: '5px 0'
      },
      className: 'input-text',
      type: 'text',
      disabled: !this.props.canEdit,
      value: this.props.editStaticIPAddress,
      placeholder: this.props.staticIPAddress,
      onChange: event => this.handleStaticIPAddressChange(event)
    }), this._label(this._i18n.t('mb:network.staticIP.netmask')), DOM.input({
      ref: 'netmask',
      style: {
        margin: '5px 0'
      },
      className: 'input-text',
      type: 'text',
      disabled: !this.props.canEdit,
      value: this.props.editNetmask,
      placeholder: this.props.netmask,
      onChange: event => this.handleNetmaskChange(event)
    }), this._label(this._i18n.t('mb:network.staticIP.gateway')), DOM.input({
      ref: 'gateway',
      style: {
        margin: '5px 0'
      },
      className: 'input-text',
      type: 'text',
      disabled: !this.props.canEdit,
      value: this.props.editGateway,
      placeholder: this.props.gateway,
      onChange: event => this.handleGatewayChange(event)
    }), this._label(this._i18n.t('mb:network.staticIP.dns')), Listbox({
      items: this.props.dnsServers,
      newItem: this.props.newDNSItem,
      placeholder: this._i18n.t('mb:network.staticIP.newPlaceholder'),
      onChange: this.handleDNSChange.bind(this),
      onAdd: this.handleDNSAdd.bind(this),
      onRemove: this.handleDNSRemove.bind(this),
      onSelect: this.handleDNSSelect.bind(this),
      disabled: !this.props.canEdit,
      emtpyLabel: this._i18n.t('mb:network.staticIP.emptyDNS')
    }), ce(WaitableButton, {
      waiting: this.props.waiting,
      buttonProps: {
        style: {
          margin: '12px 0',
          padding: '8px 12px'
        },
        type: 'primary',
        disabled: !this.props.staticIPEdited && !this.props.canEdit,
        className: 'network-button',
        label: this._i18n.t('mb:network.staticIP.apply'),
        onClick: () => this.handleApplyStaticIP()
      }
    }));
  }

  render() {
    return DOM.div({
      className: `panel-body${this.props.canEdit ? '' : ' view-disabled'}`
    }, DOM.div({
      style: {
        fontWeight: 'bold',
        marginTop: '0px'
      },
      className: 'header'
    }, this._i18n.t('mb:network.header.staticIP'), ': '), ce(Checkbox, {
      ref: 'enableStaticIPCheckbox',
      disabled: !this.props.canEdit,
      label: this._i18n.t('mb:network.checkbox.enableStaticIP'),
      labelClassNames: 'status-checkbox-label',
      onChange: () => this.handleEnableStaticIPToggle(),
      checked: this.props.enableStaticIP || this.state.enableStaticIP
    }), this.props.enableStaticIP || this.state.enableStaticIP ? this.renderInputs() : null);
  }

}

StaticIP.displayName = 'StaticIP';
StaticIP.propTypes = {
  flux: React.PropTypes.object.isRequired,
  printer: React.PropTypes.object.isRequired,
  waiting: React.PropTypes.bool.isRequired,
  path: React.PropTypes.string.isRequired,
  enableStaticIP: React.PropTypes.bool.isRequired,
  staticIPEdited: React.PropTypes.bool.isRequired,
  staticIPAddress: React.PropTypes.string.isRequired,
  netmask: React.PropTypes.string.isRequired,
  gateway: React.PropTypes.string.isRequired,
  dnsServers: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
  canEdit: React.PropTypes.bool.isRequired
};
module.exports = StaticIP;
