// MB-support-plugin/lib/ui/views/utilities/utilityTabSet.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const FluxComponent = require('eagle-print/views/flux_component');

const _ = require('lodash');

const CardContainer = require('eagle-print/views/ui/card_container');

const PrinterCardContainer = require('eagle-print/views/printers/printer_card_container');

const UtilityPanel = require('./utilityPanel');

const DEFAULTKEY = 'default';
/**
 * @classdesc Makes a collection of tabs and panels that slide in and out.
 * @example
 * const UtilityTabSet = require('./utilityTabSet');
 * const UtilityTab = require('./utilityTab');
 *
 * class MyComponent extends Component {
 *      render() {
 *          return (
 *              <UtilityTabSet title="My Tab Set" ref={(c) => this.tabset = c}>
 *                  <div tabKey="default">
 *                      Some default stuff
 *                  </div>
 *                  <UtilityTab tabKey="tab1" tabIcon="icon.svg" tabName="Tab 1" onClick={(t) => this.tabset.select(t)}>
 *                      <FirstPanel onBack={() => this.panel.back()} />
 *                  </UtilityTab>
 *                  <UtilityTab tabKey="tab2" tabIcon="icon.svg" tabName="Tab 2" onClick={(t) => this.tabset.select(t)}>
 *                      <SecondPanel onBack={() => this.panel.back()} />
 *                  </UtilityTab>
 *              </UtilityTabSet>
 *          );
 *      }
 * }
 */

class UtilityTabSet extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.logger = this.getFlux().getService('LoggingService');

    _.assign(this.state, {
      tab: DEFAULTKEY,
      swipeDirection: CardContainer.SWIPE_LEFT
    });
  }
  /**
   * @public
   * Advances the view to the specified tab
   */


  select(tab) {
    this.setState({
      tab,
      swipeDirection: CardContainer.SWIPE_LEFT
    });
  }
  /**
   * @public
   * Brings the view back to the default
   */


  back() {
    this.setState({
      tab: DEFAULTKEY,
      swipeDirection: CardContainer.SWIPE_RIGHT
    });
  }
  /**
   * @private
   * Extracts a particular child from the tabset based on its key
   * @returns {ReactElement}
   */


  _panelChild(key) {
    for (let i = this.props.children.length - 1; i >= 0; i--) {
      if (this.props.children[i] && this.props.children[i].props.tabKey === key) {
        return DOM.div({
          key: `${key}_child`
        }, this.props.children[i].props.children);
      }
    }

    this.logger.warn(`No child named ${key} in UtilityTabSet`);
    return null;
  }
  /**
   * Shows either the current tab, or the default view with the list of tabs
   * @returns {ReactElement}
   */


  _renderCard() {
    if (this.state.tab !== DEFAULTKEY) {
      return this._panelChild(this.state.tab);
    }

    return ce(UtilityPanel, {
      title: this.props.title,
      className: this.props.className,
      printer: this.props.printer,
      onBack: this.props.onBack,
      onClose: this.props.onClose
    }, DOM.div({
      className: 'bottomList'
    }, DOM.div({
      style: {
        marginBottom: 50
      }
    }, this.props.children)));
  }

  render() {
    return ce(PrinterCardContainer, _.assign({}, this.props, {
      swipeDirection: this.state.swipeDirection
    }), this._renderCard());
  }

}

UtilityTabSet.propTypes = {
  onBack: React.PropTypes.func.isRequired,
  onClose: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  children: React.PropTypes.any
};
module.exports = UtilityTabSet;
