// MB-support-plugin/lib/ui/views/utilities/utilityHeader.js
'use strict';

const React = require('react');

const DOM = React.DOM;

const path = require('path');

const PrinterUtilityHeader = React.createClass({
  displayName: "PrinterUtilityHeader",
  propTypes: {
    printer: React.PropTypes.object.isRequired
  },

  getInitialState() {
    return {
      printerName: this.props.printer.getName()
    };
  },

  onPrinterNameChange(event) {
    this.setState({
      printerName: event.target.value
    });
  },

  onKeyPress(event) {
    // On Enter (charCode 13) change printer name
    if (event.charCode === 13) {
      this.props.printer.invoke('ChangeMachineName', [event.target.value]);
    }
  },

  onNameBlur(event) {},

  displayPrinterModel(printer) {
    return printer.getInfo().model;
  },

  displayFwVersion(i18n, printer) {
    const fwVersion = printer.getInfo().firmware_version;
    return `${i18n.t('mb:utility.header.firmware')} ${fwVersion}`;
  },

  displayIpAddresss(i18n, printer) {
    const ipaddress = printer.getIpAddress();

    if (ipaddress) {
      return `${i18n.t('mb:utility.header.ip')} ${ipaddress}`;
    } else {
      return i18n.t('mb:utility.header.null_ip');
    }
  },

  printerSummary(event) {
    const i18n = this.props.flux.getService('TranslationService');
    const printer = this.props.printer;
    return DOM.div({
      className: 'utilityHeaderContainer'
    }, DOM.input({
      className: 'nameInput truncate',
      onChange: this.onPrinterNameChange,
      onKeyPress: this.onKeyPress,
      onBlur: this.onNameBlur,
      disabled: true,
      // Enable when ready for primte time
      value: this.state.printerName
    }), DOM.h6({}, this.displayPrinterModel(printer)), DOM.h6({
      style: {
        margin: '20px 0 0 0'
      }
    }, this.displayFwVersion(i18n, printer)), DOM.h6({}, this.displayIpAddresss(i18n, printer)));
  },

  render() {
    const self = this;
    const background = path.resolve(__dirname, `../images/utility_${this.props.printer.getGender().botType}.png`);
    return DOM.div({
      className: 'utilityHeader'
    }, DOM.img({
      src: background
    }), self.printerSummary());
  }

});
module.exports = PrinterUtilityHeader;
