// MB-support-plugin/lib/ui/views/utilities/utilityTab.js
'use strict';

const React = require('react');

const ce = React.createElement;

const FluxComponent = require('eagle-print/views/flux_component');

const Isvg = require('react-inlinesvg');

const DOM = React.DOM;

const path = require('path');

const classNames = require('classnames');

const util = require('../../../util');

class UtilityTab extends FluxComponent {
  render() {
    const utilityRowClass = classNames('printerRow', {
      disabled: this.props.disabled
    });
    const tabName = this.props.tabName;
    const icon = path.resolve('node_modules/MB-support-plugin/', this.props.tabIcon);
    return DOM.li({
      className: utilityRowClass,
      onClick: () => !this.props.disabled ? this.props.onClick(this.props.tabKey) : null
    }, DOM.div({
      className: 'printerRowInner'
    }, DOM.div({
      style: {
        width: 50,
        height: 50,
        maxWidth: 50,
        maxHeight: 50,
        alignSelf: 'center'
      }
    }, ce(Isvg, {
      src: util.getPathForIsvg(icon)
    })), DOM.div({
      className: 'printerInfo'
    }, DOM.div({}, DOM.div({
      className: 'printerName'
    }, tabName)))));
  }

}

UtilityTab.propTypes = {
  tabKey: React.PropTypes.any.isRequired,
  onClick: React.PropTypes.func,
  tabName: React.PropTypes.string,
  tabIcon: React.PropTypes.string,
  printer: React.PropTypes.object.isRequired,
  disabled: React.PropTypes.bool,
  children: React.PropTypes.element
};
UtilityTab.defaultProps = {
  onClick: () => {},
  disabled: false
};
module.exports = UtilityTab;
