// MB-support-plugin/lib/ui/views/utilities/utilityPanel.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const FluxComponent = require('eagle-print/views/flux_component');

const _ = require('lodash');

const Isvg = require('react-inlinesvg');

const HeaderSelectButton = require('../headerSelectButton');

class UtilityPanel extends FluxComponent {
  render() {
    return DOM.div(_.assign({}, {
      id: 'printer-utilities',
      className: `${this.props.className || ''} printerPanel`
    }), DOM.div({
      className: 'printerPanelHeader'
    }, DOM.div({
      className: 'printerPanelBackButton',
      onClick: this.props.onBack
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })), DOM.div({
      className: 'printerPanelHeaderLabel'
    }, DOM.h1({}, this.props.title)), ce(HeaderSelectButton, {
      flux: this.getFlux(),
      printer: this.props.printer
    }), DOM.div({
      className: 'printerPanelArrowDown',
      onClick: this.props.onClose
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    }))), this.props.children);
  }

}

UtilityPanel.propTypes = {
  title: React.PropTypes.string.isRequired,
  printer: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func,
  onClose: React.PropTypes.func
};
UtilityPanel.defaultProps = {
  onBack: () => {},
  onClose: () => {}
};
module.exports = UtilityPanel;
