// MB-support-plugin/lib/ui/views/utilities/zcal.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const Tooltip = require('eagle-print/views/util/tooltip');

const FluxComponent = require('eagle-print/views/flux_component');

const Button = require('eagle-print/views/ui/button');

const LinkClickUtils = require('../../../utils/linkClick');

const UrlConst = require('../../../constants/urls').url;

const PrinterStateEnum = require('../../../constants').PrinterStateEnum;

const HeaderSelectButton = require('../headerSelectButton');

class ZCal extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.i18n = this.props.flux.getService('TranslationService');
    this.logger = this.props.flux.getService('LoggingService');
    this.tracking = this.props.flux.getService('TrackingService');
  }

  onCalibrate() {
    this.props.printer.invoke('CalibrateZOffset').then(() => this.tracking.addEvent({
      name: 'z-calibration'
    })).catch(err => this.logger.error(err));
    this.props.onFinish();
  }

  render() {
    const selectedPrinter = this.props.printer;
    const status = selectedPrinter.getStatus() || {};
    const disabled = status.state === PrinterStateEnum.Busy;
    const button = ce(Button, {
      type: `secondary ${disabled ? 'disabled' : ''}`,
      onClick: disabled ? () => undefined : () => this.onCalibrate()
    }, DOM.span({}, this.i18n.t('mb:zcalibration.button')));
    const buttonTooltip = ce(Tooltip, {
      id: 'zcalTooltip',
      addTo: button,
      placement: 'left',
      content: this.i18n.t('mb:zcalibration.tooltip')
    });
    return DOM.div(_.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    }), DOM.div({
      className: 'printerPanelHeader'
    }, DOM.div({
      className: 'printerPanelBackButton',
      onClick: this.props.onBack
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })), DOM.div({
      className: 'printerPanelHeaderLabel'
    }, DOM.h1({}, DOM.a({
      className: 'supportLink',
      onClick: () => {
        LinkClickUtils.clickAction(this.getFlux(), 'Zcal_link_clicked', UrlConst.zCal);
      }
    }, this.i18n.t('mb:zcalibration.header'), ce(Isvg, {
      className: 'helperLinkIcon',
      src: 'icons/helpful-link.svg'
    })))), ce(HeaderSelectButton, {
      flux: this.getFlux(),
      printer: this.props.printer
    }), DOM.div({
      className: 'printerPanelArrowDown',
      onClick: this.props.onClose
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    }))), DOM.div({
      className: 'panel-body'
    }, this.i18n.t('mb:zcalibration.paragraph').split('\n').map(t => DOM.p({
      style: {
        marginBottom: 24
      }
    }, t)), buttonTooltip, status.state === PrinterStateEnum.Busy ? DOM.p('mb:zcalibration.busy') : null));
  }

}

ZCal.propTypes = {
  onBack: React.PropTypes.func,
  onClose: React.PropTypes.func,
  flux: React.PropTypes.object.isRequired,
  printer: React.PropTypes.object.isRequired
};
ZCal.defaultProps = {
  onBack: () => {},
  onClose: () => {},
  onFinish: () => {}
};
module.exports = ZCal;
