// MB-support-plugin/lib/ui/views/utilities/zoffsetButton.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const _ = require('lodash');

const FluxComponent = require('eagle-print/views/flux_component');

const Button = require('eagle-print/views/ui/button');
/**
 * The entire reason for this thing existing is to make the slider on ZOffset work. Basically,
 * to make the button both aware of state, and have the slider not re-render (effectively killing
 * the mouse-drag action), I had to wrap the button in its own component, so I could render just that, and
 * leave the parent (zOffset) unaffected. This is basically a hack to make the slider play nice with everything.
 */


class ZOffsetButton extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.i18n = this.props.flux.getService('TranslationService');

    _.assign(this.state, {
      disabled: this.props.disabled
    });
  }
  /**
   * Enable the button
   */


  enable() {
    this.setState({
      disabled: false
    });
  }
  /**
   * Disable the button
   */


  disable() {
    this.setState({
      disabled: true
    });
  }

  isDisabled() {
    return this.state.disabled;
  }
  /**
   * Some extra text as per the spec
   */


  _changesText() {
    if (this.state.disabled === true) {
      return DOM.div();
    }

    return DOM.div({
      style: {
        marginTop: 20
      }
    }, DOM.p({}, this.i18n.t('mb:zoffset.changes')), DOM.p({}, this.i18n.t('mb:zoffset.caution')));
  }

  render() {
    return DOM.div({
      className: 'zoffset-button-container'
    }, ce(Button, {
      type: `secondary ${this.state.disabled ? 'disabled' : ''}`,
      onClick: this.props.onClick,
      disabled: this.state.disabled
    }, this.i18n.t('mb:zoffset.apply')), this._changesText());
  }

}

module.exports = ZOffsetButton;
