// MB-support-plugin/lib/ui/views/utilities/zOffset.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const path = require('path');

const FluxComponent = require('eagle-print/views/flux_component');

const Slider = require('eagle-print/views/ui/slider');

const util = require('../../../util');

const ZOffsetButton = require('./zoffsetButton');

const HeaderSelectButton = require('../headerSelectButton');

const LinkClickUtils = require('../../../utils/linkClick');

const UrlConst = require('../../../constants/urls').url;

class ZOffset extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.i18n = this.props.flux.getService('TranslationService');
    this.logger = this.props.flux.getService('LoggingService');
    this.tracking = this.props.flux.getService('TrackingService'); // setting this as a instance variable rather than a state
    // variable, because again, setting state will cause a re-render

    this.offset = 0;

    _.assign(this.state, {
      initialOffset: 0
    });
  }
  /**
   * Basically we don't want this updating to frequently, because
   * the slider will forget that it's being dragged, essentially making it
   * so the user can only drag one tick at a time, which is insanely annoying.
   * So, we basically almost never update the state, which allows the user
   * to drag on the slider properly.
   */


  shouldComponentUpdate() {
    return !this._hasChanged();
  }

  componentDidMount() {
    this._getInitialOffset();
  }
  /**
   * Figure out what the printer's z-offset is currently set to
   */


  _getInitialOffset() {
    // we have to manually disable the apply button, rather than
    // setting state. This is kinda hacky, but we have to do it this way
    // because we can't allow the slider to re-render.
    this._apply.disable();

    return this.props.printer.invoke('GetZAdjustedOffset').then(val => this.offset = val * 10) // multiplying by ten because printer settings go from -0.8 to 0.8, but the slider goes in integers
    .then(() => this.setState({
      initialOffset: this.offset
    })).catch(err => this.logger.warn(err));
  }
  /**
   * Fires when the slider changes its value. You may notice that
   * we don't set state in here. That's because setting state
   * usually triggers a re-render, which will prevent the user
   * from being able to drag the slider. Instead we edit the UI
   * inline. It's not very React, but as it turns out, neither is the
   * slider, and that's what's really screwing us up.
   */


  _onChange(val) {
    // setting this as a instance variable rather than a state
    // variable, because again, setting state will cause a re-render
    this.offset = val.newValue;

    if (this._preview) {
      // move the preview to keep up with the slider
      this._preview.style.top = this._offsetStyle();
    }

    if (this._label) {
      // change the label to reflect the settings on the slider
      this._label.innerText = this._formatOffset();
    }

    if (this.state.initialOffset !== this.offset) {
      this._apply.enable();
    }

    if (this.state.initialOffset === this.offset) {
      this._apply.disable();
    }
  }

  _hasChanged() {
    return this.offset !== this.state.initialOffset;
  }
  /**
   * Just making the extruder icon move and fit within the space
   * in a reasonable way.
   */


  _offsetStyle() {
    return this.offset * 3 + 20;
  }
  /**
   * When the apply button is clicked
   */


  _handleClick() {
    if (!this._hasChanged()) {
      return;
    } // dividing by ten because printer settings go from -0.8 to 0.8, but the slider goes in integers


    return this.props.printer.invoke('SetZAdjustedOffset', [this.offset / 10]).then(() => this._getInitialOffset()).then(() => this.tracking.addEvent({
      name: 'z-offset_change'
    })).catch(err => this.logger.warn(err));
  }
  /**
   * Formats the label that sits in the middle of the
   * buildplate icon.
   */


  _formatOffset() {
    /* Ok so this is weird. When sending a command to the bot,
    the bot acutally thinks postive numbers should be closer to the plate, 
    where the UI represents closer to the plate as a negative number.
    So when we send an offset that looks negitive on the UI 
    we are really sending a postive number to the bot.
    
    Flip the sign here to match the UI. */
    const sign = this.offset > 0 ? '-' : '+';
    return `${sign}0.${Math.abs(this.offset)}0mm`;
  }

  render() {
    // really just UI sugar, doesn't affect how the thing works
    const ticks = _.range(7).map(() => DOM.li({}, '-'));

    return DOM.div(_.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    }), DOM.div({
      className: 'printerPanelHeader'
    }, DOM.div({
      className: 'printerPanelBackButton',
      onClick: this.props.onBack
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })), DOM.div({
      className: 'printerPanelHeaderLabel'
    }, DOM.h1({}, DOM.a({
      className: 'supportLink',
      onClick: () => {
        LinkClickUtils.clickAction(this.getFlux(), 'ZOffset_link_clicked', UrlConst.zOffset);
      }
    }, this.i18n.t('mb:zoffset.header'), ce(Isvg, {
      className: 'helperLinkIcon',
      src: 'icons/helpful-link.svg'
    })))), ce(HeaderSelectButton, {
      flux: this.getFlux(),
      printer: this.props.printer
    }), DOM.div({
      className: 'printerPanelArrowDown',
      onClick: this.props.onClose
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    }))), DOM.div({
      className: 'panel-body'
    }, DOM.div({
      style: {
        marginBottom: 50
      }
    }, DOM.p({}, this.i18n.t('mb:zoffset.paragraph')), DOM.div({
      className: 'zoffset'
    }, DOM.div({
      className: 'zoffset-preview'
    }, DOM.div({
      className: 'zoffset-extruder',
      style: {
        top: this._offsetStyle()
      },
      ref: c => this._preview = c
    }, ce(Isvg, {
      src: util.getPathForIsvg(path.resolve(__dirname, '../images/extruder-inside.svg'))
    })), DOM.div({
      className: 'zoffset-buildplate'
    }, ce(Isvg, {
      src: util.getPathForIsvg(path.resolve(__dirname, '../images/bottom-build-plate.svg'))
    }), DOM.span({
      className: 'zoffset-buildplate-label',
      ref: c => this._label = c
    }, this._formatOffset()))), DOM.div({
      className: 'zoffset-slider-container'
    }, // This is that jerk slider that causes so much grief in this component.
    // It's kind of a jerk. Don't tell it I said that though.
    ce(Slider, {
      ref: c => this._slider = c,
      onChange: this._onChange.bind(this),
      unit: 'mm',
      min: -8,
      max: 8,
      value: this.offset,
      ticks: false,
      orientation: 'vertical',
      disable: false,
      showLabels: true
    }), DOM.ul({
      className: 'zoffset-slider-ticks'
    }, ticks), DOM.ul({
      className: 'zoffset-slider-labels'
    }, DOM.li({}, '+0.80mm'), DOM.li({}, 'Default'), DOM.li({}, '-0.80mm')))), // This is that weird button I had to make to compensate
    // for the Slider being a jerk.
    ce(ZOffsetButton, {
      ref: c => this._apply = c,
      disabled: true,
      onClick: this._handleClick.bind(this),
      flux: this.props.flux
    }))));
  }

}

ZOffset.propTypes = {
  onBack: React.PropTypes.func,
  onClose: React.PropTypes.func,
  flux: React.PropTypes.object.isRequired,
  printer: React.PropTypes.object.isRequired
};
ZOffset.defaultProps = {
  onBack: () => {},
  onClose: () => {}
};
module.exports = ZOffset;
