// MB-support-plugin/lib/ui/views/printSettings/advancedSettingWidget.js
'use strict';

const React = require('react');

const Isvg = require('react-inlinesvg');

const R = require('ramda');

const path = require('path');

const Dropdown = require('eagle-print/views/ui/input_dropdown');

const Checkbox = require('eagle-print/views/ui/input_checkbox');

const FluxComponent = require('eagle-print/views/flux_component');

const EaglePrint = require('eagle-print');

const Increment = require('../input/increment');

const util = require('../../../util');

const MenuItemData = EaglePrint.views.menu.MenuItemData;

const {
  EESettingsKeys
} = require('../../../constants/print_settings');

const EEIcon = '../images/as-extruder-ee.svg';

class AdvancedSettingWidget extends FluxComponent {
  constructor(props) {
    super(props);
    this._actions = this.getFlux().actions.PrintSettingsActions;
    this._i18n = this.props.flux.getService('TranslationService');
    this._logger = this.props.flux.getService('LoggingService');
    this._trackingService = this.props.flux.getService('TrackingService');
  }

  renderInput() {
    const icon = R.contains(this.props.name, EESettingsKeys) ? /*#__PURE__*/React.createElement(Isvg, {
      src: util.fullIconPath(path.resolve(__dirname, EEIcon))
    }) : null;
    const headerElement = this.props.showTitle ? /*#__PURE__*/React.createElement("div", {
      className: "prop-hdr panel-accordion__itemWrapper"
    }, /*#__PURE__*/React.createElement("div", {
      className: "panel-close-x",
      onClick: () => this.removeSetting(this.props.name)
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/x-button.svg"
    })), /*#__PURE__*/React.createElement("div", {
      style: {
        paddingLeft: '5px'
      }
    }, this.props.title), icon) : null;

    switch (this.props.type.toLowerCase()) {
      case 'boolean':
      case 'checkbox':
        return /*#__PURE__*/React.createElement("div", {
          className: "panel-accordion__itemWrapper",
          key: this.props.name,
          id: `${this.props.name}__wrapper`,
          onContextMenu: event => this.openContextMenu(event, this.props.name)
        }, this.props.showTitle ? /*#__PURE__*/React.createElement("div", {
          className: "panel-close-x",
          onClick: () => this.removeSetting(this.props.name)
        }, /*#__PURE__*/React.createElement(Isvg, {
          src: "icons/x-button.svg"
        })) : null, /*#__PURE__*/React.createElement("div", null), /*#__PURE__*/React.createElement("div", {
          className: "input-checkbox-container-wrapper"
        }, /*#__PURE__*/React.createElement(Checkbox, R.merge(this.props, {
          checked: this.props.value === true,
          label: this.props.title,
          onChange: e => this.props.onChange(e.target.checked)
        }))), icon);

      case 'increment':
      case 'scalar':
      case 'unsigned_integer':
      case 'percent':
        return /*#__PURE__*/React.createElement("div", {
          className: "padding-top-btm-5",
          key: this.props.name,
          id: `${this.props.name}__wrapper`,
          onContextMenu: event => this.openContextMenu(event, this.props.name)
        }, headerElement, /*#__PURE__*/React.createElement("div", {
          className: "padding-sides-20"
        }, /*#__PURE__*/React.createElement(Increment, this.props)));

      case 'dropdown':
      case 'string':
        return /*#__PURE__*/React.createElement("div", {
          className: "padding-top-btm-5",
          key: this.props.name,
          id: `${this.props.name}__wrapper`,
          onContextMenu: event => this.openContextMenu(event, this.props.name)
        }, headerElement, /*#__PURE__*/React.createElement("div", {
          className: "padding-sides-20"
        }, /*#__PURE__*/React.createElement(Dropdown, R.merge(this.props, {
          onChange: e => this.props.onChange(e.value)
        }))));
    }

    this._logger.warn('Unknown widget type: ', this.props.type.toLowerCase());
  }

  openContextMenu(event, name) {
    event.preventDefault();
    event.stopPropagation();
    const menus = [new MenuItemData({
      path: 'reset_to_default',
      label: this._i18n.t('mb:printSettings.right_click.reset_to_default'),
      action: () => this.resetToDefault(name)
    }), new MenuItemData({
      path: 'reset_all_to_default',
      label: this._i18n.t('mb:printSettings.right_click.reset_all_to_default'),
      action: () => this._actions.resetAllToDefault()
    }), new MenuItemData({
      path: 'remove_setting',
      label: this._i18n.t('mb:printSettings.right_click.remove_setting'),
      action: () => this.removeSetting(name)
    }), new MenuItemData({
      path: 'remove_all_settings',
      label: this._i18n.t('mb:printSettings.right_click.remove_all_settings'),
      action: () => this._actions.removeAllSettings()
    })];
    this.props.flux.getService('MenuService').renderContextMenu(menus, event.clientX, event.clientY);
  }
  /**
   * Resets setting to Slice Config default.
   */


  resetToDefault(name) {
    const key = name.split(':').pop();

    this._actions.resetToDefault(key);
  }
  /**
   * Removes setting from advanced settings toggle and UI view
   * Triggers removal of key from Slice Config override/tray.bodyProperties object.
   * Setting for Slice Config returns to sliceConfig schema default.
   */


  removeSetting(name) {
    const key = name.split(':').pop();

    this._actions.toggleProp(key);

    this._trackingService.addEvent({
      name: 'Remove Setting',
      setting: key
    });
  }

  render() {
    return this.renderInput();
  }

}

AdvancedSettingWidget.displayName = 'AdvancedSettingWidget';
AdvancedSettingWidget.propTypes = {
  flux: React.PropTypes.object.isRequired,
  name: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  type: React.PropTypes.string.isRequired,
  value: React.PropTypes.any.isRequired,
  format: React.PropTypes.func.isRequired,
  options: React.PropTypes.array,
  step: React.PropTypes.number,
  min: React.PropTypes.number,
  max: React.PropTypes.number,
  units: React.PropTypes.string,
  disabled: React.PropTypes.bool,
  onChange: React.PropTypes.func,
  showTitle: React.PropTypes.bool
};
AdvancedSettingWidget.defaultProps = {
  onChange: () => {},
  showTitle: true
};
module.exports = AdvancedSettingWidget;
