// MB-support-plugin/lib/ui/views/printSettings/advancedSettingsAccordion.js
"use strict";

const React = require('react');

const Immutable = require('immutable');

const FluxComponent = require('eagle-print/views/flux_component');

const PrintSettingsConstants = require('../../../constants/print_settings');

const Transitions = Immutable.Map({
  WebkitTransition: 'webkitTransitionEnd',
  MozTransition: 'transitionend',
  OTransition: 'oTransitionEnd otransitionend',
  transition: 'transitionend'
});

class Accordion extends FluxComponent {
  constructor(props) {
    super(props);
    this.state = {
      isClosed: !this.props.open,
      shouldSwitchAutoOnNextCycle: false,
      height: 'auto',
      transition: 'none',
      hasBeenOpened: true
    };
  } // Taken from https://github.com/EvandroLG/transitionEnd/
  // Determines which prefixed event to listen for


  whichTransitionEnd(element) {
    return Transitions.filter((v, k) => element.style[k] != null).toArray();
  }

  componentDidMount() {
    this.refs.outer.addEventListener(this.whichTransitionEnd(this.refs.outer), this.onTransitionEnd.bind(this));
  }

  componentWillUnmount() {
    this.refs.outer.removeEventListener(this.whichTransitionEnd(this.refs.outer), this.onTransitionEnd.bind(this));
  }

  componentDidUpdate(prevProps) {
    if (this.state.shouldSwitchAutoOnNextCycle === true && this.state.isClosed === false) {
      //Set the height to auto to make compoenent re-render with the height set to auto.
      //This way the dropdown will be responsive and also change height if there is another dropdown within it.
      this.makeResponsive();
    }

    if (this.state.shouldSwitchAutoOnNextCycle === true && this.state.isClosed === true) {
      this.prepareToOpen();
    } //If there has been a change in the open prop (controlled by accordion)


    if (prevProps.open !== this.props.open) {
      console.log('Open state changed!', this.props.accordionPosition);

      if (this.props.open === true) {
        this.openAccordion();
      } else {
        this.closeAccordion();
      }
    }
  }

  onTransitionEnd() {
    if (!this.state.isClosed) {
      this.setState({
        shouldSwitchAutoOnNextCycle: true
      });
    }
  }

  handleTriggerClick(event) {
    event.preventDefault();

    if (this.props.handleTriggerClick) {
      this.props.handleTriggerClick(this.props.accordionPosition);
    } else {
      if (this.state.isClosed === true) {
        this.openAccordion();
      } else {
        this.closeAccordion();
      }
    }
  }

  closeAccordion() {
    this.setState({
      isClosed: true,
      shouldSwitchAutoOnNextCycle: true,
      height: this.refs.inner.offsetHeight
    });
  }

  openAccordion() {
    this.setState({
      height: this.refs.inner.offsetHeight,
      transition: `height ${this.props.transitionTime}ms ${this.props.easing}`,
      isClosed: false,
      hasBeenOpened: true
    });
  }

  makeResponsive() {
    this.setState({
      height: 'auto',
      transition: 'none',
      shouldSwitchAutoOnNextCycle: false
    });
  }

  prepareToOpen() {
    //The height has been changes back to fixed pixel, we set a small timeout to force the CSS transition back to 0 on the next tick.
    window.setTimeout(() => {
      this.setState({
        height: 0,
        shouldSwitchAutoOnNextCycle: false,
        transition: `height ${this.props.transitionTime}ms ${this.props.easing}`
      });
    }, 50);
  }

  handleSelectArrowToggle() {
    this.state.isClosed ? this.setState({
      isClosed: false
    }) : this.setState({
      isClosed: true
    });
  }

  renderArrow() {
    return /*#__PURE__*/React.createElement("div", {
      className: "Select-arrow",
      onClick: () => this.handleSelectArrowToggle()
    });
  }

  render() {
    const dropdownStyle = {
      height: this.state.height,
      WebkitTransition: this.state.transition,
      msTransition: this.state.transition,
      transition: this.state.transition,
      overflow: this.state.overflow
    };
    const openClass = this.state.isClosed ? 'is-closed' : 'is-open'; // If user wants different text when tray is open

    const trigger = this.state.isClosed === false && this.props.triggerWhenOpen !== undefined ? this.props.triggerWhenOpen : this.props.trigger; // Don't render children until the first opening of the Accordion if lazy rendering is enabled

    let children = this.props.children;

    if (this.props.lazyRender && !this.state.hasBeenOpened) {
      children = null;
    }

    return /*#__PURE__*/React.createElement("div", {
      className: `${this.props.classParentString}`
    }, /*#__PURE__*/React.createElement("div", {
      className: `${this.props.classParentString} panel-accordion__trigger ${openClass}`,
      onClick: event => this.handleTriggerClick(event)
    }, PrintSettingsConstants.DisplayKeys[trigger], this.renderArrow()), /*#__PURE__*/React.createElement("div", {
      className: `${this.props.classParentString} panel-accordion__contentOuter ${openClass}`,
      ref: "outer",
      style: dropdownStyle
    }, /*#__PURE__*/React.createElement("div", {
      className: `${this.props.classParentString} panel-accordion__contentInner`,
      ref: "inner"
    }, children)));
  }

}

Accordion.displayName = 'Accordion';
Accordion.propTypes = {
  transitionTime: React.PropTypes.number,
  easing: React.PropTypes.string,
  open: React.PropTypes.bool,
  classParentString: React.PropTypes.string,
  accordionPosition: React.PropTypes.number,
  handleTriggerClick: React.PropTypes.func,
  trigger: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.element]),
  triggerWhenOpen: React.PropTypes.oneOfType([React.PropTypes.string, React.PropTypes.element]),
  lazyRender: React.PropTypes.bool
};
Accordion.defaultProps = {
  transitionTime: 400,
  easing: 'linear',
  open: true,
  classParentString: 'Accordion',
  lazyRender: false
};
module.exports = Accordion;
