// MB-support-plugin/lib/ui/views/printSettings/advancedSettingToggle.js
'use strict';

const React = require('react');

const path = require('path');

const Isvg = require('react-inlinesvg');

const R = require('ramda');

const FluxComponent = require('eagle-print/views/flux_component');

const Tooltip = require('eagle-print/views/util/tooltip');

const util = require('../../../util');

const AdvancedSettingWidget = require('./advancedSettingWidget');

const AlternateCheckbox = require('../input/alternate_checkbox');

const {
  EESettingsKeys
} = require('../../../constants/print_settings');

const EEIcon = '../images/as-extruder-ee.svg';

class AdvancedSettingToggle extends FluxComponent {
  constructor(props) {
    super(props);
    this._i18n = this.getFlux().getService('TranslationService');
  }

  render() {
    const icon = R.contains(this.props.name, EESettingsKeys) ? /*#__PURE__*/React.createElement(Isvg, {
      src: util.fullIconPath(path.resolve(__dirname, EEIcon))
    }) : null;
    const checkbox = /*#__PURE__*/React.createElement(AlternateCheckbox, {
      name: this.props.name,
      on: "Hide",
      off: "Show",
      checked: this.props.active,
      onChange: () => this.props.onToggle(this.props.name)
    });
    const tooltip = /*#__PURE__*/React.createElement(Tooltip, {
      placement: "left",
      content: "Hide/Show",
      id: `${this.props.name}-tooltip`,
      addTo: checkbox
    });
    const options = Array.isArray(this.props.options) ? this.props.options : Object.keys(this.props.options);
    return /*#__PURE__*/React.createElement("div", {
      className: `advanced-toggle-item ${this.props.name}`
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("h6", {
      className: "advanced-toggle-title"
    }, this.props.title, icon), /*#__PURE__*/React.createElement("p", {
      className: "advanced-toggle-description"
    }, this.props.description), /*#__PURE__*/React.createElement("p", {
      className: "advanced-toggle-rationale"
    }, this.props.rationale), /*#__PURE__*/React.createElement(AdvancedSettingWidget, {
      flux: this.props.flux,
      id: this.props.path,
      name: this.props.path,
      title: this.props.title,
      showTitle: false,
      type: this.props.type,
      value: this.props.value,
      format: this.props.format(this._i18n),
      options: options.map(option => ({
        value: option,
        label: this.props.format(this._i18n)(option, this.props.validStrings)
      })),
      sanitize: this.props.sanitize,
      step: this.props.step,
      min: this.props.min,
      max: this.props.max,
      units: this.props.unit,
      disabled: false,
      onChange: val => this.props.onChange(this.props.path, val, this.props.type)
    })), /*#__PURE__*/React.createElement("div", {
      className: "toggle"
    }, tooltip));
  }

}

AdvancedSettingToggle.displayName = 'AdvancedSettingToggle';
AdvancedSettingToggle.propTypes = {
  flux: React.PropTypes.shape({
    store: React.PropTypes.func.isRequired,
    getService: React.PropTypes.func.isRequired
  }).isRequired,
  active: React.PropTypes.bool.isRequired,
  name: React.PropTypes.string.isRequired,
  title: React.PropTypes.string.isRequired,
  description: React.PropTypes.string,
  rationale: React.PropTypes.string,
  type: React.PropTypes.string.isRequired,
  value: React.PropTypes.any.isRequired,
  format: React.PropTypes.func.isRequired,
  options: React.PropTypes.oneOfType([React.PropTypes.object, React.PropTypes.array]),
  step: React.PropTypes.number,
  min: React.PropTypes.number,
  max: React.PropTypes.number,
  units: React.PropTypes.string,
  onChange: React.PropTypes.func,
  onToggle: React.PropTypes.func
};
module.exports = AdvancedSettingToggle;
