// MB-support-plugin/lib/ui/views/printSettings/advancedSettingsModal.js
'use strict';

const React = require('react');

const R = require('ramda');

const path = require('path');

const Isvg = require('react-inlinesvg');

const Button = require('eagle-print/views//ui/button');

const FluxComponent = require('eagle-print/views/flux_component');

const q = require('q');

const AdvancedItem = require('./advancedSettingToggle');

const util = require('../../../util');

const ui = require('../../../utils/printSettings/uiUtils');

const PrintSettingsConstants = require('../../../constants/print_settings');

const reducers = require('../../../utils/printSettings/reducers');

const GroupKeys = R.omit(['BasicSettings'], PrintSettingsConstants.GroupKeys); // don't need Basic Settings here

const SettingsKeys = PrintSettingsConstants.SettingsKeys;

const PrintModeElement = require('./printMode');

const Listbox = require('../input/listbox');

const ButtonWithDropdown = require('../buttonWithDropdown');

const Icons = Object.freeze({
  overview: '../images/quick-info-settings.svg',
  [GroupKeys.QuickSettings]: '../images/as-quick-settings.svg',
  [GroupKeys.Printer]: '../images/as-printers.svg',
  [GroupKeys.Extruder]: '../images/as-extruder.svg',
  [GroupKeys.Roofs]: '../images/as-roof.svg',
  [GroupKeys.Shells]: '../images/as-shells.svg',
  [GroupKeys.Infill]: '../images/as-infill.svg',
  [GroupKeys.Floors]: '../images/as-floors.svg',
  [GroupKeys.SupportsBridging]: '../images/as-supports-and-bridging.svg',
  [GroupKeys.Rafts]: '../images/as-rafts-base.svg',
  [GroupKeys.Brims]: '../images/as-rafts-base.svg',
  [GroupKeys.BaseLayer]: '../images/as-rafts-base.svg'
});
const EEIcon = '../images/as-extruder-ee.svg';

class AdvancedSettingsModal extends FluxComponent {
  constructor(props) {
    super(props);
    this.store = this.getFlux().store('PrintSettingsStore');
    this.actions = this.getFlux().actions.PrintSettingsActions;
    this.i18n = this.props.flux.getService('TranslationService');
    this.savePropValue = ui.savePropValue(this.getFlux(), this.props.project);
    this.onChange = this.onChange.bind(this);
    this.state = R.merge(this.store.getState(), {
      tab: this.state.currentPreset ? GroupKeys.QuickSettings : 'overview'
    });
  }

  componentDidMount() {
    this.store.addListener('change', this.onChange);
    this.onChange();
  }

  componentWillUnmount() {
    this.store.removeListener('change', this.onChange);
  }

  onChange() {
    const {
      tab
    } = this.state;
    const {
      currentPreset
    } = this.store.getState();
    const newTab = currentPreset && isOverview(tab) ? GroupKeys.QuickSettings : tab;
    const newState = R.merge(this.store.getState(), {
      tab: newTab
    });
    this.setState(newState);
  }

  onToggle(setting) {
    this.getFlux().actions.PrintSettingsActions.toggleProp(setting);
  }

  onSelectionChanged(selection) {
    // if selection is undefined, key will default to the tab in state
    const tab = R.reduce((curr, v) => v === selection ? v : curr, this.state.tab, ['overview', ...R.values(GroupKeys)]); // can't switch to "overview" if there's a preset set

    if (this.state.currentPreset && isOverview(tab)) return;
    this.setState({
      tab
    });
  }

  close() {
    this.actions.closeAdvSettingsModal(this.getFlux());
  } // this function is here because there's some confusion as to whether "close" and "done" are the same thing...


  done() {
    this.close();
  }

  jumpToSetting(setting) {
    if (R.has(setting, this.state.settings)) {
      // no need to check for undefined, because onSelectionChanged() will handle it
      const group = R.path([setting, 'groups', 0], this.state.settings);
      this.onSelectionChanged(group);
    }
  }

  renderPresetsPanel() {
    const currentPrintMode = this.state.printMode.selected; //Alphabetize print modes, to keep them in a consistent order

    const printModes = this.state.printMode.supported.sort();

    const icon = preset => reducers.presetIsExperimental(preset) ? /*#__PURE__*/React.createElement(Isvg, {
      src: util.fullIconPath(path.resolve(__dirname, EEIcon))
    }) : null;

    const mapPresets = preset => ({
      label: /*#__PURE__*/React.createElement("div", null, ui.withAsterisk(R.identity, this.store.hasOverridesForPreset, preset.name), icon(preset)),
      value: preset.name,
      selected: this.state.currentPreset === preset.name,
      disabled: preset.disabled
    });

    const presets = R.map(mapPresets, R.sortBy(R.prop('name'), this.state.presets));

    const displayPrintMode = v => this.i18n.t(`mb:printSettings.${v}`);

    const printModeSelected = mode => currentPrintMode === mode && !this.state.currentPreset;

    return /*#__PURE__*/React.createElement("div", {
      className: "print-modes"
    }, /*#__PURE__*/React.createElement("p", null, this.i18n.t(`mb:printSettings.advancedSettings.default`)), /*#__PURE__*/React.createElement("div", {
      className: "default list-box"
    }, /*#__PURE__*/React.createElement("ul", {
      className: "list-box-items"
    }, printModes.map(v => /*#__PURE__*/React.createElement("li", {
      key: v,
      className: `list-box-item${printModeSelected(v) ? ' list-box-item-selected' : ''}`,
      onClick: () => {
        this.savePropValue('occurrence:printMode', v, 'string');
        return q().then(() => this.actions.unapplyPreset()).then(() => this.actions.updatePrintMode(v)).done();
      }
    }, ui.withAsterisk(displayPrintMode, this.store.hasOverridesForPrintMode, v))))), /*#__PURE__*/React.createElement("p", null, this.i18n.t(`mb:printSettings.advancedSettings.custom`)), /*#__PURE__*/React.createElement(Listbox, {
      className: "custom",
      items: presets,
      newItem: this.state.newPresetName,
      onAdd: () => {
        this.actions.createPreset(this.state.newPresetName, currentPrintMode).then(() => this.setState({
          newPresetName: null
        }));
      },
      onRemove: () => this.actions.destroyPreset(this.state.currentPreset),
      onSelect: p => this.actions.applyPreset(p),
      onChange: s => this.setState({
        newPresetName: s
      }),
      placeholder: this.i18n.t('mb:printSettings.presets.nameModal.title.new'),
      disabled: false,
      addLabel: "+",
      removeLabel: "-",
      error: this.state.presetError.map(R.T).lift(),
      emptyLabel: this.i18n.t('mb:printSettings.presets.noneAvailable')
    }));
  }

  renderGroupsPanel() {
    const className = v => (this.state.tab === v ? `${v} active` : v) + (this.state.currentPreset && isOverview(v) ? ' disabled' : '');

    return /*#__PURE__*/React.createElement("ul", {
      className: "advanced-settings-tabs"
    }, ['overview', ...Object.values(GroupKeys)].map(groupName => /*#__PURE__*/React.createElement("li", {
      className: className(groupName),
      key: groupName,
      onClick: () => this.onSelectionChanged(groupName)
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: util.fullIconPath(path.resolve(__dirname, Icons[groupName]))
    }), this.i18n.t(`mb:printSettings.groups.${groupName}`))));
  }

  renderSettingsPanel() {
    const currentPrinter = this.getFlux().store('PrinterStore').getSelectedPrinter();
    const quickSettingsIsSelected = this.state.tab === 'quick_settings';

    const advancedItemProps = (v, k) => {
      return R.merge(v, {
        flux: this.getFlux(),
        name: k,
        title: v.name,
        active: v.visible,
        disabled: k === SettingsKeys.Supports,
        onToggle: this.onToggle.bind(this),
        onChange: this.props.onChangeProp,
        // A temporary hack, to avoid testing errors caused by the fact that the
        // base layer surface temp setting's fields are not properly populated
        value: v.value === null || v.value === undefined ? 0 : v.value
      });
    };

    const elem = (v, k) => {
      const isSupportsElement = k === 'supportType';
      if (!currentPrinter.isSixthGen() && quickSettingsIsSelected && isSupportsElement) return;else {
        return /*#__PURE__*/React.createElement("li", {
          key: k
        }, /*#__PURE__*/React.createElement(AdvancedItem, advancedItemProps(v, k)));
      }
    };

    const renderOverrides = R.pipe(R.pickBy(R.pipe(R.prop('groups'), R.contains(this.state.tab))), R.mapObjIndexed(elem), R.values);
    const printMode = this.state.printMode.selected;

    if (isOverview(this.state.tab)) {
      return /*#__PURE__*/React.createElement("div", {
        className: "print-mode-view"
      }, /*#__PURE__*/React.createElement(PrintModeElement, {
        i18n: this.i18n,
        printer: currentPrinter,
        opt: printMode,
        selected: false,
        disabled: true,
        onClick: () => {}
      }));
    } else {
      const overrides = renderOverrides(this.state.settings);
      return /*#__PURE__*/React.createElement("div", {
        className: "panel-body advanced-settings-list-container"
      }, /*#__PURE__*/React.createElement("ul", {
        className: "advanced-settings-list"
      }, overrides.length > 0 ? overrides : /*#__PURE__*/React.createElement("li", {
        className: "advanced-settings-none"
      }, "No available settings")));
    }
  }

  renderSettingsOptions() {
    return /*#__PURE__*/React.createElement(ButtonWithDropdown, {
      i18n: this._i18n,
      flux: this.props.flux,
      id: "resetSettingsButton",
      classes: "secondary1 resetSettingsButton",
      dropdownClasses: "secondary1 resetSettingsDropdown",
      label: this.i18n.t('mb:printSettings.right_click.reset_all_to_default'),
      onClick: () => this.actions.resetAllToDefault(),
      menuItems: [{
        path: 'removeAllSettings',
        label: this.i18n.t('mb:printSettings.right_click.remove_all_settings'),
        action: () => this.actions.removeAllSettings()
      }]
    });
  }

  renderPresetActions() {
    const actionsDict = {
      import: {
        onClick: () => this.actions.readPresets(),
        disabled: false
      },
      export: {
        onClick: () => this.actions.writePreset(this.state.currentPreset),
        disabled: R.isNil(this.state.currentPreset)
      },
      save: {
        onClick: () => this.actions.updatePreset(this.state.currentPreset),
        disabled: R.isNil(this.state.currentPreset) || !this.store.hasOverridesForPreset(this.state.currentPreset)
      }
    };

    const actionsMapper = (v, k) => /*#__PURE__*/React.createElement(Button, {
      type: "secondary",
      key: k,
      label: this.i18n.t(`mb:printSettings.presets.${k}`),
      onClick: v.onClick,
      disabled: v.disabled
    });

    return /*#__PURE__*/React.createElement("div", {
      className: "active-settings-preset-actions"
    }, /*#__PURE__*/React.createElement("div", {
      className: "buttonGroup"
    }, this.renderSettingsOptions(), R.values(R.mapObjIndexed(actionsMapper, actionsDict))));
  }

  renderSelectedSettings() {
    const elem = (v, k) => /*#__PURE__*/React.createElement("a", {
      key: k,
      onClick: () => this.jumpToSetting(k)
    }, v.name);

    const renderSettings = R.pipe(R.pickBy(R.prop('visible')), R.mapObjIndexed(elem), R.values);
    return /*#__PURE__*/React.createElement("div", {
      className: "advanced-settings-active"
    }, /*#__PURE__*/React.createElement("strong", null, this.i18n.t(`mb:printSettings.advancedSettings.added`)), renderSettings(this.state.settings));
  }

  render() {
    if (!this.state) return /*#__PURE__*/React.createElement("div", null);
    return /*#__PURE__*/React.createElement("div", {
      className: "advanced-settings-modal"
    }, /*#__PURE__*/React.createElement("div", {
      className: "systemPreferences"
    }, /*#__PURE__*/React.createElement("div", {
      className: "left"
    }, this.renderPresetsPanel()), /*#__PURE__*/React.createElement("div", {
      className: "left"
    }, /*#__PURE__*/React.createElement("p", null, this.i18n.t(`mb:printSettings.advancedSettings.settings`)), this.renderGroupsPanel()), /*#__PURE__*/React.createElement("div", {
      className: "right"
    }, /*#__PURE__*/React.createElement("p", {
      style: {
        textAlign: 'right'
      }
    }, this.i18n.t(`mb:printSettings.advancedSettings.settingsVisible`)), this.renderSettingsPanel())), /*#__PURE__*/React.createElement("div", {
      className: "bottom"
    }, this.renderSelectedSettings(), /*#__PURE__*/React.createElement("div", {
      className: "buttonGroup"
    }, this.renderPresetActions(), /*#__PURE__*/React.createElement(Button, {
      classNames: "doneButton",
      type: "secondary",
      label: "Done",
      onClick: this.done.bind(this)
    }))));
  }

}

const isOverview = v => v === 'overview';

AdvancedSettingsModal.displayName = 'AdvancedSettingsModal';
AdvancedSettingsModal.propTypes = {
  flux: React.PropTypes.shape({
    store: React.PropTypes.func.isRequired,
    getService: React.PropTypes.func.isRequired
  }).isRequired,
  project: React.PropTypes.object,
  onChangeProp: React.PropTypes.func
};
module.exports = AdvancedSettingsModal;
