// MB-support-plugin/lib/ui/views/printSettings/printModesModal.js
'use strict';

const React = require('react');

const Button = require('eagle-print/views/ui/button');

const PrintModes = require('../../../constants/print_settings').PrintModes;

const PrintModeElement = require('./printMode');

class PrintModesModal extends React.Component {
  constructor(props) {
    super(props);
    this.i18n = props.flux.getService('TranslationService');
    this._printSettingsStore = props.flux.store('PrintSettingsStore');
    this._printSettingsActions = props.flux.actions.PrintSettings;
    this._getCurrentPrinter = props.flux.store('PrinterStore').getSelectedPrinter();
    this.onChange = this.onChange.bind(this);
  }

  componentDidMount() {
    this._printSettingsStore.addListener('change', this.onChange);
  }

  componentWillUnmount() {
    this._printSettingsStore.removeListener('change', this.onChange);
  }

  onChange() {
    const state = this._printSettingsStore.getState();

    this.setState(state);
  }

  _getPrintModeOptions() {
    const options = [];

    if (this._getCurrentPrinter.isSixthGen()) {
      options.push(PrintModes.Balanced, PrintModes.Draft, PrintModes.HighQuality, PrintModes.Solid);
    } else {
      options.push(PrintModes.Balanced, PrintModes.Draft, PrintModes.Minfill);
    }

    return options;
  }

  _renderOptions() {
    const {
      printMode,
      currentPreset
    } = this._printSettingsStore.getState();

    const optionElements = this._getPrintModeOptions().map(opt => /*#__PURE__*/React.createElement(PrintModeElement, {
      i18n: this.i18n,
      opt: opt,
      key: opt,
      printer: this._getCurrentPrinter,
      selected: opt === printMode.selected && !currentPreset,
      disabled: !printMode.supported.includes(opt) //temp hack: user can only click on available print modes that aren't already selected -> UI should be a radial button not a checkbox
      ,
      onClick: () => printMode.supported.includes(opt) && !(opt === printMode.selected && !currentPreset) ? this.props.handlePrintModeChange(opt) : null
    }));

    return /*#__PURE__*/React.createElement("div", {
      className: "print-modes-modal__options"
    }, optionElements);
  }

  _renderFooter() {
    const addCustomSettingButton = /*#__PURE__*/React.createElement(Button, {
      id: "add-custom-print-setting",
      type: "secondary",
      label: "Custom Settings",
      onClick: () => this.props.launchAdvSettingsModal()
    });
    return /*#__PURE__*/React.createElement("div", {
      className: "print-modes-modal__footer"
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      className: "text"
    }, this.i18n.t('mb:printModes.footer')), addCustomSettingButton), /*#__PURE__*/React.createElement(Button, {
      id: "doneButton",
      type: "secondary",
      label: this.i18n.t('mb:printModes.done'),
      onClick: () => {
        this.props.flux.getService('ModalService').hide();
      }
    }));
  }

  render() {
    return /*#__PURE__*/React.createElement("div", {
      className: "print-modes-modal"
    }, this._renderOptions(), this._renderFooter());
  }

}

PrintModesModal.displayName = 'PrintModesModal';
PrintModesModal.propTypes = {
  flux: React.PropTypes.object.isRequired,
  printModes: React.PropTypes.array.isRequired,
  handlePrintModeChange: React.PropTypes.func.isRequired,
  launchAdvSettingsModal: React.PropTypes.func
};
module.exports = PrintModesModal;
