// MB-support-plugin/lib/ui/views/printSettings/printMode.js
"use strict";

const React = require('react');

const path = require('path');

const CheckBox = require('eagle-print/views/ui/input_checkbox');

const Isvg = require('react-inlinesvg');

const util = require('../../../util');

class PrintModeElement extends React.Component {
  constructor(props) {
    super(props);
  }

  printModeIcon(printMode) {
    const isSixthGenDraftMode = this.props.printer && this.props.printer.isSixthGen() && printMode === 'draft';
    return util.getPathForIsvg(path.resolve(__dirname, `../images/${isSixthGenDraftMode ? 'draft_gen6' : printMode}.svg`));
  }

  renderBox(opt, selected, disabled, onClick) {
    return /*#__PURE__*/React.createElement("div", {
      className: `box${selected ? '' : ' unselected'}`,
      onClick: onClick
    }, disabled ? null : /*#__PURE__*/React.createElement(CheckBox, {
      checked: selected,
      readOnly: true
    }), /*#__PURE__*/React.createElement(Isvg, {
      src: this.printModeIcon(opt)
    }));
  }

  render() {
    return /*#__PURE__*/React.createElement("div", {
      className: `option ${this.props.opt}${this.props.disabled ? ' disabled' : ''}`,
      key: this.props.opt
    }, this.renderBox(this.props.opt, this.props.selected, this.props.disabled, this.props.onClick), /*#__PURE__*/React.createElement("div", {
      className: "mode-name"
    }, this.props.i18n.t(`mb:printModes.${this.props.opt}.name`)), /*#__PURE__*/React.createElement("div", {
      className: "mode-description"
    }, util.pSplit(this.props.i18n.t(`mb:printModes.${this.props.opt}.description`))));
  }

}

PrintModeElement.propTypes = {
  i18n: React.PropTypes.object,
  opt: React.PropTypes.string,
  selected: React.PropTypes.bool,
  disabled: React.PropTypes.bool,
  onClick: React.PropTypes.func,
  printer: React.PropTypes.object
};
module.exports = PrintModeElement;
