// MB-support-plugin/lib/ui/views/printSettings/printSettingsPanel.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const React = require('react');

const R = require('ramda');

const Isvg = require('react-inlinesvg');

const q = require('q');

const path = require('path');

const EaglePrint = require('eagle-print');

const Dropdown = require('eagle-print/views/ui/input_dropdown');

const Button = require('eagle-print/views/ui/button');

const FluxComponent = EaglePrint.views.FluxComponent;
const PropTypes = EaglePrint.views.PropTypes;

const Widget = require('./advancedSettingWidget');

const PrintModesModal = require('./printModesModal');

const {
  ExtruderTypeEnum,
  DefaultExtruderPairs
} = require('../../../constants');

const Accordion = require('./advancedSettingsAccordion');

const ui = require('../../../utils/printSettings/uiUtils');

const {
  either
} = require('../../../utils/maybe');

const utils = require('../../../util');

const LinkClickUtils = require('../../../utils/linkClick');

const UrlConst = require('../../../constants/urls').url;

const ButtonWithDropdown = require('../buttonWithDropdown');

const UnknownMaterialString = 'generic_model';
const UnknownMaterial = 'Unknown Material';
/**
 *  This class retrieves and sets print setting for tray bodies as well as tray settings.
 *  Examples of tray body settings are layer Height, rafts, and infill Density.
 *  It is implemented by the right toolbar (propertiesToolbar.js) and is refered to as
 *  the material section.
 */

class PrintSettingsPanel extends FluxComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "_applyToAll", true);

    _defineProperty(this, "_trackingService", this.props.flux.getService('TrackingService'));

    _defineProperty(this, "_i18n", this.getFlux().getService('TranslationService'));

    _defineProperty(this, "_loggingService", this.props.flux.getService('LoggingService'));

    _defineProperty(this, "_store", this.getFlux().store('PrintSettingsStore'));

    _defineProperty(this, "_actions", this.getFlux().actions.PrintSettingsActions);

    _defineProperty(this, "savePropValue", ui.savePropValue(this.getFlux(), this.props.project));

    _defineProperty(this, "state", this._store.getState());

    _defineProperty(this, "componentWillMount", () => {
      super.componentWillMount(); // Add Listeners

      this._store.addListener('change', this.onChange);

      this._store.addListener('close_panel', this.closePanel);

      this._store.addListener('initialize_values', () => {
        this._actions.loadSchema({
          id: this.props.printer.getId()
        });
      });
    });

    _defineProperty(this, "componentDidMount", () => {
      super.componentDidMount();

      this._actions.loadSchema({
        id: this.props.printer.getId()
      });
    });

    _defineProperty(this, "componentWillUnmount", () => {
      super.componentWillUnmount(); // Remove Listeners

      this._store.removeListener('change', this.onChange);

      this._store.removeListener('close_panel', this.closePanel);

      this._store.removeListener('initialize_values', () => {
        this._actions.loadSchema({
          id: this.props.printer.getId()
        });
      });
    });

    _defineProperty(this, "closePanel", () => {
      if (this.props.isActive) {
        // NOTE:  we need to solve this in a better way at some point
        // perhaps by implementing an app wide dispatch queue.
        // For now, this solves dispatch collision thats causing error
        // and resulting in panel not closing.
        q.delay(1).then(() => this.props.closePanel());
      }
    });

    _defineProperty(this, "onChange", () => {
      const state = this._store.getState();

      this.setState(state);
    });

    _defineProperty(this, "getSelectedOccIds", () => {
      return this.props.project.getOccurrences().toArray().map(occ => occ.getId());
    });

    _defineProperty(this, "handleExtruderChange", (val, index) => {
      const displayName = this.displayExtruder(val, index);
      const currentExtruders = this.props.printer.getAttachedExtruders();
      currentExtruders[index] = val;
      this.props.printer.setAttachedExtruders(currentExtruders); // Even though we're now storing extruders in the printer object, not the tray
      // we still need savePropValue because it has the side effect of correctly
      // clearing the build.

      this.savePropValue('occurrence:extruders', currentExtruders, 'array');
      return q(this._actions.updateExtruder()).then(() => this.trackChange(displayName)).catch(err => this._loggingService.error(err)).done();
    });

    _defineProperty(this, "handleMaterialChange", (val, index) => {
      const currentMaterials = this.state.materials.selected;
      currentMaterials[index] = val;
      this.savePropValue('occurrence:materials', currentMaterials, 'array');
      return q().then(() => this._actions.updateMaterial(index, val)).then(() => this.trackChange(val)).catch(err => this._loggingService.error(err)).done();
    });

    _defineProperty(this, "handlePrintModeChange", val => {
      const displayName = this.displayPrintMode(val);
      this.savePropValue('occurrence:printMode', val, 'string');
      return q().then(() => this._actions.updatePrintMode(val)).then(() => this.trackChange(displayName)).catch(err => this._loggingService.error(err)).done();
    });

    _defineProperty(this, "trackChange", displayName => {
      const merged = R.merge({
        name: `${this.label} Change`,
        change_to: displayName
      }, this.props.printer.getTrackingInfo());

      this._trackingService.addEvent(merged);
    });

    _defineProperty(this, "displayExtruder", extruderType => {
      if (!extruderType) return null; // should fallback to just displaying the `extruderType` if there's
      // nothing specified in the translation file

      return this.props.i18n([`mb:printSettings.${extruderType}`, extruderType]);
    });

    _defineProperty(this, "displayPrintMode", v => {
      return this.props.i18n(`mb:printSettings.${v}`);
    });

    _defineProperty(this, "launchAdvSettingsModal", () => {
      this._actions.launchAdvSettingsModal(this.props.flux, {
        project: this.props.project,
        onChangeProp: this.savePropValue.bind(this),
        i18n: this.props.i18n
      });
    });

    _defineProperty(this, "advancedSettings", () => {
      const settings = R.pipe(ui.sortForPanel, R.map(this._accordion.bind(this)))(this.state.settings);
      return /*#__PURE__*/React.createElement("div", {
        className: ".advanced-settings-widgets"
      }, settings);
    });

    _defineProperty(this, "_accordion", ([group, settings]) => {
      const setting = settings.map(this._buildWidget.bind(this));
      return /*#__PURE__*/React.createElement(Accordion, {
        key: group,
        accordionPosition: null,
        classParentString: group,
        trigger: group
      }, setting);
    });

    _defineProperty(this, "_buildWidget", ([k, v]) => {
      const options = Array.isArray(v.options) ? v.options : Object.keys(v.options);
      return /*#__PURE__*/React.createElement(Widget, {
        flux: this.props.flux,
        key: k,
        id: v.path,
        name: v.path,
        title: v.name,
        type: v.type // A temporary hack, to avoid testing errors caused by the fact that the
        // base layer surface temp setting's fields are not properly populated
        ,
        value: v.value === null || v.value === undefined ? 0 : v.value,
        format: v.format(this._i18n),
        options: options.map(option => ({
          value: option,
          label: v.format(this._i18n)(option, v.validStrings)
        })),
        sanitize: v.sanitize,
        step: v.step,
        min: v.min,
        max: v.max,
        units: v.unit,
        onChange: val => this.savePropValue(v.path, val, v.type)
      });
    });

    _defineProperty(this, "renderExtruderTypesView", () => {
      const attachedExtruders = this.props.printer.getAttachedExtruders();
      const isSketchPrinter = this.props.printer.isSketchPrinter();
      return attachedExtruders.length > 1 || isSketchPrinter ? this.renderExtendedExtruderView() : this.renderSimpleExtruderView();
    });

    _defineProperty(this, "renderExtendedExtruderView", () => {
      const attachedExtruders = this.props.printer.getAttachedExtruders();
      const supportedExtruders = this.getExtruderOptions(); // hack to avoid timing issue where extruder vals  don't keep up when switching bots

      if (attachedExtruders.length !== supportedExtruders.length) {
        return this.closePanel();
      }

      return /*#__PURE__*/React.createElement(Accordion, {
        key: "extruders",
        accordionPosition: null,
        classParentString: "extruders",
        trigger: "extruders"
      }, /*#__PURE__*/React.createElement("div", null, attachedExtruders.map((extruder, index) => {
        const extruderPath = `occurrence:attached_extruder${index}`;
        return /*#__PURE__*/React.createElement("div", {
          key: extruderPath,
          id: extruderPath,
          className: "padding-sides-20"
        }, /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", {
          className: "prop-hdr"
        }, `Extruder Slot ${index + 1}`), /*#__PURE__*/React.createElement(Dropdown, {
          name: `attached_extruder${index}`,
          value: attachedExtruders[index],
          options: supportedExtruders[index],
          onChange: event => this.handlePairExtruderChange(event.value, index) // used only here to prevent selections for live printers
          ,
          isLivePrinter: this.isLivePrinter()
        }), this.renderMaterialTypesView(index), /*#__PURE__*/React.createElement("br", null));
      })));
    });

    _defineProperty(this, "renderSimpleExtruderView", () => {
      const attachedExtruders = this.props.printer.getAttachedExtruders();
      const supportedExtruders = this.getExtruderOptions();
      const extruderPath = 'occurrence:attached_extruder0';
      return /*#__PURE__*/React.createElement("div", {
        className: "padding-sides-20"
      }, /*#__PURE__*/React.createElement("h4", {
        className: "prop-hdr"
      }, "Extruder Type"), /*#__PURE__*/React.createElement("div", {
        key: extruderPath,
        id: extruderPath
      }, /*#__PURE__*/React.createElement(Dropdown, {
        name: "attached_extruder0",
        value: attachedExtruders[0],
        options: supportedExtruders[0],
        onChange: event => this.handleExtruderChange(event.value, 0) // used only here to prevent selections for live printers
        ,
        isLivePrinter: this.isLivePrinter()
      })));
    });

    _defineProperty(this, "getMaterialOptions", extruderIndex => {
      const {
        supported,
        spoolMaterials,
        translations
      } = this.state.materials;
      const isBotLive = this.props.printer.isConnected() && this.props.printer.isAuthenticated();
      const attachedExtruder = this.props.printer.getAttachedExtruders()[extruderIndex];
      const extruderIsLabs = attachedExtruder === ExtruderTypeEnum.mk14_e;
      const isSketchPrinter = this.props.printer.isSketchPrinter();
      const materialOptions = isBotLive && !extruderIsLabs && !isSketchPrinter ? [spoolMaterials[extruderIndex]] : supported[extruderIndex];
      return materialOptions.map(material => {
        const materialName = material || UnknownMaterialString; // To avoid crashing the panel, provide a fallback to material
        // translations

        const getMaterialTranslation = material => {
          if (!material) {
            return UnknownMaterial;
          } else {
            return translations[material] ? translations[material].name : material;
          }
        };

        return {
          value: materialName,
          label: getMaterialTranslation(material)
        };
      });
    });

    _defineProperty(this, "renderMaterialTypesView", extruderIndex => {
      const selected = this.state.materials.selected[extruderIndex] || UnknownMaterialString;
      const materialPath = `occurrence:material${extruderIndex}`;
      const numberOfExtruders = this.props.printer.getAttachedExtruders().length;
      const isSketchPrinter = this.props.printer.isSketchPrinter(); //there is no material selection for single-extrusion bots except 'sketch'
      //for those, material is always inferred from extruder

      if (!(numberOfExtruders > 1) && !isSketchPrinter) {
        return;
      }

      return /*#__PURE__*/React.createElement("div", {
        key: materialPath,
        id: materialPath
      }, /*#__PURE__*/React.createElement(Dropdown, {
        name: "materials",
        value: selected,
        options: this.getMaterialOptions(extruderIndex),
        onChange: event => this.handleMaterialChange(event.value, extruderIndex)
      }));
    });

    _defineProperty(this, "renderPrintModeView", () => {
      const isSixthGen = this.props.printer.isSixthGen();
      const printMode = this.state.printMode;
      const mode = either(printMode.selected, this.state.currentPreset).left(ui.withAsterisk(this.displayPrintMode.bind(this), this._store.hasOverridesForPrintMode)).right(ui.withAsterisk(utils.truncate('…', 25), this._store.hasOverridesForPreset)).lift();
      return /*#__PURE__*/React.createElement("div", {
        className: "padding-sides-20",
        key: "occurrence:printMode",
        id: "occurrence:printMode"
      }, /*#__PURE__*/React.createElement("h4", {
        className: "prop-hdr"
      }, this.props.i18n('mb:printSettings.printMode')), /*#__PURE__*/React.createElement("div", {
        id: "printModesButton",
        onClick: () => {
          {
            this.props.flux.getService('ModalService').show( /*#__PURE__*/React.createElement(PrintModesModal, {
              flux: this.props.flux,
              printModes: printMode.supported,
              handlePrintModeChange: this.handlePrintModeChange.bind(this),
              launchAdvSettingsModal: () => this.launchAdvSettingsModal()
            }), {
              title: /*#__PURE__*/React.createElement("a", {
                className: "supportLink",
                onClick: () => {
                  LinkClickUtils.clickAction(this.props.flux, 'default_print_modes_link_clicked', UrlConst.defaultPrintModes);
                }
              }, this.props.i18n('mb:printModes.title'), /*#__PURE__*/React.createElement(Isvg, {
                className: "helperLinkIcon",
                src: "icons/helpful-link-white.svg"
              })),
              modalClass: isSixthGen ? 'w800' : 'w650'
            });
          }
        }
      }, mode));
    });

    _defineProperty(this, "renderCustomSettingsButton", () => {
      return /*#__PURE__*/React.createElement(Button, {
        id: "add-custom-print-setting",
        type: "secondary",
        label: "Custom Settings",
        onClick: () => this.launchAdvSettingsModal()
      });
    });

    _defineProperty(this, "renderResetDropdown", () => {
      return /*#__PURE__*/React.createElement(ButtonWithDropdown, {
        i18n: this._i18n,
        flux: this.props.flux,
        id: "resetSettingsButton",
        classes: "secondary1 resetSettingsButton",
        dropdownClasses: "secondary1",
        label: this._i18n.t('mb:printSettings.right_click.reset_all_to_default'),
        onClick: () => this._actions.resetAllToDefault(),
        menuItems: [{
          path: 'removeAllSettings',
          label: this._i18n.t('mb:printSettings.right_click.remove_all_settings'),
          action: () => this._actions.removeAllSettings()
        }]
      });
    });

    _defineProperty(this, "render", () => {
      const i18n = this.props.i18n;
      const selectionInfo = this.props.selectionInfo;
      const extruderCount = this.state.extruders.supported.length;
      const renderExtruder = (extruderCount || 0) >= 1;

      if (!selectionInfo || selectionInfo.activeTrayOccurrenceIds.length === 0) {
        return /*#__PURE__*/React.createElement("div", R.merge(this.props, {
          className: `${this.props.className || ''} section`
        }), /*#__PURE__*/React.createElement("div", R.merge(this.props, {
          className: `errorMessage`
        }), i18n('mb:printSettings.noModelPresent')));
      }

      if (R.keys(this.state.settings).length === 0) {
        return /*#__PURE__*/React.createElement("div", R.merge(this.props, {
          className: `${this.props.className || ''} section`
        }), /*#__PURE__*/React.createElement("div", R.merge(this.props, {
          className: `errorMessage`
        }), i18n('mb:printSettings.loadingSettings')));
      }

      return /*#__PURE__*/React.createElement("div", {
        className: `${this.props.className || ''} section padding-sides-20`
      }, this.state.extruders && renderExtruder ? this.renderExtruderTypesView() : null, this.state.printMode ? this.renderPrintModeView() : null, this.advancedSettings(), /*#__PURE__*/React.createElement("div", {
        className: "custom-btn-div"
      }, this.renderCustomSettingsButton()), /*#__PURE__*/React.createElement("div", {
        className: "reset-btn-div"
      }, this.renderResetDropdown()));
    });
  }

  /**
   * Handle pair extruders change by user to avoid
   * extruder mismatch configurations; ie [mk14, mk14_hot_s] or [mk14_hot, mk14_hot_s]
   * @param {string} val - selected extruder id ie. "mk14_hot_s" or "mk13_impla"
   */
  handlePairExtruderChange(val, index) {
    const {
      supportedExtruderPairConfigs
    } = this.state.extruders;
    const otherIndex = 1 - index;
    const newExtruderPair = this.props.printer.getAttachedExtruders();
    newExtruderPair[index] = val;
    const configsWithNewExtruder = supportedExtruderPairConfigs.filter(extruderPair => {
      if (R.contains(val, extruderPair)) {
        return extruderPair;
      }
    }).sort();

    const updateExtruderPair = extruderPair => {
      this.props.printer.setAttachedExtruders(extruderPair); // Even though we're now storing extruders in the printer object, not the tray
      // we still need savePropValue because it has the side effect of correctly
      // clearing the build.

      this.savePropValue('occurrence:extruders', extruderPair, 'array');
      return q(this._actions.updateExtruderPair()).then(() => this.trackChange(this.displayExtruder(val))).catch(err => this._loggingService.error(err)).done();
    };

    if (R.contains(newExtruderPair, configsWithNewExtruder)) {
      updateExtruderPair(newExtruderPair);
    } else {
      // If the user-chosen extruder conflicts with the other, non-changed extruder,
      // update that other extruder, using the default listed in constants
      const configWithDefaultExtruder = [];
      configWithDefaultExtruder[index] = val;
      configWithDefaultExtruder[otherIndex] = DefaultExtruderPairs[val];
      updateExtruderPair(configWithDefaultExtruder);
    }
  }
  /**
   * Handle material change by user.
   * @param{string} val - selected material
   */


  getExtruderOptions() {
    const {
      supportedExtrudersPerIndex
    } = this.state.extruders;
    const attachedExtruders = this.props.printer.getAttachedExtruders(); // Why? Well because Experimental Extruder is a fake mk13 extruder
    // that requires many logic statements to build out functionality
    // throughout the app

    if (attachedExtruders.length === 1 && attachedExtruders[0] === ExtruderTypeEnum.mk13_experimental) {
      return [[{
        value: attachedExtruders[0],
        label: this.displayExtruder(attachedExtruders[0])
      }]];
    }

    return supportedExtrudersPerIndex.map(supportedExtruders => {
      const sortedExtruders = supportedExtruders.sort((extruder1, extruder2) => {
        // A hack to make sure LABS extruder is always last in the list
        if (extruder1 === ExtruderTypeEnum.mk14_e) {
          return 1;
        } else {
          if (extruder2 === ExtruderTypeEnum.mk14_e) {
            return -1;
          }
        } // The order of the non-LABS extruders doesn't matter


        return 0;
      });
      return sortedExtruders.map(extruder => ({
        value: extruder,
        label: this.displayExtruder(extruder)
      }));
    });
  }

  isLivePrinter() {
    const {
      printer
    } = this.props;
    return printer && printer.isConnected() && printer.isAuthenticated();
  }

}

PrintSettingsPanel.displayName = 'PrintSettingsPanel';
PrintSettingsPanel.propTypes = R.merge(FluxComponent.propTypes, {
  flux: React.PropTypes.object.isRequired,
  i18n: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  project: React.PropTypes.shape({
    getOccurrences: React.PropTypes.func.isRequired,
    getRootOccurrences: React.PropTypes.func.isRequired
  }).isRequired,
  tray: React.PropTypes.object.isRequired,
  selectionInfo: PropTypes.selectionInfo // optional selections

});
module.exports = PrintSettingsPanel;
