// MB-support-plugin/lib/utils/printSettings/reducers.js
"use strict";

const R = require('ramda');

const {
  roundPrec,
  escapeRegExp
} = require('../../util');

const {
  path
} = require('./format');

const L = require('./lenses');

const {
  SettingsKeys,
  EESettingsKeys,
  PresetKeyBlacklist
} = require('../../constants/print_settings'); // getGrouping :: String -> {k: v} -> [k]


const getGrouping = R.uncurryN(2, k => R.pipe(R.pickBy(R.contains(k)), R.keys)); // getOptions :: String -> {k: v} -> (Num -> Num -> Num -> [Num])

const getOptions = (k, options, setting) => {
  return R.propOr(() => R.propOr([], 'validStrings', setting), k, options);
};
/**
 * This formats a schema to add extra properties to make using the values
 * easier to use in the UI. It adds whether an element is visible, what its value is,
 * and what group it belongs to.
 * It's written out like this to make the formatting easier to test.
 *
 * @param {object<string, any>} settings - all of the settings that came from Slice Config.
 * @param {object<string, boolean>} visible - which keys in `settings` should be considered visible.
 * @param {object<string, any>} vals - the value of each item in `settings`.
 * @param {object<string, string[]>} grouping - a dictionary of the grouping for each item
 * @param {PrintSettingsConstants.Options} options - a dictionary for converting ranges to lists for dropdowns
 * @param {PrintSettingsConstants.Format} formatting - a dictionary for formatting values of settings
 * @param {PrintSettingsConstants.Sanitization} sanitization - a dictionary for validation of settings
 * @param {PrintSettingsConstants.SettingsTypes} typeOverrides - a dictionary for overriding setting type
 *
 * @returns {object<string, any>}
 */


const state = (settings, visible, vals, grouping, options, formatting, sanitization, typeOverrides) => {
  return R.mapObjIndexed((v, k) => R.merge(v, {
    path: path(k),
    visible: R.propOr(false, k, visible),
    value: R.propOr(v.initial_value, k, vals),
    groups: getGrouping(k, grouping),
    options: getOptions(k, options, v)(v.min, v.max + v.step, v.step),
    format: R.propOr(() => R.identity, k, formatting),
    sanitize: R.propOr(R.identity, k, sanitization),
    type: R.propOr(v.type, k, typeOverrides)
  }), settings);
};
/**
 * This function was made in frustration. There are a few settings from slice config whose default settings
 * (initial_value) don't match what Product wants them to be, and apparently that's the UI's problem.
 * So, instead of trying to get clever, I said "screw it" and just overrode them, since they're always
 * the same and never change.
 * @param {object} settings - the settings as they came in from slice config
 * @returns {object} - New and improved with defaults that match Product's specs
 */


const initialValueOverrides = R.over(R.compose(L.layerHeight, L.initialValue), roundPrec(1));
/**
 * Parses the available settings coming from `schema`. Most of it is just a pretty straight forward
 * list concatenation.
 * Note that extruders and print modes come in from the schema seperately from these settings.
 * @param {object} schema - the schema that comes back from the printer
 * @returns {object} - the schema parsed into the proper format for availableSettings
 */

const availableSettings = settings => {
  const s = R.omit(['basic_parameters', 'custom_parameters'], settings);
  if (R.keys(s).length === 0) return {};
  return R.pipe(initialValueOverrides)(s);
};
/**
 * Overall this a pretty mundane function, but we needed to handle some
 * special cases regarding what settings can and cannot be made visible
 * @returns {object<string, bool>} - a map of visible settings
 */


const visibleSettings = () => ({
  // Supports are a special case since it's the only advanced setting
  // that always appears in the panel by default
  [SettingsKeys.Supports]: false
});

const cleanValues = (vals, visible) => R.pipe(R.pickBy((v, k) => R.propOr(false, k, visible) === true), R.assoc('applyToAll', true))(vals);
/**
 * A fairly straight-forward helper function to determine whether the current
 * value is valid. If so, use it, if not, use the default.
 * @param {any} currentVal - the current value
 * @param {any} defaultVal - the default value
 * @param {any[]} validList - the list of valid values
 * @returns {any} - either the current value or the default
 */


const normalizeValue = (currentVal, defaultVal, validList) => currentVal && R.contains(currentVal, validList) ? currentVal : defaultVal; // getPresetByName :: String -> [Preset] -> Preset


const getPresetByName = R.curry((name, presets) => R.pipe(R.filter(R.propEq('name', name)), R.head)(presets)); // removePresetByName :: String -> [Preset] -> Preset

const removePresetByName = R.curry((name, presets) => R.reject(R.propEq('name', name), presets)); // updatePresetByName :: String -> Preset -> [Preset] -> [Preset]

const updatePresetByName = R.curry((name, update, presets) => R.pipe(R.reject(R.propEq('name', name)), R.append(update))(presets)); // valuesToPreset :: String -> String -> {k: v} -> Preset

const valuesToPreset = R.curry((name, values) => {
  return {
    name,
    settings: R.omit(PresetKeyBlacklist, values)
  };
}); // make sure it has a version number
// presetToJSON :: String -> { k: v } -> { k: v }

const presetToJSON = R.curry((scVersion, preset) => R.merge({
  version: scVersion
}, preset)); // sanitizePreset :: Preset -> Preset

const sanitizePreset = R.curry((scVersion, preset, presets) => R.pipe(R.merge({
  version: scVersion
}), // make sure it has a version number, we may use this later
R.evolve({
  version: R.identity,
  settings: R.pipe(R.omit(PresetKeyBlacklist), R.pick(R.values(SettingsKeys))),
  name: incrementPresetName(presets)
}), R.pick(['name', 'settings']))(preset));
const incrementPresetName = R.curry((presets, search) => R.pipe(R.map(R.pipe(R.prop('name'), R.match(new RegExp(`${escapeRegExp(search)}(\\s\\((\\d+)\\))*$`)))), R.filter(R.length), // remove everything that didn't match
R.map(R.pipe(R.nth(2), // grab the number as a string out of the name
R.defaultTo('0'), // if there wasn't a number, make it '0'
parseInt)), R.reduce(R.max, -1), // Find the biggest number
n => n >= 0 ? `${search} (${n + 1})` : search)(presets));
const presetHasSettings = R.curry((keys, preset) => R.reduce((acc, k) => acc || R.has(k, preset.settings), false, keys));
const presetIsExperimental = presetHasSettings(EESettingsKeys);
module.exports = {
  state,
  availableSettings,
  visibleSettings,
  cleanValues,
  normalizeValue,
  initialValueOverrides,
  getGrouping,
  getOptions,
  getPresetByName,
  updatePresetByName,
  removePresetByName,
  valuesToPreset,
  presetToJSON,
  sanitizePreset,
  incrementPresetName,
  presetIsExperimental
};
