// MB-support-plugin/lib/archetype_printer.js
'use strict';

const q = require('q');

const EventEmitter = require('eventemitter3');

const Printer = require('eagle-print/model/printer');

const util = require('./util.js');

const {
  PrinterStateEnum,
  PrinterConnTypeEnum,
  BotTypeEnum,
  GenderToType,
  FifthGenBots,
  SixthGenBots,
  PrinterGenderToDefaultExtruders
} = require('./constants'); // One of these days we'll make a unified interface for real printers and archetypes....


class ArchetypePrinter extends Printer {
  constructor(vendorName, gender) {
    const genderName = gender.getName();
    const printerId = `offline:${vendorName}:${genderName}:${genderName}`;
    const data = {
      id: printerId,
      ip_address: 'OFFLINE',
      name: genderName,
      connected: false,
      printer_status: {
        state: 'Offline',
        // TODO: real statuses here
        current_print_job: null
      },
      printer_info: {
        vendor: vendorName,
        model: genderName,
        modelName: genderName,
        firmware_version: '0.0.0'
      }
    };
    super(data, null, gender);
    const self = this;
    self._gender = gender;
    self._machineConfig = {}; // Notifications for critical updates (system, states, schema updates, etc..)

    self._updateNotifications = new EventEmitter(); // dumping things into this _info object for compatibility with
    // eagle print printer

    self._info = {
      get id() {
        return printerId;
      },

      get name() {
        return genderName;
      },

      get ip() {
        return 'OFFLINE';
      }

    }; // these are status properties generic to eagle-print's printer

    self._info.printer_status = data.printer_status;

    function getDefaultExtruders() {
      const botType = GenderToType[self.getModel()];
      return PrinterGenderToDefaultExtruders[botType];
    }

    self.attachedExtruders = getDefaultExtruders();
  }

  _initPrinter() {// Add machine config code here
    // self._getMachineConfig().done();
  }

  getLastError() {
    return undefined;
  }

  isMissingExtruder() {
    return false;
  }

  getChamberInfo() {
    return [];
  }

  getSpoolInfo() {
    if (this.isSixthGen()) {
      return q([{
        tag_uid: null
      }, {
        tag_uid: null
      }]);
    } else {
      return q(null);
    }
  }

  processMethodExists() {
    return false;
  }

  getTrackingInfo() {
    const data = {
      printer_id: this.getId(),
      printer_type: util.genderToCodeName(this.getInfo().model)
    };
    return data;
  }

  disconnect() {
    return true;
  }

  getInfo() {
    return {
      vendor: 'MakerBot',
      model: this._botGender,
      modelName: this._info.name,
      firmware_version: this.firmwareVersion
    };
  }

  getStatus() {
    return this._info.printer_status;
  }

  getDisplayStatus() {
    return 'ExportOnly';
  }

  isAuthenticated() {
    return true;
  }

  isConnected() {
    return false;
  }

  deauthenticate() {} /// The toolpath that gets here is probably an eagle-print/model/toolpath.js
  /// Toolpath object made from what is returned by 'build' in makerbot.js...


  print(toolPath) {
    const self = this;

    if (self.getStatus().state !== PrinterStateEnum.Idle) {
      // TODO: something something print queueing
      throw 'Unable to print to this printer at this time';
    } // TODO: new print flow support


    const localPath = toolPath.get('path');
    const filename = toolPath.get('name');
    const remotePath = `/current_thing/${filename}`;
    return q(self._printer.Put(localPath, remotePath).then(() => {
      self._printer.Print(filename);
    }));
  }

  cancel() {}

  _getMachineConfig() {
    const self = this;
    return q(self._printer.GetMachineConfig()).then(config => {
      self._machineConfig = config;
    });
  }

  _updateFromExtruderChange(index, config) {// TODO(shirley): make this work, when I actually figure out how to
    // get these notifications from the bot...
  }

  authenticate(authInfo) {}

  _fwVerDictToString(fwVerDict) {
    if (fwVerDict !== undefined) {
      return [fwVerDict.major, fwVerDict.minor, fwVerDict.bugfix, fwVerDict.build].join('.');
    }
  }

  getCodeName() {
    return util.genderToCodeName(this.getGender().getName());
  }

  getConnectionType() {
    return PrinterConnTypeEnum.ARCHETYPE;
  }

  getCameraFeed(cameraFrameCallback) {
    return undefined;
  }

  endCameraFeed() {} // extruderList should be an array, even if there is only one extruder


  setAttachedExtruders(extruderList, index = null) {
    if (index !== null) {
      this.attachedExtruders[index] = extruderList;
    } else {
      this.attachedExtruders = extruderList;
    }
  }

  getAttachedExtruders() {
    return this.attachedExtruders;
  } // doing this (just in case) until eagle-print's printer settles down...


  get rawData() {
    return this._info;
  }

  get machineConfig() {
    return undefined; //this._machineConfig;
  }

  get firmwareVersion() {
    return '0.0.0';
  }

  get apiVersion() {
    return '0.0.0';
  }

  get currentState() {
    return this._info;
  }

  get _botGender() {
    return this._gender.getName();
  }

  get updateNotifications() {
    return this._updateNotifications;
  }

  checkForUnsupportedExtruderConfigurations() {
    return {
      extruderSetupError: false
    };
  }

  hasExtruderMaterialMismatch() {
    return false;
  }

  isFifthGen() {
    const botType = GenderToType[this.getModel()];
    return FifthGenBots.includes(botType);
  }

  isSixthGen() {
    const botType = GenderToType[this.getModel()];
    return SixthGenBots.includes(botType);
  }

  isSketchPrinter() {
    const botType = GenderToType[this.getModel()];
    return botType === BotTypeEnum.sketch;
  }

}

module.exports = ArchetypePrinter;
