// MB-support-plugin/lib/builder.js
'use strict';

const Q = require('q');

const _ = require('lodash');

const BuildProvider = require('eagle-print/services/providers/buildProvider');

const EaglePrint = require('eagle-print');

const makerbot = require('./makerbot');

const Task = EaglePrint.scheduler.Task;

class MakerbotBuildProvider extends BuildProvider {
  constructor() {
    super();
  }

  accept(buildTray) {
    const printer = buildTray.printer;
    return printer && printer.getInfo().vendor.toLowerCase() === 'makerbot';
  } /// Describe the options for the snapshots that the builder bundles
  /// into the dot makerbot files.


  snapshotOptions(printer) {
    /**
     * Standard thumbnails settings to be used on Teams & PrintFromWeb
     * for 5th and 6th Gen printers (PEN-2160)
     */
    const stdThumbnail = {
      isometricSmall: {
        dpi: 1,
        width: 120,
        height: 120,
        projection: 'orthographic',
        camera: {
          angleAndFitToModels: true,
          position: {
            x: -1,
            y: -1,
            z: 1
          },
          up: {
            x: 0,
            y: 0,
            z: 1
          }
        },
        backgroundColor: 0xf2f2f2,
        backgroundOpacity: 0,
        boundingBoxCentered: true,
        objectsToHide: ['tray~node']
      },
      isometricMedium: {
        dpi: 1,
        width: 320,
        height: 320,
        projection: 'orthographic',
        camera: {
          angleAndFitToModels: true,
          position: {
            x: -1,
            y: -1,
            z: 1
          },
          up: {
            x: 0,
            y: 0,
            z: 1
          }
        },
        backgroundColor: 0xf2f2f2,
        backgroundOpacity: 0,
        boundingBoxCentered: true,
        objectsToHide: ['tray~node']
      },
      isometricLarge: {
        dpi: 1,
        width: 640,
        height: 640,
        projection: 'orthographic',
        camera: {
          position: {
            x: -1,
            y: -1,
            z: 1
          },
          up: {
            x: 0,
            y: 0,
            z: 1
          }
        },
        backgroundColor: 0xf2f2f2,
        backgroundOpacity: 0,
        boundingBoxCentered: true,
        objectsToHide: ['grid', 'border', 'innerBorder', 'outerBorder', 'floor', 'floorBottom', 'purgeTower']
      }
    };

    if (printer.isSixthGen()) {
      //Sombrero-style thumbnails
      return Object.assign({}, {
        large: {
          dpi: 1,
          width: 960,
          height: 1460,
          projection: 'orthographic',
          camera: {
            position: {
              x: -1,
              y: -1,
              z: 1
            },
            up: {
              x: 0,
              y: 0,
              z: 1
            }
          },
          backgroundColor: 0x000000,
          boundingBoxCentered: true,
          objectsToHide: ['grid', 'border', 'innerBorder', 'outerBorder', 'floor', 'purgeTower']
        },
        medium: {
          dpi: 1,
          width: 212,
          height: 300,
          projection: 'orthographic',
          camera: {
            position: {
              x: -1,
              y: -1,
              z: 1
            },
            up: {
              x: 0,
              y: 0,
              z: 1
            }
          },
          backgroundColor: 0x000000,
          boundingBoxCentered: true,
          objectsToHide: ['grid', 'border', 'innerBorder', 'outerBorder', 'floor', 'purgeTower']
        },
        small: {
          dpi: 1,
          width: 140,
          height: 106,
          projection: 'orthographic',
          camera: {
            angleAndFitToModels: true,
            position: {
              x: -1,
              y: -1,
              z: 1
            },
            up: {
              x: 0,
              y: 0,
              z: 1
            }
          },
          backgroundColor: 0x000000,
          boundingBoxCentered: true,
          objectsToHide: ['tray~node']
        }
      }, stdThumbnail);
    } else if (printer.isSketchPrinter()) {
      return Object.assign({}, {
        small: {
          dpi: 1,
          width: 90,
          height: 90,
          projection: 'orthographic',
          camera: {
            angleAndFitToModels: true,
            position: {
              x: -1,
              y: -1,
              z: 1
            },
            up: {
              x: 0,
              y: 0,
              z: 1
            }
          },
          backgroundColor: 0x000000,
          backgroundOpacity: 0,
          boundingBoxCentered: true,
          objectsToHide: ['tray~node', 'grid', 'border', 'innerBorder', 'outerBorder', 'floor', 'purgeTower']
        }
      }, stdThumbnail);
    } else {
      //Birdwing-style thumbnails
      return Object.assign({}, {
        large: {
          dpi: 1,
          width: 320,
          height: 200,
          projection: 'isometric',
          camera: {
            fitToModels: true
          },
          objectsToHide: []
        },
        medium: {
          dpi: 1,
          width: 110,
          height: 80,
          projection: 'isometric',
          camera: {
            fitToModels: true
          },
          objectsToHide: []
        },
        small: {
          dpi: 1,
          width: 54,
          height: 40,
          projection: 'isometric',
          camera: {
            fitToModels: true
          },
          objectsToHide: []
        }
      }, stdThumbnail);
    }
  }

  preparePrint(buildTray, flux) {
    const self = this;
    return new Task(function (done, progress, cancelToken) {
      self.getBuiltFile(buildTray, progress, cancelToken, done, flux);
    });
  }

  getBuiltFile(buildTray, progressCallback, cancelToken, callback, flux) {
    const {
      options
    } = buildTray;
    return makerbot.build(buildTray, options ? options : null, progressCallback, cancelToken, null, flux).nodeify(callback);
  } // @return {String} extension of toolpath file generated by build()
  // Always return makerbot even for 4th Gen bots since a .makerbot is
  // required to visualize the toolpath.


  getBuildFileType(printer) {
    return 'makerbot';
  }

  runHealthCheck(callback) {
    return new Q({}).nodeify(callback);
  }

}

module.exports = MakerbotBuildProvider;
