// MB-support-plugin/lib/constants.js
'use strict';

const Immutable = require('immutable');

const {
  darkenColor,
  lightenColor
} = require('./ui/views/utilities/colorManipulator');

const PRINTER_PROVIDER = 'MAKERBOT_PRINTER_PROVIDER';
const PrinterStateEnum = Object.freeze({
  // Unconnected states, managed by printer service
  Pending: 'Pending',
  // Used when we are waiting for a bot update
  Offline: 'Offline',
  // Lost contact with printer
  ExportOnly: 'Export Only',
  Unauthenticated: 'Unauthenticated',
  // Unknown printer
  Authenticating: 'Authenticating',
  // Unknown printer in process of authenticating
  Reconnecting: 'Reconnecting',
  // Known printer
  Reauthenticating: 'Reauthenticating',
  // Known printer in process of authenticating
  // Connected States, managed by printer object
  Idle: 'Idle',
  Busy: 'Busy'
});
const BotTypeEnum = Object.freeze({
  mini_8: 'mini_8',
  replicator_5: 'replicator_5',
  replicator_b: 'replicator_b',
  z18_6: 'z18_6',
  fire_e: 'fire_e',
  lava_f: 'lava_f',
  sketch: 'sketch'
});
const FifthGenBots = [BotTypeEnum.mini_8, BotTypeEnum.replicator_5, BotTypeEnum.replicator_b, BotTypeEnum.z18_6, BotTypeEnum.sketch];
const SixthGenBots = [BotTypeEnum.fire_e, BotTypeEnum.lava_f];
const ExtruderTypeEnum = Object.freeze({
  mk12: 'mk12',
  mk13: 'mk13',
  mk13_impla: 'mk13_impla',
  mk13_experimental: 'mk13_experimental',
  mk14: 'mk14',
  mk14_s: 'mk14_s',
  mk14_hot: 'mk14_hot',
  mk14_hot_s: 'mk14_hot_s',
  mk14_e: 'mk14_e',
  mk14_c: 'mk14_c',
  sketch_extruder: 'sketch_extruder'
});
const PrinterGenderToDefaultExtruders = Object.freeze({
  [BotTypeEnum.mini_8]: [ExtruderTypeEnum.mk13],
  [BotTypeEnum.replicator_5]: [ExtruderTypeEnum.mk13],
  [BotTypeEnum.replicator_b]: [ExtruderTypeEnum.mk13],
  [BotTypeEnum.z18_6]: [ExtruderTypeEnum.mk13],
  [BotTypeEnum.fire_e]: [ExtruderTypeEnum.mk14, ExtruderTypeEnum.mk14_s],
  [BotTypeEnum.lava_f]: [ExtruderTypeEnum.mk14_hot, ExtruderTypeEnum.mk14_hot_s],
  [BotTypeEnum.sketch]: [ExtruderTypeEnum.sketch_extruder]
});
const DefaultExtruderPairs = {
  [ExtruderTypeEnum.mk14]: ExtruderTypeEnum.mk14_s,
  [ExtruderTypeEnum.mk14_s]: ExtruderTypeEnum.mk14,
  [ExtruderTypeEnum.mk14_hot]: ExtruderTypeEnum.mk14_hot_s,
  [ExtruderTypeEnum.mk14_hot_s]: ExtruderTypeEnum.mk14_hot,
  [ExtruderTypeEnum.mk14_e]: ExtruderTypeEnum.mk14_hot_s,
  [ExtruderTypeEnum.mk14_c]: ExtruderTypeEnum.mk14_c
};
const ExtruderToDefaultMaterial = {
  [ExtruderTypeEnum.mk12]: 'pla',
  [ExtruderTypeEnum.mk13]: 'pla',
  [ExtruderTypeEnum.mk13_impla]: 'im-pla',
  [ExtruderTypeEnum.mk13_experimental]: 'pla',
  [ExtruderTypeEnum.mk14]: 'im-pla',
  [ExtruderTypeEnum.mk14_s]: 'pva',
  [ExtruderTypeEnum.mk14_hot]: 'abs',
  [ExtruderTypeEnum.mk14_hot_s]: 'sr30',
  [ExtruderTypeEnum.mk14_e]: 'abs',
  [ExtruderTypeEnum.mk14_c]: 'nylon-cf',
  [ExtruderTypeEnum.sketch_extruder]: 'pla'
}; // Unfortunately this information is repeated in the mb-plugin/translation.js file

const ExtruderTypeToPrettyName = Object.freeze({
  [ExtruderTypeEnum.mk12]: 'Smart Extruder',
  [ExtruderTypeEnum.mk13]: 'Smart Extruder+',
  [ExtruderTypeEnum.mk13_impla]: 'Tough Smart Extruder+',
  [ExtruderTypeEnum.mk13_experimental]: 'Experimental Extruder',
  [ExtruderTypeEnum.mk14]: 'Model 1A',
  [ExtruderTypeEnum.mk14_s]: 'Support 2A',
  [ExtruderTypeEnum.mk14_hot]: 'Model 1XA',
  [ExtruderTypeEnum.mk14_hot_s]: 'Support 2XA',
  [ExtruderTypeEnum.mk14_e]: 'LABS Extruder',
  [ExtruderTypeEnum.mk14_c]: 'Model 1C',
  [ExtruderTypeEnum.sketch_extruder]: 'Sketch Extruder'
});
const EXPERIMENTAL_EXTRUDER = 99;
const PrinterGenderString = new Immutable.Map([[BotTypeEnum.mini_8, 'Replicator Mini+'], [BotTypeEnum.replicator_5, 'Replicator 5th Gen'], [BotTypeEnum.replicator_b, 'Replicator+'], [BotTypeEnum.z18_6, 'Replicator Z18'], [BotTypeEnum.fire_e, 'Method'], [BotTypeEnum.lava_f, 'Method X'], [BotTypeEnum.sketch, 'Sketch']]);
const GenderToType = Object.freeze({
  'Replicator Mini+': BotTypeEnum.mini_8,
  'Replicator 5th Gen': BotTypeEnum.replicator_5,
  'Replicator+': BotTypeEnum.replicator_b,
  'Replicator Z18': BotTypeEnum.z18_6,
  'Method': BotTypeEnum.fire_e,
  'Method X': BotTypeEnum.lava_f,
  'Sketch': BotTypeEnum.sketch
});
const PrinterCodeNames = new Immutable.Map([['Replicator Mini+', 'wildduck'], ['Replicator 5th Gen', 'platypus'], ['Replicator+', 'horseshoe'], ['Replicator Z18', 'moose'], ['Method', 'sombrero_fire'], ['Method X', 'sombrero_lava'], ['Sketch', 'whitesmith']]);
const PrinterConnTypeEnum = Object.freeze({
  USB: 'usb',
  NETWORK: 'network',
  REFLECTOR: 'reflector',
  ARCHETYPE: 'archetype'
});
const NetworkStateEnum = Object.freeze({
  state: {
    WIFI: 'wifi',
    ETHERNET: 'ethernet',
    OFFLINE: 'offline'
  },
  wifi: {
    ENABLED: 'enabled',
    DISABLED: 'disabled'
  }
});
const PrintSettingScopeEnum = Object.freeze({
  OCCURRENCE: 'occurrence',
  TRAY: 'tray',
  TRAY_PLUGIN: 'tray_plugin'
});
const PRINTER_CONNECTION_TIMEOUT = 15000;
const PRINTER_AUTH_TIMEOUT = 120000;
const ExtruderTranslation = Object.freeze({
  extruderA: 'PlasticA',
  extruderB: 'PlasticB'
});
const PIDtoGenderName = Object.freeze({
  '0008': PrinterGenderString.get(BotTypeEnum.mini_8),
  '0005': PrinterGenderString.get(BotTypeEnum.replicator_5),
  '000b': PrinterGenderString.get(BotTypeEnum.replicator_b),
  '0006': PrinterGenderString.get(BotTypeEnum.z18_6),
  '000e': PrinterGenderString.get(BotTypeEnum.fire_e),
  '000f': PrinterGenderString.get(BotTypeEnum.lava_f),
  '1627': PrinterGenderString.get(BotTypeEnum.sketch)
});
const VendorName = 'MakerBot';
const MaterialCodesToSliceConfigStrings = {
  0: 'generic_model',
  1: 'pla',
  2: 'im-pla',
  3: 'pva',
  4: 'pet',
  5: 'abs',
  6: 'hips',
  8: 'sr30',
  9: 'asa'
};
const SliceConfigStringsToPrettyStrings = {
  'null': 'Not Detected',
  'generic_model': 'Unknown Material',
  'pla': 'PLA',
  'im-pla': 'Tough',
  'pva': 'PVA',
  'pet': 'PETG',
  'abs': 'ABS',
  'hips': 'HIPS',
  'sr30': 'SR-30',
  'asa': 'ASA',
  'nylon': 'Nylon',
  'pc-abs': 'PC-ABS',
  'pc-abs-fr': 'PC-ABS-FR',
  'nylon-cf': 'Nylon Carbon Fiber',
  'nylon12-cf': 'Nylon 12 Carbon Fiber',
  'im-pla-esd': 'ESD'
};
const StockMaterialLength = Object.freeze({
  // These are nominal spool length values in millimeter
  // used for offline material estimates
  makerbot_generation5: {
    pla_small: 75600,
    pla_large: 302400,
    abs_large: 396000
  },
  makerbot_generation6: {
    'pla': 249500,
    'im-pla': 249500,
    'generic_model': 249500,
    'pet': 246000,
    'pva': 158800,
    'abs': 310000,
    'asa': 310000,
    'sr30': 157000,
    'nylon': 259845,
    'pc-abs': 250000,
    'pc-abs-fr': 250000,
    'nylon-cf': 178000,
    'nylon12-cf': 247471,
    'im-pla-esd': 178434
  },
  makerbot_sketch: {
    'pla': 333000,
    'im-pla': 333000
  }
});
const MaterialDefaults = {
  colors: {
    'pla': 'rgb(255,0,0)',
    'generic_model': 'rgb(255,0,0)',
    'im-pla': lightenColor('rgba(43,40,40,1)', 0.55),
    'pva': darkenColor('rgba(255,212,84,0.60)', 0.45),
    'abs': darkenColor('rgba(255, 50, 0, 1)', 0.45),
    'hips': darkenColor('rgba(255,212,84,0.60)', 0.45),
    'sr30': darkenColor('rgba(204, 204, 204, 0.85)', 0.25),
    'pet': darkenColor('rgba(229,229,229,0.85)', 0.25),
    'asa': lightenColor('rgba(11,20,17,1)', 0.55),
    'nylon': lightenColor('rgba(25,25,25,1)', 0.55),
    'pc-abs': lightenColor('rgba(25,25,25,1)', 0.55),
    'pc-abs-fr': lightenColor('rgba(25,25,25,1)', 0.55),
    'nylon-cf': lightenColor('rgba(25,25,25,1)', 0.55),
    'nylon12-cf': lightenColor('rgba(11,20,17,1)', 0.55),
    'im-pla-esd': lightenColor('rgba(11,20,17,1)', 0.55)
  },
  detection_error: 'Not Detected'
};
const SketchErrorCodesToToPrettyStrings = {
  74: {
    title: 'Extruder',
    message: 'Extruder temperature out of range.'
  },
  1039: {
    title: 'Model',
    message: 'Model out of bounds.'
  },
  1041: {
    title: 'Material',
    message: 'This process requires filament to proceed.'
  },
  1043: {
    title: 'Slicing',
    message: 'Failed to slice your layout for printing. Please try again later.'
  },
  1049: {
    title: 'Model',
    message: 'This print file was sliced for a different machine type. Please reslice for the correct printer.'
  },
  1051: {
    title: 'Disk Space',
    message: "There is not enough free space on printer's storage drive."
  },
  1056: {
    title: 'File',
    message: 'Error downloading .makerbot file from cloud.'
  },
  2001: {
    title: 'Camera',
    message: 'Failed to capture image.'
  },
  default: {
    title: 'Sketch',
    message: 'This printer is experiencing an error.'
  }
};
module.exports = {
  PRINTER_PROVIDER,
  BotTypeEnum,
  FifthGenBots,
  SixthGenBots,
  PrinterStateEnum,
  PrinterGenderString,
  PrinterConnTypeEnum,
  GenderToType,
  PrinterCodeNames,
  PRINTER_CONNECTION_TIMEOUT,
  PRINTER_AUTH_TIMEOUT,
  NetworkStateEnum,
  PrintSettingScopeEnum,
  ExtruderTypeEnum,
  EXPERIMENTAL_EXTRUDER,
  ExtruderTranslation,
  PIDtoGenderName,
  VendorName,
  MaterialCodesToSliceConfigStrings,
  SliceConfigStringsToPrettyStrings,
  StockMaterialLength,
  MaterialDefaults,
  PrinterGenderToDefaultExtruders,
  ExtruderTypeToPrettyName,
  ExtruderToDefaultMaterial,
  DefaultExtruderPairs,
  SketchErrorCodesToToPrettyStrings
};
