// MB-support-plugin/lib/croissant.js
'use strict'; // Wrapper around our native croissant module to locate and
// import the correct module.

const path = require('path');

const fs = require('fs');

const paths = require('./paths');

const _croissant = paths.requireNativeCroissant();
const DIAGNOSTIC_TAG = '[croissant-diagnostic]';
const DIAGNOSTIC_WRAPPED = '__croissantDiagnosticWrapped';

function isPlainObject(value) {
  return Object.prototype.toString.call(value) === '[object Object]';
}

function redactString(value) {
  if (typeof value !== 'string') return value;
  if (/^Bearer\s+/i.test(value)) return 'Bearer [REDACTED]';
  return value;
}

function sanitize(value, depth = 0) {
  if (depth > 4) return '[MaxDepth]';
  if (value === null || value === undefined) return value;
  if (Array.isArray(value)) return value.map(item => sanitize(item, depth + 1));

  if (isPlainObject(value)) {
    const sanitized = {};
    Object.keys(value).forEach(key => {
      if (/(token|secret|password|authorization|authinfo|auth_code|code)/i.test(key)) {
        sanitized[key] = '[REDACTED]';
      } else {
        sanitized[key] = sanitize(value[key], depth + 1);
      }
    });
    return sanitized;
  }

  return redactString(value);
}

function getMappedErrorState(err) {
  if (!err) return null;
  if (_croissant.UnauthorizedError && err instanceof _croissant.UnauthorizedError) return 'UnauthorizedError';
  if (_croissant.AuthRejectedError && err instanceof _croissant.AuthRejectedError) return 'AuthRejectedError';
  if (_croissant.AuthTimedOutError && err instanceof _croissant.AuthTimedOutError) return 'AuthTimedOutError';
  return null;
}

function diagnosticLog(payload) {
  try {
    console.log(`${DIAGNOSTIC_TAG} ${JSON.stringify(payload)}`);
  } catch (err) {
    console.log(`${DIAGNOSTIC_TAG} ${payload.type || 'log_error'}`);
  }
}

function wrapPrinterAuthMethods(printer) {
  if (!printer || printer[DIAGNOSTIC_WRAPPED]) return printer;

  ['Authorize', 'Authenticate', 'Reauthorize'].forEach(methodName => {
    if (typeof printer[methodName] !== 'function') return;
    const original = printer[methodName].bind(printer);

    printer[methodName] = function () {
      const startedAt = Date.now();
      const args = Array.from(arguments);

      diagnosticLog({
        type: 'croissant_call',
        method: methodName,
        stage: 'input',
        args: sanitize(args)
      });

      const result = original.apply(null, args);

      if (result && typeof result.then === 'function') {
        return result.then(res => {
          diagnosticLog({
            type: 'croissant_call',
            method: methodName,
            stage: 'output',
            durationMs: Date.now() - startedAt,
            result: sanitize(res)
          });
          return res;
        }).catch(err => {
          diagnosticLog({
            type: 'croissant_call',
            method: methodName,
            stage: 'error',
            durationMs: Date.now() - startedAt,
            mappedErrorState: getMappedErrorState(err),
            errorName: err && err.name,
            errorMessage: err && err.message
          });
          throw err;
        });
      }

      diagnosticLog({
        type: 'croissant_call',
        method: methodName,
        stage: 'output',
        durationMs: Date.now() - startedAt,
        result: sanitize(result)
      });
      return result;
    };
  });

  printer[DIAGNOSTIC_WRAPPED] = true;
  return printer;
}

function wrapCroissantMethod(methodName, options = {}) {
  if (typeof _croissant[methodName] !== 'function') return;
  const original = _croissant[methodName].bind(_croissant);

  _croissant[methodName] = function () {
    const startedAt = Date.now();
    const args = Array.from(arguments);
    diagnosticLog({
      type: 'croissant_call',
      method: methodName,
      stage: 'input',
      args: sanitize(args)
    });

    const result = original.apply(null, args);

    if (result && typeof result.then === 'function') {
      return result.then(res => {
        const wrappedResult = options.wrapPrinter ? wrapPrinterAuthMethods(res) : res;
        diagnosticLog({
          type: 'croissant_call',
          method: methodName,
          stage: 'output',
          durationMs: Date.now() - startedAt,
          result: sanitize(wrappedResult)
        });
        return wrappedResult;
      }).catch(err => {
        diagnosticLog({
          type: 'croissant_call',
          method: methodName,
          stage: 'error',
          durationMs: Date.now() - startedAt,
          mappedErrorState: getMappedErrorState(err),
          errorName: err && err.name,
          errorMessage: err && err.message
        });
        throw err;
      });
    }

    const wrappedResult = options.wrapPrinter ? wrapPrinterAuthMethods(result) : result;
    diagnosticLog({
      type: 'croissant_call',
      method: methodName,
      stage: 'output',
      durationMs: Date.now() - startedAt,
      result: sanitize(wrappedResult)
    });
    return wrappedResult;
  };
}

wrapCroissantMethod('connectPrinter', {
  wrapPrinter: true
});
wrapCroissantMethod('reconnectPrinter', {
  wrapPrinter: true
});

_croissant.createLog = function (logsDir) {
  // set log for croissant
  fs.stat(logsDir, function (err, stats) {
    if (err) {
      // Logs dir does not exist; make it
      try {
        console.log(`Making logs dir ${logsDir}`);
        fs.mkdirSync(logsDir);
      } catch (err) {
        console.log(err);
        return;
      }
    } // check contents of dir; if it has too many croissant logs,
    // delete the oldest


    fs.readdir(logsDir, function (err, files) {
      if (err) {
        console.log(err);
        return;
      }

      try {
        let croissantLogName = `Croissant-${new Date().toISOString()}.log`;
        croissantLogName = croissantLogName.replace(/:/g, '-');

        _croissant.setLog(path.join(logsDir, croissantLogName));

        console.log(`Creating Croissant logs at ${path.join(logsDir, croissantLogName)}`);
        const logFiles = files.filter(function (filename) {
          return filename.startsWith('Croissant');
        });

        if (logFiles.length > 5) {
          logFiles.sort(function (f1, f2) {
            return fs.statSync(path.join(logsDir, f1)).birthtime < fs.statSync(path.join(logsDir, f2)).birthtime;
          });
          fs.unlink(path.join(logsDir, logFiles[logFiles.length - 1]));
        }
      } catch (err) {
        console.log(err);
        return;
      }
    });
  });
};

module.exports = _croissant;
