// MB-support-plugin/lib/errors.js
'use strict';

const _ = require('lodash');

const BirdWingError = require('./croissant').BirdWingError;

const {
  SketchErrorCodesToToPrettyStrings
} = require('./constants');

const MBPluginError = Object.freeze({
  GeneralAuthError: -9999,
  DisconnectWhileAuthingError: -9998
});
const MBPluginErrType = Object.freeze({
  MB_PLUGIN_ERROR: 'MB_PLUGIN_ERROR'
});
const MBPluginErrAction = Object.freeze({
  RETRY_AUTH: 'RETRY_AUTH',
  ACKNOWLEDGE_AND_CLOSE: 'ACKNOWLEDGE_AND_CLOSE',
  LOAD_FILAMENT: 'LOAD_FILAMENT',
  NONE: 'NONE'
});

function _getErrInfoOverrides(printer, err) {
  const bwErrors = BirdWingError.Error;

  switch (err.code) {
    case MBPluginError.GeneralAuthError:
      return {
        action: MBPluginErrAction.RETRY_AUTH
      };

    case MBPluginError.DisconnectWhileAuthingError:
      return {
        action: MBPluginErrAction.ACKNOWLEDGE_AND_CLOSE
      };

    case bwErrors.kOutOfFilament:
      return printer.currentProcess.methods.indexOf('load_filament') > -1 ? // the actual action for this error is BirdWingError.Action.CHANGE_FILAMENT
      // but that is shared with kNoFilamentLoaded and kFilamentSlip, for which we
      // don't necessarily want to change how we handle right now..? So this guy
      // gets a custom action until we maybe one day decide to handle them all
      // the same way
      {
        action: MBPluginErrAction.LOAD_FILAMENT
      } : {
        action: MBPluginErrAction.NONE
      };

    case bwErrors.kNoFilament:
    case bwErrors.kFilamentSlip:
    case bwErrors.kDoorInterlockTriggered:
    case bwErrors.kLidInterlockTriggered:
      return {
        action: MBPluginErrAction.NONE
      };

    default:
      return {};
  }
} // for making plugin-originated errors look like birdwing errors...


function getPluginError(errCode) {
  switch (errCode) {
    case MBPluginError.GeneralAuthError:
      return {
        // these have id's of -1 so as to avoid any possible collision with bw
        // error ids, and because we should never have more than one of this or
        // the other at once...
        error_id: -1,
        code: MBPluginError.GeneralAuthError,
        source: null
      };

    case MBPluginError.DisconnectWhileAuthingError:
      return {
        error_id: -1,
        code: MBPluginError.DisconnectWhileAuthingError,
        source: null
      };
  }
}

function errInfoFromPluginErr(err) {
  const errObj = {
    error: err.code,
    type: MBPluginErrType.MB_PLUGIN_ERROR
  };
  return errObj;
}

function errInfoFromBirdWingErr(err) {
  // ------------------------------------
  // PLEASE REMOVE THIS BEFORE RELEASING
  if (!err.source && err.code === 1048) {
    err.source = {
      file_tool: 'different extruder set',
      current_tool: ''
    };
  } // ------------------------------------


  return new BirdWingError(err.code, err.source);
}

function errInfoForSketchBot(err) {
  const errorCode = SketchErrorCodesToToPrettyStrings[err.code] ? err.code : 'default';
  const errObj = {
    title: `${SketchErrorCodesToToPrettyStrings[errorCode].title} Error`,
    message: SketchErrorCodesToToPrettyStrings[errorCode].message,
    // this is to avoid rendering a default 'OK' button in PrinterDetailsError.js -> _renderButtons()
    action: 'NONE'
  };
  return errObj;
}

function errIsFromPlugin(err) {
  return err.code < 0;
}

function getErrInfo(printer, err) {
  let errInfo;

  if (printer.isSketchPrinter()) {
    errInfo = errInfoForSketchBot(err);
  } else {
    errInfo = errIsFromPlugin(err) ? errInfoFromPluginErr(err) : errInfoFromBirdWingErr(err);
  }

  return _.extend(errInfo, _getErrInfoOverrides(printer, err));
}

module.exports = {
  MBPluginError,
  MBPluginErrType,
  MBPluginErrAction,
  getPluginError,
  errInfoFromPluginErr,
  errInfoFromBirdWingErr,
  getErrInfo,
  errIsFromPlugin
};
