// MB-support-plugin/lib/exportProvider.js
'use strict';

const ExportProvider = require('eagle-print/services/providers/exportProvider');

const q = require('q');

const fs = require('fs-extra');

class MakerbotExportProvider extends ExportProvider {
  constructor() {
    super();
  }

  accept(vendor) {
    return vendor.toLowerCase() === 'makerbot';
  }

  exportToFile(exportData) {
    const toolpath = exportData.currentToolpath;
    const sourceFile = toolpath.getPath();
    const destinationFile = exportData.destinationFile;
    return q.Promise(function (resolve, reject) {
      if (!toolpath || !toolpath.getPath()) {
        return reject(new Error('Cannot export invalid toolpath'));
      }

      if (sourceFile === destinationFile) {
        return reject(new Error('Can not export a toolpath that is currently open'));
      }

      fs.copy(sourceFile, destinationFile, function (error) {
        if (error) {
          return reject(error);
        }

        return resolve({
          sourceToolpathFile: sourceFile,
          destinationToolpathFile: destinationFile
        });
      });
    });
  }

  getSupportedFormats(printer) {
    return {
      name: 'Makerbot',
      extensions: ['makerbot']
    };
  }

}

module.exports = MakerbotExportProvider;
