// MB-support-plugin/lib/file_transfer_helper.js
'use strict';

class PutFileJob {
  constructor(printer, transferId, args) {
    this._printer = printer;
    this._transferId = transferId;
    this._args = args;
    this._cancelled = false;
    this._promise = null;
  }

  start() {
    this._promise = this._printer.invoke('Put', [this._transferId].concat(this._args)).catch(err => {
      if (!this._cancelled) throw err;
    });
    return this._promise;
  }

  cancel() {
    return this._printer.invoke('cancelPut', [this._transferId]).then(() => {
      this._cancelled = true;
    });
  }

  isPending() {
    if (this._promise) return this._promise.isPending();else return false;
  }

}

class FileTransferHelper {
  constructor(printer) {
    this._printer = printer;
    this._transferId = 0;
  }

  _increTransferId() {
    return this._transferId++ % 10000;
  }

  putFileJob(localPath, remotePath, blockSize, progressCallback) {
    return new PutFileJob(this._printer, this._increTransferId(), [localPath, remotePath, blockSize, progressCallback]);
  }

}

module.exports = FileTransferHelper;
