// MB-support-plugin/lib/firmwareUtils.js
'use strict';

const MessageIDs = Object.freeze({
  Waiting: 'firmware waiting message',
  Error: 'firmware error message',
  Progress: 'firmware progress message'
});

const firmwareDownloadMessages = flux => {
  const i18n = flux.getService('TranslationService');
  const message = flux.actions.message;
  return {
    error: id => {
      message.createErrorMessage({
        message: i18n.t('mb:firmware.download_error'),
        id
      });
    },
    progress: (id, percent, onCancel) => {
      message.updateProgress({
        message: i18n.t('mb:firmware.downloading'),
        id,
        progressPercentage: percent,
        isIndeterminate: false,
        actionLabel: i18n.t('updater.cancel_download'),
        actionCallback: onCancel
      });
    },
    dismiss: message.dismissMessage.bind(message)
  };
};

const updateFirmware = (printer, flux) => {
  const logger = flux.getService('LoggingService');
  const message = firmwareDownloadMessages(flux);
  printer.firmware.update().then(resp => {
    logger.info(resp);
    message.dismiss(MessageIDs.Progress);
  }, err => {
    logger.warn(err);
    message.error(MessageIDs.Error);
  }, prog => {
    if (prog.percentage >= 1) {
      message.dismiss(MessageIDs.Progress);
      return;
    }

    message.progress(MessageIDs.Progress, Math.round(prog.percentage * 100), () => {
      prog.cancel();
      message.dismiss(MessageIDs.Progress);
    });
  });
};

module.exports = {
  firmwareDownloadMessages,
  updateFirmware
};
