// MB-support-plugin/lib/firmwareListener.js
'use strict';

const EventEmitter = require('eventemitter3');

const fs = require('fs-extra');

const request = require('request');

const q = require('q');

const tmp = require('tmp');

const _ = require('lodash');

const crypto = require('crypto');

const path = require('path');

const yauzl = require('yauzl');

const {
  progress
} = require('./util');

const {
  BotTypeEnum
} = require('./constants');

const fwEndpointTargets = {
  fifthGen: 'https://s3.amazonaws.com/firmware-makerbot-com/bwfw',
  sixthGen: 'https://s3.amazonaws.com/firmware-makerbot-com/sbfw' // sketch: 'https://s3.amazonaws.com/firmware-makerbot-com/wsfw',

};
const notification = 'firmware_updates_info_change';
const vendorID = 9153;

class VerificationFailedErr extends Error {
  constructor(filename) {
    super(`${filename} didn't pass verification`);
    this.code = 'VerificationFailedErr';
  }

}

class NotZipErr extends Error {
  constructor(msg) {
    super(`Error opening firmware file. It should be a .zip: ${msg}`);
    this.code = 'NotZipErr';
  }

}

class NotMBFirmwareErr extends Error {
  constructor() {
    super('This does not appear to be a Makerbot firmware file');
    this.code = 'NotMBFirmwareErr';
  }

}

class BotNotSupportedErr extends Error {
  constructor() {
    super('This firmware does not support the current printer');
    this.code = 'BotNotSupportedErr';
  }

}
/**
 * @classdesc
 * FirmwareListener listens for firmware updates from the printer. It will also check the server
 * for new updates that the printer doesn't know about it. You need to subscribe to a listener to
 * get updates from the printer. If an update is available, you can trigger an update from here too.
 */


class FirmwareListener {
  /**
   * @constructor
   * @param {MakerBotPrinter} mbPrinter
   */
  constructor(mbPrinter) {
    this._printer = mbPrinter;
    this._availableFirmware = null;
    this._customFirmware = null;
    this._updateNotifications = new EventEmitter();
    this._updateInitiated = false;
    this._putFileJob = null;

    this._progressCallback = () => {};

    this._printer.updateNotifications.addListener(notification, (version, updateAvailable, releaseNotes) => {
      let fw = null;

      if (updateAvailable === true) {
        this._updateInitiated = false;
        fw = {
          manifest: {
            version: version.split('.'),
            description: releaseNotes
          }
        };
      }

      this._parse(fw).then(version => {
        this._updateNotifications.emit(notification, version);
      }).catch(() => {
        // if something goes wrong here, it's probably that the server didn't find
        // any relevant updates, so we'll just emit the notification with `null`
        // which should let the components know that there wasn't an update
        this._updateNotifications.emit(notification, null);
      });
    });

    this._updateNotifications.addListener(notification, fw => this._availableFirmware = fw);
  }
  /**
   * This callback is called whenever the printer sends a "firmware_updates_info_change" notification
   * @callback fwCallback
   * @param {Object} firmware - The latest firmware. It is `null` if there's no new firmware
   */

  /**
   * Adds a notification listener for any available firmware updates
   * @param {fwCallback} callback
   */


  listen(callback) {
    this._updateNotifications.addListener(notification, callback);
  }
  /**
   * Removes a notification listener
   * @param {fwCallback} callback
   */


  remove(callback) {
    this._updateNotifications.removeListener(notification, callback);
  }
  /**
   * Asks the printer if it knows about any available firmware.
   * The printer will send back a notification, which will either have info
   * about an available update, or we'll check the web endpoint.
   * IMPORTANT: The returned Promise won't have any information about available updates.
   * To get that information, you must add a listener. See `listen()`
   * @returns {q.Promise}
   */


  check() {
    if (this._printer._printer.UpdateAvailableFirmwareIsSupported()) {
      return this._printer.invoke('UpdateAvailableFirmware');
    } else {
      this._parse(null).then(version => {
        this._updateNotifications.emit(notification, version);
      }).catch(() => {
        // if something goes wrong here, it's probably that the server didn't find
        // any relevant updates, so we'll just emit the notification with `null`
        // which should let the components know that there wasn't an update
        this._updateNotifications.emit(notification, null);
      });
    }
  }
  /**
   * Tells the printer to update its firmware
   * @returns {q.Promise}
   */


  update() {
    this._updateInitiated = true;

    if (this._availableFirmware && this._availableFirmware.remote) {
      const checksum = this._availableFirmware.md5sum || ''; // prevent bypassing checksum for downloads

      return this._download(this._availableFirmware.path, checksum).then(tmp => this.upload(tmp)).then(tmp => fs.unlinkSync(tmp)).catch(err => console.log('err:', err));
    } // despite the name `DownloadAndInstallFirmware` doesn't download anything,
    // it only installs. Thus, this will throw an error when it tries to update
    // from remote


    return this._printer.invoke('DownloadAndInstallFirmware');
  }
  /**
   * Uploads firmware file (.zip) to the printer to be installed
   * @param {string} localPath - Local path to file
   * @param {bool} [verify] - whether to perform a verification on the file. This is not necessary if the firmware was downloaded via FirmwareListener#download.
   * @returns {q.Promise}
   */


  upload(localPath, verify) {
    if (this._putFileJob && this._putFileJob.isPending()) throw 'Already transferring a firmware file';
    this._updateInitiated = true;
    const remotePath = `/firmware/firmware.zip`;
    this._putFileJob = this._printer.fileTransferHelper.putFileJob(localPath, remotePath);

    if (verify) {
      return this.verify(localPath).then(() => {
        this._printer.invoke('BrooklynUpload', [remotePath, true]);

        return this._putFileJob.start();
      });
    } // hehe, surprise!! BrooklynUpload doesn't actually upload!
    // We call this now with `transfer_wait` (the second parameter) set to
    // true, which will make the printer wait until we've actually finished
    // uploading the localPath (which is done with `PutInit`), before uploading the
    // localPath to the Brooklyn hardware


    this._printer.invoke('BrooklynUpload', [remotePath, true]);

    return this._putFileJob.start().then(() => localPath);
  }

  managingFwUpdate() {
    return this._putFileJob && this._putFileJob.isPending();
  }

  cancelUpdate() {
    let transferCancelPromise = q();
    let processCancelPromise = q();
    if (this._putFileJob && this._putFileJob.isPending()) transferCancelPromise = this._putFileJob.cancel();
    if (_.get(this._printer, 'currentProcess.name') === 'FirmwareBurningProcess') processCancelPromise = this._printer.invoke('Cancel');
    return q.allSettled([transferCancelPromise, processCancelPromise]).then(res => {
      const errs = [];

      for (const r of res) {
        if (r.reason) errs.push(r.reason);
      }

      if (errs.length) throw errs;
    });
  }
  /**
   * Checks to see if there's a cached update before asking the
   * printer again. Its intended use is to reduce lag time in
   * communication between the client and the printer
   * @returns {Boolean}
   */


  hasUpdate() {
    const current = this.currentVersion;
    return this._availableFirmware != null && current !== _latestVersion(current, this._availableFirmware) && !this._updateInitiated;
  }
  /**
   * Checks the hash of the firmware and compares it to the checksum
   * @param {string} filename - path to the file on disk you want to veryify
   * @param {string} [checksum] - md5 checksum for relevant firmware
   * @returns {q.Promise}
   */


  verify(filename, checksum) {
    const hash = crypto.createHash('md5');
    return new q.Promise((resolve, reject) => {
      const file = fs.readFileSync(filename); // need to do this synchronously for this._isSupported

      if (checksum != null) {
        hash.update(file);
        const digest = hash.digest('hex'); // get checksum for input file

        if (digest !== checksum) {
          return reject(new VerificationFailedErr(filename));
        }
      } // check if this machine is supported by this firmware


      return this._isSupported(file).then(resolve).catch(reject);
    });
  }
  /**
   * Formats the current firware version from the printer in the same schema
   * as the JSON response from the server. This just makes comparison easier.
   * @returns {Object} The current firmware version formatted
   */


  get currentVersion() {
    const fw = this._printer.currentState.firmware_version;
    return {
      manifest: {
        version: [fw.major, fw.minor, fw.bugfix, fw.build]
      }
    };
  }
  /**
   * Returns the cached firmware update
   * @returns {Object}
   */


  get availableUpdate() {
    return this._availableFirmware;
  }
  /**
   * Indicates whether an update has begun
   * @returns {Boolean}
   */


  get updateInitiated() {
    return this._updateInitiated;
  }
  /**
   * @private
   * Checks the response from the printer looking for new firmware. If not, we'll ping the website, just to
   * make sure.
   * @param {Object} [newFirmware] - information about known firmware from the printer. If the printer
   * doesn't know about new firmware, it is `null`. In that case we ping the website for
   * new firmware.
   * @returns {q.Promise}
   */


  _parse(newFirmware) {
    // If nothing comes back from the printer, let's check the web endpoint
    if (newFirmware == null) {
      return this._checkFirmwareRemote();
    }

    return new q.Promise(resolve => {
      // otherwise, set the version from what the printer already has
      resolve(_latestVersion(this.currentVersion, newFirmware));
    });
  }
  /**
   * @private
   * Hits a JSON endpoint to see what firmware is available remotely. The JSON response will
   * have every available firmware, which may or may not support the current machine, so we
   * have to filter it down, and then find the most recent of the applicable firmware.
   * @returns {q.Promise}
   */


  _checkFirmwareRemote() {
    let fwEndpoint; // skipping sketch until it's ready & explicitly ignoring it in the constants map to be extra safe
    // if (this._printer.isFifthGen() && this._printer._printerInfo.info.bot_type === BotTypeEnum.sketch) {
    //     fwEndpoint = fwEndpointTargets.sketch;
    // }

    if (this._printer.isFifthGen() && !this._printer._printerInfo.info.bot_type === BotTypeEnum.sketch) {
      fwEndpoint = fwEndpointTargets.fifthGen;
    } else {
      fwEndpoint = fwEndpointTargets.sixthGen;
    }

    return new q.Promise((resolve, reject) => {
      request(`${fwEndpoint}/list.json`, (err, res, body) => {
        if (err) {
          reject(err);
          return;
        }

        const updates = JSON.parse(body);

        const pid = this._printer.getGender().pid;

        const available = this._filterAvailableFirmware(pid, updates);

        if (available.length > 0) {
          const latest = this._filterLatestFirmware(available);

          latest.remote = true;
          this._updateInitiated = false;
          resolve(latest);
        } else {
          reject();
        }
      });
    });
  }
  /**
   * @private
   * In the event that there isn't an availble update on the printer, but there is one on the server, we
   * have to download it to a temporary location, and upload it to the printer ourselves. This method downloads
   * the file and puts it in a tmp location
   * @param {string} filepath - the file name for the firmware update on the server
   * @returns {q.Promise}
   */


  _download(filepath, checksum) {
    const out = tmp.tmpNameSync({
      postfix: '.zip'
    });
    let fwEndpoint; // skipping sketch until it's ready & explicitly ignoring it in the constants map to be extra safe
    // if (this._printer.isFifthGen() && this._printer._printerInfo.info.bot_type === BotTypeEnum.sketch) {
    //     fwEndpoint = fwEndpointTargets.sketch;
    // }

    if (this._printer.isFifthGen() && !this._printer._printerInfo.info.bot_type === BotTypeEnum.sketch) {
      fwEndpoint = fwEndpointTargets.fifthGen;
    } else {
      fwEndpoint = fwEndpointTargets.sixthGen;
    }

    return new q.Promise((resolve, reject, notify) => {
      progress(request(`${fwEndpoint}/${filepath}`)).on('progress', state => notify(state)).pipe(fs.createWriteStream(out)).on('finish', () => {
        this.verify(out, checksum).then(() => resolve(out)).catch(reject);
      }).on('error', err => {
        reject(err);
      });
    });
  }
  /**
   * @private
   * When we get a list of firmware back from the JSON endpoint, not every version will be compatible
   * with the current printer, thus we need to filter it down based on the `pid` of the printer.
   * @param {Number} pid - The Product ID of the current printer
   * @param {Object[]} allUpdates - All of the unfiltered firmware updates parsed from the JSON endpoint
   * @returns {Object[]} Filtered results
   */


  _filterAvailableFirmware(pid, allUpdates) {
    return allUpdates.filter(u => {
      return u.manifest.supported_machines.filter(m => m.pid === pid).length > 0;
    });
  }
  /**
   * @private
   * Cycles through the filtered firmware updates from the JSON endpoint and returns the
   * latest firmware update.
   * @param {Object[]} updates - filtered firmware updates fro the JSON endpoint
   * @returns {Object} The most recent update
   */


  _filterLatestFirmware(updates) {
    let latest = updates[0];
    updates.map(u => {
      latest = _latestVersion(latest, u);
    });
    return latest;
  }
  /**
   * @private
   * Checks to see if a particular firmware supports this machine
   * @param {Buffer} file - the zip file containing the firmware
   * @returns {q.Promise}
   */


  _isSupported(file) {
    return this._extractManifest(file).then(manifest => {
      // loop through supported machines
      for (let i = 0; i < manifest.supported_machines.length; i++) {
        const supported = manifest.supported_machines[i]; // if we find that this machine is supported, resolve and return

        if (supported.pid === this._printer.getGender().pid && supported.vid === vendorID) {
          return manifest;
        }
      }

      throw new BotNotSupportedErr();
    });
  }

  _extractManifest(file) {
    const deferred = q.defer();
    let foundManifest = false; // unzip the firmware file

    yauzl.fromBuffer(file, (err, zipFile) => {
      if (err) {
        deferred.reject(new NotZipErr(err.message));
        return;
      }

      zipFile.on('entry', entry => {
        // find the manifest file
        if (path.basename(entry.fileName) === 'manifest.json') {
          foundManifest = true; // stream it

          zipFile.openReadStream(entry, (err, stream) => {
            if (err) {
              deferred.reject(err);
              return;
            }

            let json = '';
            stream.on('readable', () => {
              const data = stream.read();

              if (data) {
                json += data;
              } else {
                // parse and return the manifest
                deferred.resolve(JSON.parse(json));
              }
            }).on('error', deferred.reject);
          });
        }
      });
      zipFile.on('end', () => {
        if (!foundManifest) {
          deferred.reject(new NotMBFirmwareErr());
        }
      });
    });
    return deferred.promise;
  }

}
/**
 * Takes two versions and recursively determines which is the
 * latest of the two.
 * @param {Object} v1 - A version of firmware
 * @param {Object} v2 - Another version of firmware
 * @param {Number} [n=0] - Depth of the iteration. Used for recursion
 * @returns {Object} The latest of the two provided versions
 */


function _latestVersion(v1, v2, n) {
  if (!n) n = 0; // apparently default parameters don't work?
  // Will throw an error if n is out of bounds. If that happens,
  // it most likely means that the two versions are equal, so
  // return either one.

  try {
    if (v1.manifest.version[n] > v2.manifest.version[n]) {
      return v1;
    }

    if (v2.manifest.version[n] > v1.manifest.version[n]) {
      return v2;
    }

    return _latestVersion(v1, v2, n + 1);
  } catch (e) {
    return v1; // both equal, doesn't matter
  }
}

module.exports = FirmwareListener;
