// MB-support-plugin/lib/gcode_toolpath_parser.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GcodeParserStream = void 0;

const {
  Transform
} = require('stream');
/**
 * GCode ToolPath Parser
 *
 * ### How It Works:
 * It takes a gcode readable file nodejs stream and converts the readable
 * chunks into layer chunks. This task is executed in LayerChunksTransform
 *
 * Each layer chunks contains comments on the top portion follow by linear
 * move commands.
 *
 * ie. Layer Chunk:
 * """
 * ; Layer Section 0 (1)
 * ; Material 0
 * ; Lower Position  0
 * ; Upper Position  0.4
 * ; Thickness       0.4
 * ; Width           1.4
 * G1 X-3.55 Y-3.55 Z0.4 F15000; Travel Move
 * ..
 * G1 X-9.71 Y-9.71 Z0.4 E60.85656 F900; Inset
 * ..
 * G1 X0.16 Y-2.705 Z0.4 F1380; Travel Move
 * """
 *
 * Next, each layer chunk will get transformed by the GcodeCmdParserTransform
 * nodejs stream. It will parse each layer chunk for command moves,
 * lower/upper positions, and thickness/width values, then build out the
 * necessary object for the eagle 3D previewer.
 *
 * Each output GcodeCmdParser chunk will return:
 * {
 *    totalNrMoves: number, (all moves)
 *    totalNrHeadMoves: number, (moves that are just head moves)
 *    totalLength: number,
 *    movesModel: { type, [Move]}
 *    movesSupport: { type, [Move]}
 * }
 **/


class LayerChunksTranform extends Transform {
  constructor() {
    super({
      objectMode: true
    });
    /** regex expression to match the following:
     * case 1:
     * """
     * ; Layer Section 12 (13)
     * ...
     * G1 X12 Y12 Z123 F123; Infill
     * ...
     * G1 X42 Y42 Z123 F123; Infill
     * ; Layer Section 124 (14) <= *** Excludes this line ***
     *
     * -- OR --
     *
     * case 2:
     * """
     * ; Layer Section 123 (1234)
     * ...
     * G1 X12 Y12 Z123 F123; Travel Move
     * ...
     * G1 X42 Y42 Z123 F123; Travel Move
     * ; End of print  <= *** Excludes this line ***
     *
     * ### Explaination
     *
     * The regex will capture the top section comments [ie ; Layer Section 12 (13)] +
     * the linear move commands [ie. G1 X12 Y12 Z123; Infill] **UNTIL**
     * it hits another layer comment [; Layer Section 124 (14)] or
     * end of print comment [; End of print  ].
     *
     * With this regex we can split the buffer. The trailing end of the split will be
     * the new buffer that doesn't match the regex expression, meaning it didn't have
     * the end layer comment or end of print comment.
     **/

    this._delimiter = /(; Layer Section \d* \(\d*\) (\r\n?|\n)){1}(.|\n|\r\n)*?((?=; Layer Section \d* \(\d*\) (\r\n?|\n))|(?=; End of print (\r\n?|\n))){1}/g;
    this._encoding = 'utf8';
    this._buffer = '';
  }

  _transform(chunk, encoding, callback) {
    // convert input chunk encoding into output
    // encoding and append to internal buffer
    if (encoding === 'buffer') {
      this._buffer += chunk.toString(this._encoding);
    } else if (encoding === this._encoding) {
      this._buffer += chunk;
    } else {
      this._buffer += Buffer.from(chunk, encoding).toString(this._encoding);
    } // Check if chuck from file read stream contains matching pattern
    // based on the delimiter


    if (this._delimiter.test(chunk)) {
      // parse out the chunk for matching layer sections within
      const sections = this._buffer.match(this._delimiter); // any trailing end that doesn't match up, return to buffer
      // for the next file read stream chunk


      this._buffer = this._buffer.split(this._delimiter).pop(); // push out the layer sections

      sections.forEach(this.push, this);
    }

    callback();
  }

  _flush(callback) {
    callback(null, this._buffer);
  }

}

class GcodeCmdParserTransform extends Transform {
  constructor(buildVolume, totalRaftLayers) {
    super({
      objectMode: true
    });
    this._encoding = 'utf8';
    this._buffer = '';
    this._lowPosRegex = /; Lower Position.+\d+[.]?\d*.(\r\n?|\n)/gi;
    this._upPosRegex = /; Upper Position.+\d+[.\d+]?.+/gi;
    this._thicknessRegex = /; Thickness.+\d+[.]?\d*.(\r\n?|\n)/gi;
    this._widthRegex = /; Width.+\d+[.]?\d*.(\r\n?|\n)/gi;
    this._linearCmdRegex = /G1.+/gi;
    this._linearCmdTypeRegex = /(; )(.*)/gi;
    this._xPositionRegex = /X[-]?\d+[.]?\d+/;
    this._yPositionRegex = /Y[-]?\d+[.]?\d+/;
    this._zPositionRegex = /Z[-]?\d+[.]?\d+/;
    this._floatRegex = /[-]?\d+[.]?\d*/;
    this._xOffset = buildVolume.x / 2;
    this._yOffset = buildVolume.y / 2;
    this._totalRaftLayers = totalRaftLayers;
    this._inches_per_mm = 0.0393701;
    this._layerCount = 0;
    this._lowerPosition = 0;
    this._upperPosition = 0;
    this._thickness = 0;
    this._width = 0;
    this._totalMoves = 0;
    this._totalTravelMoves = 0;
    this._isTravel = true;
    this._isModel = true;
    this._movesModel = [];
    this._movesSupport = [];
  }

  _parseTag(tag, prevCmd, nextCmd) {
    switch (tag) {
      case 'Trailing Extrusion Move':
        this._isTravel = false;
        break;

      case 'Anchor':
      case 'Bridge':
      case 'Infill':
      case 'Inset':
      case 'Outline':
      case 'Spur':
        this._isTravel = false;
        this._isModel = true;
        break;

      case 'Support':
      case 'Purge':
      case 'Invalid Move':
        this._isTravel = false;
        this._isModel = false;
        break;

      case 'Backlash Compensation':
      case 'Connection':
      case 'Long Restart':
      case 'Long Retract':
      case 'Restart':
      case 'Retract':
      case 'Leaky Travel Move':
      case 'Travel Move':
        this._isTravel = true;
        break;

      default:
        if (this.logger) {
          this.logger.info(`tag not found: ${tag}`);
        }

    } // all raft layers are tagged as support


    if (this._layerCount <= this._totalRaftLayers) {
      this._isModel = false;
    } // Check previous and next command moves if either tags equals
    // "support". This should help with certain support features
    // not rendering because it lacks the prev cmd to generate preview.


    const prevTag = prevCmd && prevCmd.match(this._linearCmdTypeRegex)[0].replace(/; |\s/, '').trim();
    const nextTag = nextCmd && nextCmd.match(this._linearCmdTypeRegex)[0].replace(/; |\s/, '').trim();

    if (['Support', 'Purge', 'Invalid Move'].includes(prevTag) || ['Support', 'Purge', 'Invalid Move'].includes(nextTag)) {
      this._isModel = false;
    }
  }

  _parseCommentCommands() {
    // ; Lower Position  0
    //                   ^
    this._lowerPosition = parseFloat(this._buffer.match(this._lowPosRegex)[0].match(this._floatRegex)[0]) * this._inches_per_mm; // ; Upper Position  0
    //                   ^

    this._upperPosition = parseFloat(this._buffer.match(this._upPosRegex)[0].match(this._floatRegex)[0]) * this._inches_per_mm; // ; Thickness       0.4
    //                    ^

    this._thickness = parseFloat(this._buffer.match(this._thicknessRegex)[0].match(this._floatRegex)[0]) * this._inches_per_mm; // ; Width           1.4
    //                    ^

    this._width = parseFloat(this._buffer.match(this._widthRegex)[0].match(this._floatRegex)[0]) * this._inches_per_mm;
  }

  _parseTagsAndCoordinates(cmd) {
    // G1 X5.09 Y5.09 Z0.4 E11.59722 F900; Inset
    //     [x]   [y]   [z]                 [tag]
    const cmdTag = cmd.match(this._linearCmdTypeRegex)[0].replace(/; |\s/, '').trim();

    const x = (parseFloat(cmd.match(this._xPositionRegex)[0].match(this._floatRegex)[0]) + this._xOffset) * this._inches_per_mm;

    const y = (parseFloat(cmd.match(this._yPositionRegex)[0].match(this._floatRegex)[0]) + this._yOffset) * this._inches_per_mm;

    const z = parseFloat(cmd.match(this._zPositionRegex)[0].match(this._floatRegex)[0]) * this._inches_per_mm;

    return {
      x,
      y,
      z,
      cmdTag
    };
  }

  _parseMoveCommands() {
    // All 'G1 ...' are linear moves (https://marlinfw.org/docs/gcode/G000-G001.html)
    const linearMoves = this._buffer.match(this._linearCmdRegex);

    this._totalMoves = linearMoves.length;
    linearMoves.forEach((cmd, index) => {
      const {
        x,
        y,
        z,
        cmdTag
      } = this._parseTagsAndCoordinates(cmd); // Update this._isTravel & this._isModel based on tag


      this._parseTag(cmdTag, linearMoves[index - 1], linearMoves[index + 1]); // For all non-travel moves, area equals thickness * width;


      const area = this._isTravel ? 0 : this._thickness * this._width;

      if (this._isTravel) {
        this._totalTravelMoves += 1;
      } // All raft layers count as support layers


      if (this._isModel) {
        this._movesModel.push({
          x,
          y,
          z,
          area,
          index: index += 1
        });
      } else {
        this._movesSupport.push({
          x,
          y,
          z,
          area,
          index: index += 1
        });
      }
    });
  }

  _resetToDefault() {
    this._lowerPosition = 0;
    this._upperPosition = 0;
    this._thickness = 0;
    this._width = 0;
    this._totalMoves = 0;
    this._totalTravelMoves = 0;
    this._isTravel = true;
    this._isModel = true;
    this._movesModel = [];
    this._movesSupport = [];
  }

  _transform(chunk, encoding, callback) {
    // convert input encoding into output encoding
    // and update the buffer
    if (encoding === 'buffer') {
      this._buffer = chunk.toString(this._encoding);
    } else if (encoding === this._encoding) {
      this._buffer = chunk;
    } else {
      this._buffer = Buffer.from(chunk, encoding).toString(this._encoding);
    } // track layer count


    this._layerCount += 1; // parse the header portion of the buffer layer chunk

    this._parseCommentCommands(); // parse the linear commands of the buffer layer chunk


    this._parseMoveCommands(); // build out the payload


    this._layer = {
      totalNrMoves: this._totalMoves,
      totalNrHeadMoves: this._totalTravelMoves,
      movesMaterial: {
        '199': this._movesModel
      },
      movesSupport: {
        '201': this._movesSupport
      },
      upperPosition: this._upperPosition,
      lowerPosition: this._lowerPosition
    }; // push the payload

    this.push(this._layer); // reset certain values for the next incoming chunk

    this._resetToDefault();

    callback();
  }

  _flush(callback) {
    callback(null, this._buffer);
  }

}
/**
 * A function that pipes the two transform streams to a Gcode readable
 * file nodejs stream, then outputs the resulting stream out
 *
 * @param {stream} fileStream - A Gcode readable file nodejs stream
 * @param {obj} parsingParams - Obj that contains buildVolume & total
 *                              raft layers information
 **/


const GcodeParserStream = (fileStream, parsingParams) => {
  return fileStream.pipe(new LayerChunksTranform()).pipe(new GcodeCmdParserTransform(parsingParams.buildVolume, parsingParams.totalRaftLayers));
};

exports.GcodeParserStream = GcodeParserStream;
