// MB-support-plugin/lib/makerbot.js
'use strict';

const q = require('q');

const path = require('path');

const fs = require('fs-extra');

const R = require('ramda');

const tmp = require('tmp');

const JOB_TYPE = require('eagle-print/stores/buildStore').JOB_TYPE;

const {
  SettingsUISchema,
  decodeKey
} = require('@makerbot/settingsui');

const Sliceconfig = require('sliceconfig');

const thing = require('./thing');

const paths = require('./paths'); // The "|| null" is necessary to prevent the grunt tests from throwing an error
// saying that they cannot find module "settingsUI".


const NodeGrue = paths.requireNodeGrue();

const {
  ExtruderTypeEnum,
  EXPERIMENTAL_EXTRUDER
} = require('./constants'); // Configure temp directory behavior


tmp.setGracefulCleanup(); // The module's configuration.

const configuration = createDefaultConfiguration(); /// Returns the module's configuration.  Useful for unit tests and debugging

function config() {
  return configuration;
}
/**
 * Returns an nconf configuration based off of the OS that the software is
 * running on.
 *
 * @returns {nconf} - the configuration
 */


function createDefaultConfiguration() {
  console.info('using default builder configuration');

  const nconf = require('nconf');

  const configData = {
    makerbot: {
      pass_full_profile: require('os').type() != 'Windows_NT'
    }
  };
  nconf.defaults(configData);
  return nconf;
}
/**
 * Return the entire list of 'supported_configurations' from sliceconfig
 **/


function getSupportedConfigurations() {
  console.log('getting "supported_configurations" from sliceconfig');
  const scRunner = new Sliceconfig(paths.getBinDir());
  return scRunner.supportedConfigs();
} /// Does this printer have a machine config?


function hasMachineConfig(p) {
  return p.machineConfig && Object.keys(p.machineConfig).length !== 0;
}
/**
 * Parse the attached_extruders's tool ids into extruder strings
 * found in supported_extruders's list via machineConfig.
 *
 * 07.29.2019 - SliceConfig no longer accepts machineConfig as a
 * parameter. Instead send only the extruders array parsed from
 * machineConfig.
 * 08.28.2019 - Had to add an exception to Experimental Extruder (toolid 99)
 * In machine_config, 99 is labeled as a mk13 extruder.
 *
 * - Ray
 * @param {machineConfig} tray.printer.machineConfig object
 */


function parseExtrudersFromMachineConfig(machineConfig) {
  const {
    attached_extruders,
    supported_extruders
  } = machineConfig.extruder_profiles;
  return attached_extruders.map(({
    id
  }) => {
    return id === EXPERIMENTAL_EXTRUDER ? ExtruderTypeEnum.mk13_experimental : supported_extruders[parseInt(id)];
  });
}
/**
 * Builds a tray
 * @param {Tray} tray - a tray object should have the following
 * <p> printer: the target printer object </p>
 * <p> stlFiles: array of stl file paths </p>
 * <p> transforms: array of transforms for each STL file </p>
 * @param {hash} options - optional arguments
 * @param {function} callback - a node.js style callback function
 * @returns {promise} promise - promise result contains a tray object containing
    the following
 * <p> name: name and extension of the toolpath </p>
 * <p> path: location of the toolpath i.e path or url </p>
 */


function build(tray, options, progressCallback, cancelToken, callback, flux) {
  options = options || {}; /// get a thumbnail from the tray and add it to a temp directory (not in use
  /// yet);

  const createThumbnails = function (tray) {
    try {
      // list of names the thumbnails will save to and where to get the image
      let thumbnailData;

      if (tray.printer.isSixthGen()) {
        // sombrero-style thumbnails
        thumbnailData = [{
          name: 'thumbnail_960x1460.png',
          snapshot: tray.traySnapshots.large
        }, {
          name: 'thumbnail_212x300.png',
          snapshot: tray.traySnapshots.medium
        }, {
          name: 'thumbnail_140x106.png',
          snapshot: tray.traySnapshots.small
        }];
      } else if (tray.printer.isSketchPrinter()) {
        // sketchy-style thumbnail
        thumbnailData = [{
          name: 'thumbnail_90x90.png',
          snapshot: tray.traySnapshots.small
        }];
      } else {
        // standard thumbnails
        thumbnailData = [{
          name: 'thumbnail_320x200.png',
          snapshot: tray.traySnapshots.large
        }, {
          name: 'thumbnail_110x80.png',
          snapshot: tray.traySnapshots.medium
        }, {
          name: 'thumbnail_55x40.png',
          snapshot: tray.traySnapshots.small
        }];
      } // append standard thumbnails used in Teams & PrintFromWeb to  all 5th & 6th Printers


      thumbnailData.push({
        name: 'isometric_thumbnail_120x120.png',
        snapshot: tray.traySnapshots.isometricSmall
      }, {
        name: 'isometric_thumbnail_320x320.png',
        snapshot: tray.traySnapshots.isometricMedium
      }, {
        name: 'isometric_thumbnail_640x640.png',
        snapshot: tray.traySnapshots.isometricLarge
      }); // create a directory for the thumbnails

      const thumbnailDir = tmp.dirSync(); // write a file for a thumbnail into the thumbnail directory
      // @thumbnailInfo:
      //  * name <string> - name for the thumbnail file to be saved
      //  * snapshot <imageURL> - b64 string for the png to be saved

      const write = function (thumbnailInfo) {
        if (!thumbnailInfo) return;
        const thumbnailPath = path.join(thumbnailDir.name, thumbnailInfo.name);
        const imageURL = thumbnailInfo.snapshot.valueOf().dataURL;
        const thumbnailBuffer = new Buffer(imageURL.substring('data:image/png;base64,'.length), 'base64');
        return q.nfapply(fs.writeFile, [thumbnailPath, thumbnailBuffer]);
      }; // Create an array of thumbnail promises to settle/write


      const thumbnailPromises = [];
      thumbnailData.forEach(thumbnail => {
        thumbnailPromises.push(write(thumbnail));
      }); // create a promise for each thumbnail and wrap them all together to return

      return q.all(thumbnailPromises).then(function () {
        return thumbnailDir;
      });
    } catch (error) {
      return q.reject(error);
    }
  }; /// add a input and output filepaths, printer type, and slice settings to args


  const addPathsToArgs = function (args, modelFilePath) {
    // create a temp dir for the model and toolpath files
    const tempDir = tmp.dirSync(); // Copy STL file into temporary directory

    const modelName = path.basename(modelFilePath);
    args.inputPath = path.join(tempDir.name, modelName);
    fs.copySync(modelFilePath, args.inputPath);
    const toolpathIsJtp = !tray.printer.isSketchPrinter(); // path to which we will write the toolpath output from nodegrue

    args.tpPath = path.join(tempDir.name, toolpathIsJtp ? 'print.jsontoolpath' : 'print.gcode'); // path to which we will write the meta.json output from nodegrue; also
    // path of metadata that will be input to sliceconfig's `package_makerbot`
    // command;

    args.metaPath = path.join(tempDir.name, 'meta.json');
    const mbName = `${tray.name || 'untitled'}.makerbot`; // path to the final .makerbot;

    args.outputPath = path.join(tempDir.name, mbName);
    args.makerbotName = mbName;
    args.configPath = path.join(tempDir.name, 'config.json');
    return args;
  };

  const getAccelConfig = flux => {
    const printer = flux.store('PrinterStore').getSelectedPrinter();
    if (printer.isSketchPrinter()) return null;
    const printSettings = flux.store('PrintSettingsStore').getState();
    const {
      botType
    } = printer.getGender();
    const extruders = printer.getAttachedExtruders() ? printer.getAttachedExtruders().join('.') : null;
    const materials = printSettings.materials.selected.join('.');
    const mbirRootDir = paths.getInstallArtifactsRootDir(); // There are no accel configs w/ 5th gen & Experimental Extruder

    if (!extruders || extruders === ExtruderTypeEnum.mk13_experimental) {
      console.log(`exiting accel configs`);
      return null;
    }

    return path.join(mbirRootDir, 'public', botType, extruders, `${materials}.json`) // For production scenario where profiles are nested in app.asar.unpacked
    // and not in app.asar
    .replace(/app\.asar/, 'app.asar.unpacked');
  };

  function handleProgress(progressData) {
    try {
      // turn progress into an array so we treat single and multiple entries
      // the same
      let progressJson = progressData.split('\n');

      if (progressJson.length > 1 && progressJson[progressJson.length - 1] === '') {
        progressJson = progressJson[progressJson.length - 2];
      } else {
        progressJson = progressJson[progressJson.length - 1];
      }

      if (progressJson && progressCallback) {
        progressJson = JSON.parse(progressJson); // progress callback wants progress as decimals

        progressCallback(progressJson.totalPercentComplete / 100, undefined);
      }
    } catch (error) {
      console.log(`Slice Progress Error for line ${progressData}: ${error}`);
    }
  } // if we don't have stl files in our tray
  // - OR - if we return an undefined value within the stlFiles array
  // just return a rejection


  if (tray.stlFiles === undefined || tray.stlFiles.includes(undefined)) {
    return q.reject(new Error('Unsupported tray object')).nodeify(callback);
  } // build a promise that will do preprocessing and then slice the file.


  return q.Promise(function (resolve, reject, notify) {
    let args = {
      printerType: tray.printer.getGender().botType
    }; // start print process on bot if necessary

    let startPrintPromise;

    if (tray.parameters && tray.parameters.jobType === JOB_TYPE.PRINT) {
      // (printing is only allowed for bots that will be using .makerbots)
      const mbName = `${tray.name || 'untitled'}.makerbot`;
      startPrintPromise = tray.printer.startPrint(mbName, cancelToken);
    } else {
      startPrintPromise = q();
    }
    /**
     * Generate the list of extruder to model selections based on the
     * list of bodyProperties from the tray.
     *
     * -- NOTE --
     * getProperty/getValue is GC's hack ontop of ImmutableJs therefore
     * breaks mb-plugin's unit tests
     *
     * - Ray
     */


    const assignedExtruderList = tray.bodyProperties.map(bodyProperty => {
      return bodyProperty.toJSON().info.extruder.value;
    }); // call slice config with the slice command

    startPrintPromise // start by creating a dot thing file for this tray
    .then(function () {
      return thing.createThingFile(tray.stlFiles, assignedExtruderList, tray.transforms, tray.name);
    }, function (error) {
      reject(error);
    }) // then fetch the profile with which to slice from sliceconfig
    .then(function (thingFilePath) {
      args.command = 'get_profile';
      args = addPathsToArgs(args, thingFilePath);
      args = R.mergeAll([args, getProfileArgs(tray, tray.printer, options, flux, cancelToken, callback)]);
      return runSliceConfig(args, cancelToken, callback);
    }, function (error) {
      reject(error);
    }) // then dump it into a file to pass to nodegrue
    .then(function (profileJson) {
      return q.nfapply(fs.writeFile, [args.configPath, JSON.stringify(profileJson)]);
    }, function (error) {
      reject(error);
    }).then( // create the toolpath and base metadata using the newly created dot thing
    // if there is an error in createThingFile, reject the build
    function () {
      const toolpathIsJtp = !tray.printer.isSketchPrinter();
      const toolpather = new NodeGrue.NodeToolpather();
      const accelConfigFilePath = getAccelConfig(flux);

      if (cancelToken && cancelToken.isCancelled()) {
        // if the cancel token has already been used (e.g. cancel was hit
        // while generating the dot thing to slice), reject.
        const err = new Error('Cancelled');
        err.type = 'cancel';
        reject(err);
      } else {
        const {
          sliceCompletionPromise,
          cancelFunc
        } = toolpather.runMiracleGrue(args.configPath, args.inputPath, handleProgress, toolpathIsJtp ? args.tpPath : null, args.metaPath, toolpathIsJtp ? null : args.tpPath, accelConfigFilePath);

        if (cancelToken) {
          // if the cancel token is used while slicing, reject.
          cancelToken.on('cancel', () => {
            cancelFunc();
          });
        }

        return sliceCompletionPromise;
      }
    }, function (error) {
      reject(error);
    }).then( // create thumbnails
    function () {
      return createThumbnails(tray);
    }, function (error) {
      // (Cancelling a slice will reject the sliceCompletionPromise)
      if (error.message == 'Cancelled') {
        error.type = 'cancel';
      }

      reject(error);
    }).then( // add the thumbnail directory to the args list and run slice config
    // to package the final .makerbot. if createThumbnails failed,
    // log it and run slice config without the thumbnail directory argument
    function (thumbnailDir) {
      args.command = 'package_makerbot';
      args.thumbnailDir = thumbnailDir.name; // rename the `tpPath` property to `inputPath` so
      // sliceconfig.js is able to direct it properly to the
      // `-i` arg to sliceconfig command line.

      args.inputPath = args.tpPath;
      delete args.tpPath;
      return runSliceConfig(args, cancelToken, callback);
    }, function (error) {
      console.log(error);
      args.command = 'package_makerbot';
      return runSliceConfig(args, cancelToken, callback);
    }).then( // when slicing via slice config finishes up, resolve with the makerbot
    // file name and path. reject on errors. report progress back to the
    // progress callback.
    function () {
      resolve({
        name: args.makerbotName,
        path: args.outputPath
      });
    }, function (error) {
      if (tray.parameters && tray.parameters.jobType === JOB_TYPE.PRINT) {
        tray.printer.cancel();
      }

      reject(error);
    });
  }).nodeify(callback);
}
/**
 * get args needed to generate a miracle grue profile
 *
 * @param {Tray} tray - a tray object should have the following
 *    <p> printer: the target printer object </p>
 *    <p> bodyProperties: an array of list of overrides
 * @param {Printer} printer - a machine that contains the following
 *    <p> machineConfig - a description of that machines properties </p>
 *    <p> gender - the type of machine that this is</p>
 * @param {hash} options - optional arguments
 *
 * @returns {object} args
 */


function getProfileArgs(tray, printer, options, flux) {
  const args = {};
  printer = printer || tray.printer; // add the materials, print mode and overrides

  const configuration = configurationFromTray(tray);
  const printSettingsStore = flux.store('PrintSettingsStore');
  const printSettings = printSettingsStore.getState();
  const storeOverrides = printSettingsStore.getOverrides();
  options = options || {};
  args.overrides = options.overrides ? R.mergeAll([storeOverrides, options.overrides]) : storeOverrides;
  args.materials = options.materials ? options.materials.selected : printSettings.materials.selected;
  args.printMode = printSettings.printMode.selected;
  args.profilePath = configuration.profilePath; // add the machine config or bot name depending on whats there
  // NOTE: machine config differs on live & archetype Method, do not pass machine config to Method bot

  args.printerType = printer.getGender().botType;

  if (printer.isSketchPrinter()) {
    // Grab the only supported extruder
    args.extruders = printSettings.extruders.supported[0];
  } else if (hasMachineConfig(printer) && printer.isConnected() && printer.isFifthGen()) {
    args.extruders = parseExtrudersFromMachineConfig(printer.machineConfig);
  } else {
    args.extruders = printer.getAttachedExtruders();
  }

  return args;
}
/**
 * generate a miracle grue profile
 *
 * @param {Tray} tray - a tray object should have the following
 *    <p> printer: the target printer object </p>
 *    <p> bodyProperties: an array of list of overrides
 * @param {Printer} printer - a machine that contains the following
 *    <p> machineConfig - a description of that machines properties </p>
 *    <p> gender - the type of machine that this is</p>
 * @param {hash} options - optional arguments
 * @param {function} callback - a node.js style callback function
 *
 * @returns {promise} promise - promise result contains a Profile JSON object
 */


function getProfile(tray, printer, options, flux, cancelToken, callback) {
  let args = {
    command: 'get_profile'
  }; // call slice config with the slice command

  args = R.mergeAll([args, getProfileArgs(tray, printer, options, flux)]);
  return runSliceConfig(args, cancelToken, callback);
}
/**
 * create the slice settings (as sliceconfig expects them) for a tray
 *
 * @param {Tray} tray - an object representing a tray of models
 *    <p> bodyProperties - a list of properties, one for each body
 *    <p> printer - an object representing a printer
 *
 *  @returns {Dict} - a json dict of settings formatted for slice config
 */


function configurationFromTray(tray) {
  if (!tray.bodyProperties || tray.bodyProperties.length === 0) return {}; // setup a configuration object to fill out

  const configuration = {
    materials: [null, null],
    printMode: null,
    overrides: {}
  }; // convert from an immutable map to a js object. is there a better way?

  const properties = tray.bodyProperties[0].toJSON(); // helper function that makes sure we care about this property

  function realProperty(prop) {
    return prop !== undefined && prop.value !== undefined;
  } // loop through all the properties


  for (const property in properties) {
    if (realProperty(properties[property])) {
      if (property === 'materials') {
        configuration.materials = properties.materials.value;
      } else if (property === 'printMode') {
        configuration.printMode = properties.printMode.value;
      } else if (property === 'customProfilePath') {
        configuration.profilePath = properties.customProfilePath.value[0];
      } else if (property === 'rotationLock' || property === 'orientation' || property === 'extruders') {
        // We skip rotationLock and orientation because we never want to hand them to sliceConfig
        // at some point, let's actually remove rotationLock from the tray, instead of just ignoring it
        // We skip extruders because we get them from the printer object itself
        continue;
      } else {
        // All other properties are assumed to be overrides. Use the decodeKey
        // function from the sliceconfig interface to convert the key to use the
        // correct delimitors and set the override.
        const decodedProperty = decodeKey(property);
        configuration.overrides[decodedProperty] = properties[property].value;
      }
    }
  }

  return configuration;
}
/**
 * get a schema describing available parameters that UI can set
 *
 * @param {Tray} tray - a tray object should have the following
 *    <p> printer: the target printer object </p>
 *    <p> bodyProperties: an array of list of overrides
 * @param {Printer} printer - a tray object should have the following
 *
 * @returns {promise} - promise result contains a json schema for available
 *  parameters.
 */


function getSchema(tray, printer, ignoreExtruderArgs = false, flux, cancelToken, callback) {
  const args = {
    command: 'get_schema'
  };
  printer = printer || tray.printer; // add the materials, print mode and overrides

  const configuration = configurationFromTray(tray);
  const printSettingsStore = flux.store('PrintSettingsStore');
  const printSettings = printSettingsStore.getState();
  args.overrides = printSettingsStore.getOverrides();
  args.materials = printSettings.materials.selected;
  args.printMode = printSettings.printMode.selected;
  args.profilePath = configuration.profilePath; // add the machine config or bot name depending on whats there

  args.printerType = printer.getGender().botType;

  if (printer.isSketchPrinter()) {
    // Grab the only supported extruder
    args.extruders = printSettings.extruders.supported[0];
  } else if (hasMachineConfig(printer) && printer.isConnected() && printer.isFifthGen()) {
    args.extruders = parseExtrudersFromMachineConfig(printer.machineConfig);
  } else {
    const attachedExtruders = printer.getAttachedExtruders();
    const hasExperimentalExtruder = attachedExtruders && R.contains(ExtruderTypeEnum.mk13_experimental, attachedExtruders); // Experimental Extruder should only be assigned if connected to the
    // printer. Otherwise, don't pass extruder argument and have
    // sliceconfig return back default selected extruder.
    //
    // To revert back to default extruders going from Method => to Method X,
    // don't build out the extruders arg. (Offline/Archetypes only)

    if (!ignoreExtruderArgs && !hasExperimentalExtruder && !attachedExtruders.includes(undefined)) {
      args.extruders = attachedExtruders;
    }
  }

  return runSliceConfig(args, cancelToken, callback).then(schema => {
    // In rare cases, sliceconfig returns emptry string, (""), resulting
    // errors with SettingsUISchema. Instead, lets exit out early before
    // things break.
    // -- Ray
    if (R.isEmpty(schema)) return;
    const sc = new Sliceconfig(paths.getBinDir());
    schema.supported_configurations = sc.supportedConfigs();
    return new SettingsUISchema(schema);
  });
}
/**
 * get the current version of sliceconfig
 * @returns {promise} - promise result contains a string containing the version
 */


function getVersion(callback) {
  const args = {
    command: 'get_version'
  };
  return runSliceConfig(args, null, callback);
}
/**
 * runs sliceconfig for a command with arguments
 *
 * for more information on what arguments are needed with each command, see the
 * command line interface documentation in the slice config repo at
 * github.com/makerbot/sliceconfig/blob/develop/doc/command_line_interface.md
 *
 * @param {list} args - optional arguments for sliceconfig
 * @param {function} callback - a node.js style callback function
 *
 * @returns {promise} promise - promise result depends on the command passed in.
 *  if command['notify'] is false, resolve contains stdout from slice config. if
 *  command['notify'] is true, resolve is empty.
 */


function runSliceConfig(args, cancelToken, callback) {
  return q.Promise(function (resolve, reject, notify) {
    // call slice config using the given command and arguments
    const scRunner = new Sliceconfig(paths.getBinDir());

    if (args && args.command === 'slice') {
      console.log(args);
    }

    const sliceConfigProcess = scRunner.execute(args);

    if (cancelToken) {
      // if the cancel token has already been used (e.g. cancel was hit
      // while generating the dot thing to slice), reject.
      if (cancelToken.isCancelled()) {
        sliceConfigProcess.kill();
        const err = new Error('Cancelled');
        err.type = 'cancel';
        reject(err);
      } else {
        // if the cancel token is used while slicing, reject.
        cancelToken.on('cancel', () => {
          sliceConfigProcess.kill();
          const err = new Error('Cancelled');
          err.type = 'cancel';
          reject(err);
        });
      }
    }

    sliceConfigProcess.on('data', data => {
      switch (data.type) {
        case 'error':
          reject(data.content);
          break;

        case 'info':
          // figure out how to get this into the 'end' call
          resolve(data.content);
          break;

        case 'logging':
          // do some logging?
          break;

        case 'progress':
          // Only reporting for type: slice, not others
          // else pre-birdwing bots get 2 progress bars.
          if ((data.content.substring(0, 1) === '{' || data.content.substring(0, 1) === '[') && JSON.parse(data.content).type && JSON.parse(data.content).type === 'slice') {
            notify(data.content);
            break;
          }

          break;

        case 'default':
          console.log(`unsupported type\n:${data}`);
      }
    });
  }).nodeify(callback);
}
/**
 * Shutdown the builder (cleaning up any resources)
 */


function shutdown() {// tbd what do we need to shutdown here? existing processes?
}

module.exports = {
  build,
  getProfile,
  getSchema,
  getSupportedConfigurations,
  shutdown,
  config,
  getVersion
};
