// MB-support-plugin/lib/makerbot-printer.js
'use strict';

const q = require('q');

const path = require('path');

const Printer = require('eagle-print/model/printer');

const EventEmitter = require('eventemitter3');

const _ = require('lodash');

const croissant = require('./croissant');

const MBPluginError = require('./errors').MBPluginError;

const util = require('./util.js');

const PrintJobHelper = require('./print_job_helper.js');

const FirmwareListener = require('./firmwareListener');

const FileTransferHelper = require('./file_transfer_helper');

const {
  maybe
} = require('./utils/maybe');

const {
  PrinterStateEnum,
  PRINTER_CONNECTION_TIMEOUT,
  EXPERIMENTAL_EXTRUDER,
  PrinterGenderToDefaultExtruders,
  ExtruderTypeToPrettyName,
  FifthGenBots,
  SixthGenBots,
  BotTypeEnum,
  ExtruderTypeEnum
} = require('./constants');

const {
  parseConnType
} = require('./utils/printerConnectionUtil');

class MakerBotPrinter extends Printer {
  constructor(printerInfo, gender, machineConfig, connectionManager, flux) {
    const data = {
      id: printerInfo.uid,
      connected: true,
      printer_status: {
        state: 'Unauthenticated'
      }
    };
    super(data, null, gender);
    const self = this;
    self._flux = maybe(flux); // Notifications for critical updates (system, states, machine_config updates, etc..)

    self._updateNotifications = new EventEmitter(); // hook up to notifications from connectionManager

    self._connectionManager = connectionManager;

    self._connectionManager.on('update', self.updateFromPrinterInfo, this);

    self._connectionManager.on('connection_change', function (croissantPrinter) {
      // whenever connection changes, emit message
      self._updateNotifications.emit('conn_type_change', this._connectionManager.currentConnType);

      if (!croissantPrinter) {
        return self._disconnectPrinter();
      } else {
        self._initPrinter(croissantPrinter);
        /**
         * When a sixth gen printer comes back online, attempt to
         * update the spool materials. However, only the current
         * selected printer will complete the flow of updating the
         * spool materials. Check printSettingsActions.js'
         * loadSchema
         *
         * This listener does have a bug where ethernet connection
         * and wifi connection are not treated as the same
         * 'network' connection.
         *
         * --Ray
         **/


        self.updateSpoolMaterialsOnceAlive(); // once printer is authenticated, emit message

        self._updateNotifications.emit('printer_authenticated', this._connectionManager.currentConnType);
      }
    }, this);

    self._connectionManager.on('check_life', self._checkAlive, this);

    self._lastSignOfLife = null;
    self._printerInfo = printerInfo; // let's deprecate usage of this...

    self._printer = null; // croissant printer

    self._disconnectEvent = null; // resolves when printer disconnects
    // obj to hold the callback that gets called when disconnectEvent resolves

    self._disconnectCallbackObj = null;
    self._firmware = null;
    self._networkState = null;
    self._machineConfig = null;
    if (machineConfig) self._machineConfig = machineConfig; // these are status properties generic to eagle-print's printer

    self._info = {};
    self._info.printer_status = {};
    self.setState(PrinterStateEnum.Unauthenticated);
    self._info.printer_status.current_print_job = {};
    self._info.printer_status.current_print_job.percent_done = 0; // these ones are makerbot specific, in case they may come in handy

    self._info.printer_status.current_process = null;
    self._info.printer_status.toolheads = null; // Errors from error_notifications and from processes go in here;
    // an error will get removed when acknowledgeError gets called for it
    // or when an error_acknowledged is received

    self._errors = new Map(); // this should only be available if the bot's process is in the 'failed' state
    // should only ever have one entry - this is incorrect and has been changed now
    // It will contain all state change errors.

    self._info.printer_status.state_change_errors = new Map();
    self._info.update_firmware_available = null;
    self._cameraFeedActive = false;
    self._printJobHelper = new PrintJobHelper(self);
    self.updateFromPrinterInfo(printerInfo);
  }

  addError(error) {
    this._errors.set(error.error_id, error);

    this._updateNotifications.emit('error');
  }

  removeError(errorId) {
    this._errors.delete(errorId);

    this._updateNotifications.emit('error');
  }

  saveStateError(err) {
    err.isStateErr = true;
    this.addError(err);
  }

  clearStateErrors() {
    for (const errEntry of this._errors) {
      const errId = errEntry[0];
      const err = errEntry[1];

      if (err.isStateErr) {
        this.removeError(errId);
      }
    }
  }

  setState(state) {
    if (state !== this.getStatus().state) {
      this._info.printer_status.state = state;

      this._updateNotifications.emit('status', {});

      this._flux.map(flux => flux.actions.PrinterDetailsActions.status(this.getId(), {}));
    }
  }

  isAuthenticating() {
    const state = this.getStatus().state;
    return state === PrinterStateEnum.Authenticating || state === PrinterStateEnum.Reauthenticating;
  }

  _checkAlive(callback) {
    const self = this;
    const now = new Date(); // have heard from this printer recently; alive

    if (now - this._lastSignOfLife < PRINTER_CONNECTION_TIMEOUT) return callback(true); // may be alive; ping to check

    if (this._printer) {
      return q(this._printer.Ping()).timeout(PRINTER_CONNECTION_TIMEOUT).then(function () {
        callback(true);
      }).catch(function (err) {
        // So, zipping the logs makes the printer not respond to
        // ping requests for some reason (or something like that).
        // This is a subject of hot debate as to why. Regardless,
        // we want to ignore timeouts when zipping logs.
        if (self.currentProcess && self.currentProcess.name === 'ZipLogsProcess') return;

        self._updateNotifications.emit('dying_scream', err);

        console.error(`Printer ${self.getName()} disconnected due to error`, err);

        self._flux.map(flux => {
          flux.actions.CopyBotLogs.timeout(self.getId(), err);
          flux.actions.message.printerDisconnect(self.getId());
        });

        callback(false);
      }).done();
    } // have not heard from this printer, and nothing to ping; probably not alive


    if (this.isAuthenticating()) {
      const err = new Error();
      err.type = MBPluginError.DisconnectWhileAuthingError;
      err.message = 'Disconnected while authenticating';
      return callback(false, err);
    }

    return callback(false);
  }

  disconnect() {
    // _printer's disconnect event triggers _disconnectPrinter
    if (this._printer) {
      console.log('Disconnecting', this.getName());

      this._printer.disconnect();
    }

    this._disconnectPrinter();
  } // (sets a bunch of stuff back to default state when printer disconnects


  _disconnectPrinter() {
    this._lastSignOfLife = null;
    this._printer = null;
    this._disconnectEvent = null;
    this._disconnectCallbackObj = null; // TODO: handle this stuff in a more JS-y way with prototypes and stuff?

    this._info.printer_status = {};
    this._info.printer_status.current_print_job = {};
    this._info.printer_status.current_print_job.percent_done = 0;
    this._info.printer_status.current_process = null;
    this._info.printer_status.toolheads = null;
    this._printJobHelper = new PrintJobHelper(this);
    this._errors = new Map();

    this._connectionManager.setInitState();

    this.setState(PrinterStateEnum.Offline);
  }

  _initPrinter(croissantPrinter) {
    const self = this; // When we change _printers, we don't want the old _printer's disconnectEvent
    // continuation causing the MBPrinter object to change to the Offline state,
    // so here we take advantage of some js object stuff and make the old callback
    // null before we set the continuation for the new _printer's disconnectEvent

    if (self._disconnectCallbackObj) self._disconnectCallbackObj.callback = null;
    self._printer = croissantPrinter;
    const disconnObj = {
      callback: self._disconnectPrinter
    };
    self._disconnectCallbackObj = disconnObj; // this will resolve when the printer disconnects (hopefully)

    self._disconnectEvent = q(self._printer.disconnectEvent).catch(err => {
      // disconnected because of error
      self._updateNotifications.emit('dying_scream', err);

      console.error(`Printer ${self.getName()} disconnected because of error:`, err);
    });

    self._updateNotifications.on('doDisconnect', disconnectFlag => {
      /** ** Monkey Patching **
       *
       * To avoid killing the printer when attempting to reconnect via
       * another connType after a "dying_scream" event happened,
       * printerManager.js will emit a "doDisconnect" and pass a
       * flag either to disconnect the printer or wait while auth'ing.
       *
       * -- Ray
       **/
      if (disconnectFlag) {
        if (disconnObj.callback) disconnObj.callback.call(self);
      } else {
        return;
      }
    }); // clear any errors that happened during auth, since we've succeeded
    // at this point


    self._errors = new Map(); // firmware update listener & check

    this._firmware = new FirmwareListener(this);

    this._firmware.check();

    this._fileTransferHelper = new FileTransferHelper(this);

    const doProcessChecks = info => {
      const step = _.get(info, 'current_process.step');

      const cancelled = _.get(info, 'current_process.cancelled');

      if (cancelled && step === 'failed') {
        this._updateNotifications.emit('process_failed', info);
      } else if (cancelled && step === 'done') {
        this._updateNotifications.emit('process_cancelled', info);
      }

      if (step === 'completed' && !cancelled) this._updateNotifications.emit('process_completed', info);

      const id = _.get(info, 'current_process.id');

      if (id && id !== this.currentProcess.id) {
        this._updateNotifications.emit('process_started', info);
      }
    }; // hook up notifications


    self._printer.setSystemNotificationCallback(function (info) {
      self.updateFromSystemInfo(info);

      self._updateNotifications.emit('system', info);

      self._flux.map(flux => flux.actions.PrinterDetailsActions.system(self.getId(), info));
    });

    self._printer.setStateNotificationCallback(function (info) {
      doProcessChecks(info);
      self.updateFromSystemInfo(info);

      self._flux.map(flux => flux.actions.PrinterDetailsActions.state(self.getId(), info)); // this is a thing that should be happening here, but some undesirable
      // behavior with the handling of printer errors becomes visible with it.....
      // to be uncommented when the error handling gets fixed...
      //self._updateNotifications.emit('system', info);

    });

    self._printer.setErrorNotificationCallback(function (errCode, errId, source) {
      self._errorNotificationCallback(errCode, errId, source);

      self._flux.map(flux => flux.actions.PrinterDetailsActions.error(self.getId(), errCode, errId, source));
    });

    self._printer.setErrorAcknowledgedCallback(function (errId) {
      self._errorAcknowledgedCallback(errId);
    });

    self._printer.setExtruderChangeCallback(function (index, config) {
      self._updateFromExtruderChange(index, config);

      self._flux.map(flux => flux.actions.PrinterDetailsActions.extruderChange(self.getId(), index, config));
    });

    self._printer.setNetworkStateChangeCallback(function (state) {
      self._updateFromNetworkStateChange(state); // removing this because apparently nothing uses this action
      // self._flux.map(flux => {
      //     flux.actions.PrinterDetailsActions.networkChange(self.getId(), state);
      // });

    }); // If you thought to yourself, "Hey, 'Firmware' isn't spelled right", you'd be right! There's a typo in Kaiten,
    // so when we build Croissant, the typo carries over:
    // https://github.com/makerbot/Birdwing-Software/blob/master/firmware/kaiten/src/kaiten/firmwareupdates.py#L136


    self._printer.setFirwareUpdatesInfoChangeCallback((version, updateAvailable, isOnline, err, releaseDate, releaseNotes) => {
      self._updateNotifications.emit('firmware_updates_info_change', version, updateAvailable, releaseNotes);
    }); // Change to idle to break out of isAuth check in UpdatePrinterStatus


    self.setState(PrinterStateEnum.Pending); // // once printer is authenticated, emit message
    // self._updateNotifications.emit('printer_authenticated', this._connectionManager.currentConnType);

    self.updatePrinterStatus();
  }

  _errorNotificationCallback(errorCode, errorId, source) {
    const err = {
      code: errorCode,
      id: errorId,
      source
    };
    this.addError(err);
  }

  _errorAcknowledgedCallback(errorId) {
    this.removeError(errorId);
  }

  acknowledgeError(errorId) {
    // If this error was not internally-generated (ie. error code <= 0),
    // acknowledge the error with the printer; the error should then eventually
    // get cleared on this side from feedback from kaiten...
    if (errorId > -1) {
      let ackPromise = q();
      if (this.processMethodExists('acknowledge_error')) ackPromise = this.invoke('ProcessMethod', ['acknowledge_error']);else if (this.processMethodExists('acknowledge_failure')) ackPromise = this.invoke('ProcessMethod', ['acknowledge_failure']);
      return ackPromise.then(() => {
        return this.invoke('Acknowledged', [errorId]);
      }).then(() => {
        // clear the error after calling AcknowledgeError in case for
        // some reason we miss the error_acknowledged notification
        this._errorAcknowledgedCallback(errorId);
      }).done();
    } else {
      // else, just remove it
      this.removeError(errorId);
    }
  } // override getters for properties that can change


  getInfo() {
    const gender = this.getGender();
    return {
      vendor: gender.getVendor(),
      model: gender.getName(),
      firmware_version: this._fwVerDictToString(this._printerInfo.info.firmware_version)
    };
  }

  getSerialNumber() {
    return util.printerIDtoSerial(this.getId());
  }

  getName() {
    return this._printerInfo.name;
  }

  getCodeName() {
    return util.genderToCodeName(this.getGender().getName());
  }

  getConnectionType() {
    if (this._connectionManager) {
      return this._connectionManager.currentConnType;
    }

    return null;
  }

  getNetworkState() {
    if (this._networkState) {
      return this._networkState;
    }

    return null;
  }

  getIpAddress() {
    return this._printerInfo.info.ip;
  }

  getExtruderInfo() {
    return _.get(this._printerInfo, 'info.toolheads.extruder', []);
  }

  getExtruderTrackingInfo() {
    const extruderTrackingInfo = [];

    if (this._extruderStats && this._extruderStats.length > 0) {
      const attached = this.getAttachedExtruders();
      attached.map((extruder, extruderIndex) => {
        const extruderInfo = {
          extruder_id: null,
          extruder_version: null,
          extruder_print_time: null
        }; // Update extruderInfo stats if present...

        if (this._extruderStats[extruderIndex]) {
          extruderInfo.extruder_id = this._extruderStats[extruderIndex].serial;
          extruderInfo.extruder_version = this._extruderStats[extruderIndex].extruder_type;
          extruderInfo.extruder_print_time = this._extruderStats[extruderIndex].extruder_print_time;
        }

        extruderTrackingInfo[extruderIndex] = extruderInfo;
      });
      return extruderTrackingInfo;
    }
  }
  /**
   * Build all the extruder(s) information from systeminfo &
   * get_tool_usage_stats kaiten method.
   */


  getAllExtruderInfo() {
    const extruders = this.getAttachedExtruders();
    const isDual = extruders.length > 1;
    const toolheadInfo = this.getExtruderInfo();
    const toolUsageStatsPromises = [];
    const extruderStats = [];
    return q(extruders.map((extruder, index) => {
      toolUsageStatsPromises[index] = this.invoke('GetToolUsageStats', isDual ? [index] : null);
    })).then(() => Promise.all(toolUsageStatsPromises)).then(results => {
      results.map((result, index) => {
        const extruderType = this._getExtruderTypeName(extruders[index], index);

        const formattedPrintTime = result.extrusion_time_s ? this._getFormattedPrintTime(result.extrusion_time_s) : null;
        extruderStats[index] = _.merge(result, {
          extruder_type: extruderType,
          extruder_print_time: formattedPrintTime
        });
      });
      return _.merge([], extruderStats, toolheadInfo);
    });
  }

  getChamberInfo() {
    return _.get(this._printerInfo, 'info.toolheads.chamber', []);
  }

  getSystemInfo() {
    return this.invoke('GetSystemInformation').then(systemInfo => {
      const redundantInfo = ['current_process', 'firmware_version', 'has_been_connected_to', 'ip', 'machine_name', 'sound'];
      systemInfo = _.omit(systemInfo, redundantInfo);
      return systemInfo;
    });
  }

  getSpoolInfo() {
    if (!this.isSixthGen()) {
      return q(null);
    }

    if (!this.isConnected() || !this.isAuthenticated()) {
      return q([{
        tag_uid: null
      }, {
        tag_uid: null
      }]);
    } // 6th gens on old firmware
    // GetSpoolInfo(bayIndex) = individual spoolInfo


    if (this.compareCurrApiVersionWith('1.11.0') < 0) {
      return q([0, 1].map(spoolIndex => this.invoke('UpdateSpoolInfo', [spoolIndex]).then(() => this.invoke('GetSpoolInfo', [spoolIndex])))).then(spoolInfoPromises => q.all(spoolInfoPromises)).then(spoolInfos => {
        return spoolInfos.map(spoolInfo => {
          const {
            material_type
          } = spoolInfo;
          spoolInfo.material_name = util.getMaterialSliceConfigStringFromCode(material_type);
          return spoolInfo;
        });
      });
    } else {
      // 6th gens on new firmware
      // GetSpoolsInfo() = array of spool infos from both bays
      // remaining spoolInfo is also available via getSystemInfo, but doesn't
      // include other spool data for UI purposes (color, name, etc)
      return this.invoke('GetSpoolsInfo');
    }
  }

  getDisabledErrors() {
    return _.get(this._printerInfo, 'info.disabled_errors', []);
  }

  getCurrentExtruderType(extruderIndex = 0) {
    const extruderProfiles = _.get(this.machineConfig, 'extruder_profiles');

    const attachedExtrudersArray = _.get(extruderProfiles, 'attached_extruders');

    const supportedExtrudersObj = _.get(extruderProfiles, 'supported_extruders');

    if (attachedExtrudersArray && attachedExtrudersArray.length && supportedExtrudersObj) {
      const currentExtruderId = attachedExtrudersArray[extruderIndex].id;
      return supportedExtrudersObj[currentExtruderId];
    }
  } // sigh... I don't really want to put this here, but it looks like it would
  // be the cleanest for now.....


  isMissingExtruder() {
    const extruderInfos = this.getExtruderInfo();

    for (const exInfo of extruderInfos) {
      if (!exInfo.tool_present) return true;
    }

    return false;
  }

  getStatus() {
    return this._info.printer_status;
  }

  getDisplayStatus() {
    return this.getStatus().state;
  }

  getLastError() {
    // gets the most recent error
    let lastErr;

    for (const val of this._errors.values()) lastErr = val;

    return lastErr;
  }

  getStateError() {
    // this doesn't account very well for the possibility that there would be
    // more than one state error, but during correct functioning, there should
    // only be one....
    for (const value of this._errors.values()) {
      if (value.isStateErr) return value;
    }
  }

  acknowledgeLastError() {
    const lastErr = this.getLastError();
    this.acknowledgeError(lastErr.error_id);
  }

  setAttachedExtruders() {
    // This exists for consistency with archetype printers
    // and purposely does nothing.
    return;
  } // Translate an array of toolhead Ids ([100, 101] or [14]) to an
  // array of toolhead names (['mk14', 'mk14_s'] or ['mk13_impla'])


  getAttachedExtruders() {
    const extruders = [];
    const supportedExtruders = this._machineConfig ? this._machineConfig.extruder_profiles.supported_extruders : [];
    /**
     * Hacky Hack
     *
     * There will be a point in time on the first-time connection and/or reconnection
     * that _machineConfig's attached_extruder can conflict with sliceconfig schema's
     * attached_extruder. This hack addresses _machineConfig with mk14_c extruder which
     * has a different purge tower dimension, but in this state will retain the [63, 0]
     * position.
     *
     * -- Ray
     **/

    if (this.getStatus().state === PrinterStateEnum.Reconnecting) {
      const botType = this._machineConfig.bot_type;
      return PrinterGenderToDefaultExtruders[botType];
    } // On first-time connection and possibily on reconnection,
    // this._printerInfo.info.toolheads does not exist. It will
    // result in this.getExtruderInfo() to be empty [].
    // Use this._machineConfig if available and this.getExtruderInfo()
    // returns empty [].


    const attachedExtruders = _.isEmpty(this.getExtruderInfo()) && this._machineConfig ? this._machineConfig.extruder_profiles.attached_extruders : this.getExtruderInfo();
    attachedExtruders.map(attachedExtruder => {
      const id = attachedExtruder.tool_id; // FW issue (fw 2.6.0) - For experimental extruder, we're borrowing
      // one of the 'mk13' tool_ids that weren't being used, 99, which
      // on the firmware supportedExtruder dict is known as 'mk13'.
      // Here we assign the tool_id 99 to 'mk13_experimental':

      if (id === EXPERIMENTAL_EXTRUDER) {
        extruders.push('mk13_experimental');
      } else {
        extruders.push(supportedExtruders[id]);
      }
    });
    return extruders;
  }

  isAuthenticated() {
    return !!this._printer && this.getStatus().state !== PrinterStateEnum.Reconnecting && this.getStatus().state !== PrinterStateEnum.Unauthenticated;
  }

  isConnected() {
    return this.getStatus().state !== PrinterStateEnum.Offline && this.getStatus().state !== PrinterStateEnum.Reconnecting;
  }

  startPrint(filename, cancelToken) {
    const self = this;
    return self._printJobHelper.startPrint(filename, cancelToken);
  }

  acknowledgeBuildPlateCleared() {
    return this.invoke('ProcessMethod', ['build_plate_cleared']);
  } /// toolPath object made from what is returned by 'build' in makerbot.js...


  sendPrintFile(toolPath, progressCallback, cancelToken) {
    const self = this;
    const localPath = toolPath.get('path');
    const filename = path.basename(localPath);
    const remotePath = `/current_thing/${filename}`;
    return self._printJobHelper.sendPrintFile(localPath, remotePath, progressCallback, cancelToken);
  }

  cancel() {
    if (this._printJobHelper.managingPrintJob()) return q(this._printJobHelper.cancel());
    if (this._firmware.managingFwUpdate()) return this._firmware.cancelUpdate();

    if (this._printer) {
      return this.invoke('Cancel');
    }

    return q();
  }

  updateMachineConfig() {
    const self = this;
    return q(self._printer.GetMachineConfig()).then(function (config) {
      self._machineConfig = config;

      self._updateNotifications.emit('machine_config_change', self._machineConfig);
    });
  }

  _updateFromExtruderChange(index, config) {
    // 6th gens on new firmware
    if (this.isSixthGen() && this.isConnected() && this.isAuthenticated() && this.compareCurrApiVersionWith('1.11.0') >= 0) {
      this.invoke('GetExtrudersConfigs').then(configs => {
        this._machineConfig.extruder_profiles.attached_extruders = configs;

        this._updateNotifications.emit('machine_config_change', this._machineConfig);

        this._flux.map(flux => flux.actions.PrintSettingsActions.loadSchema({
          id: this.getId()
        }));
      });
    } else {
      // 5th-gens, and 6th-gens on earlier firmware that doesn't support getting
      // extruder configs from kaiten.
      // Firmware issue (fw 2.6.0)
      // Tool id 99 shares the same supported_extruder name, mk13.
      // However for EE, sliceconfig treats 99 as mk13_experiemental.
      //
      // Let updateFromSystemInfo handle the change for that scenario.
      if (config.id === EXPERIMENTAL_EXTRUDER) return;
      this._machineConfig.extruder_profiles.attached_extruders[index] = config;

      this._updateNotifications.emit('machine_config_change', this._machineConfig); // For all other extruder changes (Smart Extruder+ and Tough Extruder), loadSchema
      // must be called, so that the print setting store can have the new settings.


      this._flux.map(flux => flux.actions.PrintSettingsActions.loadSchema({
        id: this.getId()
      }));
    }
  }

  _updateFromNetworkStateChange(state) {
    this._networkState = state;

    this._updateNotifications.emit('network_state_change', state);
  }

  deauthenticate(username) {
    let deauthPromise = q();
    if (username && this.isAuthenticated() && this._printer.DeauthorizeIsSupported()) deauthPromise = this.invoke('Deauthorize', [username]);
    deauthPromise.then(() => {
      this.disconnect();
    }).done();
  }

  _fwVerDictToString(fwVerDict) {
    if (fwVerDict) {
      return [fwVerDict.major, fwVerDict.minor, fwVerDict.bugfix, fwVerDict.build].join('.');
    }
  }

  _getFormattedPrintTime(timeInSeconds) {
    // Convert seconds to HH:MM:SS
    const hours = Math.floor(timeInSeconds / 3600);
    const minutes = Math.floor(timeInSeconds % 3600 / 60);
    const seconds = timeInSeconds % 60;
    const printTime = `${hours}:${minutes}:${seconds}`;
    return printTime;
  }

  _getExtruderTypeName(extruderCode, extruderIndex) {
    // Use the current extruderCode or grab the default extruderCode
    // from PrinterGenderToDefaultExtruders enum
    const currBotType = this._printerInfo.info.bot_type;
    extruderCode = extruderCode || PrinterGenderToDefaultExtruders[currBotType][extruderIndex]; // Convert extruderCode to pretty name

    const extruderType = ExtruderTypeToPrettyName[extruderCode];
    return extruderType;
  }

  updatePrinterStatus() {
    const self = this;
    self._lastSignOfLife = new Date();
    const state = this.getStatus().state;

    if (state === PrinterStateEnum.Busy || state === PrinterStateEnum.Pending || state === PrinterStateEnum.Idle) {
      if ('current_process' in self.currentState && self.currentState.current_process !== null) {
        const cp = self.currentState.current_process;
        self.setState(PrinterStateEnum.Busy);
        self._info.printer_status.current_process = cp;

        if (cp.name == 'PrintProcess' && cp.step == 'printing') {
          self._info.printer_status.current_print_job.percent_done = cp.progress;
        } // The above logic seems to be incorrect. From the Kaiten API doc,
        // it appears that state change errors do not get cleared out by
        // error_acknowledged notification. So, if we put them in errors map
        // we will need to have special logic to clear them out. Infact,
        // we may not ever clear state change errors because the only way
        // we clear state change errors is when we get a state notification
        // with null errors...with no indication of the error_id to clear
        // Experimenting with following solution for this messy situation
        //     - put all errors (error notifications and state change errors)
        //       in errors map
        //     - also remember state change errors separately. When we get
        //       a state notification with no errors, use the stored error id
        //       to clear from the error map


        if (cp.error) {
          self.saveStateError(cp.error);
        } else {
          self.clearStateErrors();
        }

        self._updateNotifications.emit('status', {});

        this._flux.map(flux => flux.actions.PrinterDetailsActions.status(self.getId(), {}));
      } else {
        self.setState(PrinterStateEnum.Idle);
        self._info.printer_status.current_print_job = {};
        self._info.printer_status.current_process = null;
        self._info.printer_status.current_print_job.percent_done = 0;
        self.clearStateErrors();
      }

      if ('toolheads' in self.currentState) {
        self._info.printer_status.toolheads = self.currentState.toolheads;
      }
    }
  }

  updateFromPrinterInfo(printerInfo) {
    const self = this;

    if (self._printer) {
      // Separate kaiten call to get CloudService information
      q(self._printer.GetCloudServicesInfo()).then(cloudServicesInfo => {
        self._printerInfo.info = _.assign(self._printerInfo.info, cloudServicesInfo);
      });
    }

    self._printerInfo.info = _.assign(self._printerInfo.info, printerInfo.info);

    _.forOwn(printerInfo, (value, key) => {
      // this disallows empty bot names/addresses/uids, but that
      // would be silly anyways...
      if (_.get(printerInfo, key) && key !== 'info') {
        self._printerInfo[key] = value;
      }
    });

    self.updatePrinterStatus();
  }

  getCameraFeed(cameraFrameCallback) {
    if (this._printer) {
      const self = this;
      self._cameraFeedActive = true;

      self._printer.setCameraFrameNotification(function (frame) {
        const frameBuf = new Buffer(new Uint8Array(frame), 'binary'); // convert binary data to base64 for displaying

        const frameAsB64 = `data:image/jpeg;base64,${frameBuf.toString('base64')}`;

        if (util.checkJpeg(frameBuf)) {
          cameraFrameCallback(frameAsB64);
        } else {
          console.log(`Received corrupt camera frame: ${frameAsB64}`);
        }
      });

      q(self.invoke('RequestCameraStream')).catch(function (err) {
        self._cameraFeedActive = false;
        throw err;
      }).done();
    }
  }

  endCameraFeed() {
    if (this._printer) {
      const self = this;
      q(self.invoke('EndCameraStream')).then(function () {
        self._cameraFeedActive = false;

        self._printer.unsetCameraFrameNotification();
      }).catch(console.error).done();
    }
  }

  checkAuthenticated() {
    const self = this;

    if (!self.isAuthenticated()) {
      const err = new Error();
      err.type = 'UnauthenticatedError';
      err.message = 'Printer is Unauthenticated or Offline';
      throw err;
    }
  }

  updateFromSystemInfo(info) {
    /**
     * (Currently in Sombrero firmware 0.54.1.216 and in the distant
     * future...) 'updateSpoolMaterialsFromSystemInfo' is another hack
     * around since kaiten is not emitting a notification for spool
     * changes.
     *
     * For now, update spoolInfo in printSettingsStore when there is a
     * difference between tag_uid(s) or when a printer comes back online.
     * Check 'updateSpoolMaterialsOnceAlive'.
     *
     * ::eelroll::
     *
     * -- Ray xoxoxo
     **/
    this.updateSpoolMaterialsFromSystemInfo(info); // Current Birdwing Firmware (2.6.0) doesn't recongize experimental
    // extruder (tool_id: 99) as an extruder change and won't throw an
    // extruder change notification. This is a work around for that issue.

    info.toolheads.extruder.map(extruder => {
      // escape hatch for when there's no machine config
      if (!this._machineConfig) return; // this is a pointer to the extruder! mutating it will mutate
      // _machineConfig!

      const prevTool = this._machineConfig.extruder_profiles.attached_extruders[extruder.index];
      const currToolId = extruder.tool_id;

      if (currToolId !== 0 && currToolId !== prevTool.id) {
        // Looks like machine_config updates the attached_extruder
        // calibrated flag even though the toolhead ids are different
        prevTool.id = currToolId;

        this._updateNotifications.emit('machine_config_change', this._machineConfig); // Callout to loadSchema to filter out any user_settings vs
        // available_settings in printSettingsStore


        this._flux.map(flux => flux.actions.PrintSettingsActions.loadSchema({
          id: this.getId()
        }));
      }

      const hasExperimental = this._machineConfig.extruder_profiles.attached_extruders.reduce((acc, e) => acc || e.id === EXPERIMENTAL_EXTRUDER, false);

      if (hasExperimental) {
        this._flux.map(flux => flux.actions.PrinterDetailsActions.showEETermsAndConditions());
      }
    });
    const printerInfo = new croissant.PrinterInfo();
    printerInfo.name = info.machine_name;
    printerInfo.address = this._printerInfo.address;
    printerInfo.uid = this._printerInfo.uid;
    printerInfo.info = info;
    printerInfo.connType = parseConnType(this._printerInfo);
    this.updateFromPrinterInfo(printerInfo);
  }
  /**
   * Temporary/Permanent hack for detecting spool changes
   * based on systemInfo updates.
   *
   * There are issues with this hack. On spool changes, it
   * does get called twice in a row. The second call, it will
   * detect an empty spool, but the selectedMaterials
   * in PrintSettingsStore will remain unaffected.
   *
   * -- Ray
   **/


  updateSpoolMaterialsFromSystemInfo(info) {
    if (!this.isSixthGen()) return; // patch work to bypass this check for spool material until
    // the old _printerInfo.info gets updated with spool material
    // --Ray

    let oldTagUids = [];

    if (this._printerInfo.info.filamentbays) {
      oldTagUids = this._printerInfo.info.filamentbays.map(bay => bay.tag_uid);
    } else {
      return;
    }

    const newTagUids = info.filamentbays.map(bay => bay.tag_uid);

    if (!_.isEqual(oldTagUids, newTagUids)) {
      /** OLD 1.3v Method/MethodX support **/
      if (this.isSixthGen() && this.isConnected() && this.isAuthenticated() && this.compareCurrApiVersionWith('1.11.0') < 1) {
        const spoolsInfoPromises = newTagUids.map((tag_uid, index) => {
          const checkSpoolTagUid = () => {
            const spoolInfoPromise = this.invoke('GetSpoolInfo', [index]);

            if (spoolInfoPromise) {
              return spoolInfoPromise.then(result => {
                if (result.tag_uid !== tag_uid) {
                  return checkSpoolTagUid();
                } else {
                  return result;
                }
              });
            } else {
              return null;
            }
          };

          return checkSpoolTagUid();
        });
        q.all(spoolsInfoPromises).then(spoolsInfo => {
          this._flux.map(flux => {
            flux.actions.PrintSettingsActions.updateMaterials({
              id: this.getId(),
              spoolsInfo
            });
          });
        });
      } else {
        // keep invoking 'GetSpoolsInfo' until the RFID reader
        // returns a result with matching tag_uid
        const checkSpoolTagUids = () => {
          return this.invoke('GetSpoolsInfo').then(spoolsInfo => {
            const resultTagUids = spoolsInfo.map(spoolInfo => spoolInfo.tag_uid);

            if (!_.isEqual(resultTagUids, newTagUids)) {
              return checkSpoolTagUids();
            } else {
              return spoolsInfo;
            }
          });
        };

        checkSpoolTagUids().then(spoolsInfo => {
          this._flux.map(flux => {
            flux.actions.PrintSettingsActions.updateMaterials({
              id: this.getId(),
              spoolsInfo
            });
          });
        });
      }
    }
  } // Compare Printer's Croissant API version with a "cutoff" version


  compareCurrApiVersionWith(cutoffApiStr) {
    const currApiStr = this._printerInfo.info.api_version;
    return this.compareVersions(currApiStr, cutoffApiStr);
  }
  /*
   * Takes current and cutoff version numbers in major.minor.bugfix format, and
   * breaks them down and compares them.
   *
   * return 1  if current version greater than the cutoff version
   * return 0  if current version matches the cutoff version
   * return -1 if current version less than the cuttoff version
   */


  compareVersions(current, cutoff) {
    if (cutoff === current) {
      return 0;
    }

    const cutoffArr = cutoff.split('.');
    const currArr = current.split('.');

    for (let i = 0; i < currArr.length; i++) {
      // Current version bigger than Cutoff version
      if (parseInt(currArr[i]) > parseInt(cutoffArr[i])) {
        return 1;
      } // Cutoff version bigger than Current version


      if (parseInt(currArr[i]) < parseInt(cutoffArr[i])) {
        return -1;
      }
    } // Otherwise they are the same.


    return 0;
  }
  /**
   * A method to call out to onGetPrinterSchema in PrintSettingsStore via
   * PrintSettingsAction.
   *
   * This will be called upon connection changed and should cover the
   * scenario of when a printer comes online.
   *
   * -- Ray
   **/


  updateSpoolMaterialsOnceAlive() {
    if (!this.isSixthGen()) return;
    /** OLD 1.3v Method/MethodX support **/

    if (this.isSixthGen() && this.isConnected() && this.isAuthenticated() && this.compareCurrApiVersionWith('1.11.0') < 0) {
      const spoolsInfoPromise = [0, 1].map(bayIndex => {
        return this.invoke('GetSpoolInfo', [bayIndex]);
      });
      q.all(spoolsInfoPromise).then(spoolsInfo => {
        this._flux.map(flux => {
          flux.actions.PrintSettingsActions.updateMaterials({
            id: this.getId(),
            spoolsInfo
          });
        });
      });
    } else {
      this.invoke('GetSpoolsInfo').then(spoolsInfo => {
        this._flux.map(flux => {
          flux.actions.PrintSettingsActions.updateMaterials({
            id: this.getId(),
            spoolsInfo
          });
        });
      });
    }
  }

  checkForUnsupportedExtruderConfigurations() {
    if (!this.isSixthGen() || !this.isConnected() || !this.isAuthenticated()) {
      return {
        extruderSetupError: false
      };
    } else {
      return this._flux.map(flux => {
        const payload = {};
        const printerStore = flux.store('PrinterStore');
        const selectedPrinter = printerStore.getSelectedPrinter();
        /** Update Nov 7th 2019
         * As of firmware 1.4.0.344, get_extruders_configs() will
         * now provide information if extruder slot has incorrect
         * extruder/material.
         *
         * EXECPT ATM material information is not available with
         * the latest 1.4 RC.
         *
         * FYI - extruder update notification will update the
         * machine_config.extruder_profiles.attached_extruders
         * profiles. Check _updateFromExtruderChange
         **/

        const hasUnsupportedExtruderConfigs = () => {
          if (!this._machineConfig) return false;
          return this._machineConfig.extruder_profiles.attached_extruders.reduce((res, extruderConfig) => {
            if (!extruderConfig.can_pair_tools) res = true;
            return res;
          }, false);
        };

        const hasMismatchedExtruderConfigs = () => {
          if (this._machineConfig === null) return false;
          return this._machineConfig.extruder_profiles.attached_extruders.reduce((res, extruderConfig) => {
            if (!extruderConfig.can_dual_print) res = true;
            return res;
          }, false);
        };

        const hasMissingToolheads = () => {
          // Null checking with printerInfo.info.toolheads
          // Happens on first time connections with sombrero printers
          if (!this._printerInfo.info.toolheads) return false;
          return this._printerInfo.info.toolheads.extruder.reduce((res, extruder) => {
            if (!extruder.tool_present) {
              res = true;
            }

            return res;
          }, false);
        };

        const getSupportedPairedExtrudersConfigs = () => {
          return this._machineConfig.extruder_profiles.attached_extruders.reduce((res, extruderInfo, index) => {
            const {
              all_dual_tool_types
            } = extruderInfo; // Remove LABS extruder from the config list, since
            // it has its own rules

            const extrudersForThisIndex = all_dual_tool_types.filter(extruder => {
              if (extruder !== ExtruderTypeEnum.mk14_e) return extruder;
            });
            extrudersForThisIndex.forEach((extruder, index) => {
              if (res[index]) {
                res[index].push(extruder);
              } else {
                res[index] = [extruder];
              }
            });
            return res;
          }, []);
        };

        const getSupportedPerExtruderConfigs = () => {
          return this._machineConfig.extruder_profiles.attached_extruders.map(extruderConfig => extruderConfig.all_dual_tool_types);
        };

        const getErrorTypePerExtruder = () => {
          // Order of Error Priority
          // Missing Extruder > Unsupported Extruder > Mismatched Extruder
          const toolheadErrorPriority = ['missing', 'unsupported', 'mismatched'];
          return this._printerInfo.info.toolheads.extruder // map each extruderInfo/extruderConfig to the three possible
          // error status contained in one array:
          // [    0  ,     1     ,      2     ]
          // [missing, unsupported, mismatched] => slot 1 extruder
          // [missing, unsupported, mismatched] => slot 2 extruder
          .map((extruderInfo, extruderIndex) => {
            const extruderConfig = this._machineConfig.extruder_profiles.attached_extruders[extruderIndex];
            return [1, 2, 3].map(val => {
              let result;

              switch (val) {
                // Check for missing extruder
                case 1:
                  result = !extruderInfo.tool_present;
                  break;
                // Check for unsupported extruder

                case 2:
                  result = !extruderConfig.can_pair_tools && extruderConfig.can_dual_print;
                  break;
                // Check for mismatch extruder

                case 3:
                  result = !extruderConfig.can_pair_tools && !extruderConfig.can_dual_print;
                  break;

                default:
                  result = false;
              }

              return result;
            });
          }) // translate each extruder's array results to toolhead error string
          .map(results => toolheadErrorPriority[results.findIndex(res => !!res)] || 'none');
        }; // only block user if current bot contains unsupported
        // extruder configuration...


        if (selectedPrinter && selectedPrinter.getId() !== this.getId()) {
          payload.extruderSetupError = false;
        } else if (this.isSixthGen() && this.isConnected() && this.isAuthenticated() && this.compareCurrApiVersionWith('1.11.0') < 0) {
          /** OLD 1.3v Method/MethodX support **/
          const hasUnsupportedExtruderConfigs = () => {
            return this._printerInfo.info.toolheads.extruder.reduce((res, extruder) => {
              if (!extruder.tool_type_correct || !extruder.tool_present) {
                res = true;
              }

              return res;
            }, false);
          };

          if (this._printerInfo.info.toolheads && hasUnsupportedExtruderConfigs()) {
            // Order of Error Priority
            // Missing Extruder > Unsupported Extruder > Mismatched Extruder
            const toolheadErrorPriority = ['missing', 'unsupported', 'mismatched'];
            payload.extruderSetupError = true;
            payload.perExtruderErrorType = this._printerInfo.info.toolheads.extruder.map(extruder => {
              return [1, 2, 3].map(val => {
                let result;

                switch (val) {
                  // Error handling for missing extruders
                  case 1:
                    result = !extruder.tool_present;
                    break;
                  // 1.3v fw doesn't report unsupported extruders

                  case 2:
                    result = false;
                    break;
                  // Error handling for mismatch extruders

                  case 3:
                    result = !extruder.tool_type_correct;
                    break;

                  default:
                    result = false;
                }

                return result;
              });
            }) // translate each extruder's array results to toolhead error string
            .map(results => toolheadErrorPriority[results.findIndex(res => !!res)] || 'none'); // Hardcode the supported extruder configurations

            payload.supportedConfigs = this._printerInfo.info.bot_type === BotTypeEnum.fire_e ? {
              pairedExtrudersConfigs: [['mk14', 'mk14_s']],
              perExtruderConfigs: [['mk14'], ['mk14_s']]
            } : {
              pairedExtrudersConfigs: [['mk14', 'mk14_s'], ['mk14_hot', 'mk14_hot_s']],
              perExtruderConfigs: [['mk14', 'mk14_hot'], ['mk14_s', 'mk14_hot_s']]
            };
          } else {
            payload.extruderSetupError = false;
          }
        } else {
          if (this.isSixthGen() && this.isConnected() && this.isAuthenticated() && this._printerInfo.info.toolheads && this._machineConfig && (hasMissingToolheads() || hasMismatchedExtruderConfigs() || hasUnsupportedExtruderConfigs())) {
            payload.extruderSetupError = true;
            payload.perExtruderErrorType = getErrorTypePerExtruder();
            payload.supportedConfigs = {
              pairedExtrudersConfigs: getSupportedPairedExtrudersConfigs(),
              perExtruderConfigs: getSupportedPerExtruderConfigs()
            };
          } else {
            payload.extruderSetupError = false;
          }
        }

        return payload;
      }).lift();
    }
  }

  hasExtruderMaterialMismatch() {
    const supportedMaterialsByExtruder = [];
    const extruders = this._machineConfig ? this._machineConfig.extruder_profiles.attached_extruders : this.getAttachedExtruders();
    extruders.map(extruder => {
      if (extruder && extruder.supported_materials) {
        supportedMaterialsByExtruder.push(extruder.supported_materials);
      }
    });
    return this._flux.map(flux => {
      return flux.store('PrintSettingsStore').hasExtruderMaterialMismatch(supportedMaterialsByExtruder);
    }).lift();
  }

  invoke(kaitenMethod, params) {
    const self = this;
    self.checkAuthenticated();
    if (!(kaitenMethod in self._printer)) throw `Method ${kaitenMethod} does not exist.`;

    try {
      return q(self._printer[kaitenMethod].apply(self._printer, params)).catch(function (err) {
        // jank
        if (err.message === 'null\n') {
          console.log(`Something went wrong when calling method ${kaitenMethod}; printer most likely disconnected`);
        } else {
          throw err;
        }
      });
    } catch (err) {
      console.error(err);
      return err;
    }
  }

  getCurrentJob() {
    if (this.currentState && this.currentState.current_process) {
      const cp = this.currentState.current_process; // TODO GC-26618 - This is dirty fix for 0.9 release. This fixes/enables print jobs in Schedule view in ED.

      return {
        // elapsed_time helps to make sure that we are displaying printing progress (not homing or heating progress)
        percent_done: cp.elapsed_time > 0 ? cp.progress : 0,
        elapsed_build_time: cp.elapsed_time ? cp.elapsed_time / 60 : 0,
        estimated_build_time: cp.time_estimation ? cp.time_estimation / 60 : 0,
        submit_time: cp.start_time ? new Date(cp.start_time).valueOf() / 1000 : 0,
        job_id: cp.id ? cp.id : 0,
        filename: cp.filename ? cp.filename.split('/').pop() : null
      };
    }

    return null;
  }

  updateExtruderStats() {
    this._extruderStats = [];
    const isDual = this.getAttachedExtruders().length > 1;
    const extruderStatsPromises = [];
    q(this.getAttachedExtruders().map((extruder, index) => {
      extruderStatsPromises[index] = this.invoke('GetToolUsageStats', isDual ? [index] : null);
    })).then(() => q.all(extruderStatsPromises)).then(results => {
      results.map((result, index) => {
        const extruderType = this._getExtruderTypeName(this.getAttachedExtruders()[index], index);

        const formattedPrintTime = this._getFormattedPrintTime(results[index].extrusion_time_s);

        result.extruder_type = extruderType;
        result.extruder_print_time = formattedPrintTime;
        this._extruderStats[index] = result;
      });
    });
  }

  getTrackingInfo() {
    const data = util.trackingInfoForPrinter(this);
    data.extruders = this.getExtruderTrackingInfo();
    return data;
  }

  processMethodExists(method) {
    const processMethods = _.get(this.getStatus(), 'current_process.methods', []);

    return processMethods.indexOf(method) >= 0;
  }
  /** @private
   * Handles loading and unloading of the filament.
   */


  _changeFilament(processMethod, idleMethod, extruderIndex) {
    const extruder = extruderIndex === undefined ? 0 : extruderIndex;

    if (this.getStatus().state === PrinterStateEnum.Busy) {
      return this.invoke('ProcessMethod', [processMethod, extruder]);
    }

    return this.invoke(idleMethod, [extruder]);
  }
  /**
   * Getting the printer to load/unload filament while it's idle vs. when it's busy are different,
   * hence the distinction between processMethod and idleMethod.
   */


  loadFilament(extruderIndex) {
    const processMethod = 'load_filament';
    const idleMethod = 'LoadFilament';
    return this._changeFilament(processMethod, idleMethod, extruderIndex);
  }

  unloadFilament(extruderIndex) {
    const processMethod = 'unload_filament';
    const idleMethod = 'UnloadFilament';
    return this._changeFilament(processMethod, idleMethod, extruderIndex);
  }
  /**
   * Handles stopping the filament from extruding after it has been loaded. As in `_changeFilament()`,
   * stopping the filament is different while it's idle vs when it's busy. Furthermore, there seems to
   * be a bug in which when you invoke `stop_filament` while the printer is idle, it will cause the screen
   * to freeze, if it has one.
   */


  stopFilament() {
    if (this.getStatus().state === PrinterStateEnum.Busy) {
      return this.invoke('ProcessMethod', ['stop_filament']);
    } // Can't use "stop_filament" now as it causes a the printer screen to crash.
    // See PEN-1104 for more details.


    return this.cancel();
  }

  isFifthGen() {
    const currentBotType = this._printerInfo.info.bot_type;
    return FifthGenBots.includes(currentBotType);
  }

  isSixthGen() {
    const currentBotType = this._printerInfo.info.bot_type;
    return SixthGenBots.includes(currentBotType);
  }

  isSketchPrinter() {
    return false;
  }

  get machineConfig() {
    return this._machineConfig;
  }

  get currentState() {
    return this._printerInfo.info;
  }

  get cameraFeedActive() {
    return this._cameraFeedActive;
  }

  get currentProcess() {
    return this._info.printer_status.current_process || {};
  }

  get updateNotifications() {
    return this._updateNotifications;
  }

  get networkState() {
    return this._networkState;
  }

  get firmware() {
    return this._firmware;
  }

  get fileTransferHelper() {
    return this._fileTransferHelper;
  }

  get soundState() {
    return _.get(this._printerInfo, 'info.sound');
  }

  get analyticsState() {
    return _.get(this._printerInfo, 'info.analytics_enabled');
  } // /stores/printerStore.js


  injectFlux(flux) {
    this._flux = maybe(flux);
    return this;
  }

}

module.exports = MakerBotPrinter;
