// MB-support-plugin/lib/mb-project-validator-provider.js
'use strict';

const R = require('ramda');

const ProjectValidatorProvider = require('eagle-print/services/providers/projectValidatorProvider');

const {
  SettingsKeys,
  ValidatorSettingsTypes: SettingsTypes
} = require('./constants/print_settings');

class MBProjectValidatorProvider extends ProjectValidatorProvider {
  constructor() {
    super();
  }

  getPluginKey() {
    return 'mb';
  }

  getPropertyValidationSchema(schemaVersion, versions) {
    const mbPropertySchema = {};
    const basicSettings = ['Extruder', 'PrintModes', 'Supports', 'InfillDensity', 'LayerHeight', 'NumShells', 'Rafts'];
    /** If a new print settings option's type was not listed in
     * SettingsTypes, then getOptionalPluginRef will default to
     * 'GeneralElemNumber' which there is a high chance that
     * is correct type.
     */

    const getOptionalPluginRef = setting => {
      let type;

      switch (R.prop(setting, SettingsTypes)) {
        case 'boolean':
          type = 'GeneralElemBoolean';
          break;

        case 'string':
          type = 'GeneralElemString';
          break;

        case 'unsigned_integer':
          type = 'GeneralElemNumber';
          break;

        case 'array':
          type = 'GeneralElemArray';
          break;

        default:
          type = 'GeneralElemNumber';
      }

      return this.optionalPluginRef('print', type);
    };

    if (schemaVersion >= versions.INITIAL_VERSION.schema) {
      mbPropertySchema.rootOccurrence = R.zipObj(R.map(key => key === 'InfillDensity' ? 'infillDensity' : SettingsKeys[key], basicSettings), R.map(key => getOptionalPluginRef(SettingsKeys[key]), basicSettings));
    }

    if (schemaVersion >= versions.R1_9_DEVELOPMENT_START.schema) {
      mbPropertySchema.rootOccurrence = R.zipObj(R.map(key => {
        switch (key) {
          case SettingsKeys.Extruders:
            return 'attached_extruder0';

          case SettingsKeys.ExtruderTemp:
            return 'extruderTemp0';

          case SettingsKeys.OutlineFeedRate:
            return 'extrusionProfiles>outlines>feedrate';

          case SettingsKeys.InfillFillThickness:
            return 'extruderProfiles>0>maxSparseFillThickness';

          case SettingsKeys.SparseFeedRate0:
            return 'extrusionProfiles>infill>feedrate';

          case SettingsKeys.RaftBaseFeedRate:
            return 'extrusionProfiles>raftBase>feedrate';

          case SettingsKeys.RaftBaseFanSpeed:
            return 'extrusionProfiles>raftBase>fanSpeed';

          case SettingsKeys.BaseLayerSurfaceFeedRate0:
            return 'extrusionProfiles>firstModelLayer>feedrate';

          case SettingsKeys.FirstModelLayerCoolingFanSpeed:
            return 'extrusionProfiles>firstModelLayer>fanSpeed';

          default:
            return key;
        }
      }, R.values(SettingsKeys)), R.map(getOptionalPluginRef, R.values(SettingsKeys)));
      mbPropertySchema.rootOccurrence.attached_extruder0 = this.optionalPluginRef('print', 'GeneralElemString');
    }

    if (schemaVersion >= versions.R2_6_DEVELOPMENT_START.schema) {
      mbPropertySchema.rootOccurrence = R.zipObj(R.values(SettingsKeys), R.map(getOptionalPluginRef, R.values(SettingsKeys)));
    }

    if (schemaVersion >= versions.R2_9_DEVELOPMENT_START.schema) {
      mbPropertySchema.rootOccurrence = R.zipObj(R.values(SettingsKeys), R.map(getOptionalPluginRef, R.values(SettingsKeys)));
    }

    return mbPropertySchema;
  }

  getValidationSchema(schemaVersion, versions, properties) {
    const mbProjectSchema = {
      properties: {},
      definitions: {}
    };
    return mbProjectSchema;
  }

}

module.exports = MBProjectValidatorProvider;
