// MB-support-plugin/lib/mb-project-version-provider.js
'use strict';

const ProjectVersionProvider = require('eagle-print/services/providers/projectVersionProvider');

class MBProjectVersionProvider extends ProjectVersionProvider {
  constructor() {
    super();
    this._versionNumberMap = {
      1: '0.12',
      100: '0.13',
      110: '0.14',
      150: '0.15',
      160: '0.16',
      170: '1.0',
      1010: '1.7',
      1090: '1.9',
      2060: '2.6',
      2070: '2.7',
      2080: '2.9'
    };
  }

  getCurrentVersionNumber() {
    return {
      pluginKey: 'mb',
      version: 2090
    };
  }

  mapProjectVersionToPrintVersion(projectVersion) {
    return this._versionNumberMap[projectVersion];
  }

  getVersions() {
    return {
      INITIAL_VERSION: {
        version: 1,
        schema: 1,
        comment: 'Start of history'
      },
      R013_DEVELOPMENT_START: {
        version: 100,
        schema: 1,
        comment: 'R0.13 Development'
      },
      R014_DEVELOPMENT_START: {
        version: 110,
        schema: 1,
        comment: 'R0.14 Development'
      },
      R015_DEVELOPMENT_START: {
        version: 150,
        schema: 1,
        comment: 'R0.15 Development'
      },
      R016_DEVELOPMENT_START: {
        version: 160,
        schema: 1,
        comment: 'R0.16 Development'
      },
      R017_DEVELOPMENT_START: {
        version: 170,
        schema: 1,
        comment: 'R0.17 Development'
      },
      R1_1_DEVELOPMENT_START: {
        version: 1010,
        schema: 1,
        comment: 'R1.1 Development'
      },
      R1_9_DEVELOPMENT_START: {
        version: 1090,
        schema: 2,
        comment: 'Advanced Settings and Print Mode properties added'
      },
      R2_6_DEVELOPMENT_START: {
        version: 2060,
        schema: 3,
        comment: '2.6 Development'
      },
      R2_7_DEVELOPMENT_START: {
        version: 2070,
        schema: 4,
        comment: 'Experimental Extruder and Custom Print Modes'
      },
      R2_9_DEVELOPMENT_START: {
        version: 2090,
        schema: 5,
        comment: 'Handling dual extrusion'
      }
    };
  }

}

module.exports = MBProjectVersionProvider;
