// MB-support-plugin/lib/mb_toolpath_parser.js
'use strict';

const ToolpathParser = require('eagle-print/services/providers/toolpathParser');

const q = require('q');

const JSZip = require('jszip'); // let stream = require('stream');


const JSONStream = require('JSONStream');

const es = require('event-stream');

const {
  isEqual
} = require('lodash');

const ParamsProvider = require('./parameters');

const {
  BotTypeEnum
} = require('./constants');

const {
  GcodeParserStream
} = require('./gcode_toolpath_parser.js');

const INCHES_PER_MM = 0.0393701;
const EXTRUDER_A = '199';
const EXTRUDER_B = '198';
const SUPPORT = {
  [EXTRUDER_A]: '201',
  [EXTRUDER_B]: '207'
}; /// Object that holds state of the parser and updates it based on new commands.
///
/// @buildVolume - the bounds of the machine, used to offset toolpath commands
///   correctly.
/// @totalRaftLayers - number of layers that are in the raft
/// @extruderCount - the number of extruders used in this print

const ParserState = function (buildVolume, totalRaftLayers, extruderCount) {
  // we expect each layer to have the following format:
  //  {
  //    totalNrMoves: number, (all moves)
  //    totalNrHeadMoves: number, (moves that are just head moves)
  //    totalLength: number
  //    movesModel: { type, [Move]}
  //    movesSupport: { type, [Move]}
  //  }
  //
  this.totalMoves = 0;
  this.totalTravelMoves = 0;
  this.totalLength = 0; // NOTE: at some point we're going to want to break this down in a finer way
  // where we have lists of for each model on the plate. that will require a
  // change to how we encode moves though.

  this.movesModel = [];
  this.movesSupport = []; // variables that we'll use to compute the layer data

  this.index = 0;
  this.isModel = true;
  this.isTravel = true;
  this.lastMove = {};
  this.preCmdTag = null;
  this.totalLayers = 0;
  this.lastExtruderMoves = {
    a: 0,
    b: 0
  };
  this.currExtruder = null; // EXTRUDER_A, EXTRUDER_B, or null
  // store the meta data on the current layer

  this.lowerPosition = 0;
  this.upperPosition = 0;
  this.thickness = 0;
  this.width = 0; // generate x and y offsets. in eagle, (0, 0) is the lower left corner of
  // the buildplate. in miracle grue, it is the center of the build plate.

  this.xOffset = buildVolume.x / 2;
  this.yOffset = buildVolume.y / 2;
  this.totalRaftLayers = totalRaftLayers;
  this.extruderCount = extruderCount;
};

ParserState.prototype = {
  constructor: ParserState,

  /// parse a commands function out and pass it to the appropriate parsing
  /// method.
  ///
  /// @command - the jsontoolpath command to parse as an Object
  ///   * function <string> - what this commmands purpose is
  parseCommand(command) {
    if (command.function == 'move') {
      return this.parseMoveCommand(command);
    } else if (command.function == 'comment') {
      return this.parseCommentCommand(command);
    }
  },

  /// parse a comment command, emitting a new layer, and updating our layer
  /// meta data
  ///
  /// for each layer, we will have a comment declaring its chunk number, the
  /// upper bounds, the lower bounds, the width, and the thickness for that
  /// layer. there is also a comment for the end of a print.
  ///
  /// @command - the jsontoolpath comment command to parse as an Object
  ///   * parameters.comment <string> - a comment adding info about this layer
  parseCommentCommand(command) {
    const comment = command.parameters.comment; // regex used to pull a float out of a string

    const floatRegex = /\d+[.]?\d*/; // check to see if this comment is for the lower position

    const lowerPosRegex = /lower position.*\d+[.]?\d*/i;
    const lowerPositionMatch = comment.match(lowerPosRegex);

    if (lowerPositionMatch) {
      // note on syntax: the +(<string>) will cast the match as a number.
      this.lowerPosition = +comment.match(floatRegex)[0] * INCHES_PER_MM;
      return;
    } // check to see if this comment is for the upper position


    const upperPosRegex = /upper position.*\d+[.]?\d*/i;
    const upperPositionMatch = comment.match(upperPosRegex);

    if (upperPositionMatch) {
      this.upperPosition = +comment.match(floatRegex)[0] * INCHES_PER_MM;
      return;
    } // check to see if this comment is for the thickness


    const thicknessRegex = /thickness.*\d+[.]?\d*/i;
    const thicknessMatch = comment.match(thicknessRegex);

    if (thicknessMatch) {
      this.thickness = +comment.match(floatRegex)[0] * INCHES_PER_MM;
      return;
    } // check to see if this comment is for the width


    const widthRegex = /width.*\d+[.]?\d*/i;
    const widthMatch = comment.match(widthRegex);

    if (widthMatch) {
      this.width = +comment.match(floatRegex)[0] * INCHES_PER_MM;
      return;
    } // check to see if this comment is for the chunk number


    const chunkRegex = /chunk.*\d/i;
    const layerSectionRegex = /layer section.*\d.*[(]\d*[)]/i;

    if (comment.match(chunkRegex) || comment.match(layerSectionRegex)) {
      this.totalLayers++;
      return this.reachedEndOfLayer();
    }

    const materialRegex = /material.*(\d)/i;
    const materialMatch = comment.match(materialRegex);

    if (materialMatch) {
      const extruderID = materialMatch[1];

      if (extruderID === '0') {
        this.currExtruder = EXTRUDER_A;
      } else if (extruderID === '1') {
        this.currExtruder = EXTRUDER_B;
      } else {
        console.error('Unknown extruder id:', extruderID);
      }

      return;
    } // check to see if this comment is for the end of the print


    if (comment === 'End of print') {
      return this.reachedEndOfLayer();
    }
  },

  /// parse a move command and update our moves, travel moves, length, and
  /// moves arrays accordingly.
  ///
  /// @command - the jsontoolpath command to parse as an Object
  ///   * parameters - x, y, z, a, b and feedrate for this command
  ///   * tags - array of strings tagging this move
  parseMoveCommand(command) {
    // update the move index, current move, and current tag
    this.index++;
    this.currentMove = {
      x: (command.parameters.x + this.xOffset) * INCHES_PER_MM,
      y: (command.parameters.y + this.yOffset) * INCHES_PER_MM,
      z: command.parameters.z * INCHES_PER_MM,
      index: this.index
    };
    this.loadCurrentTag(command.tags[0]); // update the total moves and total travel moves

    if (this.isTravel) {
      this.totalTravelMoves++;
    }

    this.totalMoves++; // if this is a dual-material print, use extruder A vs extruder B logic

    if (this.extruderCount > 1) {
      if (this.currExtruder === EXTRUDER_A) {
        this.movesModel.push(this.currentMove);
      } else {
        this.movesSupport.push(this.currentMove);
      } // if this is a single-material print, use model vs support logic

    } else {
      if (this.isModel) {
        this.movesModel.push(this.currentMove);
      } else {
        this.movesSupport.push(this.currentMove);
        this.cleanupPrevRestartWithSupportMoves(command.tags[0]);
      }
    } // keep track of preview command tag for possible cleanup


    this.preCmdTag = command.tags[0]; // set the last move to the one we just parsed

    this.lastMove = this.currentMove;
  },

  /**
   * Bugfix/Patch (PEN-2878) -- Ray
   *
   * The assumption with threeEngineToolpathLoader is that the first supportMoves will
   * always have an area of zero.
   *
   * With the mb jsontp, it safe to assume that every support island has a "restart"
   * before it starts the extrusion of the support. The "restart" move in "this.movesModel"
   * has an area of zero. We'll just pop this move out and prepend it into "this.movesSupport".
   *
   * @param {string} currCmdTag
   */
  cleanupPrevRestartWithSupportMoves(currCmdTag) {
    if (this.preCmdTag === 'Restart' && currCmdTag === 'Support' && this.movesModel.length > 0) {
      const restartMove = this.movesModel.pop();
      const indexMove = this.movesSupport.findIndex(mv => isEqual(mv, this.currentMove));

      if (indexMove === 0) {
        this.movesSupport.unshift(restartMove);
      } else {
        this.movesSupport.splice(indexMove - 1, 0, restartMove);
      }
    }
  },

  /// update this.isTravelMove and this.isModel based on @tag{string}, a tag
  /// from miracle_grue.
  loadCurrentTag(tag) {
    switch (tag) {
      case 'Trailing Extrusion Move':
        this.isTravel = false;
        break;

      case 'Anchor':
      case 'Bridge':
      case 'Infill':
      case 'Inset':
      case 'Outline':
      case 'Spur':
        this.isTravel = false;
        this.isModel = true;
        break;

      case 'Support':
      case 'Purge':
      case 'Invalid Move':
        this.isTravel = false;
        this.isModel = false;
        break;

      case 'Backlash Compensation':
      case 'Connection':
      case 'Long Restart':
      case 'Long Retract':
      case 'Restart':
      case 'Retract':
      case 'Leaky Travel Move':
      case 'Travel Move':
        this.isTravel = true;
        break;

      default:
        if (this.logger) {
          this.logger.info(`tag not found: ${tag}`);
        }

    } // all raft layers are tagged as support


    if (this.totalLayers <= this.totalRaftLayers) {
      this.isModel = false;
    } // use the width and thickness of this layer to create the area.
    // TODO(zap) - edit the drawing of extrusions to better reflect width
    // and thickness instead of just area.


    this.currentMove.area = this.isTravel ? 0 : this.width * this.thickness;
  },

  /// update the deferred promise with the data on the finished layer and
  /// reinitialize layer data.
  reachedEndOfLayer() {
    let layer;
    const extruder = this.currExtruder || EXTRUDER_A; // check that we have actual moves in this layer

    if (this.totalMoves - this.totalTravelMoves !== 0) {
      // create layer object
      layer = {
        totalNrMoves: this.totalMoves,
        totalNrHeadMoves: this.totalTravelMoves,
        movesMaterial: {
          [extruder]: this.movesModel
        },
        movesSupport: {
          [SUPPORT[extruder]]: this.movesSupport
        },
        upperPosition: this.upperPosition,
        lowerPosition: this.lowerPosition
      };
    } // reset our state


    this.totalMoves = 0;
    this.totalTravelMoves = 0;
    this.totalLength = 0;
    this.movesModel = [];
    this.movesSupport = [];
    this.index = 0;
    this.isModel = true;
    this.isTravel = true;
    this.lastMove = {};
    return layer;
  }

};

class MBToolpathParser extends ToolpathParser {
  constructor(loggingService, miracleConfig, buildVolume) {
    super();
    this.logger = loggingService;
    /*
     *  {
     *    header: { sliceHeight: number },
     *    layers: [ {
     *      totalNrMoves: number, (all moves)
     *      totalNrHeadMoves: number, (moves that are just head moves)
     *      totalLength: number,
     *      movesModel: { type, [Move]}
     *      movesSupport: { type, [Move]}
     *    }]
     *  }
     */

    this.parsedData = {
      header: {
        sliceHeight: null
      },
      layers: []
    };
    this.paramsProvider = new ParamsProvider();
  } /// @return the extention for mb slice files


  getExt() {
    return '.makerbot';
  }

  getExtensions() {
    return ['.makerbot'];
  }

  unzipToolpath(read, filePath) {
    const result = {};
    return read(filePath).then(zippedFileData => {
      return JSZip.loadAsync(zippedFileData);
    }).then(zippedFile => {
      result.tpStream = zippedFile.files['print.jsontoolpath'] ? zippedFile.file('print.jsontoolpath').nodeStream() : zippedFile.file('print.gcode').nodeStream();
      return zippedFile.file('meta.json').async('string');
    }).then(metaData => {
      result.metaData = JSON.parse(metaData);
      return result;
    });
  } /// @metadata - the data from a meta.json file as a JSON object


  loadToolpath(metaData, tpStream, messaging) {
    const self = this; // reset parsed data option

    self.parsedData = {
      header: {
        sliceHeight: null
      },
      layers: []
    }; /// extract heights, and initialize parser state

    self.parsedData.header.sliceHeight = metaData.miracle_config.gaggles.default.layerHeight * INCHES_PER_MM; /// @miracleConfig - the config Object
    ///   * raftBaseLayers
    ///   * raftInterfaceLayers
    ///   * raftSurfaceLayers
    /// @return raft thickness in number of layers

    const calculateRaftLayers = function (miracleConfig) {
      // rafts are turned off, the height is 0
      if (!miracleConfig.doRaft) {
        return 0;
      } // start at zero and add up the height of the raft given the
      // number of layers and their thicknesses


      const {
        numberOfLayers: raftBaseLayers
      } = miracleConfig.raftProfiles.base;
      const {
        numberOfLayers: raftInterfaceLayers
      } = miracleConfig.raftProfiles.interface;
      const {
        numberOfLayers: raftSurfaceLayers
      } = miracleConfig.raftProfiles.surface;
      return raftBaseLayers + raftInterfaceLayers + raftSurfaceLayers;
    }; // Sometimes the meta.json has two extruder distances,
    // but the second one is 0.0. This accounts for that.


    const extruderCount = metaData.extrusion_distances_mm.length < 2 || metaData.extrusion_distances_mm[1] <= 0 ? 1 : metaData.extrusion_distances_mm.length;
    const buildVolumeParams = self.paramsProvider.getParamsByBotType(metaData.bot_type).buildVolume.envelopeSize;

    const getBuildVolumeInMM = function (buildVolumeInMeters) {
      const buildVolume = {};
      Object.keys(buildVolumeInMeters).forEach(axis => {
        buildVolume[axis] = buildVolumeParams[axis] * 1000;
      });
      return buildVolume;
    };

    const isSketchPrinter = metaData.bot_type === BotTypeEnum.sketch;
    const parsingParams = {
      totalRaftLayers: calculateRaftLayers(metaData.miracle_config.gaggles.default),
      buildVolume: getBuildVolumeInMM(buildVolumeParams),
      extruderCount,
      isSketchPrinter
    }; /// Parse a @toolpathNodeStream. when finished, resolve the @deferred promise
    /// with the parsed and formatted data.

    const parseToolpath = function (toolpathNodeStream, parsingParams) {
      messaging.display();
      const defered = new q.defer();
      const parserState = new ParserState(parsingParams.buildVolume, parsingParams.totalRaftLayers, parsingParams.extruderCount);

      const parseCommand = function (command) {
        return parserState.parseCommand(command);
      };

      if (!parsingParams.isSketchPrinter) {
        // pipe stream to a jsonstream parser that parses out each command.
        // pipe parsed out json blobs to the parseCommand function above.
        toolpathNodeStream.pipe(JSONStream.parse('*.*')).pipe(es.mapSync(parseCommand)).on('data', function (layer) {
          self.parsedData.layers.push(layer);
        }).on('end', function (data) {
          messaging.dismiss();
          defered.resolve(self.parsedData);
        });
      } else {
        // For Sketch printers, the toolpathNodeStream is a gcode
        // readable file nodejs stream.
        // Check "unzipToolpath" for more info
        const parserStream = GcodeParserStream(toolpathNodeStream, parsingParams);
        parserStream.on('data', chunk => {
          self.parsedData.layers.push(chunk);
        }).on('end', () => {
          messaging.dismiss();
          defered.resolve(self.parsedData);
        });
      } // return the promise from the parser state.


      return defered.promise;
    };

    return parseToolpath(tpStream, parsingParams).catch(function (err) {
      throw new Error(err);
    });
  }

}

module.exports = MBToolpathParser;
