// MB-support-plugin/lib/mbImporter.js
'use strict';

const fs = require('fs-extra');

const q = require('q');

const path = require('path');

const ImportProvider = require('eagle-print/services/providers/importProvider');

const ImportableFormat = require('eagle-print/services/importService').ImportableFormat;

const Toolpath = require('eagle-print/model/toolpath').default;

const MBToolpathProvider = require('./mb_toolpath_parser');

const PrinterGenderString = require('./constants').PrinterGenderString;

class MbImporter extends ImportProvider {
  constructor() {
    super(new ImportableFormat(['makerbot'], 'Makerbot Files (*.makerbot)'), Toolpath);
  }

  importFromData(data, sourceUrl) {
    const loader = new MBToolpathProvider();

    const readFunc = filepath => {
      return q.nfapply(fs.readFile, [filepath]);
    };

    return loader.unzipToolpath(readFunc, sourceUrl).then(parsedData => {
      const idToolpath = `toolpath:${sourceUrl}`; // TODO:  What is the significance of key `targetPrinterModel`?
      // Changed to using the field to pass slicedBotType instead of botType in order
      // to do bot_type check before print.
      //
      // Previous code below.
      // `let botType = 'Replicator 5th Gen';`

      const slicedBotType = PrinterGenderString.get(parsedData.metaData.bot_type); // TODO: Unsure what toolpath.fileSize provides.
      // HACK: setting as 1 to bypass for now.

      const fileSize = 1;
      const toolpath = new Toolpath({
        path: sourceUrl,
        name: path.basename(sourceUrl),
        fileSize,
        targetPrinterModel: slicedBotType,
        targetPrinterVendor: 'MakerBot',
        extraInfo: 'more info!',
        parser: idToolpath
      });
      return {
        toolpath,
        parsedData
      };
    });
  }

}

module.exports = MbImporter;
