// MB-support-plugin/lib/parameters.js
'use strict';

const _ = require('lodash');

const q = require('q');

const ParameterProvider = require('eagle-print/services/providers/parameterProvider');

const EaglePrint = require('eagle-print');

const UNITS = EaglePrint.model.UNITS;
const Gender = EaglePrint.model.Gender;

const PrinterGenderString = require('./constants').PrinterGenderString;

const BotTypeEnum = require('./constants').BotTypeEnum;

const BASE_CAPABILITIES = {
  // Ignored by MakerBot Print. Irrelevant for GrabCAD Print as MakerBot printers are not supported.
  can_print: true,
  can_supply_progress: true,
  can_cancel_build: true,
  can_cancel_queueing: true
};
const printerData = {
  [BotTypeEnum.replicator_5]: {
    pid: 5,
    buildVolume: {
      envelopeSize: {
        x: 25.2 * UNITS.CM.factor,
        y: 19.9 * UNITS.CM.factor,
        z: 14.8 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.replicator_b]: {
    pid: 11,
    buildVolume: {
      envelopeSize: {
        x: 29.5 * UNITS.CM.factor,
        y: 19.5 * UNITS.CM.factor,
        z: 16.5 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.mini_8]: {
    pid: 8,
    buildVolume: {
      envelopeSize: {
        x: 10.1 * UNITS.CM.factor,
        y: 12.6 * UNITS.CM.factor,
        z: 12.6 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.z18_6]: {
    pid: 6,
    buildVolume: {
      envelopeSize: {
        x: 30 * UNITS.CM.factor,
        y: 30.5 * UNITS.CM.factor,
        z: 45.7 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.fire_e]: {
    pid: 8082,
    buildVolume: {
      envelopeSize: {
        x: 15.24 * UNITS.CM.factor,
        y: 19.05 * UNITS.CM.factor,
        z: 19.685 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.lava_f]: {
    pid: 8082,
    buildVolume: {
      envelopeSize: {
        x: 15.24 * UNITS.CM.factor,
        y: 19.05 * UNITS.CM.factor,
        z: 19.685 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: true
  },
  [BotTypeEnum.sketch]: {
    buildVolume: {
      envelopeSize: {
        x: 15 * UNITS.CM.factor,
        y: 15 * UNITS.CM.factor,
        z: 15 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    // change this for public release for now
    isReleased: true
  },
  // generic bot type
  generic: {
    buildVolume: {
      envelopeSize: {
        x: 29.5 * UNITS.CM.factor,
        y: 19.5 * UNITS.CM.factor,
        z: 16.5 * UNITS.CM.factor
      }
    },
    capabilities: BASE_CAPABILITIES,
    isReleased: false
  }
};

class MBGender extends Gender {
  constructor(data, map) {
    super(data, map);
  }

  get botType() {
    return this.get('botType');
  }

  get pid() {
    // sketchy js is sketchy -- it seems for this parseInt, it will parse from the
    // beginning of the passed string up until the first non-hex char, and then
    // convert that to an int:
    // eg. parseInt('dafdgfdg', 16) === parseInt('dafd', 16) === 56061...
    //
    // so we may see something like, if we have a botType 'bot_bonobo', this would
    // end up telling us the pid is 11, which is the same as rep+...
    // we don't make any guarantees of behavior for bots with unknown botTypes
    // though, and currently we only use only use this pid to check compatibility
    // for fw update so it shouldn't be that much of a problem, but here's a warning
    // just in case...
    return this.get('pid') || parseInt(this.botType.split('_')[1], 16);
  }

}

function constructGender(name, botType) {
  const pData = printerData[botType] || printerData['generic'];
  let newGender = new MBGender(_.merge({
    vendor: 'MakerBot',
    name
  }, pData));
  newGender = newGender.set('botType', botType);
  return newGender;
}

class MakerbotParameterProvider extends ParameterProvider {
  constructor() {
    super(); // for now, store all parameters in memory.
    // this initialization of the set of printer models
    // is temporary until we obtain this data from insight

    this._genders = {};

    for (const botType in BotTypeEnum) {
      const name = PrinterGenderString.get(BotTypeEnum[botType]);
      this._genders[name] = constructGender(name, BotTypeEnum[botType]);
    }
  }

  setLogger(logger) {
    this._logger = logger;
  }

  accept(vendor) {
    return vendor.toLowerCase() === 'makerbot';
  }

  getGender(vendor, genderName, botType) {
    let genderObj = this._genders[genderName];

    if (!genderObj) {
      this._logger.warn(`Using generic gender for printer ${genderName}; ${botType}`);

      genderObj = constructGender('Printer', botType);
    }

    return q(genderObj);
  }

  getVendors() {
    const vendors = {};
    vendors['MakerBot'] = {
      displayName: 'MakerBot',
      genders: this._genders
    };
    return vendors;
  }

  getParamsByBotType(botType) {
    return printerData[botType];
  }

}

module.exports = MakerbotParameterProvider;
