// MB-support-plugin/lib/paths.js
'use strict';

const path = require('path'); // TODO: These paths are duplicated from the makerbot build system...


const os_type = require('os').type(); // Where the grunt task unpacks our artifacts


const installRoot = `${__dirname}/../mb_ir`;

function getInstallArtifactsRootDir() {
  return installRoot;
}

function getBinDir() {
  switch (os_type) {
    case 'Linux':
      return path.join(installRoot, 'usr/bin');

    case 'Darwin':
      return path.join(installRoot, 'Library/MakerBot');

    case 'Windows_NT':
      return path.join(installRoot, 'MakerWare');

    default:
      throw `Unsupported OS type: ${os_type}`;
  }
}

function getConfigDir() {
  let remote, app;

  try {
    remote = require('electron').remote;
    app = remote.app;
  } catch (err) {
    // if this is being run from the main process, then 'remote' doesn't exist and will throw an error
    // in that case, simply require app
    try {
      app = require('electron').app;
    } catch (err) {// this can happen if running in unit test environment
    }
  }

  return app ? app.getPath('userData') : '';
}

function getWinUsbMonitor() {
  if (os_type !== 'Windows_NT') throw `Unsupported OS type: ${os_type}`; // this needs to be changed if we ever decide to support 32bit (lol)

  const programfiles = process.env.programfiles;
  return {
    path: path.join(programfiles, 'MakerBot Usb Support', 'MakerWare', 'UsbMonitor'),
    bin: 'UsbMonitor.exe'
  };
}

function requireNativeCroissant() {
  let Croissant;
  const platform = process.platform;
  const binary = path.basename(process.argv[0]);

  switch (os_type) {
    case 'Linux':
      Croissant = require('../mb_ir/usr/bin/croissantjs');
      break;

    case 'Darwin':
      if (binary === 'Electron Helper') Croissant = require('../mb_ir/Library/MakerBot/croissantjs_electron');else Croissant = require('../mb_ir/Library/MakerBot/croissantjs');
      break;

    case 'Windows_NT':
      // On windows the library loading mechanism prevents us from using the
      // same native module with both the node and electron binaries.  We try
      // to only use the electron specific module when we have to because we
      // don't ever test the electron module.
      if (platform == 'win32' && (binary == 'electron.exe' || binary == 'GrabCADPrint.exe' || binary == 'makerbot-print.exe')) {
        Croissant = require('../mb_ir/MakerWare/croissantjs_electron');
      } else {
        Croissant = require('../mb_ir/MakerWare/croissantjs');
      }

      break;

    default:
      throw `Unsupported OS type: ${os_type}`;
  }

  return Croissant;
}

function requireSliceConfig() {
  let SliceConfig; // Note: .js suffix needed for require() to disambiguate from sliceconfig Python script that exists in same directory

  switch (os_type) {
    case 'Linux':
      SliceConfig = require('../mb_ir/usr/bin/sliceconfig.js');
      break;

    case 'Darwin':
      SliceConfig = require('../mb_ir/Library/MakerBot/sliceconfig.js');
      break;

    case 'Windows_NT':
      // On windows the library loading mechanism prevents us from using the
      // same native module with both the node and electron binaries.  We try
      // to only use the electron specific module when we have to because we
      // don't ever test the electron module.
      SliceConfig = require('../mb_ir/MakerWare/sliceconfig.js');
      break;

    default:
      throw `Unsupported OS type: ${os_type}`;
  }

  return SliceConfig;
}

function requireNodeGrue() {
  let NodeGrue;
  const platform = process.platform;
  const binary = path.basename(process.argv[0]);

  switch (os_type) {
    case 'Linux':
      // NOTE: untested but it's not like we build mb print for linux anyways
      NodeGrue = require('../mb_ir/usr/share/makerbot/node_modules/nodegrue_addon');
      break;

    case 'Darwin':
      if (binary === 'Electron Helper') {
        NodeGrue = require('../mb_ir/Library/MakerBot/nodegrue_addon_electron');
      } else {
        NodeGrue = require('../mb_ir/Library/MakerBot/nodegrue_addon');
      }

      break;

    case 'Windows_NT':
      // On windows the library loading mechanism prevents us from using the
      // same native module with both the node and electron binaries.  We try
      // to only use the electron specific module when we have to because we
      // don't ever test the electron module.
      if (platform == 'win32' && (binary == 'electron.exe' || binary == 'GrabCADPrint.exe' || binary == 'makerbot-print.exe')) {
        NodeGrue = require('../mb_ir/MakerWare/nodegrue_addon_electron');
      } else {
        NodeGrue = require('../mb_ir/MakerWare/nodegrue_addon');
      }

      break;

    default:
      throw `Unsupported OS type: ${os_type}`;
  }

  return NodeGrue;
}

module.exports = {
  getInstallArtifactsRootDir,
  getBinDir,
  getConfigDir,
  getWinUsbMonitor,
  requireNativeCroissant,
  requireSliceConfig,
  requireNodeGrue
};
