// MB-support-plugin/lib/printerManager.js
'use strict';

const Immutable = require('immutable');

const q = require('q');

const EventEmitter = require('eventemitter3');

const _ = require('lodash');

const request = require('request');

const croissant = require('./croissant');

const ArchetypePrinter = require('./archetype_printer');

const MakerBotPrinter = require('./makerbot-printer');

const SketchPrinter = require('./sketchPrinter');

const Errors = require('./errors');

const PrinterStateEnum = require('./constants').PrinterStateEnum;

const PrinterConnTypeEnum = require('./constants').PrinterConnTypeEnum;

const BotTypeEnum = require('./constants').BotTypeEnum;

const PRINTER_AUTH_TIMEOUT = require('./constants').PRINTER_AUTH_TIMEOUT;

const ConnectionManager = require('./printerConnectionManager');

const util = require('./util');

const {
  parseConnType,
  compareConnType
} = require('./utils/printerConnectionUtil'); // TODO: Move this out to default-config to support staging URL in the future


const reflectorUrl = 'https://reflector.makerbot.com';
/*
Events:
    add_printer: request that a printer be added to My Printers
    remove_printer: request that a printer be removed from My Printers
    update_printer_list: request that the printer list in printer store be updated
*/

class PrinterManager extends EventEmitter {
  constructor(parameterService, loggingService, tvApiClient, trackingService, userInfo, token, storedPrintersManager) {
    super();
    this._parameterService = parameterService;
    this._logger = loggingService;
    this._tvApiClient = tvApiClient;
    this._trackingService = trackingService;
    this._storedPrintersManager = storedPrintersManager;
    this._printerCache = new Immutable.Map();
    this._connectionManagers = new Map();
    this._finder = null;
    this._findingPromise = null;
    this._startFinderTimeout = null;
    this._stopFinderTimeout = null;
    this._pollingReflectorId = null;
    this._userInfo = userInfo;
    this._userToken = token;
    this._currentPrinters = [];
    this._authPromises = new Map();
    this.reconnectPrinters = this.reconnectPrinters.bind(this);
    this._pollReflectorForSketchPrinters = this._pollReflectorForSketchPrinters.bind(this);
    this._reconnectMakerBotPrinter = this._reconnectMakerBotPrinter.bind(this);
    this._reconnectSketchPrinter = this._reconnectSketchPrinter.bind(this);
    this.sketchPrinterList = [];
  }

  init() {
    const self = this; // load up printers from stored

    let loadPrintersPromise;

    if (this._userToken) {
      loadPrintersPromise = this._userPrintersFromStored(this._userToken);
    } else {
      loadPrintersPromise = this._loadAnonymousPrinters();
    }

    return loadPrintersPromise.then(function () {
      return self._loadArchetypePrinters(self._userToken);
    }).then(function () {
      for (const printer of self.getPrinters()) self.emit('add_printer', printer); // force update of printer store


      self.emit('update_printer_list');
    });
  }

  _loadArchetypePrinters(token) {
    const archetypePrinterPromises = [];

    const archetypeIds = this._storedPrintersManager.getArchetypePrinters(this._userToken);

    for (const id of archetypeIds) {
      const info = this._storedPrintersManager.getPrinter(id);

      archetypePrinterPromises.push(this._createArchetypePrinter(info.vendor, info.gender));
    }

    return q.allSettled(archetypePrinterPromises);
  }

  stop() {
    for (const promise of this._authPromises.values()) {
      promise.cancel();
    }

    for (const uid of this._currentPrinters) {
      const printer = this._printerCache.get(uid);

      if (printer) printer.disconnect();
    } // disable the continuation that schedules find to start up again...


    if (this._findingPromise) this._findingPromise.cancelFindingContinuation();

    if (this._startFinderTimeout) {
      // if startFnderTimeout is set, finder has already stopped, and the
      // stop promise resolved
      clearTimeout(this._startFinderTimeout);
      return q();
    } else if (this._stopFinderTimeout) {
      // if stopFinderTimeout is set, we are waiting to stop the detector
      // clear the timeout and stop immediately
      const self = this;
      clearTimeout(this._stopFinderTimeout);
      return this._finder.stopFindingPrinters().then(function () {
        return self._findingPromise;
      });
    } else {
      // if neither are set, stop has been called; return the promise waiting
      // for stop to finish
      return this._findingPromise;
    }
  }

  _getUserPrintersFromWeb() {
    // these promises return maps of uid -> printerInfo
    const printerInfosPromise = q(this._tvApiClient.getUserPrintersFromReflector(this._userToken).then(userPrinterListRes => {
      const userPrinters = userPrinterListRes.printers;
      const supportedReflectorBots = {};
      Object.keys(userPrinters).map(printerId => {
        // filter out no-longer-supported bots
        const botType = userPrinters[printerId].status.bot_type;

        if (BotTypeEnum.hasOwnProperty(botType)) {
          supportedReflectorBots[printerId] = userPrinters[printerId];
        }
      });
      return this._getReflectorPrinterInfos(supportedReflectorBots);
    }, err => {
      // failed to fetch from reflector, but that's okay...
      this._logger.warn('Failed to fetch printer list from reflector for user: ', err);

      throw err;
    })); // treat the list gotten from tvApiClient as the absolute truth
    // of printers associated with the account; if printers are in
    // this list but not in the stored, add them; if printers are not
    // in this list but are in stored, clean them out

    return printerInfosPromise.then(results => {
      // if we failed to fetch printers from reflector,
      // don't bother cleaning; we don't want to mistakenly deauthenticate
      // all the user's printers....
      if (!results) return; // clean

      const storedUserPrinters = this._storedPrintersManager.getUserPrinters(this._userToken);

      for (const uid of storedUserPrinters) {
        /**
         * If storedUserPrinters contain a printer (uid) not present in
         * tvApiClient's list, remove it.
         *
         * If tvApiClient has a printer (uid) not present in
         * storedUserPrinters list, then it will be added to the stored
         * list via createPrinter flow.
         *
         * However, if a printer (uid) is found in both tvApiClient and
         * in storedUserPrinters lists, use storedUserPrinters info.
         * This will avoid automatically setting the printer to
         * reflector.
         *
         * -- Ray
         **/
        if (!Object.keys(results).includes(uid)) {
          this._storedPrintersManager.removePrinterForUser(uid, this._userToken);

          this.emit('remove_printer', this._printerCache.get(uid));
        } else {
          // TODO - Try LAN before attempting reflector on app-load
          // but only when we're capable of falling back to reflector
          // with LAN connections...
          const cachedPrinter = this._storedPrintersManager.getPrinter(uid);

          results[uid] = this._printerInfoFromSavedInfo({
            printer_info: cachedPrinter.printer_info
          });
        }
      }

      return this._offlinePrintersFromPrinterInfosList(results);
    });
  }

  start(clientSecret) {
    // if there is a logged-in user, try loading their printer list from web
    let loadPrintersPromise = q();
    if (this._userToken) loadPrintersPromise = this._getUserPrintersFromWeb();
    return loadPrintersPromise.then(() => {
      return this._loadArchetypePrinters();
    }).then(() => {
      // ensure that all our printers make it onto the My Printers list...
      for (const uid of this._currentPrinters) {
        this.emit('add_printer', this._printerCache.get(uid));
      }
    }).finally(() => {
      // Find Croissant printers
      // start detection regardless of success/failure of printer loading
      // there's some weirdness with making a new finder before the last one
      // has been stopped, so make it here, where the previous should
      // definitely have been stopped...
      this._finder = new croissant.PrinterFinder();

      this._finder.setClientSecret(clientSecret);

      if (this._userInfo && this._userToken) {
        this._finder.setThingiverseInfo(this._userInfo.name, this._userToken);
      } else {
        // The new auth flow _always_ requires a username
        // TODO: Get the local username when not logged in
        const username = this._userInfo ? this._userInfo.name : 'ANON';

        this._finder.setUsernameOnly(username);
      }

      this._findPrinters(); // Poll Reflector for Sketch Printers every 5 seconds


      this._pollReflectorForSketchPrinters(5);
    });
  }
  /**
   * Polls Reflector via GET-REQUEST for Sketch Printers
   * @param {int} timeInterval - Delay in seconds between each GET-request
   */


  _pollReflectorForSketchPrinters(timeInterval) {
    if (!this._pollingReflectorId) {
      const options = {
        headers: {
          'Authorization': `Bearer ${this._userToken}`,
          'Content-Type': 'application/json'
        },
        url: `${reflectorUrl}/printers`
      };

      this._logger.info(`Polling Reflector-GET for Sketch Printers`);

      this._pollingReflectorId = setInterval(() => {
        request(options, (err, res, body) => {
          if (err || ![200, 201].includes(res.statusCode)) {
            this._logger.warn(`REQUEST Error. Unable to fetch printer list. ${err}`);
          } else {
            const {
              printers
            } = JSON.parse(body);
            const sketchPrinterIds = Object.keys(printers).filter(id => id.match(/^(28571627)/g));
            const sketchUids = sketchPrinterIds.map(id => this._uidFromBotTypeAndIserial(BotTypeEnum.sketch, id));
            this.sketchPrinterList = sketchUids;
            sketchPrinterIds.forEach(printerId => {
              const tvPrinterInfo = printers[printerId];
              /*
               * reflector may send back a response with an ID but no status until fully
               * declaring the bot offline (for ~5 mins). if that happens we don't want to add this
               * entry to the printer list
               */

              if (_.isEmpty(tvPrinterInfo.status)) return q(null);

              const printerInfo = this._formatReflectorPrinterInfo(tvPrinterInfo);

              if (!this._printerCache.get(`2857:1627:${printerId}`)) {
                this._logger.info(`New Sketch Printer detected [${printerId}]. Adding printer.`);

                this._createPrinter(printerInfo).then(printer => {
                  // emit change to update printers list
                  this.emit('add_printer', printer);
                  this.emit('update_printer_list');
                }).catch(err => this._logger.log(err));
              } else {
                if (process.env.NODE_ENV && process.env.NODE_ENV !== 'production') {
                  console.log(`Existing Sketch Printer detected:`, printerId);
                } // otherwise check the cache & bring it back online


                q(this._printerCache.get(`2857:1627:${printerId}`)).then(printer => {
                  printer.updateFromPrinterInfo(printerInfo);
                });
              }
            });
          }
        });
      }, timeInterval * 1000);
    }
  }

  _loadAnonymousPrinters() {
    const anonPrinterUids = this._storedPrintersManager.getAnonymousPrinters();

    return this._offlinePrintersFromUidList(anonPrinterUids);
  }

  _offlinePrintersFromUidList(uids) {
    const printerPromises = [];

    const handleErr = err => {
      this._logger.warn(err);
    };

    for (const uid of uids) {
      const savedInfo = this._storedPrintersManager.getPrinter(uid);

      const printerInfo = this._printerInfoFromSavedInfo(savedInfo);

      printerPromises.push(this._createOfflinePrinter(printerInfo, savedInfo.machineConfig).catch(handleErr));
    }

    return q.allSettled(printerPromises);
  }

  _offlinePrintersFromPrinterInfosList(printerInfos) {
    const printerPromises = [];

    const handleErr = err => {
      this._logger.warn(err);
    };

    Object.keys(printerInfos).map(printerId => {
      const savedInfo = this._storedPrintersManager.getPrinter(printerId) || {};
      printerPromises.push(this._createOfflinePrinter(printerInfos[printerId], savedInfo.machineConfig).catch(handleErr));
    });
    return q.allSettled(printerPromises);
  }

  _getReflectorPrinterInfos(userPrinters) {
    const printerInfos = {};
    const printerIds = Object.keys(userPrinters);
    printerIds.map(id => {
      // _formatReflectorPrinterInfo assumes that the printer will be of new enough
      // firmware to have a status.bot_type; we will ignore printers that
      // do not have it
      try {
        const printerInfo = this._formatReflectorPrinterInfo(userPrinters[id]);

        printerInfos[printerInfo.uid] = printerInfo;
      } catch (err) {
        this._logger.warn(`Failed to add user printer ${printersList[id].name}: ${err}`);
      }
    });
    return printerInfos;
  }

  _userPrintersFromStored(token) {
    const storedUserPrinters = this._storedPrintersManager.getUserPrinters(token);

    return this._offlinePrintersFromUidList(storedUserPrinters);
  }

  _uidFromBotTypeAndIserial(bot_type, iserial) {
    let vid;
    let pid;

    if (bot_type === BotTypeEnum.sketch) {
      vid = '2857';
      pid = '1627';
    } else {
      vid = '23c1';
      pid = `0000${bot_type.split('_')[1]}`.slice(-4);
    }

    return [vid, pid, iserial].join(':');
  }

  _formatReflectorPrinterInfo(tvPrinterInfo) {
    const printerInfo = new croissant.PrinterInfo();
    printerInfo.name = tvPrinterInfo.name;
    printerInfo.address = tvPrinterInfo.status.ip;
    printerInfo.uid = this._uidFromBotTypeAndIserial(tvPrinterInfo.status.bot_type, tvPrinterInfo.printer_id);
    printerInfo.info = tvPrinterInfo.status;
    printerInfo.connType = parseConnType(tvPrinterInfo);
    return printerInfo;
  }

  _printerInfoFromSavedInfo(savedInfo) {
    const printerInfo = new croissant.PrinterInfo();
    printerInfo.name = savedInfo.printer_info.name;
    printerInfo.address = savedInfo.printer_info.address;
    printerInfo.uid = savedInfo.printer_info.uid;
    printerInfo.info = savedInfo.printer_info.info;
    printerInfo.connType = parseConnType(savedInfo.printer_info);
    return printerInfo;
  }

  _createOfflinePrinter(printerInfo, machineConfig) {
    const printer = this._printerCache.get(printerInfo.uid);

    if (printer) {
      // assume the info we're receiving here is the latest for this printer;
      // if this printer object has already been made, update its info in
      // MBPrinter and stored info
      printer.updateFromPrinterInfo(printerInfo);
      return this._storedPrintersManager.updateStoredInfo(printerInfo).then(() => {
        return printer;
      });
    }

    return this._createPrinter(printerInfo, machineConfig).then(printer => {
      printer.setState(PrinterStateEnum.Reconnecting);

      this._currentPrinters.push(printer.getId());

      return printer;
    });
  }

  getPrinter(uid) {
    return this._printerCache.get(uid);
  }

  getPrinters() {
    return this._printerCache.toArray();
  }

  _findPrinters() {
    // start finding printers; this will stop the finder after 3 secs, and
    // then start it 3 secs after _findingPromise resolves
    const self = this;

    if (!self._findingPromise || !self._findingPromise.isPending()) {
      self._stopFinderTimeout = setTimeout(function () {
        self._stopFinderTimeout = null;
        q(self._finder.stopFindingPrinters()).done();
      }, 3000);
      self._findingPromise = q(self._finder.findPrinters(function (printerInfo) {
        let printerPromise; // first detection

        if (!self._printerCache.get(printerInfo.uid)) {
          printerPromise = self._createPrinter(printerInfo).catch(function (err) {
            self._logger.log(err);
          });
        } else {
          printerPromise = q(self._printerCache.get(printerInfo.uid));
        }

        printerPromise.then(function (printer) {
          const connectionManager = self._connectionManagers.get(printerInfo.uid); // check if this printer is coming back online


          if (printer.getStatus().state === PrinterStateEnum.Offline) {
            if (self._currentPrinters.indexOf(printerInfo.uid) !== -1) {
              printer.setState(PrinterStateEnum.Reconnecting);
            } else {
              printer.setState(PrinterStateEnum.Unauthenticated);
            }
          }

          connectionManager.updateInfo(printerInfo);
        });
      })); // hack to make this continuation cancellable....

      self._findingPromise.afterFinding = function () {
        self._checkConnections();

        self._startFinderTimeout = setTimeout(function () {
          self._startFinderTimeout = null;

          self._findPrinters();
        }, 3000);
      };

      self._findingPromise.cancelFindingContinuation = function () {
        self._findingPromise.afterFinding = null;
      };

      self._findingPromise.then(function () {
        if (self._findingPromise.afterFinding) {
          self._findingPromise.afterFinding();
        }
      }).catch(function (err) {
        self._logger.error(err);
      }).done();
    }
  }

  _checkConnections() {
    const self = this;

    for (const pair of self._connectionManagers) {
      const key = pair[0];
      const val = pair[1];
      val.checkConnections().catch(function (err) {
        if (err.type) {
          const printer = self._printerCache.get(key); // TODO: doublecheck this...


          printer.addError(Errors.getPluginError(err.type));

          self._logger.error(err);
        }
      }).done();
    }
  }

  addArchetypePrinter(vendorName, genderName) {
    const self = this;
    return q(this._createArchetypePrinter(vendorName, genderName)).then(function (printer) {
      self._storedPrintersManager.storeArchetype(printer.getId(), vendorName, genderName, self._userToken);

      return printer;
    });
  }

  _createArchetypePrinter(vendorName, genderName) {
    return this._parameterService.getGender('MakerBot', genderName).then(gender => {
      const printer = new ArchetypePrinter(vendorName, gender);

      this._currentPrinters.push(printer.getId());

      this._printerCache = this._printerCache.set(printer.getId(), printer);
      return printer;
    });
  }

  _attachTrackingEvents(printer) {
    printer.updateNotifications.on('conn_type_change', () => {
      const trackingInfo = util.trackingInfoForPrinter(printer);
      const newConnType = printer.getConnectionType();

      if (newConnType === PrinterConnTypeEnum.USB) {
        this._trackingService.addEvent(_.merge({
          name: 'Printer Added USB'
        }, trackingInfo));
      }
    });
    printer.updateNotifications.on('dying_scream', err => {
      console.log(`${printer.getName()} | ${err} | Connection: ${printer.getConnectionType()}`);
      /** ** Monkey Patching **
       *
       * When a USB connection dies, there is a chance that the croissant
       * network discovery will pick up the printer. BUT there is a
       * higher chance that it won't be picked up by the network
       * discovery.
       *
       * Let's attempt to connect via IP address before deciding to label
       * the printer as "OFFLINE".
       *
       * -- Ray
       **/

      if (printer.getConnectionType() === PrinterConnTypeEnum.USB) {
        if (printer.getIpAddress()) {
          this.addByIp(printer.getIpAddress());
          printer.setState(PrinterStateEnum.Reconnecting);

          this._logger.log(`${PrinterConnTypeEnum.USB} Connection dropped out. Switching ${printer.getName()} to ${PrinterConnTypeEnum.NETWORK} at ${printer.getIpAddress()}...`);

          return printer.updateNotifications.emit('doDisconnect', false);
        }

        return printer.updateNotifications.emit('doDisconnect', true);
      }

      if (printer.getConnectionType() === PrinterConnTypeEnum.NETWORK) {
        const trackingInfo = util.trackingInfoForPrinter(printer);

        this._trackingService.addEvent(_.merge({
          name: 'Printer Disconnected from Network'
        }, trackingInfo));

        if (this._userToken) {
          // Reconstruct printerInfo w/ reflector connection
          // Croissant only cares about the userToken & uid
          // BUT for the app to properly assign reflector connection,
          // the IP address needs to be parsed out from 'tcp:<ipaddr>
          // :port#' to '<ipaddr>'
          const currPrinterInfo = this._connectionManagers.get(printer.getId())._printerInfos.get(PrinterConnTypeEnum.NETWORK);

          const {
            address,
            info
          } = currPrinterInfo;
          printer.setState(PrinterStateEnum.Reconnecting);
          currPrinterInfo.address = /^(?:[0-9]{1,3}\.){3}[0-9]{1,3}$/.test(address) ? address : address.split(':')[1];
          currPrinterInfo.info = {
            bot_type: info.bot_type
          };
          currPrinterInfo.connType = PrinterConnTypeEnum.REFLECTOR;

          this._logger.log(`${PrinterConnTypeEnum.NETWORK} Connection dropped out. Switching ${printer.getName()} over to Reflector...`);

          this._createPrinter(currPrinterInfo).then(printer => {
            this._connectionManagers.get(currPrinterInfo.uid).updateInfo(currPrinterInfo);

            try {
              this.authFromStored(currPrinterInfo);
              return printer.updateNotifications.emit('doDisconnect', false);
            } catch (err) {
              console.log(err);
            }
          });
        }

        return printer.updateNotifications.emit('doDisconnect', true);
      }
    }); // http://jira.makerbot.net/browse/PEN-1255
    // Commenting out until we figure out when exactly we want these events
    // to be sent from our side...

    /*
    let pp = 'PrintProcess';
      let sendPrintTrackingEvent = (name) => {
        let trackingInfo = util.trackingInfoForPrinter(printer);
        trackingInfo.extruder_type = printer.getCurrentExtruderType();
        this._trackingService.addEvent(
            _.merge({ name: name }, trackingInfo));
    };
      printer.updateNotifications.on(
        'process_completed',
        (info) => {
            if (info.current_process.name === pp)
                sendPrintTrackingEvent("Print Completed");
        });
      printer.updateNotifications.on(
        'process_cancelled',
        (info) => {
            if (info.current_process.name === pp)
                sendPrintTrackingEvent("Print Cancelled");
        });
      printer.updateNotifications.on(
        'process_failed',
        (info) => {
            if (info.current_process.name === pp)
                sendPrintTrackingEvent("Print Failed");
        });
    */
  }

  _createPrinter(printerInfo, machineConfig) {
    const self = this;

    if (printerInfo.info.bot_type === BotTypeEnum.sketch) {
      return self._getGender(printerInfo.info.bot_type).then(function (gender) {
        const printer = new SketchPrinter(printerInfo, gender, self._userToken, self);
        self._printerCache = self._printerCache.set(printerInfo.uid, printer);

        self._attachTrackingEvents(printer);

        return printer;
      });
    }

    const connectionManager = new ConnectionManager();
    connectionManager.updateInfo(printerInfo);
    self._connectionManagers = self._connectionManagers.set(printerInfo.uid, connectionManager); // hook up events

    connectionManager.on('need_auth', function (printerInfo, doneCallback) {
      self.authFromStored(printerInfo).nodeify(doneCallback);
    });
    connectionManager.on('update', function (printerInfo) {
      // only update stored info for bot if we're storing info for it
      if (self._storedPrintersManager.getPrinter(printerInfo.uid)) self._storedPrintersManager.updateStoredInfo(printerInfo);
    }, this);
    return self._getGender(printerInfo.info.bot_type).then(function (gender) {
      const printer = new MakerBotPrinter(printerInfo, gender, machineConfig, connectionManager);
      self._printerCache = self._printerCache.set(printerInfo.uid, printer);

      self._attachTrackingEvents(printer);

      return printer;
    });
  }

  _getGender(bot_type) {
    const genderName = util.botTypeToGender(bot_type);
    return this._parameterService.getGender('MakerBot', genderName, bot_type);
  }

  addByIp(ip) {
    const self = this;
    return q(self._finder.findByIp(ip)).then(function (printerInfo) {
      // filter out non-supported bots from being connected
      if (!BotTypeEnum.hasOwnProperty(printerInfo.info.bot_type)) return;
      let printer = undefined;

      const cachedPrinter = self._printerCache.get(printerInfo.uid);

      printer = compareConnType(cachedPrinter, printerInfo) >= 0 ? cachedPrinter : printer;
      const printerPromise = printer ? q(printer) : self._createPrinter(printerInfo);
      return printerPromise.then(function (printer) {
        if (printer.isAuthenticated()) return printer;

        self._connectionManagers.get(printerInfo.uid).updateInfo(printerInfo); // attempt to auth from stored info; if it fails, just
        // return the printer and auth will be requested for it


        try {
          return self.authFromStored(printerInfo);
        } catch (err) {
          return printer;
        }
      });
    });
  }

  reconnectPrinters() {
    this._logger.info(`Reconnecting Printers...`); // List through all the current printers and try to reauth using
    // the latest printerInfo from the _connectionManager.


    this._currentPrinters.forEach(uid => {
      const printer = this._printerCache.get(uid);

      if (printer.getConnectionType() === PrinterConnTypeEnum.ARCHETYPE) return;

      if (!printer.isSketchPrinter()) {
        this._reconnectMakerBotPrinter(uid, printer);
      } else {
        this._reconnectSketchPrinter(printer);
      }
    });
  }

  _reconnectMakerBotPrinter(uid, printer) {
    const currConnManager = this._connectionManagers.get(uid);

    const {
      _currentConnType
    } = currConnManager;

    const printerInfo = currConnManager._printerInfos.get(_currentConnType);

    if (printerInfo) {
      // ONLY reconnect to printers with PrinterInfo
      // OFFLINE Printers will not have a croissant.printerInfo
      // stored in connectionManager
      this._logger.info(`Reconnecting ${printer.getName()} printer via ${printerInfo.connType}`);

      this.authFromStored(printerInfo);
    }
  }

  _reconnectSketchPrinter(printer) {
    const options = {
      headers: {
        'Authorization': `Bearer ${this._userToken}`,
        'Content-Type': 'application/json'
      },
      url: `${reflectorUrl}/printers/${printer.getIserial()}`
    };
    request(options, (err, res, body) => {
      if (err || ![200, 201].includes(res.statusCode)) {
        this._logger.info(`Error! Unable to reconnect via Reflector for Sketch Printer ${printer.getName()}`);

        printer.setState(PrinterStateEnum.Offline);
      } else {
        const jsonBody = JSON.parse(body);

        const printerInfo = this._formatReflectorPrinterInfo(jsonBody);

        this._logger.info(`Reconnecting Sketch printer, ${printer.getName()}, via ${printerInfo.connType}`);

        printer.setState(PrinterStateEnum.Reconnecting);
        printer.updateFromPrinterInfo(printerInfo);
      }
    });
  } // -------------AUTH STUFF-------------------


  deauthenticate(printer) {
    const uid = printer.getId();

    const connManager = this._connectionManagers.get(uid);

    if (connManager) {
      printer.deauthenticate(this._userInfo ? this._userInfo.name : 'ANON');
      connManager.changeConnection(null, null);
    }

    const index = this._currentPrinters.indexOf(uid);

    if (index != -1) this._currentPrinters.splice(index, 1);

    this._storedPrintersManager.removePrinterForUser(uid, this._userToken);
  }

  _doAuthenticate(connectPromise, printer, printerInfo) {
    const self = this;
    const authResult = {};

    connectPromise.continuation = function (res) {
      self._logger.log(`Successfully connected to ${printerInfo.name} using connection type: ${printerInfo.connType}`);

      authResult.printer = res[0];
      authResult.authInfo = res[1]; // alert the printer's connectionManager that auth has completed

      self._connectionManagers.get(printerInfo.uid).changeConnection(authResult.printer, printerInfo);

      let addAccountPromise = q();
      if (printerInfo.connType === PrinterConnTypeEnum.USB && self._userToken) addAccountPromise = printer.invoke('AddMakerbotAccount', [self._userInfo ? self._userInfo.name : 'ANON', self._userToken]);
      return addAccountPromise.then(function () {
        self._currentPrinters.push(printer.getId());

        self.emit('add_printer', printer); // sigh... we should look into whether we really need to call this here...
        //
        // Right now it gets called so that printer.getTrackingInfo will
        // have some extruder stats-- but those aren't even guaranteed
        // to be correct, since we don't update it at all after getting
        // it the first time (and it used to be called in updateSchema too, WHY???)
        //
        // Is it necessary in the first place for ALL tracking info for a printer
        // to contain extruder stats? If so, this should be called in the
        // getTrackingInfo method of MBPrinter (which I'm wary of having
        // in the printer class in the first place), which would necessitate
        // getTrackingInfo to be a promise, which would require changes to
        // all the current tracking event-handling code...
        //
        //If not, this should only be called where necessary (and just
        // invoke, rather than storing the extruder stats as a property
        //of the printer object....)

        return printer.updateExtruderStats();
      }).then(function () {
        // Temporary(?): do not fail auth if updateMachineConfig fails.
        // Auth has succeeded at this point though, so we should present
        // that properly... Hopefully we will soon be moving away from
        // getting the schema in this function, in which case, we can
        // either keep this the way it is now, or go back to how it was before...
        printer.updateMachineConfig().catch(function (err) {
          self._logger.warn(err);
        }).finally(function () {
          const machineConfig = self._printerCache.get(printerInfo.uid).machineConfig;

          return self._storedPrintersManager.updateStoredInfo(printerInfo, machineConfig, authResult.authInfo, self._userToken);
        }).done();
        return printer;
      });
    };

    connectPromise.cancel = function () {
      this.continuation = function (res) {// do nothing?
      };
    };

    self._authPromises.set(printer.getId(), connectPromise);

    return connectPromise.then(function (res) {
      return q.all(res);
    }).timeout(PRINTER_AUTH_TIMEOUT).then(function (res) {
      return connectPromise.continuation(res);
    }).catch(function (err) {
      if (err.code === 'ETIMEDOUT') {
        const trackingInfo = util.trackingInfoForPrinter(printer);

        self._trackingService.addEvent(_.merge(trackingInfo, {
          name: 'Reauthentication Timeout',
          connection_type: printerInfo.connType
        }));
      }

      self._logger.error(`Failed to authenticate to printer ${printerInfo.name} over connection type ${printerInfo.connType}: ${err}`);

      const connectionManager = self._connectionManagers.get(printerInfo.uid); // This auth request may be happening because we're trying to move
      // from a lower priority one to a higher priority one; if it fails
      // because eg. auth code is no longer valid, we don't want MBPrinter
      // to think it's become unauthenticated when it still has its lower
      // priority connection (so we only change to unauth'd when we have
      // no fallback)


      if (!connectionManager.currentConnType) {
        printer.setState(PrinterStateEnum.Unauthenticated);
      } else {
        printer.setState(PrinterStateEnum.Pending);
      }

      throw err;
    }).finally(function () {
      self._authPromises.delete(printer.getId());
    });
  }

  authFromStored(printerInfo) {
    const printer = this._printerCache.get(printerInfo.uid);

    if (printer.isAuthenticating()) return q(printer); // check for stored auth information; only relevant for network printers

    let authInfo;

    if (printerInfo.connType === PrinterConnTypeEnum.NETWORK) {
      if (this._currentPrinters.indexOf(printerInfo.uid) !== -1) {
        if (!this._userToken) authInfo = this._storedPrintersManager.getPrinter(printerInfo.uid).authInfo;
      } else {
        throw 'No stored auth info for network printer';
      }
    }

    this._logger.debug(`Attempting reauth ${printerInfo.name} over connection type: ${printerInfo.connType}`);

    printer.setState(PrinterStateEnum.Reauthenticating);
    const self = this;
    const connectPromise = q.try(function () {
      return authInfo ? self._finder.reconnectPrinter(printerInfo, authInfo) : self._finder.reconnectPrinter(printerInfo);
    }).then(function (printer) {
      return [printer, q(authInfo)];
    });
    return this._doAuthenticate(connectPromise, printer, printerInfo).catch(err => {
      if (err instanceof croissant.UnauthorizedError) {
        this._logger.log(`Reauth to ${printerInfo.name} rejected as Unauthorized; ` + `removing local printer info for user`);

        this.deauthenticate(printer);
      }

      throw err;
    });
  }

  requestNewAuth(printerInfo) {
    const printer = this._printerCache.get(printerInfo.uid);

    if (printer.isAuthenticating() || printer.isAuthenticated()) return q(printer);

    if (printerInfo.connType === PrinterConnTypeEnum.NETWORK) {
      this._logger.debug(`Requesting new auth to ${printerInfo.name} over connection type: ${printerInfo.connType}`);

      printer.setState(PrinterStateEnum.Authenticating);
    } else {
      this._logger.debug(`Attempting reauth ${printerInfo.name} over connection type: ${printerInfo.connType}`);

      printer.setState(PrinterStateEnum.Reauthenticating);
    }

    const connectPromise = q(this._finder.connectPrinter(printerInfo));
    return this._doAuthenticate(connectPromise, printer, printerInfo);
  } // try to auth with the highest-available-priority printerInfo, without using
  // stored authInfo


  authenticate(printer) {
    const authResObj = {
      printer,
      wasNewAuth: false
    };

    const printerInfo = this._connectionManagers.get(printer.getId()).getHighestPriorityConnInfo();

    if (!printerInfo) throw 'Cannot authenticate to an offline printer';
    const isNetworkBot = printerInfo.connType === PrinterConnTypeEnum.NETWORK; // If the bot is a network bot and we don't think it belongs to the current
    // user, go straight to button-press auth

    if (isNetworkBot && this._currentPrinters.indexOf(printerInfo.uid) === -1) {
      authResObj.wasNewAuth = true;
      return this.requestNewAuth(printerInfo).then(() => {
        return authResObj;
      });
    } // otherwise, try authing from stored


    return this.authFromStored(printerInfo).catch(err => {
      // if our reauth to a network bot was denied, try button auth
      if (isNetworkBot && err instanceof croissant.UnauthorizedError) {
        authResObj.wasNewAuth = true;
        return this.requestNewAuth(printerInfo).then(() => {
          return authResObj;
        });
      }

      throw err;
    });
  }

}

module.exports = PrinterManager;
