// MB-support-plugin/lib/print_job_helper.js
'use strict';

const path = require('path');

const q = require('q');

const PrinterStateEnum = require('./constants').PrinterStateEnum; // a little class to help us keep track of print jobs and their associated
// slicing and file-sending stuff...


class PrintJobHelper {
  constructor(printer) {
    const self = this;
    self._printer = printer;
    self._printStarted = false;
    self._cancelToken = undefined;
    self._managingJob = false;
    self._printId = 0;

    self._printer.updateNotifications.on('process_cancelled', function (info) {
      if (self.managingPrintJob() && info.current_process.name === 'PrintProcess') // do not call cancel again
        self._cancelToken.removeListener('cancel', self._cancelPrint, self); // cancel every/anything else

      self.cancel();

      self._cleanUp();
    });
  }

  managingPrintJob() {
    return this._managingJob;
  }

  startPrint(filename, cancelToken) {
    const self = this;
    self._managingJob = true;
    self._cancelToken = cancelToken; // set callback to clean up printprocess on cancel

    self._cancelToken.once('cancel', self._cancelPrint, self); // check if printprocess has been started already (ie. by slicing)


    let startPrintPromise = q();

    if (!self._printStarted) {
      if (self._printer.getStatus().state !== PrinterStateEnum.Idle) throw 'Unable to print to this printer at this time'; // Print args: filename, [ensure_build_plate_clear, transfer_wait]

      startPrintPromise = q(self._printer.invoke('Print', [filename, undefined, true])).then(function () {
        self._printStarted = true;
      });
    }

    return startPrintPromise;
  }

  sendPrintFile(localPath, remotePath, progressCallback, cancelToken) {
    const self = this;
    self._managingJob = true;
    self._cancelToken = cancelToken;
    const filename = path.basename(localPath);
    return self.startPrint(filename, cancelToken).then(function () {
      self._cancelToken.once('cancel', function () {
        self._printer.invoke('cancelPut', [self._printId]);
      });

      return self._printer.invoke('Put', [++self._printId, localPath, remotePath, undefined, progressCallback]);
    }).catch(function (err) {
      if (cancelToken.isCancelled()) {
        const cancelledErr = new Error('Cancelled');
        cancelledErr.type = 'cancel';
        throw cancelledErr;
      } else {
        // cancel printprocess if we somehow fail putting
        // (if put was in fact cancelled, the printprocess should
        // be getting cancelled by the cancelToken...
        self._cancelPrint();

        throw err;
      }
    }).finally(function () {
      // after print has sent (or failed), we no longer care about it
      self._cleanUp();
    });
  }

  _cancelPrint() {
    try {
      this._cancelToken.removeListener('cancel', this._cancelPrint, this);

      return q(this._printer.invoke('Cancel', []));
    } catch (err) {
      console.log('Failed to cancel print');
      if (err.type !== 'UnauthenticatedError') throw err;
    }
  }

  cancel() {
    if (this._cancelToken) this._cancelToken.cancel();
  }

  _cleanUp() {
    // reset states
    this._cancelToken = undefined;
    this._printStarted = false;
    this._managingJob = false;
  }

}

module.exports = PrintJobHelper;
