// MB-support-plugin/lib/printerConnectionManager.js
'use strict';

const EventEmitter = require('eventemitter3');

const q = require('q');

const PRINTER_CONNECTION_TIMEOUT = require('./constants').PRINTER_CONNECTION_TIMEOUT;

const PrinterConnTypeEnum = require('./constants').PrinterConnTypeEnum;

const PRIORITY = [PrinterConnTypeEnum.USB, PrinterConnTypeEnum.NETWORK, PrinterConnTypeEnum.REFLECTOR];
/*
Events:
    update: emitted when we think printer info should be updated
    connection_change: emitted when the connection type should change (eg.
        switching from network to usb connection)
    need_auth: emitted to request that authentication be done for the given
        printerInfo
    check_life: emitted to ask MBPrinter to check that its current connection
        is still alive
*/

class ConnectionManager extends EventEmitter {
  constructor() {
    super();
    this.setInitState();
  } // there's probably a better way to handle this (architecturally) such that
  // this isn't needed, but I guess it's okay for now...?


  setInitState() {
    this._printerInfos = new Map();
    this._currentConnType = null; // TODO: keep previous printer around for some time?

    this._lastReaped = null;
    this._prevInfos = new Map();
  }

  _infosAreEqual(printerInfo1, printerInfo2) {
    // ignore timestamps in our equality-checking (make copies of the
    // printerInfos so we don't end up changing them)
    const pInfo1 = Object.assign({}, printerInfo1);
    delete pInfo1.timestamp;
    const pInfo2 = Object.assign({}, printerInfo2);
    delete pInfo2.timestamp;
    return JSON.stringify(pInfo1) === JSON.stringify(pInfo2);
  }

  hasConnType(connType) {
    return !!this._printerInfos.get(connType);
  } // returns the printerInfo for the highest priority connection available


  getHighestPriorityConnInfo() {
    for (const connType of PRIORITY) {
      const conn = this._printerInfos.get(connType);

      if (conn) return conn;
    }

    return null;
  } // return true if connType1 is of higher priority than connType2 (invalid being
  // lowest priority);
  // return false if connType1 is invalid (regardless of whether connType2 is valid)


  static isHigherPriority(connType1, connType2) {
    const c1index = PRIORITY.indexOf(connType1);
    const c2index = PRIORITY.indexOf(connType2);
    if (c1index < 0) return false;else if (c2index < 0) return true;else return c1index < c2index;
  }

  updateInfo(printerInfo) {
    const connType = printerInfo.connType;
    const highestPriorityConnInfo = this.getHighestPriorityConnInfo(); // should request infos be updated if:
    // * this is the first printerInfo we're receiving for this bot or
    // * if printer is connected, and the info is for the current connection or
    // * the printer is not connected, and the info is for the highest priority
    // * of connType currently available
    // AND
    // * info has changed
    // * (otherwise, info that is connType specific, like address, will get
    // * changed every time a bot gets detected on a different interface)

    if ((!highestPriorityConnInfo || this._currentConnType && connType === this._currentConnType || !this._currentConnType && highestPriorityConnInfo.connType === connType) && !this._infosAreEqual(this._printerInfos.get(connType), printerInfo)) this.emit('update', printerInfo);
    printerInfo.timestamp = new Date();

    this._printerInfos.set(connType, printerInfo);
  }

  changeConnection(croissantPrinter, printerInfo) {
    if (printerInfo) {
      this._currentConnType = printerInfo.connType;
    } else {
      this._currentConnType = null;
    }

    this.emit('connection_change', croissantPrinter);
  }

  checkConnections() {
    const self = this; // check if connections have disappeared or if current conn should change

    return this._reap().finally(function () {
      if (!self._printerInfos.size) {
        // printer has gone completely offline
        self.changeConnection(null, null);
      } else {
        // try to auth to highest priorty conn possible
        self._tryConnectionChange(self._prevInfos);
      } // copy current state of _printerInfos map


      self._prevInfos = new Map();

      for (const pair of self._printerInfos) {
        const key = pair[0];
        const val = pair[1];

        self._prevInfos.set(key, val);
      }
    });
  }

  _tryConnectionChange(prevPrinterInfos) {
    const self = this;

    const shouldRequestAuth = function (printerInfo) {
      if (!printerInfo) return false; // if the printerInfo is for a higher priority than the currently
      // used, check that it wasn't around previously (otherwise, we
      // would have already attempted auth to it, and failed)

      const higherThanCurrent = ConnectionManager.isHigherPriority(printerInfo.connType, self._currentConnType);
      const isNewlyDiscovered = !prevPrinterInfos.get(printerInfo.connType);
      const justLostConnection = prevPrinterInfos.size > self._printerInfos.size && !self._currentConnType;
      return higherThanCurrent && isNewlyDiscovered || justLostConnection;
    };

    const _requestAuth = function (printerInfo) {
      if (shouldRequestAuth(printerInfo)) {
        self.emit('need_auth', printerInfo, function (err) {
          if (err) {
            // if we fail, try again with a lower priority connection until
            // it works, or we give up
            let i = PRIORITY.indexOf(printerInfo.connType) + 1;

            while (!self._printerInfos.get(PRIORITY[i]) && i < PRIORITY.length) i++;

            const pInfo = self._printerInfos.get(PRIORITY[i]);

            _requestAuth(pInfo);
          }
        });
      }
    };

    const highestPriorityConnInfo = this.getHighestPriorityConnInfo();

    _requestAuth(highestPriorityConnInfo);
  }

  _reap() {
    // get rid of old printerInfos
    const now = new Date();
    let reapDonePromise;

    for (const pair of this._printerInfos) {
      const connectionType = pair[0];
      const printerInfo = pair[1]; // if the printerInfo was not updated during this last round of detection
      // (ie. was timestamped before the last reap), AND it has exceeded the
      // timeout limit, this printer has likely disappeared from this interface
      // check if the printer is detected and it is passed its timeout or was recently connected

      if (printerInfo.timestamp < this._lastReaped && now - printerInfo.timestamp > PRINTER_CONNECTION_TIMEOUT || this._currentConnType != null) {
        if (connectionType === this._currentConnType) {
          // If connection type is being used, ask the printer if it's
          // is still alive
          const checkDoneDeferred = new q.defer();
          reapDonePromise = checkDoneDeferred.promise;
          this.emit('check_life', this._checkAliveDoneCallback(this, checkDoneDeferred));
        } else {
          // else delete mercilessly
          this._printerInfos.delete(connectionType);
        }
      }
    }

    this._lastReaped = now;
    if (!reapDonePromise) reapDonePromise = q();
    return reapDonePromise;
  }

  _checkAliveDoneCallback(self, checkDoneDeferred) {
    return function (isAlive, err) {
      if (isAlive) {
        self._printerInfos.get(self._currentConnType).timestamp = new Date();
      } else {
        self._printerInfos.delete(self._currentConnType);

        self.changeConnection(null, null);
      }

      if (err) {
        checkDoneDeferred.reject(err);
      } else {
        checkDoneDeferred.resolve();
      }
    };
  }

  get currentConnType() {
    return this._currentConnType;
  }

}

module.exports = ConnectionManager;
