// MB-support-plugin/lib/printer-service-provider.js
'use strict';

const q = require('q');

const path = require('path');

const _ = require('lodash');

const analyticUtils = require('eagle-print/util/analyticUtils');

const PrinterProvider = require('eagle-print/services/providers/printerProvider').default;

const PrinterManager = require('./printerManager');

const StoredPrintersManager = require('./stored_printers_manager');

const Errors = require('./errors');

const croissant = require('./croissant');

const constants = require('./constants');

const util = require('./util');

class MBPrinterProvider extends PrinterProvider {
  constructor(clientSecret) {
    super();
    this._logger = console;
    this._clientSecret = clientSecret;
    this._storedPrintersManager = new StoredPrintersManager();
    this._fluxMethods = {}; // because for a real start, we need the flux methods to have been set,
    // but that happens after start for printerProvider gets called...

    const self = this;
    this._initDoneDeferred = q.defer();

    this._initDoneDeferred.promise.then(function () {
      self._fluxMethods.clearPrintersMethod();

      return self._getPrinterManager().start(clientSecret);
    }).done();
  }

  _getPrinterManager() {
    if (!this._printerManager) {
      const printerManager = new PrinterManager(this._parameterService, this._logger, this._tvApiClient, this._trackingService, this._userInfo, this._userToken, this._storedPrintersManager);
      printerManager.on('add_printer', printer => {
        if (this._fluxMethods.addPrinterMethod) this._fluxMethods.addPrinterMethod(printer);
      });
      printerManager.on('remove_printer', printer => {
        if (this._fluxMethods.removePrinterMethod) this._fluxMethods.removePrinterMethod(printer);
      });
      printerManager.on('update_printer_list', () => {
        if (this._fluxMethods.getPrintersMethod) this._fluxMethods.getPrintersMethod();
      });
      this._printerManager = printerManager;
    }

    return this._printerManager;
  }

  setTvApiClient(tvApiClient) {
    this._tvApiClient = tvApiClient;
  }

  setParameterService(parameterService) {
    this._parameterService = parameterService;
  }

  setTrackingService(trackingService) {
    this._trackingService = trackingService;
  }

  stop() {
    // stop the stored printers manager; all write requests after this gets
    // called will be ignored.
    return this._storedPrintersManager.stop();
  }

  init(userInfo, token) {
    const self = this;
    this._userInfo = userInfo;
    this._userToken = token;
    return this._storedPrintersManager.loadSavedPrinters().then(function () {
      return self._getPrinterManager().init();
    });
  }

  userChanged(userInfo, token) {
    const self = this;
    this._userInfo = userInfo;
    this._userToken = token;
    const oldPrinterManager = this._printerManager;
    this._printerManager = null;

    this._fluxMethods.clearPrintersMethod();

    this._getPrinterManager().init().then(function () {
      if (oldPrinterManager) return oldPrinterManager.stop();
    }).then(function () {
      return self._printerManager.start(self._clientSecret);
    }).done();
  }

  finishInit(addPrinterMethod, removePrinterMethod, clearPrintersMethod, getPrintersMethod) {
    this._fluxMethods.addPrinterMethod = addPrinterMethod;
    this._fluxMethods.removePrinterMethod = removePrinterMethod;
    this._fluxMethods.clearPrintersMethod = clearPrintersMethod;
    this._fluxMethods.getPrintersMethod = getPrintersMethod;

    this._initDoneDeferred.resolve();
  }

  setLogger(logger) {
    this._logger = logger;

    this._storedPrintersManager.setLogger(logger);

    this._logger.info('MakerBot PrinterServiceProvider set logger'); // set log for croissant


    const logPath = this._logger.bunyanLogger.getLogFilePath(); // Note: path.dirname(undefined) === '.'
    // If inside asar -> can cause Croissant native code to crash


    const logDir = path.dirname(logPath);

    if (!logPath || !logDir || '.' === logDir) {
      // GC-30551 Avoid feeding values to native module that can crash it
      this._logger.error(`Cannot determine normal directory for Croissant logs.  Logs may be lost.  path=${logPath} dir=${logDir}`);
    } else {
      croissant.createLog(logDir);
    }
  }

  getPrinter(uid) {
    return q(this._printerManager.getPrinter(uid));
  }

  getPrinters(refresh, callback) {
    // looks like sometimes this gets called before start has finished...
    if (!this._printerManager) return q([]);
    return q(this._printerManager.getPrinters()).nodeify(callback);
  }
  /**
   * @param printerUrl {string} - URL to the printer that should be added to this provider, if acceptable.
   * @param callback {function}
   */


  addPrinter(ip, callback) {
    // this should be fixed somewhere else and not hacked into here.
    // please don't let this stay here :(
    ip = ip.replace('ip://', '');
    return this._printerManager.addByIp(ip).nodeify(callback);
  }

  addOfflinePrinter(vendorName, genderName, callback) {
    return this._printerManager.addArchetypePrinter(vendorName, genderName).nodeify(callback);
  }

  authenticate(printer) {
    // Remove any previous auth errors
    printer.removeError(Errors.MBPluginError.GeneralAuthError.error_id);
    return this._printerManager.authenticate(printer).then(authRes => {
      if (authRes.wasNewAuth && printer.getConnectionType() === constants.PrinterConnTypeEnum.NETWORK) {
        // we're not returning this guy because we don't care to wait for
        // this to pass or fail before reporting the result of auth
        printer.invoke('NetworkState').then(networkState => {
          const trackingInfo = util.trackingInfoForPrinter(printer);
          trackingInfo.name = 'Printer Authenticated ';

          if (networkState.state === constants.NetworkStateEnum.state.WIFI) {
            trackingInfo.name += 'Wi-Fi';
          } else if (networkState.state === constants.NetworkStateEnum.state.ETHERNET) {
            trackingInfo.name += 'Ethernet';
          }

          this._trackingService.addEvent(trackingInfo);
        }).done();
      }
    }).catch(function (err) {
      printer.addError(Errors.getPluginError(Errors.MBPluginError.GeneralAuthError));
      throw 'Authorization error or manually cancelled';
    });
  }

  deauthenticate(printer) {
    return this._printerManager.deauthenticate(printer);
  }

  print(printer, toolPath, progressCallback, cancelToken, flux = {}) {
    const trackingInfo = util.trackingInfoForPrinter(printer);
    const printSettings = analyticUtils.getPrintSettingsForAnalyticEvents(flux, printer);
    trackingInfo.name = 'Print File Transfer Started';

    this._trackingService.addEvent(_.merge(trackingInfo, printSettings));

    cancelToken.once('cancel', () => {});
    return printer.sendPrintFile(toolPath, progressCallback, cancelToken).then(() => {
      const trackingInfo = util.trackingInfoForPrinter(printer);
      trackingInfo.name = 'Print File Transfer Completed';

      this._trackingService.addEvent(_.merge(trackingInfo, printSettings));
    }).catch(err => {
      const trackingInfo = util.trackingInfoForPrinter(printer);
      trackingInfo.name = `Print File Transfer ${err.type === 'cancel' ? 'Cancelled' : 'Failed'}`;

      this._trackingService.addEvent(_.merge(trackingInfo, printSettings));

      throw err;
    });
  }
  /**
   * @param printerUrl {string} - URL to a printer.
   * @returns {boolean} - True  = The given printer URL can be added to this printer provider.
   *                      False = The given printer URL cannot be added to this printer provider.
   */


  accept(ip) {
    // check that it's a valid ip
    const splitIp = ip.split('.');
    if (splitIp.length != 4) return false;

    for (let i = 0; i < splitIp.length; i++) {
      if (parseInt(splitIp[i]) < 0 || parseInt(splitIp[i]) > 256) {
        return false;
      }
    }

    return true;
  }

  convertToArchetypeId(printerId) {
    const PID = printerId.split(':').slice(1, 2);
    const genderName = constants.PIDtoGenderName[PID];
    const vendorName = constants.VendorName;
    return `offline:${vendorName}:${genderName}:${genderName}`;
  }

}

module.exports = MBPrinterProvider;
