// MB-support-plugin/lib/sketchPrinter.js
'use strict';

const q = require('q');

const Printer = require('eagle-print/model/printer');

const EventEmitter = require('eventemitter3');

const request = require('request');

const path = require('path');

const fs = require('fs-extra');

const util = require('./util.js');

const {
  maybe
} = require('./utils/maybe');

const {
  PrinterConnTypeEnum,
  PrinterStateEnum,
  FifthGenBots,
  SixthGenBots,
  GenderToType,
  ExtruderTypeEnum
} = require('./constants'); // TODO: Move this out to default-config to support staging URL in the future


const reflectorUrl = 'https://reflector.makerbot.com';

class SketchPrinter extends Printer {
  constructor(printerInfo, gender, userToken, printerManager, flux) {
    const data = {
      id: printerInfo.uid,
      connected: true,
      printer_status: {
        state: 'Unauthenticated'
      }
    };
    super(data, null, gender); // NOTE - Flux gets injected through PrinterStore via
    // this.injectFlux method. Check line 112-115 in
    // /stores/printerStore.js

    this._flux = maybe(flux);
    this._gender = gender;
    this._machineConfig = {}; // Notifications for critical updates (system, states, schema updates, etc..)

    this._updateNotifications = new EventEmitter();
    this.attachedExtruders = [ExtruderTypeEnum.sketch_extruder];
    this._printerInfo = printerInfo; // these are status properties generic to eagle-print's printer

    this._info = {};
    this._info.printer_status = {};
    this.setState(PrinterStateEnum.Unauthenticated);
    this._info.printer_status.current_print_job = {};
    this._info.printer_status.current_print_job.percent_done = 0; // these ones are makerbot specific, in case they may come in handy

    this._info.printer_status.current_process = null;
    this._info.printer_status.toolheads = null; // Errors from error_notifications and from processes go in here;
    // an error will get removed when acknowledgeError gets called for it
    // or when an error_acknowledged is received

    this._errors = new Map(); // this should only be available if the bot's process is in the 'failed' state
    // should only ever have one entry - this is incorrect and has been changed now
    // It will contain all state change errors.

    this._info.printer_status.state_change_errors = new Map();
    this._info.update_firmware_available = null;
    this._cameraFeedActive = false;
    this.userToken = userToken;
    this.cameraFeedIntervalId = null;
    this.updatePrinterStatus();
    this._sendCmdToReflector = this._sendCmdToReflector.bind(this);
    this._pollCmdFromReflector = this._pollCmdFromReflector.bind(this);
    this.emptyCameraImageBase64 = null;
    const emptyCameraImagePath = path.resolve(__dirname, './ui/views/images/empty.png');
    fs.readFile(emptyCameraImagePath, (err, data) => {
      if (err) {
        throw err;
      }

      this.emptyCameraImageBase64 = data;
    });
    this.printerManager = printerManager;
    this.checkAlive();
  }

  _initPrinter() {// Add machine config code here
    // this._getMachineConfig().done();
  }

  checkAlive() {
    setInterval(() => {
      const uid = this._printerInfo.uid;
      const printersList = this.printerManager.sketchPrinterList;
      const printersListHasBot = printersList.includes(uid);

      if (printersListHasBot && this.getStatus().state === PrinterStateEnum.Offline) {
        this.updatePrinterStatus();
      } else if (!printersListHasBot && this.getStatus().state !== PrinterStateEnum.Offline) {
        this.disconnect('Printer is not on Reflector printer list');
      }
    }, 5 * 1000); //every 5 seconds
  }

  disconnect(error) {
    this._updateNotifications.emit('dying_scream', error); // Put this back when we figure out getting flux in here
    //this.flux.actions.message.printerDisconnect(this._printerInfo.uid);


    this.setState(PrinterStateEnum.Offline);
  }

  addError(error) {
    this._errors.set(error.code, error);

    this._updateNotifications.emit('error');
  }

  removeError(errorId) {
    this._errors.delete(errorId);

    this._updateNotifications.emit('error');
  }

  saveStateError(err) {
    err.isStateErr = true;
    this.addError(err);
  }

  clearStateErrors() {
    for (const errEntry of this._errors) {
      const [errId, err] = errEntry;

      if (err.isStateErr) {
        this.removeError(errId);
      }
    }
  }

  getLastError() {
    // gets the most recent error
    let lastErr;

    for (const val of this._errors.values()) lastErr = val;

    return lastErr;
  }

  acknowledgeLastError() {
    const lastErr = this.getLastError();
    this.acknowledgeError(lastErr.error_id);
  }

  isMissingExtruder() {
    return false;
  }

  getChamberInfo() {
    return [];
  }

  getSpoolInfo() {
    return q(null);
  }

  processMethodExists(method) {
    const {
      current_process
    } = this._printerInfo.info;
    const processMethods = current_process ? current_process.methods : [];
    return processMethods.includes(method);
  }

  getTrackingInfo() {
    const data = {
      printer_id: this.getId(),
      printer_type: util.genderToCodeName(this.getInfo().model)
    };
    return data;
  }

  _errorAcknowledgedCallback(errorId) {
    this.removeError(errorId);
  }

  acknowledgeError(errorId) {
    // If this error was not internally-generated (ie. error code <= 0),
    // acknowledge the error with the printer; the error should then eventually
    // get cleared on this side from feedback from kaiten...
    if (errorId > -1) {
      let ackPromise = q();
      if (this.processMethodExists('acknowledge_error')) ackPromise = this.invoke('ProcessMethod', ['acknowledge_error']);else if (this.processMethodExists('acknowledge_failure')) ackPromise = this.invoke('ProcessMethod', ['acknowledge_failure']);
      return ackPromise.then(() => {
        return this.invoke('Acknowledged', [errorId]);
      }).then(() => {
        // clear the error after calling AcknowledgeError in case for
        // some reason we miss the error_acknowledged notification
        this._errorAcknowledgedCallback(errorId);
      }).done();
    } else {
      // else, just remove it
      this.removeError(errorId);
    }
  }

  getInfo() {
    return {
      vendor: 'MakerBot',
      model: this._botGender,
      modelName: this._info.name,
      firmware_version: this.firmwareVersion
    };
  }

  getStatus() {
    return this._info.printer_status;
  }

  getDisplayStatus() {
    return this.getStatus().state;
  }

  isAuthenticated() {
    const {
      state
    } = this.getStatus();
    return ![PrinterStateEnum.Reconnecting, PrinterStateEnum.Unauthenticated].includes(state);
  }

  isConnected() {
    const {
      state
    } = this.getStatus();
    return ![PrinterStateEnum.Offline, PrinterStateEnum.Reconnecting].includes(state);
  }

  deauthenticate() {}

  cancel() {
    return this.invoke('cancel');
  }

  _getMachineConfig() {
    return q(this._printer.GetMachineConfig()).then(config => {
      this._machineConfig = config;
    });
  }

  _updateFromExtruderChange(index, config) {// TODO(shirley): make this work, when I actually figure out how to
    // get these notifications from the bot...
  }

  authenticate(authInfo) {}

  getCodeName() {
    return util.genderToCodeName(this.getGender().getName());
  }

  getConnectionType() {
    return PrinterConnTypeEnum.REFLECTOR;
  }

  getCameraImage() {
    const options = {
      method: 'GET',
      url: `${reflectorUrl}/printers/${this.getIserial()}/image`,
      headers: {
        'Authorization': `Bearer ${this.userToken}`,
        'Content-Type': 'image/jpeg'
      },
      encoding: null
    };

    const convertImageData = imageData => {
      const imageBuffer = new Buffer(imageData);
      imageData = imageBuffer.toString('base64');
      return `data:image/jpeg;base64,${imageData}`;
    };

    return q.Promise((resolve, reject, notify) => {
      request(options, (err, res, body) => {
        if (err) {
          console.error(err);
          resolve(convertImageData(this.emptyCameraImageBase64));
        } else if (!res || ![200, 201].includes(res.statusCode) || !res.body || res.body === '' || res.body.length < 1) {
          resolve(convertImageData(this.emptyCameraImageBase64));
        } else {
          const dataUrl = convertImageData(body);
          resolve(dataUrl);
        }
      });
    });
  }

  getCameraFeed(cameraFrameCallback) {
    if (!this.cameraFeedIntervalId) {
      // Grab the first image, then...
      this.getCameraImage().then(imageData => {
        // Return imageData back via callback
        cameraFrameCallback(imageData); // Set up setInterval to ping for imageData

        this.cameraFeedIntervalId = setInterval(() => {
          this.getCameraImage().then(imageData => {
            cameraFrameCallback(imageData);
          });
        }, 30 * 1000);
      });
    }
  }

  endCameraFeed() {
    if (this.cameraFeedIntervalId) {
      clearInterval(this.cameraFeedIntervalId);
      this.cameraFeedIntervalId = null;
    }
  } // extruderList should be an array, even if there is only one extruder


  setAttachedExtruders(extruderList, index = null) {
    if (index !== null) {
      this.attachedExtruders[index] = extruderList;
    } else {
      this.attachedExtruders = extruderList;
    }
  }

  getAttachedExtruders() {
    return this.attachedExtruders;
  }

  _fwVerDictToString() {
    if (!this._printerInfo.info) return '0.0.0';
    const {
      firmware_version
    } = this._printerInfo.info;
    const {
      major,
      minor,
      bugfix,
      build
    } = firmware_version;
    return `${major}.${minor}.${bugfix}.${build}`;
  } // doing this (just in case) until eagle-print's printer settles down...


  get rawData() {
    return this._info;
  }

  get machineConfig() {
    return undefined; //this._machineConfig;
  }

  get firmwareVersion() {
    return this._fwVerDictToString();
  }

  get apiVersion() {
    return '0.0.0';
  }

  get currentState() {
    return this._info;
  }

  get _botGender() {
    return this._gender.getName();
  }

  get updateNotifications() {
    return this._updateNotifications;
  }

  checkForUnsupportedExtruderConfigurations() {
    return {
      result: false
    };
  }

  hasExtruderMaterialMismatch() {
    return false;
  }

  isFifthGen() {
    const botType = GenderToType[this.getModel()];
    return FifthGenBots.includes(botType);
  }

  isSixthGen() {
    const botType = GenderToType[this.getModel()];
    return SixthGenBots.includes(botType);
  }

  setState(state) {
    if (state !== this.getStatus().state) {
      this._info.printer_status.state = state;
    }
  }

  getName() {
    return this._printerInfo.name;
  }

  updatePrinterStatus() {
    const {
      current_process: currProcess
    } = this._printerInfo.info;

    if (!currProcess) {
      this.setState(PrinterStateEnum.Idle);
      this._info.printer_status = Object.assign({}, this._info.printer_status, {
        current_process: null,
        current_print_job: {
          percent_done: 0
        }
      });
    } else {
      this.setState(PrinterStateEnum.Busy);
      this._info.printer_status = Object.assign({}, this._info.printer_status, {
        current_process: currProcess,
        current_print_job: {
          percent_done: currProcess.name === 'PrintProcess' && currProcess.step === 'printing' ? currProcess.progress : null
        }
      });
    }

    const {
      state
    } = this.getStatus();
    const {
      Busy,
      Pending,
      Idle
    } = PrinterStateEnum;

    if ([Busy, Pending, Idle].includes(state)) {
      if (currProcess && currProcess.error) {
        this.saveStateError(currProcess.error);
      } else {
        this.clearStateErrors();
      }
    }

    this._flux.map(flux => flux.actions.PrinterDetailsActions.status(this.getId(), {}));
  }

  updateFromPrinterInfo(printerInfo) {
    this._printerInfo = printerInfo;
    this.updatePrinterStatus();
  }

  invoke(kaitenMethod, params = {}) {
    if (!kaitenMethod) return q(null); // Not elegant at all, but to truly capture the state of which the
    // polling of the command + polling with printer status, we'll need to
    // record the previous current_process (if null) or step (if currPros
    // not null)

    const {
      current_process
    } = this._printerInfo.info;
    const oldCurrProcOrStep = current_process && current_process.step;
    return this._sendCmdToReflector(kaitenMethod, params).then(cmdId => {
      return this._pollCmdFromReflector(cmdId, oldCurrProcOrStep, 2);
    });
  }
  /**
   * Send Kaiten Cmds to Reflector via POST-REQUEST for Sketch Printers
   * @param {str} method - kaiten method (snakecasing please!)
   * @param {obj/array} params - (obj) params contains additional arguments listed in
   *                             kaiten-API.md document. (array) - contains the process
   *                             method name needed to pass into params object...
   */


  _sendCmdToReflector(method, params) {
    const defer = q.defer();
    const options = {
      headers: {
        'Authorization': `Bearer ${this.userToken}`,
        'Content-Type': 'application/json'
      },
      url: `${reflectorUrl}/printers/${this.getIserial()}/commands`,
      method: 'POST',
      body: JSON.stringify({
        // ARGGGHHH! camelcasing vs snakecasing
        method: method === 'ProcessMethod' ? 'process_method' : method,
        params: Array.isArray(params) ? {
          method: params[0]
        } : params
      })
    };
    request(options, (err, res, body) => {
      if (err || ![200, 201].includes(res.statusCode)) {
        console.error(`Reflector REST API Error: Unable to process '${method}' command.`, err);
        defer.reject(err || res.body);
      } else {
        console.log(`Reflector REST API: Sent '${method}' command success`);
        const jsonRes = JSON.parse(body); // ATM, we're not invoking multiple cmds via reflector,
        // therefore return the first element of cmdIds array

        defer.resolve(jsonRes[0]);
      }
    });
    return defer.promise;
  }
  /**
   * Poll Reflector for status update on a command via GET-REQUEST for Sketch Printers
   * @param {str} cmdId - command id return reflector after sending a cmd request
   * @param {str/null} oldCurrProcOrStep - old state for comparing during polling
   * @param {int} timeInterval - time delay between polling for cmd status (seconds)
   **/


  _pollCmdFromReflector(cmdId, oldCurrProcOrStep, timeInterval) {
    const defer = q.defer();
    const options = {
      headers: {
        'Authorization': `Bearer ${this.userToken}`,
        'Content-Type': 'application/json'
      },
      url: `${reflectorUrl}/printers/${this.getIserial()}/commands/${cmdId}`,
      method: 'GET'
    };
    const pollingInterval = setInterval(() => {
      request(options, (err, res, body) => {
        if (err || ![200, 201].includes(res.statusCode)) {
          console.error(`Reflector REST API Error: Unable to poll command id: [${cmdId}].`, err);
          clearInterval(pollingInterval);
          defer.reject(err || res.body);
        } else {
          const jsonRes = JSON.parse(body);
          const {
            status
          } = jsonRes; // Not elegant at all, but to truly capture the state of which the
          // polling of the command + polling with printer status, we'll need to
          // record the previous current_process (if null) or step (if currPros
          // not null)

          const {
            current_process
          } = this._printerInfo.info;
          const newCurrProcOrStep = current_process && current_process.step; // TODO - find another logic condition to keep the spinner
          // loading because _printerInfo is still not up-to-date

          if (status === 'ack' && oldCurrProcOrStep !== newCurrProcOrStep) {
            clearInterval(pollingInterval);
            defer.resolve();
          }
        }
      });
    }, timeInterval * 1000);
  } // after slice -> upload the .makerbot


  uploadFileToCloudslicer(dotMakerbotFile) {
    const options = {
      method: 'POST',
      url: `${reflectorUrl}/printers/${this.getIserial()}/upload`,
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization': `Bearer ${this.userToken}`
      },
      body: dotMakerbotFile
    };
    return q.Promise((resolve, reject) => {
      return request(options, (error, response, body) => {
        // how should errors be handled?
        if (error) {
          throw new Error(error);
        } else {
          // return the GCP file location URL
          const makerbotFileURL = JSON.parse(response.body).url;
          resolve(this.startCloudPrint(makerbotFileURL));
        }
      });
    });
  } // send print command with stored .makerbot


  startCloudPrint(makerbotFileURL) {
    const params = {
      url: makerbotFileURL,
      ensure_build_plate_clear: false
    };
    this.invoke('external_print', params);
  } /// Toolpath object made from what is returned by 'build' in makerbot.js...


  sendPrintFile(dotMakerbotFile) {
    if (this.getStatus().state !== PrinterStateEnum.Idle) {
      throw 'Unable to print to this printer at this time';
    }

    const filePath = path.resolve(__dirname, dotMakerbotFile.get('path'));
    return q(fs.readFile(filePath, (err, data) => {
      if (err) {
        throw err;
      } else {
        return this.uploadFileToCloudslicer(data);
      }
    }));
  }

  startPrint() {
    // This is here just for consistency with materbot-printer.js, and does nothing.
    return q();
  }

  changeFilament(processMethod, idleMethod) {
    if (this.getStatus().state === PrinterStateEnum.Busy) {
      return this.invoke('process_method', {
        method: processMethod
      });
    } else {
      return this.invoke(idleMethod);
    }
  }

  loadFilament() {
    const processMethod = 'load_filament';
    const idleMethod = 'load_filament';
    return this.changeFilament(processMethod, idleMethod);
  }

  unloadFilament() {
    const processMethod = 'unload_filament';
    const idleMethod = 'unload_filament';
    return this.changeFilament(processMethod, idleMethod);
  }

  stopFilament() {
    if (this.getStatus().state === PrinterStateEnum.Busy) {
      return this.invoke('process_method', {
        method: 'stop_filament'
      });
    }

    return this.cancel();
  }

  getExtruderInfo() {
    return this._printerInfo.info.toolheads ? this._printerInfo.info.toolheads.extruder : [];
  }

  getAllExtruderInfo() {
    return q(this.getExtruderInfo());
  }

  getSystemInfo() {
    return this._printerInfo.info || {};
  }

  getNetworkState() {
    return {};
  }

  getIpAddress() {
    return this._printerInfo.info ? this._printerInfo.info.ip : null;
  }

  getPlatformInfo() {
    return this._printerInfo.info.toolheads ? this._printerInfo.info.toolheads.platform : [];
  }

  getIserial() {
    return this.getId().split(':')[2];
  }

  getCurrentExtruderType() {
    return ExtruderTypeEnum.sketch_extruder;
  }

  isSketchPrinter() {
    return true;
  } // /stores/printerStore.js


  injectFlux(flux) {
    this._flux = maybe(flux);
    return this;
  }

}

module.exports = SketchPrinter;
