// MB-support-plugin/lib/stored_printers_manager.js
'use strict';

const q = require('q');

const fs = require('fs');

const _ = require('lodash');

const paths = require('./paths');

const ANON_TOKEN = 'ANON';
const PrinterType = {
  ARCHETYPE: 0,
  REAL: 1
};

class StoredPrintersManager {
  constructor() {
    this._savedPrinters = {};
    this._writingPromise = null;
    this._printersJsonNeedsUpdate = false;
    this._stopped = false;
  }

  setLogger(logger) {
    this._logger = logger;
  }

  getPrinter(uid) {
    return this._getRealPrinters()[uid] || this._getArchetypePrinters()[uid];
  }

  _isArchetype(uid) {
    return uid.split(':')[0] === 'offline';
  }

  _getRealPrinters() {
    return this._savedPrinters.realPrinters;
  }

  _getArchetypePrinters() {
    return this._savedPrinters.archetypes;
  }

  getUserPrinters(token) {
    return this._getPrintersForToken(token, PrinterType.REAL);
  }

  getAnonymousPrinters() {
    return this._getPrintersForToken(null, PrinterType.REAL);
  }

  getArchetypePrinters(token) {
    return this._getPrintersForToken(token, PrinterType.ARCHETYPE);
  }

  _getPrintersForToken(token, realOrArchetype) {
    const printerIds = [];

    if (!token) {
      if (realOrArchetype === PrinterType.REAL) {
        // real printers in printers.json that have an authInfo attached
        // to them, which should only be stored if they had been authenticated
        // to with no logged in user
        const printers = this._getRealPrinters();

        for (const uid in printers) {
          if (printers[uid].authInfo) printerIds.push(uid);
        }

        return printerIds;
      } else {
        token = ANON_TOKEN;
      }
    } // printers in printers.json that have the given user token attached to them


    const printers = realOrArchetype === PrinterType.REAL ? this._getRealPrinters() : this._getArchetypePrinters();

    for (const uid in printers) {
      const userTokens = printers[uid].userTokens;
      if (userTokens && userTokens.indexOf(token) !== -1) printerIds.push(uid);
    }

    return printerIds;
  }

  _getPrinterListFile() {
    const subFolder = paths.getConfigDir() + (process.platform == 'win32' ? '\\printers.json' : '/printers.json');
    return subFolder;
  }

  loadSavedPrinters() {
    const self = this;
    const readFile = q.denodeify(fs.readFile);
    return readFile(self._getPrinterListFile()).catch(function (err) {
      // If printer.json does not exist, return {};
      // warning here: We're banking on that the printers.json will
      // then get created when it gets written on upgrade -- ie. it
      // won't get created until the next request to write if it doesn't
      // upgrade. It is very unlikely that that will be the case, and
      // this._savedPrinters, which is what we actually care about, will
      // be correct, so it shouldn't be a problem -- but in case any
      // strange behavior crops up because of this...
      if (err.code !== 'ENOENT') throw err;
      return '{}';
    }).then(function (data) {
      try {
        self._savedPrinters = JSON.parse(data);
      } catch (err) {
        if (err.name === 'SyntaxError') self._logger.warn('Failed to parse printers.json');
      }

      if (self._upgradePrintersJson()) self._writeToPrinterJSON();
    });
  }

  _upgradePrintersJson() {
    // nice place to put all format changes to printers.json
    let updated = false; // as of PEN-1057: store both real printers and archetypes in printers.json

    if (!this._savedPrinters.realPrinters && !this._savedPrinters.archetypes) {
      this._savedPrinters = {
        realPrinters: this._savedPrinters,
        archetypes: {}
      };
      updated = true;
    }

    return updated;
  }

  _writeToPrinterJSON() {
    if (this._stopped) {
      this._logger.warn('Stored printers manager is stopped; ignoring write');

      return;
    }

    if (this._isWriting) throw 'Already writing to printers.json!';

    const file = this._getPrinterListFile();

    const writeFile = q.denodeify(fs.writeFile);

    const doWrite = () => {
      this._printersJsonNeedsUpdate = false;
      return writeFile(file, JSON.stringify(this._savedPrinters, null, '\t')).then(() => {
        this._logger.log('Updated stored printer data'); // if more updates got added while we were waiting for write to
        // finish, write again


        if (this._printersJsonNeedsUpdate) return doWrite();
      });
    };

    this._writingPromise = doWrite();
    return this._writingPromise;
  }

  _updatePrintersJson() {
    this._printersJsonNeedsUpdate = true;

    if (!this._isWriting) {
      try {
        fs.mkdirSync(paths.getConfigDir());
      } catch (e) {
        // Folder already existing is not a fail case, overlook
        if (e.code != 'EEXIST') throw e;
      }

      return this._writeToPrinterJSON();
    } else {
      return q();
    }
  }

  removePrinterForUser(uid, token) {
    let savedInfo;
    if (this._isArchetype(uid)) savedInfo = this._getArchetypePrinters()[uid];else savedInfo = this._getRealPrinters()[uid];

    if (savedInfo) {
      if (this._isArchetype(uid) && !token) token = ANON_TOKEN;

      if (token) {
        const tokens = savedInfo.userTokens;
        const index = tokens.indexOf(token);

        if (index != -1) {
          tokens.splice(index, 1);
        }
      } else {
        delete this._getRealPrinters()[uid].authInfo;
      } // remove from printers.json if this printer no longer belongs to anyone


      if ((!savedInfo.userTokens || !savedInfo.userTokens.length) && !savedInfo.authInfo) {
        // hopefully it will never be the case that we may have archetypes
        // and real printers with the same id...
        delete this._getRealPrinters()[uid];
        delete this._getArchetypePrinters()[uid];
      }

      this._updatePrintersJson().done();
    }
  }

  updateStoredInfo(printerInfo, machineConfig, authInfo, userToken) {
    // only write if something we're storing changed
    const savedInfo = this._getRealPrinters()[printerInfo.uid];

    const savedInfoCopy = Object.assign({}, savedInfo || {
      printer_info: {}
    }); // this is pretty janky looking, but I blame it on how js handles objects....

    savedInfoCopy.printer_info = {
      uid: printerInfo.uid || savedInfoCopy.printer_info.uid,
      name: printerInfo.name || savedInfoCopy.printer_info.name,
      address: printerInfo.address || savedInfoCopy.printer_info.address,
      info: printerInfo.info || savedInfoCopy.printer_info.info
    };
    if (machineConfig) savedInfoCopy.machineConfig = machineConfig;

    if (userToken) {
      savedInfoCopy.userTokens = _.union(savedInfoCopy.userTokens, [userToken]);
    } else if (authInfo) {
      savedInfoCopy.authInfo = authInfo;
    }

    if (JSON.stringify(savedInfo) !== JSON.stringify(savedInfoCopy)) {
      this._getRealPrinters()[printerInfo.uid] = savedInfoCopy;
      return this._updatePrintersJson();
    } else {
      return q();
    }
  }

  storeArchetype(printerId, vendor, gender, userToken) {
    let savedInfo = this._getArchetypePrinters()[printerId];

    if (!savedInfo) {
      savedInfo = {
        vendor,
        gender,
        userTokens: []
      };
      this._getArchetypePrinters()[printerId] = savedInfo;
    }

    if (!userToken) userToken = ANON_TOKEN;

    const tokens = _.union(savedInfo.userTokens, [userToken]);

    if (tokens.length != savedInfo.userTokens.length) {
      savedInfo.userTokens = tokens;
      return this._updatePrintersJson();
    } else {
      return q();
    }
  }

  stop() {
    return this._writingPromise.then(() => {
      this._stopped = true;
    });
  }

  get _isWriting() {
    return this._writingPromise && this._writingPromise.isPending();
  }

}

module.exports = StoredPrintersManager;
