// MB-support-plugin/lib/thing.js
"use strict";

const fs = require('fs');

const JSZip = require('jszip');

const path = require('path');

const q = require('q');

const tmp = require('tmp');

const THREE = require('three');

const constants = require('./constants'); // Configure temp directory behavior


tmp.setGracefulCleanup(); // setup thing extension

const THING_EXTENSION = '.thing';
/**
 * Helper function to split a single dimension array into
 * bite size chunks
 *
 * @param (array) the array you want to break down
 * @param (size) the size of the chunks
 *
 * @return an array x by x size
 */

const createMultiChunksArray = (array, size) => {
  const result = [];

  for (let i = 0; i < array.length; i += size) {
    result.push(Array.from(array.slice(i, i + size)));
  }

  return result;
};
/**
 * Take the single dimension 16-element array representation of THREE.Matrix4
 * transform and convert it into a usable 4x4 array for libThing. The scaling
 * factor needs to be in mm.
 *
 * @param transform (array) single dimension 16-element array that is row-major
 *  Check trayBody.getBuildTransform() => transform.toArray()
 *
 * @return transform formatted into 4 x 4 array
 * [
 *   [ A11, A12, A13, A14 ],
 *   [ A21, A22, A23, A24 ],
 *   [ A31, A32, A33, A34 ],
 *   [ A41, A42, A43, A44 ]
 * ]
 */


const modifyTransformForManifest = transform => {
  /**
   * THREE.Matrix4().fromArray expects array set to be column-major. The
   * transform return from trayBodies.getBuildTransform() returns row-major
   * single dimension array.
   *
   * It requires a transpose() after creating the transform matrix using
   * row-major array input (@param transform).
   *
   */
  const threeTransform = new THREE.Matrix4().fromArray(transform).transpose(); // extract rotation, position, and scale matricies from the transformation

  const position = new THREE.Vector3();
  const quaternion = new THREE.Quaternion();
  const scale = new THREE.Vector3();
  threeTransform.decompose(position, quaternion, scale); // all eagle-desktop transforms are in meters, but libthing deals in
  // millimeters, so scale everything up by 1000

  position.multiplyScalar(1000);
  scale.multiplyScalar(1000);
  const scaledTransform = new THREE.Matrix4().compose(position, quaternion, scale);
  /**
   * --NOTES--
   * To build the proper represenation of the Manifest Transformation matrix,
   * output array needs to be row-major, then chunk into 4 x 4 array.
   *
   * From [ A11, A12, A13, A14, A21, A22, A23, A24 ,A31, A32, A33, A34,
   *   A41, A42, A43, A44 ]
   *
   * To [
   *   [ A11, A12, A13, A14 ],
   *   [ A21, A22, A23, A24 ],
   *   [ A31, A32, A33, A34 ],
   *   [ A41, A42, A43, A44 ]
   * ]
   *
   * THREE.Matrix4().set takes in row-major array, but .elements display
   * column-major array. To return row-major array from .elements, transpose
   * the matrix before returning the array:
   *
   * - scaledTransform.elements will return col-major array
   * - scaledTransform.transpose().elements will return row-major array
   *
   */

  return createMultiChunksArray(scaledTransform.transpose().elements, 4);
};

const createThingFile = (stlFiles, extruderList, transforms, outputFileName, callback) => {
  // check that we have stl files and that its the same size as transforms
  if (!stlFiles || stlFiles.length === 0) {
    return q.Promise.reject(new Error('No STL files to create dot thing!'));
  } else if (!transforms || transforms.length !== stlFiles.length) {
    return q.Promise.reject(new Error('Invalid transforms supplied to STL merge'));
  } // create the empty manifest


  const manifest = {
    // where to find the spec
    namespace: 'http://spec.makerbot.com/ns/thing.0.1.1.1',
    // we can probably kill this section
    attribution: {
      author: 'Unknown',
      license: 'Unknown'
    },
    // this is where we'll put the stl file names
    objects: {},
    // this is where we'll store info on each instance of our objects
    instances: {},
    // this is where we'll store the transformations for each instance
    transformations: {},
    // this is where we store the settings to use on our instances
    // for now, we're only printing on single extrusion machines with uniform
    // settings, so we'll leave it at just the traditional plastic a
    construction: ['PlasticA', 'PlasticB']
  }; // spin up a tmp directory and a tmp filepath for the thing

  const tempDir = tmp.dirSync();
  const outputFilePath = path.join(tempDir.name, outputFileName + THING_EXTENSION);
  const zipFile = new JSZip(); // setup an array to keep track of what stl's have already been processed

  const alreadyZipped = [];
  const zipPromises = []; // method to add transforms for each instance to the thing manifest

  const addTransformToManifest = (id, transform) => {
    const transformName = `transform${id}`;
    const matrix = modifyTransformForManifest(transform);
    manifest.transformations[transformName] = {
      matrix
    };
    return transformName;
  }; // function to add a mesh file to the zip folder. returns a promise.


  const addFileToZip = (filePath, fileName) => {
    // create a defered object
    const d = q.defer(); // in the readfile callback function, use the deferred object

    fs.readFile(filePath, (err, data) => {
      if (err) d.reject(err);
      zipFile.file(fileName, data);
      d.resolve();
    }); // return the deferred object's promise

    return d.promise;
  }; // function to add the thing manifest to the sip folder. returns a promise.


  const addManifest = () => {
    return q.Promise(resolve => {
      zipFile.file('manifest.json', JSON.stringify(manifest, null, 2));
      resolve();
    });
  }; // compress the thing file. return promise to compress the zipped file


  const compressZipFile = () => {
    // create a defered object
    const d = q.defer(); // options for zipping the file

    const zipOptions = {
      type: 'nodebuffer',
      compression: 'DEFLATE',
      compressionOptions: {
        level: 6
      },
      platform: 'UNIX'
    }; // generate the zipped content and write it to our output file

    zipFile.generateNodeStream(zipOptions).pipe(fs.createWriteStream(outputFilePath)).on('finish', () => {
      d.resolve(outputFilePath);
    }); // return the deffered object's promise

    return d.promise;
  }; // translate from extruderA/B to PlasticA/B


  const translateExtruder = extruder => {
    return constants.ExtruderTranslation[extruder] ? constants.ExtruderTranslation[extruder] : 'PlasticA';
  }; // process each model file


  stlFiles.forEach((stlFile, i) => {
    // get the filepath and transform for this mesh
    const meshFilepath = stlFile;
    const meshBasename = path.basename(meshFilepath);
    const meshTransform = transforms[i];
    const instanceName = `instance${i}`;
    const transformName = addTransformToManifest(i, meshTransform);
    const plasticSelection = translateExtruder(extruderList[i]); // add the instance to the manifest with the correct values

    manifest.instances[instanceName] = {
      construction: plasticSelection,
      object: meshBasename,
      scale: 'mm',
      xform: transformName
    };
    /**
     * if this mesh hasn't been added to the zip file yet, then create a
     * promise to do so and add it to the array of zip promises.
     */

    if (alreadyZipped.indexOf(meshBasename) === -1) {
      alreadyZipped.push(meshBasename);
      zipPromises.push(addFileToZip(meshFilepath, meshBasename));
      manifest.objects[meshBasename] = {};
    }
  }); // settle all the adding promises, then add the meta data, then zip up the dot
  // thing file

  return q.allSettled(zipPromises).then(addManifest).then(compressZipFile).nodeify(callback);
};

module.exports = {
  createThingFile
};
