// MB-support-plugin/lib/thingiverse-api-client.js
'use strict';

const Eagle = require('eagle-print');

const TypeCheck = Eagle.util.TypeCheck;

const EventEmitter = require('eventemitter3');

const q = require('q');

const url = require('url');

const request = require('request');

const fs = require('fs-extra');

const path = require('path');

const util = require('./util');

const ErrorsEnum = require('./thingiverse-api-errors').ErrorsEnum;

const TVApiError = require('./thingiverse-api-errors').ThingiverseApiError;

const client_id = '23c5eabe7a566d3a53c0';
const client_secret = 'c30f532bcc67bb65d3476daedc0e60f4';
const redirectUri = 'makerbotprint://callback';
const baseOauthUrl = 'https://login.makerbot.com'; // TODO: Move this out to default-config to support staging URL in the future

const reflectorUrl = 'https://reflector.makerbot.com';

class ThingiverseApiClient extends EventEmitter {
  /**
   * @param {string} serverConfig
   * @param {LoggingService} loggerService
   */
  constructor(serverConfig, loggerService, trackingService, configService) {
    TypeCheck.isType(serverConfig, Object);
    TypeCheck.isType(loggerService, Object);
    TypeCheck.isType(trackingService, Object);
    TypeCheck.isType(configService, Object);
    super();
    this._trackingService = trackingService;
    this._configService = configService;
    this.serverConfig = serverConfig;
    this.log = loggerService;
    this.suppressedHeaders = ['Authorization'];
  }

  _sanitizeForLogging(headers) {
    //Suppressed headers will be replaced with 'omitted' in debug logs
    const sanitizedHeaders = Object.assign({}, headers);
    this.suppressedHeaders.map(function (headerField) {
      if (sanitizedHeaders[headerField]) {
        sanitizedHeaders[headerField] = 'omitted';
      }
    });
    return sanitizedHeaders;
  }

  _request(token, options) {
    return q.Promise((resolve, reject) => {
      options.url = options.url || this.getUrl(options.path);
      delete options.path;
      options.headers = options.headers || {};

      if (token) {
        options.headers['Authorization'] = `Bearer ${token}`;
        options.headers['Content-Type'] = 'application/json';
      }

      const proxySettings = this._configService.get('proxySettings');

      if (proxySettings) {
        let hostInfo;
        let userInfo = '';

        if (proxySettings.address && proxySettings.port) {
          hostInfo = `${proxySettings.address}:${proxySettings.port}`; // Username and Password can be optional!

          if (proxySettings.username && proxySettings.password) {
            userInfo = `${proxySettings.username}:${proxySettings.password}@`;
          }

          options.proxy = proxySettings.protocol + userInfo + hostInfo;
        }
      }

      this.log.info('Thingiverse Api request: ', {
        url: options.url,
        method: options.method,
        headers: this._sanitizeForLogging(options.headers),
        encoding: options.encoding,
        body: options.body
      }); // clarity for future selves -> this is using the imported 'request' library here inside _request()

      request(options, (error, response, body) => {
        const parsedError = this._parseErrorResponse(options, error, response, body);

        if (parsedError) {
          this._handleError(reject, parsedError);

          this.log.info('Thingiverse Api request error: ', parsedError);
        } else {
          this._handleSuccessfulResponse(resolve, response, body);

          this.log.debug('Thingiverse Api request success: ', response.status_code);
        }
      });
    });
  }

  _parseErrorResponse(options, error, response, body) {
    // add more error codes if we know of them
    if (error) {
      console.log(`Connection Error ${error.message}`);
      return new TVApiError(error.message, ErrorsEnum.ConnectionError);
    } else if (response.statusCode == 403) {
      return new TVApiError('Access Denied', ErrorsEnum.AccessDenied);
    } else if (response.statusCode == 404) {
      return new TVApiError('Not Found', ErrorsEnum.FileNotFound);
    } else if (response.statusCode == 405) {
      return new TVApiError('Method Not Allowed', ErrorsEnum.MethodNotAllowed);
    } else if (response.statusCode >= 500) {
      return new TVApiError('Server Error', ErrorsEnum.ServerError);
    } else if (response.statusCode >= 400) {
      return new TVApiError('Client Error', ErrorsEnum.ClientError);
    }
  }

  _handleSuccessfulResponse(resolve, response, body) {
    let parsedResponse;

    if (response.headers['content-type'] && response.headers['content-type'].indexOf('json') > -1) {
      parsedResponse = JSON.parse(body);
    } else {
      parsedResponse = body; // if the response is url-encoded, access_token is sent in the body
      // and is of the format ^access_token=<token>&token_type=Bearer.
      // token will need to be extracted out of that

      const access_token = /^access_token=.*&token_type=Bearer$/.test(body) ? body.match(/^access_token\=([\S\s]*?)\&/)[1] : null;
      if (access_token) parsedResponse = access_token;
    }

    resolve(parsedResponse);
    this.emit('successfulResponse', parsedResponse);
  }

  _handleError(reject, error) {
    this.emit('error', error);
    reject(error);
  }

  getUrl(path) {
    // for auth (login), we user the auth url
    // after successful auth, further api calls are with the api url
    if (path) {
      const serverUrl = /\.*login\.*/.test(path) ? this.serverConfig['server']['auth_url'] : this.serverConfig['server']['api_url'];
      return url.resolve(serverUrl, path);
    }

    return undefined;
  }
  /**
   * Authenticate user on server and return token to be used in other requests.
   * @param {string} username
   * @param {string} password
   * @returns {Promise} resolves to {string} token or one of errors
   */


  authenticate(username, password) {
    return this._request(null, {
      method: 'POST',
      path: `login/oauth/access_token`,
      form: {
        grant_type: 'password',
        client_id,
        client_secret,
        username,
        password
      }
    });
  }
  /**
   * Invalidate token on server. NB: this method will fail if token is invalid,
   * so you might need to ignore its result if you just need to log user out.
   * @param {string} token
   * @returns {Promise}
   */


  logout(token) {
    return this._request(token, {
      method: 'DELETE',
      path: `sessions`
    });
  }

  getUser(token) {
    return this._request(token, {
      path: `users/me`
    });
  }

  getAvatar(token, url) {
    /* Ideally, I should have used the request call we have for all the api requests
       but, it doesn't work with the current way request is setup.
       Can't have a callback as well as pipe to stream...besides, I'm not sure
       I want the response always piped to a file.
     */
    let remote, app;

    try {
      remote = require('electron').remote;
      app = remote.app;
    } catch (err) {
      // if this is being run from the main process, then 'remote' doesn't exist and will throw an error
      // in that case, simply require app
      try {
        app = require('electron').app;
      } catch (err) {// this can happen if running in unit test environment
      }
    }

    const userDataPath = app ? app.getPath('userData') : '';
    const avatarFile = path.resolve(userDataPath, 'avatar.jpg');
    return q.Promise((resolve, reject) => {
      const avatarFileStream = fs.createWriteStream(avatarFile);
      avatarFileStream.on('finish', function () {
        resolve(avatarFile);
      });
      request.get(url).on('error', err => this._handleError(reject, err)).pipe(avatarFileStream).on('error', err => this._handleError(reject, err));
    });
  }

  getUserPrintersFromReflector(token) {
    const options = {
      url: `${reflectorUrl}/printers`
    };
    return this._request(token, options);
  }

  getClientId() {
    return client_id;
  }

  getRedirectUri() {
    return redirectUri;
  }

  getBaseOauthUrl() {
    return baseOauthUrl;
  }

}

module.exports = ThingiverseApiClient;
