// MB-support-plugin/lib/toolpather_debug_menu_items.js
'use strict';

const fs = require('fs');

const MenuItemProvider = require('eagle-print').views.menu.MenuItemProvider;

const q = require('q');

const analyticUtils = require('eagle-print/util/analyticUtils');

const R = require('ramda');

const util = require('./util'); /// function for exporting files from @tmpPath to @dstPath, pretty much the
/// mb-plugin's export service
/// @returns a q.resolve or a q.reject depending on error status.
// eslint-disable-next-line no-unused-vars


const exportFile = function (tmpPath, dstPath) {
  // flag and function to log errors
  const done = function (err) {
    if (err) {
      return q.reject(err);
    }

    return q.resolve();
  }; // create a read stream from tmpPath, log on 'error'


  const rd = fs.createReadStream(tmpPath);
  rd.on('error', function (err) {
    done(err);
  }); // create write stream from dstPath, log on 'error'

  const wr = fs.createWriteStream(dstPath);
  wr.on('error', function (err) {
    done(err);
  });
  wr.on('close', function (ex) {
    done();
  }); // pipe the read path to the write path

  rd.pipe(wr);
}; /// function to log debug menu errors. for now just console log.


const logError = function (err) {
  console.error(err);
}; /// is @printer a makerbot printer


const isMakerbotPrinter = function (printer) {
  return printer && printer.getInfo().vendor.toLowerCase() === 'makerbot';
};

const progressReport = function (flux, fraction, action, actionLabel, actionCallback, closeAction, cursorStyle) {
  const idMessage = `project_${action}`;
  const textMessageProgress = `messages.project.${action}.progress`;
  const textMessageSuccess = `messages.project.${action}.success`;
  const i18n = flux.getService('TranslationService');

  if (fraction < 0) {
    // dismiss message (failed)
    flux.actions.message.dismissMessage(idMessage);
  } else if (fraction === 100) {
    // end message
    flux.actions.message.createSuccessMessage({
      id: idMessage,
      message: i18n.t(textMessageSuccess)
    });
  } else {
    // progress message
    flux.actions.message.updateProgress({
      progressPercentage: fraction,
      isIndeterminate: true,
      id: idMessage,
      message: i18n.t(textMessageProgress),
      actionLabel,
      actionCallback,
      closeAction,
      cursorStyle
    });
  }
};

module.exports = {
  /// the toolpather menu item, all other items will go inside of this one
  MBToolpathDebugMenu: MenuItemProvider.createSimpleMenuItemClass({
    displayName: 'DebugMBToolpath',
    label: 'MB Toolpather...'
  }, {
    unlocalized: true
  }),
  /// use a custom profile to slice the build plate and export the dot makerbot
  /// file. eventually, this should maybe be a print.
  MBCustomSliceMenuItem: MenuItemProvider.createSimpleMenuItemClass({
    displayName: 'MBCustomSlice',
    label: 'Custom MB Slice',

    action() {
      // use election.remote.dialogs to get the profile and save locations
      const remote = require('electron').remote;

      const flux = this.getFlux();
      const exportService = flux.getService('ExportService');
      const printer = flux.stores.PrinterStore.getSelectedPrinter(); // Must have selected an auth'ed printer.

      if (!isMakerbotPrinter(printer)) {
        flux.actions.message.createErrorMessage({
          message: 'This menu option requires selecting an authenticated MakerBot Printer'
        });
        logError('This menu option requires selecting a MakerBot Printer');
        return;
      } // ask the user for a config path, exit if not provided


      const configFilters = [{
        name: 'config',
        extensions: ['json']
      }];
      const configPath = remote.dialog.showOpenDialog(remote.getCurrentWindow(), {
        title: 'choose a profile',
        filters: configFilters
      });

      if (!configPath) {
        flux.actions.message.createErrorMessage({
          message: 'Please provide a valid profile path.'
        });
        return;
      } // Selected profile bot_type must match auth'ed printer bot_type.


      const parsedProfile = JSON.parse(fs.readFileSync(configPath[0], 'utf8'));

      if (printer.getGender().botType !== parsedProfile._bot) {
        flux.actions.message.createErrorMessage({
          message: 'Selected printer bot_type does not match profile bot_type.'
        });
        return;
      } // Ask the user for a save location, exit if not provided.


      const savePath = remote.dialog.showSaveDialog(remote.getCurrentWindow(), {
        title: 'choose a save location',
        filters: [exportService.getSupportedFormats(printer)]
      });

      if (!savePath) {
        flux.actions.message.createErrorMessage({
          message: 'Please provide a valid file save path.'
        });
        return;
      }

      const project = flux.stores.ProjectStore.getProject();
      const tray = project.getActiveTray();
      let customTray = null;
      const trayBodies = tray.getTrayBodies().toArray();
      trayBodies.forEach(trayBody => {
        const customOccurrence = trayBody.getOccurrence().setProperty('customProfilePath', configPath);
        const customBody = trayBody.setOccurrence(customOccurrence);
        customTray = tray.setTrayBody(customBody);
      }); // Wait for the current build to be cleared and then re-slice

      const buildStore = flux.stores.BuildStore;
      buildStore.once('change', () => {
        setImmediate(() => {
          flux.actions.printer.exportToFile(printer, customTray, project, savePath);
        });
      });
      flux.actions.printer.cancelBuildsForTray(customTray.getId());
    }

  }, {
    unlocalized: true
  }),
  /// export the current tray as a dot thing file
  ExportThingMenuItem: MenuItemProvider.createSimpleMenuItemClass({
    displayName: 'ExportThing',
    label: 'Export Thing',

    action() {
      // ask the user for an export path, exit if not provided
      const remote = require('electron').remote;

      const saveFilters = [{
        name: 'model',
        extensions: ['thing']
      }];
      const savePath = remote.dialog.showSaveDialog(remote.getCurrentWindow(), {
        title: 'choose a save location',
        filters: saveFilters
      });

      if (!savePath) {
        return;
      }
      /**
       * Use flux to get the current project and printer
       **/


      const flux = this.getFlux();
      const project = flux.stores.ProjectStore.getProject();
      const printer = flux.stores.PrinterStore.getSelectedPrinter();

      if (!isMakerbotPrinter(printer)) {
        logError('This menu option requires selecting a MakerBot Printer');
        return;
      }

      const ProjectService = flux.getService('ProjectService');
      return q().then(() => {
        progressReport(flux, 0, 'save');
      }).then(() => {
        ProjectService.save(project, savePath, flux);
      }).then(() => {
        progressReport(flux, 100, 'save');
      });
    }

  }, {
    unlocalized: true
  }),
  /// use the current settings and printer to generate / save a custom profile
  ExportCustomProfileMenuItem: MenuItemProvider.createSimpleMenuItemClass({
    displayName: 'ExportCustomProfile',
    label: 'Export Custom Profile',

    action() {
      // ask the user for an export path, exit if not provided
      const remote = require('electron').remote;

      const saveFilters = [{
        name: 'configuration',
        extensions: ['json']
      }];
      const savePath = remote.dialog.showSaveDialog(remote.getCurrentWindow(), {
        title: 'choose a save location',
        filters: saveFilters
      });

      if (!savePath) {
        return;
      } // use the flux stores to get the active tray and selected printer


      const flux = this.getFlux();
      const project = flux.stores.ProjectStore.getProject();
      const tray = project.getActiveTray();
      const printer = flux.stores.PrinterStore.getSelectedPrinter();

      if (!isMakerbotPrinter(printer)) {
        logError('This menu option requires selecting a MakerBot Printer');
        return;
      } // generate build info from the current project's active tray


      const rootOccurrences = project.getRootOccurrences();
      const tessellationMap = project.getTessellationMap();
      const buildInfo = tray.getBuildInfo(printer, {
        tessellationMap,
        rootOccurrences
      }); // turn the profile into a pretty json string and write it

      const saveProfile = function (profile) {
        const jsonString = JSON.stringify(profile, null, 2);
        fs.writeFileSync(savePath, jsonString);
      }; // use sliceconfig to generate a custom profile with the settings on the
      // current tray. save the resulting json to the desired save path


      const sliceconfig = require('./makerbot');

      sliceconfig.getProfile(buildInfo, printer, null, flux).then(saveProfile).catch(logError);
    }

  }, {
    unlocalized: true
  }),
  /// Opens dialog where user selects a dot makerbot (.makerbot) file to import
  /// and send to the selected printer.
  PrintFromToolpathMenuItem: MenuItemProvider.createSimpleMenuItemClass({
    displayName: 'PrintFromToolpath',
    label: 'Print from .makerbot',

    action() {
      // Make sure an authenticated (live) printer is selected.
      const remote = require('electron').remote;

      const flux = this.getFlux();
      const printer = flux.stores.PrinterStore.getSelectedPrinter();
      let loadPath = null; // A valid printer needs to be a makerbot printer, and not `Offline`

      const isValidPrinter = isMakerbotPrinter(printer) && printer.getStatus().state !== 'Offline';

      if (!isValidPrinter) {
        flux.actions.message.createErrorMessage({
          message: 'This menu option requires selecting an authenticated MakerBot Printer'
        });
        logError('This menu option requires selecting an authenticated MakerBot Printer');
        return;
      } // If printer is valid...


      return q() // Pop dialog, let user choose import dot makerbot file
      .then(() => {
        const loadFilters = [{
          name: 'makerbot',
          extensions: ['makerbot']
        }];
        loadPath = remote.dialog.showOpenDialog(remote.getCurrentWindow(), {
          title: 'choose a .makerbot file to print',
          filters: loadFilters
        });
        if (!loadPath) logError('No valid load path for .makerbot!');
        return;
      }).then(() => {
        const trackingService = this.props.flux.getService('TrackingService');
        const trackingInfo = util.trackingInfoForPrinter(printer);
        trackingInfo.extruder_type = printer.getCurrentExtruderType();
        const printSettings = analyticUtils.getPrintSettingsForAnalyticEvents(this.props.flux, printer);
        trackingService.addEvent(R.merge({
          name: 'Print Started From .makerbot'
        }, trackingInfo, printSettings));
      }) // Create a new project to clear the build plate.
      .then(() => flux.actions.file.newProject()) // Import toolpath, and send to selected printer.
      .then(() => flux.actions.file.importToolpath(loadPath[0])).catch(logError).done();
    }

  }, {
    unlocalized: true
  })
};
