// MB-support-plugin/lib/util.js
'use strict';

const path = require('path');

const R = require('ramda');

const {
  DOM
} = require('react');

const constants = require('./constants');

const jpegSOI = 'ffd8'; // starting bytes indicating the data is a jpeg

const jpegEOI = 'ffd9'; // ending bytes

function checkJpeg(frameBuf) {
  // assume that if the start and end are good, the middle is okay too
  return (frameBuf.slice(0, 2).toString('hex') === jpegSOI && frameBuf.slice(frameBuf.length - 2, frameBuf.length).toString('hex')) === jpegEOI;
}

function getPathForIsvg(path) {
  // some stupid code to fix paths before passing them to react-inlinesvg--
  // react-inlinesvg is using a module called httpplease that will url-encode
  // incoming paths, which breaks with windows' backslash-separated paths...
  // Luckily, these paths will be handled correctly if we change the backslashes
  // to forward slashes though....
  const os_type = require('os').type();

  if (os_type === 'Windows_NT') return path.split('\\').join('/');
  return path;
}

function fullIconPath(uri) {
  return getPathForIsvg(path.resolve('node_modules/MB-support-plugin/', uri));
}

function botTypeToGender(botType) {
  return constants.PrinterGenderString.get(botType);
}

function genderToCodeName(gender) {
  return constants.PrinterCodeNames.get(gender);
}

function printerIDtoSerial(id) {
  return id.split(':')[2];
}

function connTypeToString(connType) {
  switch (connType) {
    case constants.PrinterConnTypeEnum.USB:
      return 'USB';

    case constants.PrinterConnTypeEnum.NETWORK:
      return 'Network';

    case constants.PrinterConnTypeEnum.REFLECTOR:
      return 'Reflector';

    case constants.PrinterConnTypeEnum.ARCHETYPE:
      return 'Archetype';
  }
}

function trackingInfoForPrinter(printer) {
  return {
    printer_id: printerIDtoSerial(printer.getId()),
    printer_type: genderToCodeName(printer.getGender().getName()),
    connection_type: printer.getConnectionType()
  };
}

function sendPrinterAnalytics(flux, event) {
  const trackingService = flux.getService('TrackingService');
  const printer = flux.store('PrinterStore').getSelectedPrinter();
  const trackingInfo = printer ? printer.getTrackingInfo() : null;
  const trackingData = R.merge({
    name: event
  }, trackingInfo);
  trackingService.addEvent(trackingData);
}
/**
 * A wrapper for `request` to allow progress reporting during
 * a large download.
 *
 * @example
 * progress(request("http://example.com/example"))
 *  .on('progress', prog => console.log(`Downloading: ${prog.percentage}%`))
 *  .catch(err => console.log('something went wrong:', err));
 *
 * @returns {Request} the same request object passed as a parameter, but with extra listeners attached
 */


function progress(request, options) {
  // prevent us from wrapping it twice
  if (request.progressContext) {
    return request;
  }

  if (request.response) {
    throw new Error("Request has already responded. Can't get progress");
  }

  const context = {};
  options = options || {};
  context.request = request;
  context.options = {};
  context.options.lengthHeader = options.lengthHeader || 'content-length'; // we've initiated a request

  const onRequest = context => {
    context.state = context.request.progressState;
  }; // server responeded in entirety


  const onResponse = (context, response) => {
    // zero out everything
    context.state = context.request.progressState = {
      cancel: context.request.abort.bind(context.request),
      percentage: 0,
      size: {
        total: Number(response.headers[context.options.lengthHeader]) || null,
        transferred: 0
      }
    };
  }; // data has come down, report the progress


  const onData = (context, data) => {
    context.state.size.transferred += data.length;
    reportState(context);
  }; // clean up everything


  const onEnd = context => {
    context.request.progressState = context.request.progressContext = null;
  }; // actually report the state of the download


  const reportState = context => {
    if (!context.request.progressState) {
      return;
    }

    const state = context.state;

    if (state.size.total != null) {
      state.percentage = Math.min(state.size.transferred, state.size.total) / state.size.total;
    }

    context.request.emit('progress', state);
  };

  request.on('request', () => onRequest(context)).on('response', res => onResponse(context, res)).on('data', data => onData(context, data)).on('end', () => onEnd(context));
  return request;
}

function getMaterialPrettyStringFromCode(materialName) {
  return constants.SliceConfigStringsToPrettyStrings[materialName];
}

function getMaterialSliceConfigStringFromCode(materialType) {
  return constants.MaterialCodesToSliceConfigStrings[materialType];
}

function filamentLengthToMass(length, linearDensity) {
  return length * linearDensity;
}

const getTracking = flux => flux.getService('TrackingService');

const getLogging = flux => flux.getService('LoggingService');

const getI18n = flux => flux.getService('TranslationService');

const getModal = flux => flux.getService('ModalService');

const getConfig = flux => flux.getService('ConfigService'); // Converts an unsave JS string (with ':' and '.') into a safe JS string for
// object keys.
// ex: safeKey('a.b:c') // returns 'a_b_c'
// safeKey :: String -> String


const safeKey = R.replace(/[:|.]/g, '_'); // Gets a value from an object whose keys have been converted to safe keys.
// You can use an unsafe key to search and it will be converted
// getSafeKeyProp :: String -> {k: v} -> a

const getSafeKeyProp = R.uncurryN(2, R.pipe(safeKey, R.prop)); // Sets a value in an object whose keys have been converted to safe keys.
// You can use an unsafe key to search and it will be converted
// setSafeKeyProp :: String -> a -> {k: v} -> {k: v}

const setSafeKeyProp = R.uncurryN(3, R.pipe(safeKey, R.assoc)); // removeSafeKeyProp :: String -> {k: v} -> {k: v}

const removeSafeKeyProp = R.uncurryN(2, R.pipe(safeKey, R.dissoc)); // appendSafeKey :: String -> [String] -> [String]

const appendSafeKey = R.uncurryN(2, R.pipe(safeKey, R.append)); // removeSafeKey :: String -> [String] -> [String]

const removeSafeKey = (elem, array) => R.without([safeKey(elem)], array); // split into paragraph tags on newlines
// pSplit :: String -> [React.Element]


const pSplit = R.pipe(R.split('\n'), R.map(R.partial(DOM.p, [{}])));

const rangeStep = (step, start, stop) => {
  return R.map(n => start + step * n, R.range(0, 1 + (stop - start) / step >>> 0));
}; // createRound :: (Num -> Num) -> Num -> Num -> Num


const createRound = R.curry((f, precision, num) => {
  const p = Math.pow(10, precision);
  return f(num * p) / p;
}); // roundPrec :: Num -> Num -> Num

const roundPrec = createRound(Math.round); // this gives R.map access to collection's index like native Array.prototype.map

const mapIndexed = R.addIndex(R.map);
const reduceIndexed = R.addIndex(R.reduce);
const mergeLists = R.curry((a, b) => reduceIndexed((acc, v, i) => {
  const acc_ = acc;
  acc_[i] = v;
  return acc_;
}, a || [], b));
const fileSafeString = R.pipe(R.trim, R.toLower, R.split(/[^a-zA-Z0-9]/g), R.filter(R.prop('length')), R.join('_'));
const truncate = R.curry((suffix, length, s) => s.length > length ? s.substring(0, length) + suffix : s);

const escapeRegExp = string => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

module.exports = {
  checkJpeg,
  getPathForIsvg,
  fullIconPath,
  botTypeToGender,
  genderToCodeName,
  printerIDtoSerial,
  connTypeToString,
  trackingInfoForPrinter,
  sendPrinterAnalytics,
  progress,
  getMaterialPrettyStringFromCode,
  getMaterialSliceConfigStringFromCode,
  filamentLengthToMass,
  getTracking,
  getLogging,
  getI18n,
  getModal,
  getConfig,
  safeKey,
  getSafeKeyProp,
  setSafeKeyProp,
  removeSafeKeyProp,
  appendSafeKey,
  removeSafeKey,
  pSplit,
  rangeStep,
  roundPrec,
  mapIndexed,
  fileSafeString,
  truncate,
  escapeRegExp,
  reduceIndexed,
  mergeLists
};
