// MB-support-plugin/lib/ui/views/buttonWithDropdown.js
'use strict';

const React = require('react');

const MenuItemData = require('eagle-print/views/menu/menu_item_data');

class ButtonWithDropdown extends React.Component {
  constructor(props) {
    super(props);
  }

  openContextMenu(event) {
    event.preventDefault();
    event.stopPropagation();
    const menuItems = this.props.menuItems.map(menuItem => {
      return new MenuItemData({
        path: menuItem.path,
        label: menuItem.label,
        action: menuItem.action
      });
    });
    this.props.flux.getService('MenuService').renderContextMenu(menuItems, event.clientX, event.clientY);
  }

  getDropdown(disabled, dropdownClasses) {
    return /*#__PURE__*/React.createElement("div", {
      id: "dropdown",
      className: `buttonDropdown ${dropdownClasses}${disabled ? ' disabled' : ''}`,
      onClick: disabled ? undefined : event => this.openContextMenu(event)
    }, /*#__PURE__*/React.createElement("div", {
      className: "buttonDropdownLabel"
    }, "..."));
  }

  render() {
    const disabled = this.props.disabled;
    const classes = this.props.classes ? this.props.classes : '';
    return /*#__PURE__*/React.createElement("div", {
      className: "buttonWithDropdownContainer"
    }, /*#__PURE__*/React.createElement("div", {
      id: this.props.id,
      className: `button ${classes}`,
      onClick: disabled ? undefined : this.props.onClick
    }, /*#__PURE__*/React.createElement("div", {
      className: "buttonLabel"
    }, this.props.label)), this.getDropdown(disabled, this.props.dropdownClasses));
  }

}

ButtonWithDropdown.displayName = 'ButtonWithDropdown';
ButtonWithDropdown.propTypes = {
  flux: React.PropTypes.object.isRequired,
  id: React.PropTypes.string.isRequired,
  label: React.PropTypes.string.isRequired,
  onClick: React.PropTypes.func.isRequired,
  disabled: React.PropTypes.bool,
  menuItems: React.PropTypes.array.isRequired,
  classes: React.PropTypes.string,
  dropdownClasses: React.PropTypes.string
};
module.exports = ButtonWithDropdown;
