// MB-support-plugin/lib/ui/views/cameraFeed.js
'use strict';

const React = require('react');

const ce = React.createElement;
const DOM = React.DOM;

const ReactDOM = require('react-dom');

const path = require('path');

const styles = require('./styles');

const PrinterStateEnum = require('../../constants').PrinterStateEnum;

const BotTypeEnum = require('../../constants').BotTypeEnum; // width:height of unrotated camera frame


const ASPECT_RATIO = 4 / 3;
module.exports = React.createClass({
  displayName: 'CameraFeed',
  propTypes: {
    printer: React.PropTypes.object.isRequired,
    props: React.PropTypes.object
  },

  cameraFrameCallback() {},

  getInitialState() {
    const self = this;
    return {
      width: 360,

      get height() {
        return this.width * 1 / ASPECT_RATIO;
      },

      get rotatedHeight() {
        return this.width * ASPECT_RATIO;
      },

      get shouldRotate() {
        return self.props.printer.getGender().botType === BotTypeEnum.z18_6;
      }

    };
  },

  componentDidMount() {
    const width = ReactDOM.findDOMNode(this).offsetWidth;
    this.setState({
      width
    });
  },

  componentWillReceiveProps(nextProps) {
    const self = this;

    if (nextProps.printer !== self.props.printer || !nextProps.printer.cameraFeedActive) {
      self.cameraFrameCallback = function (cameraImageData) {
        self.setState({
          cameraImage: cameraImageData
        });
      };

      if (nextProps.printer.isAuthenticated()) {
        nextProps.printer.getCameraFeed(function (cameraImageData) {
          self.cameraFrameCallback(cameraImageData);
        });
      }
    }
  },

  componentWillUnmount() {
    const self = this; // make this do nothing so that frames we receive between invoking
    // end camera stream and the bot actually stopping sending frames does
    // not trigger setState

    self.cameraFrameCallback = function () {};

    if (this.props.printer.isAuthenticated()) {
      this.props.printer.endCameraFeed();
    }
  },

  _getLoadingSpinner() {
    // magic numbers; should be the same as the busy printer badge...
    const radius = 50;
    const drawRadius = 41;
    const style = styles.dashedCircle(drawRadius, 4);
    return DOM.div({
      style: {
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: styles.offWhite(),
        width: this.state.width,
        height: this.state.shouldRotate ? this.state.rotatedHeight : this.state.height,
        overflow: 'hidden'
      }
    }, ce('svg', {
      style: {
        width: 2 * radius,
        height: 2 * radius
      }
    }, ce('circle', {
      cx: radius,
      cy: radius,
      r: drawRadius,
      strokeWidth: '6px',
      className: 'makerbot-grey-stroke',
      style
    })));
  },

  _getCameraImage() {
    const style = {
      width: '100%'
    };

    if (this.props.printer.getGender().botType === BotTypeEnum.sketch) {
      style.height = '100%';
    }

    if (this.state.shouldRotate) {
      style.transform = `${'rotate(-90deg) ' + 'translate('}${0.5 * (this.state.height - this.state.rotatedHeight)}px) ` + `scale(${ASPECT_RATIO})`;
    }

    return DOM.img({
      src: this.state.cameraImage,
      style
    });
  },

  _getAuthImage() {
    const botType = this.props.printer.getGender().botType;
    let imageSrc = 'images/authorize_rep5.png';

    if (botType === BotTypeEnum.mini_8) {
      imageSrc = 'images/authorize_mini.png';
    } else if (this.props.printer.isSixthGen()) {
      imageSrc = 'images/authorize_fire_e.png';
    }

    imageSrc = path.resolve(__dirname, imageSrc);
    const props = {
      id: 'printerImg',
      src: imageSrc
    };
    return /*#__PURE__*/React.createElement("img", props);
  },

  render() {
    const self = this;
    const selectedPrinter = this.props.printer;
    let printerState;
    if (selectedPrinter) printerState = selectedPrinter.getStatus().state;

    if (printerState === PrinterStateEnum.Authenticating) {
      return self._getAuthImage();
    } else if (selectedPrinter && self.state.cameraImage) {
      return DOM.div({
        style: {
          backgroundColor: 'white',
          height: self.state.shouldRotate ? self.state.rotatedHeight + 3 : self.state.height + 3,
          overflow: 'auto',
          draggable: false
        }
      }, self._getCameraImage());
    } else {
      return self._getLoadingSpinner();
    }
  }

});
