// MB-support-plugin/lib/ui/views/cloudPrintDialog.js
'use strict';

const FluxComponent = require('eagle-print/views/flux_component');

const React = require('react');

const Button = require('eagle-print/views/ui/button');

const path = require('path');

const listensToClickOutside = require('react-onclickoutside/decorator');

class CloudPrintDialog extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.flux = this.getFlux();
    this.i18n = this.flux.getService('TranslationService');
  }

  renderConfirmButton() {
    const openCloudPrint = () => {
      require('electron').remote.shell.openExternal('https://cloudprint.makerbot.com/');
    };

    return /*#__PURE__*/React.createElement(Button, {
      id: "cloudPrintButton",
      label: this.i18n.t('mb:cloudPrintModal.cloudPrintButton'),
      onClick: openCloudPrint
    });
  }

  render() {
    const cloudPrintLogo = path.resolve(__dirname, 'images/cloudPrintLogo.svg');
    const cloudPrintImage = path.resolve(__dirname, 'images/cloudPrintImage.png');
    return /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintModal"
    }, /*#__PURE__*/React.createElement("img", {
      src: cloudPrintLogo,
      alt: "Cloud Print Logo",
      className: "cloudPrintLogo"
    }), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintModalBody"
    }, /*#__PURE__*/React.createElement("img", {
      src: cloudPrintImage,
      alt: "Cloud Print Image",
      className: "cloudPrintImage"
    }), /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintText"
    }, /*#__PURE__*/React.createElement("div", null, this.i18n.t('mb:cloudPrintModal.text1')), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintTextHeader"
    }, this.i18n.t('mb:cloudPrintModal.header1')), /*#__PURE__*/React.createElement("div", null, this.i18n.t('mb:cloudPrintModal.text2')), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintTextHeader"
    }, this.i18n.t('mb:cloudPrintModal.header2')), /*#__PURE__*/React.createElement("div", null, this.i18n.t('mb:cloudPrintModal.text3')), /*#__PURE__*/React.createElement("br", null), /*#__PURE__*/React.createElement("div", {
      className: "cloudPrintTextHeader"
    }, this.i18n.t('mb:cloudPrintModal.header3')), /*#__PURE__*/React.createElement("div", null, this.i18n.t('mb:cloudPrintModal.text4')))), this.renderConfirmButton());
  }

}

module.exports = CloudPrintDialog;
