// MB-support-plugin/lib/ui/views/dualFilament.js
'use strict';

const React = require('react');

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const path = require('path');

const HeaderSelectButton = require('./headerSelectButton');

const DualFilamentPanel = React.createClass({
  displayName: "DualFilamentPanel",
  propTypes: {
    onBack: React.PropTypes.func.isRequired,
    onClose: React.PropTypes.func.isRequired,
    printer: React.PropTypes.object.isRequired,
    flux: React.PropTypes.object.isRequired,
    className: React.PropTypes.string
  },

  render() {
    const i18n = this.props.flux.getService('TranslationService');

    const props = _.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    });

    return /*#__PURE__*/React.createElement("div", props, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeader"
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerPanelBackButton",
      onClick: this.props.onBack
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    })), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelHeaderLabel"
    }, /*#__PURE__*/React.createElement("h1", null, i18n.t('mb:filament.header'))), /*#__PURE__*/React.createElement(HeaderSelectButton, {
      flux: this.props.flux,
      printer: this.props.printer
    }), /*#__PURE__*/React.createElement("div", {
      className: "printerPanelArrowDown",
      onClick: this.props.onClose
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/back-arrow.svg"
    }))), /*#__PURE__*/React.createElement("div", {
      className: "bottomList"
    }, /*#__PURE__*/React.createElement("img", {
      className: "methodFilamentImage",
      src: path.resolve(__dirname, './images/methodFilamentPrompt.png')
    }), /*#__PURE__*/React.createElement("div", {
      className: "methodFilamentText"
    }, ' ', "Tap material on the touchscreen located on the printer to load or unload material.", ' ')));
  }

});
module.exports = DualFilamentPanel;
