// MB-support-plugin/lib/ui/views/eeTermsAndConditionsDialog.js
'use strict';

const FluxComponent = require('eagle-print/views/flux_component');

const React = require('react');

const ce = React.createElement;
const DOM = React.DOM;

const Button = require('eagle-print/views/ui/button');

class EETermsAndConditionsDialog extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.i18n = this.getFlux().getService('TranslationService');
    this.loggingService = this.getFlux().getService('LoggingService');
    this.modalService = this.getFlux().getService('ModalService');
    this.configService = this.getFlux().getService('ConfigService');
  }

  close() {
    this.configService.set('eeTermsSeen', true);
    this.modalService.hide();
  }

  renderConfirmButton() {
    return ce(Button, {
      className: 'termsAcceptButton',
      label: this.i18n.t('mb:ee_terms_and_conditions.accept_button'),
      onClick: this.close.bind(this)
    });
  }

  render() {
    return DOM.div({
      className: 'terms-and-conditions-modal experimental-extruder'
    }, DOM.p({
      className: 'terms-and-conditions-text',
      dangerouslySetInnerHTML: {
        __html: this.i18n.t('mb:ee_terms_and_conditions.text')
      }
    }), DOM.div({}), this.renderConfirmButton());
  }

}

module.exports = EETermsAndConditionsDialog;
