// MB-support-plugin/lib/ui/views/estimates.js
'use strict';

var _react = _interopRequireDefault(require("react"));

var _lodash = _interopRequireDefault(require("lodash"));

var _path = _interopRequireDefault(require("path"));

var _flux_component = _interopRequireDefault(require("eagle-print/views/flux_component"));

var _printMaterialBadge = _interopRequireDefault(require("./printMaterialBadge"));

var _loadingPlaceholder = _interopRequireDefault(require("./loadingPlaceholder"));

var _uiUtils = require("./utilities/timeAndEstimates/uiUtils");

var _util = require("../../util");

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Estimates extends _flux_component.default {
  constructor(props) {
    super(props);
    this._i18n = this.props.flux.getService('TranslationService');
  }

  _isNil(value) {
    return value === null || value === undefined;
  }

  getMaterialEstimates(bayIndex = 0) {
    const materialResults = {};
    materialResults.gram = (0, _uiUtils.formatMaterialAmount)(this.props.estimateInfo.mass[bayIndex], 2);
    materialResults.distance = this.props.estimateInfo.distance[bayIndex];
    /**
     * derive stock material length for offline estimates & create supplemental text about < 6th gen spool size...
     * 6th gen: check bot type & toolpath material estimate in each bayIndex -> assign stock material length
     * in a way we can use to compare with toolpath's extrusion_distance
     * 5th gen: check if bot is a mini or standard size -> assign stock pla length
     * TODO: abstract supplemental spool text into separate function to clean up logic
     **/

    const botType = this.props.estimateInfo.botType;
    const material = this.props.estimateInfo.material[bayIndex];

    if (this.props.printer.isSixthGen()) {
      materialResults.spoolLength = _constants.StockMaterialLength.makerbot_generation6[material];
    } else if (botType === _constants.BotTypeEnum.sketch) {
      materialResults.spoolLength = _constants.StockMaterialLength.makerbot_sketch[material];
    } else if (botType === _constants.BotTypeEnum.mini_8) {
      materialResults.spoolLength = _constants.StockMaterialLength.makerbot_generation5.pla_small;

      if (materialResults.gram < 73) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.small_spool.1/3');
      } else if (materialResults.gram < 145) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.small_spool.2/3');
      } else if (materialResults.gram < 210) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.small_spool.3/3');
      } else {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.small_spool.3+/3');
      }
    } else {
      materialResults.spoolLength = _constants.StockMaterialLength.makerbot_generation5.pla_large;

      if (materialResults.gram < 302) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.large_spool.1/3');
      } else if (materialResults.gram < 604) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.large_spool.2/3');
      } else if (materialResults.gram < 909) {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.large_spool.3/3');
      } else {
        materialResults.spoolFraction = this._i18n.t('mb:estimates.material_estimates.large_spool.3+/3');
      }
    }

    return materialResults;
  }

  getTimeAMPM(date) {
    let hours = date.getHours();
    let minutes = date.getMinutes();
    const ampm = hours >= 12 ? 'pm' : 'am';
    hours = hours % 12;
    hours = hours ? hours : 12; // the hour '0' should be '12'

    minutes = minutes < 10 ? `0${minutes}` : `${minutes}`;
    const strTime = `${hours}:${minutes}`;
    return {
      strTime,
      ampm
    };
  }

  getDuration(seconds) {
    const days = Math.floor(seconds / (3600 * 24));
    seconds = seconds - days * 3600 * 24;
    const hours = Math.floor(seconds / 3600);
    seconds = seconds - hours * 3600;
    const minutes = Math.floor(seconds / 60);
    let duration = '';

    if (days > 0) {
      duration += `${days}d `;
    }

    if (hours > 0) {
      duration += `${hours}h `;
    }

    return `${duration + minutes}m`;
  }

  getTimeEstimates() {
    const timeResults = {};
    const now = new Date();
    const estimateDate = new Date(now.getTime() + 1000 * this.props.estimateInfo.time); //DAY i.e. today, tomorrow or whatever day of the week

    if (now.getDay() == estimateDate.getDay() && this.props.estimateInfo.time < 24 * 60 * 60) {
      // same day as today & less than 24 hours
      timeResults.day = this._i18n.t('mb:estimates.time_estimates.today');
    } else if (now.getDay() + 1 == estimateDate.getDay() && this.props.estimateInfo.time < 48 * 60 * 60) {
      // same day as tomorrow & less than 48 hours
      timeResults.day = this._i18n.t('mb:estimates.time_estimates.tomorrow');
    } else {
      timeResults.day = this._i18n.t(`mb:estimates.time_estimates.${estimateDate.getDay()}`);
    } //TIME hh:mm


    timeResults.time = this.getTimeAMPM(estimateDate); //DURATION d h m

    timeResults.duration = this.getDuration(this.props.estimateInfo.time);
    return timeResults;
  }

  getMaterialInfo(spool, bayIndex) {
    /**
     * Check spool for smart or dumb data -> otherwise default to 5th gen.
     * Pull material color from spool, and process that info to create needed color.
     * If no material type or color are present(dumb spool), show fallback.
     * For a 6th-gen LABS extruder, we ignore the loaded spool material, and show
     * the user-selected material.
     **/
    const hasSmartSpool = spool && spool.tag_uid !== null;

    const hasLabsExtruder = this.props.printer.getAttachedExtruders()[0] === _constants.ExtruderTypeEnum.mk14_e;

    const materialInfoResults = {};

    if (hasSmartSpool && !hasLabsExtruder) {
      materialInfoResults.type = spool.material_type ? _constants.SliceConfigStringsToPrettyStrings[spool.material_name] : _constants.SliceConfigStringsToPrettyStrings['generic_model'];
      materialInfoResults.materialColorName = spool.material_color_name !== '' ? spool.material_color_name : `Color ${_constants.MaterialDefaults.detection_error}`;
      materialInfoResults.remainingColor = (0, _uiUtils.formatMaterialColor)(spool).remainingColor;
      materialInfoResults.neededColor = (0, _uiUtils.formatMaterialColor)(spool).neededColor;
      return materialInfoResults;
    } else {
      bayIndex = !this._isNil(bayIndex) ? bayIndex : 0;
      const material = this.props.estimateInfo.material[bayIndex];
      materialInfoResults.type = _constants.SliceConfigStringsToPrettyStrings[material];
      materialInfoResults.neededColor = _constants.MaterialDefaults.colors[material];
      return materialInfoResults;
    }
  }

  constructPayload(spool, bayIndex) {
    return {
      showSpoolFraction: !this.props.spoolInfo,
      remainingMaterialAmount: spool && spool.amount_remaining >= 0 ? (0, _util.filamentLengthToMass)(spool.amount_remaining, spool.linear_density) : null,
      materialInfo: this.getMaterialInfo(spool, bayIndex),
      spool
    };
  }

  renderOneMaterialEstimate(payload, bayIndex) {
    const {
      materialInfo,
      showSpoolFraction,
      remainingMaterialAmount,
      spool
    } = payload;

    const hasLabsExtruder = this.props.printer.getAttachedExtruders()[bayIndex] === _constants.ExtruderTypeEnum.mk14_e; // Material estimate info


    const materialEstimates = this.getMaterialEstimates(bayIndex);

    const neededGram = this._i18n.t('mb:estimates.material_estimates.g', {
      g: materialEstimates.gram
    });

    const remainingGram = this._i18n.t('mb:estimates.material_estimates.g', {
      g: remainingMaterialAmount !== null ? Math.floor(remainingMaterialAmount) : null
    }); // Filament color, passed to the small filament circles


    const remainingMaterialDonutStyle = {
      borderColor: materialInfo.remainingColor
    };
    const neededMaterialDonutStyle = {
      borderColor: materialInfo.neededColor
    };

    const remainingMaterialDonut = /*#__PURE__*/_react.default.createElement("span", {
      className: "primary-donut",
      style: remainingMaterialDonutStyle
    });

    const neededMaterialDonut = /*#__PURE__*/_react.default.createElement("span", {
      className: "secondary-donut",
      style: neededMaterialDonutStyle
    });

    const materialRemainingPercentage = spool && spool.original_amount > 0 ? Math.ceil(spool.amount_remaining / spool.original_amount * 100) : null;
    const materialNeededPercentage = spool && spool.original_amount > 0 ? Math.ceil(materialEstimates.distance / spool.original_amount * 100) : Math.ceil(materialEstimates.distance / materialEstimates.spoolLength * 100); // this accounts for dumb spools where we don't have spool original_value

    const isSixthGen = this.props.printer.isSixthGen();
    const hasSmartSpool = spool && spool.tag_uid !== null;
    const materialError = spool && (spool.amount_remaining === 0 || materialNeededPercentage >= materialRemainingPercentage);
    const spoolCount = Math.ceil(materialEstimates.distance / materialEstimates.spoolLength);
    const PrintMaterialBadgeProps = {
      printer: this.props.estimateInfo.botType,
      printerIsConnected: this.props.printerIsConnected,
      materialRemainingPercentage,
      materialNeededPercentage,
      remainingMaterialColor: materialInfo.remainingColor,
      neededMaterialColor: materialInfo.neededColor,
      showMaterialError: isSixthGen && hasSmartSpool && materialError,
      showSpoolCount: spoolCount > 1 ? spoolCount : null,
      strokeWidth: 5.67,
      r: 24
    };

    const neededMaterial = /*#__PURE__*/_react.default.createElement("div", null, neededMaterialDonut, neededGram, /*#__PURE__*/_react.default.createElement("span", {
      style: {
        marginLeft: '3px'
      }
    }, this._i18n.t('mb:estimates.material_estimates.status.needed')));

    const remainingMaterial = // For a 6th-gen LABS extruder, we ignore the loaded spool
    !hasLabsExtruder && (!this._isNil(materialRemainingPercentage) || materialRemainingPercentage >= 0) ? /*#__PURE__*/_react.default.createElement("div", null, remainingMaterialDonut, remainingGram, /*#__PURE__*/_react.default.createElement("span", {
      style: {
        marginLeft: '3px'
      }
    }, this._i18n.t('mb:estimates.material_estimates.status.remaining'))) : null;

    const renderMaterialType = () => /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-title"
    }, materialInfo.type);

    const renderMaterialColorName = () => /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-title"
    }, materialInfo.materialColorName);

    const renderNeededMaterial = () => /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-subtitle"
    }, neededMaterial);

    const renderRemainingMaterial = () => /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-subtitle"
    }, hasSmartSpool && remainingMaterial ? remainingMaterial : null);

    const rendershowSpoolFraction = () => /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-subtitle"
    }, showSpoolFraction ? materialEstimates.spoolFraction : null);

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "material-info-wrapper",
      key: `materialEstimate_${bayIndex}`
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "spool"
    }, /*#__PURE__*/_react.default.createElement(_printMaterialBadge.default, PrintMaterialBadgeProps)), /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-details"
    }, renderMaterialType(), renderMaterialColorName(), renderNeededMaterial(), rendershowSpoolFraction(), renderRemainingMaterial()));
  }

  renderMaterial() {
    const spoolInfo = this.props.spoolInfo;
    const estTitle = 'mb:estimates.material_estimates.title';
    const estTitlePlural = 'mb:estimates.material_estimates.title_plural';
    let payload;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "material-estimate"
    }, /*#__PURE__*/_react.default.createElement("span", {
      className: "estimate-title-modal"
    }, spoolInfo ? this._i18n.t(estTitlePlural) : this._i18n.t(estTitle)), spoolInfo ? spoolInfo.map((spool, bayIndex) => {
      payload = this.constructPayload(spool, bayIndex);
      return this.renderOneMaterialEstimate(payload, bayIndex);
    }) : (payload = this.constructPayload(), this.renderOneMaterialEstimate(payload)));
  }

  renderTime() {
    const timeEstimates = this.getTimeEstimates();
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "time-estimate"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "ready-by"
    }, this._i18n.t('mb:estimates.time_estimates.ready_by')), /*#__PURE__*/_react.default.createElement("div", {
      className: "calendar"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "calendar-details"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "estimated-day"
    }, timeEstimates.day), /*#__PURE__*/_react.default.createElement("div", {
      className: "time"
    }, timeEstimates.time.strTime), /*#__PURE__*/_react.default.createElement("div", {
      className: "ampm"
    }, timeEstimates.time.ampm)), /*#__PURE__*/_react.default.createElement("img", {
      className: "image-calendar",
      src: _path.default.resolve(__dirname, 'images/date-time.svg')
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-details"
    }, /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-title"
    }, " ", this._i18n.t('mb:estimates.time_estimates.title')), /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-subtitle"
    }, timeEstimates.duration)));
  }

  render() {
    const isSixthGen = this.props.printer.isSixthGen(); // The estimateInfo may need time to catch up to the newly selected bot

    if (isSixthGen && (!this.props.spoolInfo || this.props.estimateInfo.mass.length < 2)) {
      return /*#__PURE__*/_react.default.createElement(_loadingPlaceholder.default, null);
    }

    return /*#__PURE__*/_react.default.createElement("div", {
      className: "estimate-widget"
    }, this.renderMaterial(), /*#__PURE__*/_react.default.createElement("div", {
      className: "mid-margin"
    }), this.renderTime());
  }

}

Estimates.displayName = 'Estimates';
Estimates.propTypes = _lodash.default.merge({}, _flux_component.default.propTypes, {
  flux: _react.default.PropTypes.object.isRequired,
  printer: _react.default.PropTypes.object.isRequired,
  estimateInfo: _react.default.PropTypes.object.isRequired //{time, mass, botType}

});
module.exports = Estimates;
