// MB-support-plugin/lib/ui/views/filament.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const path = require('path');

const HeaderSelectButton = require('./headerSelectButton');

const UtilityTab = require('./utilities/utilityTab');

const PrinterFilamentPanel = React.createClass({
  displayName: "PrinterFilamentPanel",
  propTypes: {
    onBack: React.PropTypes.func.isRequired,
    onClose: React.PropTypes.func.isRequired,
    printer: React.PropTypes.object.isRequired,
    flux: React.PropTypes.object.isRequired,
    className: React.PropTypes.string
  },

  onLoad() {
    const trackingService = this.props.flux.getService('TrackingService');

    const merged = _.merge({
      name: 'Load Filament'
    }, this.props.printer.getTrackingInfo());

    trackingService.addEvent(merged);
    this.props.printer.loadFilament().catch(err => {
      this.props.flux.getService('LoggingService').warn(err);
    });
    this.props.onBack();
  },

  onUnload() {
    const trackingService = this.props.flux.getService('TrackingService');

    const merged = _.merge({
      name: 'Unload Filament'
    }, this.props.printer.getTrackingInfo());

    trackingService.addEvent(merged);
    this.props.printer.unloadFilament().catch(err => {
      this.props.flux.getService('LoggingService').warn(err);
    });
    this.props.onBack();
  },

  render() {
    const buttons = [];
    const i18n = this.props.flux.getService('TranslationService');
    buttons.push(ce(UtilityTab, {
      key: 'unload',
      tabName: i18n.t('mb:filament.unload'),
      tabIcon: path.resolve(__dirname, 'images/unload-filament.svg'),
      onClick: this.onUnload
    }));
    buttons.push(ce(UtilityTab, {
      key: 'load',
      tabName: i18n.t('mb:filament.load'),
      tabIcon: path.resolve(__dirname, 'images/load-filament.svg'),
      onClick: this.onLoad
    }));
    return DOM.div(_.assign({}, this.props, {
      className: `${this.props.className || ''} printerPanel`
    }), DOM.div({
      className: 'printerPanelHeader'
    }, DOM.div({
      className: 'printerPanelBackButton',
      onClick: this.props.onBack
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    })), DOM.div({
      className: 'printerPanelHeaderLabel'
    }, DOM.h1({}, i18n.t('mb:filament.header'))), ce(HeaderSelectButton, {
      flux: this.props.flux,
      printer: this.props.printer
    }), DOM.div({
      className: 'printerPanelArrowDown',
      onClick: this.props.onClose
    }, ce(Isvg, {
      src: 'icons/back-arrow.svg'
    }))), DOM.div({
      className: 'bottomList'
    }, DOM.div({
      style: {
        marginBottom: 50
      }
    }, buttons)));
  }

});
module.exports = PrinterFilamentPanel;
