// MB-support-plugin/lib/ui/views/firmwareUpdateBanner.js
'use strict';

const React = require('react');

const FluxComponent = require('eagle-print/views/flux_component');

const Button = require('eagle-print/views/ui/button.js');

const Isvg = require('react-inlinesvg');

const firmwareUtils = require('../../firmwareUtils');

class FirmwareUpdateBanner extends FluxComponent {
  /** @constructor **/
  constructor(props) {
    super(props);
    this.i18n = this.getFlux().getService('TranslationService');
    this.tracking = this.getFlux().getService('TrackingService');
    Object.assign(this.state, {
      show: this.props.printer.firmware && this.props.printer.firmware.hasUpdate() && !this.props.printer.firmware.updateInitiated
    });
  }

  componentDidMount() {
    if (this.props.printer.firmware) {
      this.props.printer.firmware.listen(this._handleNotif.bind(this));

      this._handleNotif();
    }
  }

  componentWillUnmount() {
    if (this.props.printer.firmware) {
      this.props.printer.firmware.remove(this._handleNotif.bind(this));
    }
  }
  /**
   * Updates the printer with either the latest version the printer knows about, or with
   * a custom firmware file that the user chooses.
   */


  _handleUpdate() {
    const flux = this.getFlux();
    firmwareUtils.updateFirmware(this.props.printer, flux);
    this.tracking.addEvent({
      name: 'firmware_update_initialized'
    });
    this.setState({
      show: false
    });
    this.props.onBack();
  }

  _handleClose() {
    this.setState({
      show: false
    });
  }

  _handleNotif() {
    this.setState({
      show: this.props.printer.firmware.hasUpdate()
    });
  }

  renderUpdate() {
    return /*#__PURE__*/React.createElement("div", {
      className: "message-list",
      style: styles.container
    }, /*#__PURE__*/React.createElement("div", {
      className: "progress-message-block warning",
      style: styles.banner
    }, /*#__PURE__*/React.createElement("label", {
      style: {
        marginLeft: 13,
        marginTop: 2
      }
    }, this.i18n.t('mb:firmware.available')), /*#__PURE__*/React.createElement(Button, {
      style: {
        marginLeft: 38,
        padding: '7 17'
      },
      type: "warning",
      onClick: () => this._handleUpdate(),
      ref: "updateButton"
    }, /*#__PURE__*/React.createElement("span", null, this.i18n.t('mb:firmware.update'))), /*#__PURE__*/React.createElement("div", {
      className: "closeButton closePanel",
      style: {
        marginRight: 4
      },
      onClick: this._handleClose.bind(this),
      ref: "close"
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/x-button.svg"
    }))));
  }

  render() {
    if (this.props.printer.firmware && this.state.show && !this.props.printer.firmware.updateInitiated) {
      return this.renderUpdate();
    }

    return /*#__PURE__*/React.createElement("div", null);
  }

}

FirmwareUpdateBanner.propTypes = Object.assign({}, FluxComponent.propTypes, {
  printer: React.PropTypes.object.isRequired,
  onBack: React.PropTypes.func
});
FirmwareUpdateBanner.defaultProps = {
  onBack: () => {}
};
const styles = {
  container: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0
  },
  banner: {
    padding: 10,
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between'
  }
};
FirmwareUpdateBanner.displayName = 'FirmwareUpdateBanner';
module.exports = FirmwareUpdateBanner;
