// MB-support-plugin/lib/ui/views/headerSelectButton.js
'use strict';

const React = require('react');

const DOM = React.DOM;

const FluxComponent = require('eagle-print/views/flux_component');

class HeaderSelectButton extends FluxComponent {
  onSelectClicked() {
    const flux = this.getFlux();
    const project = flux.store('ProjectStore').getProject();
    flux.actions.printer.selectPrinter(project, this.props.printer);
  }

  render() {
    const flux = this.getFlux();
    const i18n = flux.getService('TranslationService');
    const selectedPrinter = flux.store('PrinterStore').getSelectedPrinter();
    const isSelectedPrinter = selectedPrinter && selectedPrinter.getId() === this.props.printer.getId();
    return DOM.div({
      id: 'selectButtonContainer'
    }, DOM.div({
      id: 'selectButton',
      className: isSelectedPrinter ? 'disabled' : undefined,
      onClick: isSelectedPrinter ? undefined : () => this.onSelectClicked()
    }, isSelectedPrinter ? i18n.t('printer_details.selected') : i18n.t('printer_details.select')));
  }

}

HeaderSelectButton.displayName = 'HeaderSelectButton';
HeaderSelectButton.propTypes = {
  printer: React.PropTypes.shape({
    getId: React.PropTypes.func.isRequired
  }).isRequired
};
module.exports = HeaderSelectButton;
