// MB-support-plugin/lib/ui/views/mb_account_sidebar_widget.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const React = require('react');

const Isvg = require('react-inlinesvg');

const _ = require('lodash');

const path = require('path');

const listensToClickOutside = require('react-onclickoutside/decorator');

const PluginComponent = require('eagle-print').views.PluginComponent;

const pluginName = require('../../../index').pluginName;

class MBAccountSidebarWidget extends PluginComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      showUserPanel: false
    });

    _defineProperty(this, "getPluginNames", () => {
      return [pluginName];
    });

    _defineProperty(this, "getStateFromFlux", () => {
      const plugin = this.getFlux().getPlugin(pluginName);
      const userInfo = plugin.getUserInfo();
      let showUserPanel = this.state ? this.state.showUserPanel : undefined;

      if (!userInfo) {
        showUserPanel = false;
      }

      return {
        userInfo,
        showUserPanel
      };
    });

    _defineProperty(this, "componentDidMount", () => {
      super.componentDidMount();
      const plugin = this.getFlux().getPlugin(pluginName);
      const userInfo = plugin.getUserInfo();
      this.state.userInfo = userInfo;
    });

    _defineProperty(this, "handleClickOutside", event => {
      if (this.state.showUserPanel) {
        this.setState({
          showUserPanel: false
        });
      }
    });

    _defineProperty(this, "handleIconClick", () => {
      if (this.state.userInfo) {
        // TODO Add code for displaying slide out panel
        this.setState({
          showUserPanel: !this.state.showUserPanel
        });
      } else {
        this._renderLoginModal();
      }
    });

    _defineProperty(this, "handleLogOut", () => {
      const flux = this.getFlux();
      const plugin = flux.getPlugin(pluginName);
      plugin.remove_auth_info();
      if (plugin.shouldPromptForLogin()) this._renderLoginModal(true);
    });

    _defineProperty(this, "_renderUserInfoPopup", () => {
      const i18n = this.getFlux().getService('TranslationService');
      const plugin = this.getFlux().getPlugin(pluginName);
      const userInfo = plugin.getUserInfo();
      const userName = userInfo.full_name.trim().length > 0 ? userInfo.full_name : userInfo.name;
      return /*#__PURE__*/React.createElement("div", {
        ref: "popout",
        className: "accountPanel"
      }, /*#__PURE__*/React.createElement("div", this.props, /*#__PURE__*/React.createElement("div", {
        className: "closeButton closePanel",
        onClick: this.handleIconClick
      }, /*#__PURE__*/React.createElement(Isvg, {
        src: "icons/x-button.svg"
      })), /*#__PURE__*/React.createElement("div", {
        className: "userContainer"
      }, /*#__PURE__*/React.createElement("img", {
        src: this.state.userInfo.userAvatar,
        className: "avatar"
      }), /*#__PURE__*/React.createElement("span", {
        className: "userInfo"
      }, /*#__PURE__*/React.createElement("span", {
        className: "header"
      }, i18n.t('mb:accountSignIn.loggedInAs')), /*#__PURE__*/React.createElement("div", {
        className: "userName"
      }, userName))), /*#__PURE__*/React.createElement("div", {
        className: "buttonGroup"
      }, /*#__PURE__*/React.createElement("button", {
        ref: "logoutButton",
        className: "button secondary",
        onClick: this.handleLogOut
      }, i18n.t('mb:accountSignIn.logout')))));
    });

    _defineProperty(this, "_renderLoginModal", isMandatory => {
      const flux = this.getFlux();
      const mbPlugin = flux.getPlugin(pluginName);
      mbPlugin.showLoginPopup(this.getFlux(), isMandatory);
    });

    _defineProperty(this, "render", () => {
      const plugin = this.getFlux().getPlugin(pluginName);
      const logger = this.getFlux().getService('LoggingService');
      plugin.initializeUserInfoForPendo().then(result => {
        if (result) {
          pendo.loadGuides(); // Attempting to reload guides

          logger.info(`PENDO::Pendo configured with Visitor ID: ${pendo.getVisitorId()}`);
        }
      });
      const isLoginDisabled = this.getFlux().getService('ConfigService').get('disableLogin', false);

      if (!isLoginDisabled) {
        const avatarSrc = this.state.userInfo ? path.resolve(__dirname, this.state.userInfo.userAvatar) : path.resolve(__dirname, 'images/nav-login-icon.svg');
        return /*#__PURE__*/React.createElement("div", _.assign({}, this.props, {
          className: `${this.props.className || ''} item account`,
          key: 'UserPanel'
        }), /*#__PURE__*/React.createElement("img", {
          src: avatarSrc,
          ref: "accountWidget",
          className: "loginButton",
          onClick: this.handleIconClick
        }), this.state.showUserPanel && this.state.userInfo ? this._renderUserInfoPopup() : null);
      }

      return null;
    });
  }

}

MBAccountSidebarWidget.displayName = 'MBAccountSidebarWidget';
module.exports = listensToClickOutside(MBAccountSidebarWidget);
