// MB-support-plugin/lib/ui/views/mb_account_login_modal.js
'use strict';

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const FluxComponent = require('eagle-print/views/flux_component');

const React = require('react');

const path = require('path');

const Isvg = require('react-inlinesvg');

const {
  ipcRenderer
} = require('electron');

const pluginName = require('../../../index').pluginName;

const ErrorsEnum = require('../../thingiverse-api-errors').ErrorsEnum;

const ProxySettingsModal = require('./proxySettingsModal');

class MBAccountLoginModal extends FluxComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "state", {
      username: this.props.email || '',
      password: '',
      inProgress: false,
      showProxy: false,

      /**
       * @type {undefined|string}
       */
      lastError: undefined,
      // 'This is a test';
      ssoTimeoutId: undefined
    });

    _defineProperty(this, "trackingService", this.getFlux().getService('TrackingService'));

    _defineProperty(this, "componentWillMount", () => {
      this.getFlux().getService('MenuService').setDisabled(true);
      super.componentWillMount();
      ipcRenderer.on('received-url', this.handleLoginOauthPKCE);
    });

    _defineProperty(this, "componentWillUnmount", () => {
      super.componentWillUnmount();
      this.getFlux().getService('MenuService').setDisabled(false);
      ipcRenderer.removeAllListeners(['received-url']);
    });

    _defineProperty(this, "componentDidMount", () => {
      super.componentDidMount();
      this.refs.emailInput.select();
    });

    _defineProperty(this, "componentDidUpdate", (prevProps, prevState) => {
      // When changing from a non-error state -> error state, focus the username input (i.e. failed login)
      if (this.state.lastError && !prevState.lastError) {
        this.refs.emailInput.select();
      }
    });

    _defineProperty(this, "handleUsernameChange", e => {
      this.setState({
        username: e.target.value
      });
    });

    _defineProperty(this, "handlePasswordChange", e => {
      this.setState({
        password: e.target.value
      });
    });

    _defineProperty(this, "authenticate", (username, password) => {
      const plugin = this.getFlux().getPlugin(pluginName);
      return plugin.authenticate(username, password).then(thingiverseToken => {
        // save out the auth info...this function should eventually save it to the json file
        return plugin.save_auth_info(username, thingiverseToken).then(userInfo => {
          return userInfo;
        });
      });
    });

    _defineProperty(this, "handleLogin", () => {
      const flux = this.getFlux();
      this.setState({
        lastError: undefined
      });

      if (this.validateCredentials()) {
        this.setState({
          inProgress: true
        });
        const log = flux.getService('LoggingService');
        log.debug('Thingiverse sign in request for: ', this.state.username);
        const authenticationPromise = this.authenticate(this.state.username, this.state.password);
        authenticationPromise.then(userInfo => {
          log.info('authenticate success');
          flux.getService('ModalService').hide();
          return userInfo;
        }).catch(err => {
          const i18n = flux.getService('TranslationService');
          let errorMessage;
          this.trackingService.addEvent({
            name: 'Failed Log In'
          });
          log.info('authenticate failed');

          if (err.errorEnum === ErrorsEnum.AccessDenied) {
            log.warn('Login failed because of invalid credentials');
            errorMessage = i18n.t('mb:accountSignIn.mismatch_login_error');
          } else if (err.errorEnum === ErrorsEnum.ConnectionError) {
            log.warn('Login failed because of connection error');
            errorMessage = i18n.t('mb:accountSignIn.connection_login_error');
          } else if (err.errorEnum === ErrorsEnum.MethodNotAllowed) {
            log.warn('Login failed');
            errorMessage = i18n.t('mb:accountSignIn.unknown_login_error');
          } else if (err.errorEnum === ErrorsEnum.FileNotFound) {
            log.warn('File Not Found');
            errorMessage = i18n.t('mb:accountSignIn.not_found');
          } else if (err.errorEnum === ErrorsEnum.ServerError) {
            log.warn('Server Error');
            errorMessage = i18n.t('mb:accountSignIn.server_error');
          } else if (err.errorEnum === ErrorsEnum.ClientError) {
            log.warn('Client Error');
            errorMessage = i18n.t('mb:accountSignIn.client_error');
          } else {
            log.error('Login failed for unknown reason', err);
            errorMessage = i18n.t('mb:accountSignIn.unknown_login_error');
          }

          this.setState({
            inProgress: false,
            lastError: errorMessage
          });

          if (this.props.onAuthenticationError) {
            this.props.onAuthenticationError();
          }
        });
      }
    });

    _defineProperty(this, "handleLoginOauthPKCE", (_event, data) => {
      const mbPlugin = this.getFlux().getPlugin(pluginName);
      const i18n = this.getFlux().getService('TranslationService');
      const ModalService = this.getFlux().getService('ModalService');
      const LoggingService = this.getFlux().getService('LoggingService');
      const urlParams = new URLSearchParams(data);
      const codeValidationRegex = /^[a-zA-Z0-9]{16}$/;
      const code = urlParams.get('code');
      const error = urlParams.get('error');
      let accessToken = '';

      if (error) {
        LoggingService.error('Server Error');
        this.setState({
          inProgress: false,
          lastError: error
        });
        this.props.onAuthenticationError();
      } else {
        // validate the 'code' query param
        if (codeValidationRegex.test(code)) {
          mbPlugin // Request authorization token to MB-Services (tv/reflector)
          ._loginOauthPKCE(code) // Request user info given the access token
          .then(jsonRes => {
            LoggingService.info('authenticate success');
            accessToken = jsonRes.access_token;
            return mbPlugin._thingiverseApiClient.getUser(accessToken);
          }) // Update userinfo and access token within the app
          .then(info => {
            LoggingService.info('get user info success');
            const {
              name
            } = info;
            return mbPlugin.save_auth_info(name, accessToken);
          }) // Close login modal
          .then(() => {
            ModalService.hide();
          }).catch(e => {
            LoggingService.error('Login failed during the Oauth process', e);
            this.setState({
              inProgress: false,
              lastError: i18n.t('mb:accountSignIn.client_error')
            });
            this.props.onAuthenticationError();
          });
        } else {
          LoggingService.error('Unrecognized Code Value Error');
          this.setState({
            inProgress: false,
            lastError: i18n.t('mb:accountSignIn.client_error')
          });
          this.props.onAuthenticationError();
        }
      } // Clear the timeout with the login modal incase
      // if client runs into errors


      const {
        ssoTimeoutId
      } = this.state;
      ssoTimeoutId ? clearTimeout(ssoTimeoutId) : null;
      this.setState({
        ssoTimeoutId: undefined
      });
    });

    _defineProperty(this, "handleKeyDown", e => {
      this.setState({
        lastError: undefined
      });

      if (e.keyCode === 13) {
        this.handleLogin();
      }
    });

    _defineProperty(this, "validateCredentials", () => {
      return this.state.username.length > 0 && this.state.password.length > 0;
    });

    _defineProperty(this, "handleForgottenPassword", () => {
      this.trackingService.addEvent({
        name: 'Forgot Password Hit'
      });
      const forgottenPasswordURL = 'https://accounts.makerbot.com/forgot';

      require('electron').remote.shell.openExternal(forgottenPasswordURL);
    });

    _defineProperty(this, "handleSignup", e => {
      this.trackingService.addEvent({
        name: 'Sign Up Button Hit'
      });
      const signupURL = 'https://accounts.makerbot.com/register?active=1&theme=store&utm_source=mbp&utm_campaign=createaccount';

      require('electron').remote.shell.openExternal(signupURL);
    });

    _defineProperty(this, "handleTerms", e => {
      this.trackingService.addEvent({
        name: 'Terms of Service Hit'
      });
      const signupURL = 'http://www.makerbot.com/legal/terms';

      require('electron').remote.shell.openExternal(signupURL);
    });

    _defineProperty(this, "handlePrivacyPolicy", e => {
      this.trackingService.addEvent({
        name: 'Privacy Policy Hit'
      });
      const signupURL = 'http://www.makerbot.com/legal/privacy';

      require('electron').remote.shell.openExternal(signupURL);
    });

    _defineProperty(this, "handleGoogleSso", () => {
      const i18n = this.getFlux().getService('TranslationService');
      const plugin = this.getFlux().getPlugin(pluginName);

      const signInUrl = plugin._generateOauthAuthReqUrl();

      this.trackingService.addEvent({
        name: 'Google SSO Hit'
      });
      this.setState({
        inProgress: true,
        lastError: undefined
      }); // Set timeout with Google SSO

      const ssoTimeoutId = setTimeout(() => {
        if (this.refs.loginButton) {
          this.setState({
            inProgress: false,
            lastError: i18n.t('mb:accountSignIn.timed_out')
          });
          this.props.onAuthenticationError();
        }
      }, 2 * 60 * 1000);
      this.setState({
        ssoTimeoutId
      });

      require('electron').remote.shell.openExternal(signInUrl);
    });

    _defineProperty(this, "getErrorDiv", () => {
      const width = 25,
            height = 25,
            viewBox = `0 0 ${width} ${height}`;
      return /*#__PURE__*/React.createElement("div", {
        className: "error"
      }, this.state.lastError ? /*#__PURE__*/React.createElement("svg", {
        ref: "errorIcon",
        className: "error-icon",
        width: width,
        height: height,
        viewBox: viewBox
      }, /*#__PURE__*/React.createElement("image", {
        xlinkHref: path.resolve(__dirname, 'images/alert-icon-white.svg'),
        height: height,
        width: width
      })) : null, this.state.lastError ? /*#__PURE__*/React.createElement("span", {
        className: "error-text"
      }, " ", this.state.lastError, " ") : null);
    });

    _defineProperty(this, "renderUsernameField", i18n => {
      return /*#__PURE__*/React.createElement("div", {
        className: `form-group auth-input ${this.state.lastError ? 'has-error' : ''}`
      }, /*#__PURE__*/React.createElement("label", null, (null, i18n.t('mb:accountSignIn.username'))), /*#__PURE__*/React.createElement("input", {
        ref: "emailInput",
        className: `input-text ${this.state.lastError ? 'has-error' : ''}`,
        onChange: this.handleUsernameChange,
        onKeyDown: this.handleKeyDown,
        disabled: this.state.inProgress,
        placeholder: i18n.t('mb:accountSignIn.username'),
        keyboardType: "email-address",
        autoFocus: true
      }));
    });

    _defineProperty(this, "renderPasswordField", i18n => {
      return /*#__PURE__*/React.createElement("div", {
        className: "form-group auth-input "
      }, /*#__PURE__*/React.createElement("label", null, (null, i18n.t('mb:accountSignIn.password'))), /*#__PURE__*/React.createElement("a", {
        onClick: this.handleForgottenPassword,
        className: "forgot-pwd-link"
      }, i18n.t('mb:accountSignIn.forgot_password')), /*#__PURE__*/React.createElement("input", {
        ref: "passwordInput",
        type: "password",
        className: `input-text ${this.state.lastError ? 'has-error' : ''}`,
        onChange: this.handlePasswordChange,
        onKeyDown: this.handleKeyDown,
        placeholder: i18n.t('mb:accountSignIn.password'),
        disabled: this.state.inProgress
      }));
    });

    _defineProperty(this, "renderLinks", i18n => {
      return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("label", null, (null, i18n.t('mb:accountSignIn.signing_terms_header'))), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("a", {
        onClick: this.handleTerms
      }, i18n.t('mb:accountSignIn.terms_service')), /*#__PURE__*/React.createElement("label", null, (null, ' • ')), /*#__PURE__*/React.createElement("a", {
        onClick: this.handlePrivacyPolicy
      }, i18n.t('mb:accountSignIn.privacy_policy'))));
    });

    _defineProperty(this, "renderLoginButtonGroup", (i18n, loginButtonClass) => {
      return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
        className: "buttonGroup loginButtonGroup"
      }, /*#__PURE__*/React.createElement("button", {
        ref: "loginButton",
        className: loginButtonClass,
        onClick: this.handleLogin,
        disabled: this.state.inProgress
      }, this.state.inProgress ? /*#__PURE__*/React.createElement("div", {
        className: "refreshSpinnerWhite spinning"
      }, /*#__PURE__*/React.createElement(Isvg, {
        src: "icons/spinner-small.svg",
        key: "printerRefreshSpinner"
      })) : i18n.t('mb:accountSignIn.login')), /*#__PURE__*/React.createElement("button", {
        ref: "signupButton",
        className: "button secondary",
        onClick: this.handleSignup
      }, i18n.t('mb:accountSignIn.signup'))), this.renderGoogleSsoButtonGroup(i18n, 'buttonGroup loginButtonGroup'), /*#__PURE__*/React.createElement("div", {
        className: "buttonGroup loginButtonGroup"
      }, /*#__PURE__*/React.createElement("a", {
        onClick: () => this.setState({
          showProxy: true
        })
      }, i18n.t('mb:accountSignIn.proxy_link'))));
    });

    _defineProperty(this, "renderGoogleSsoButtonGroup", (i18n, loginButtonClass) => {
      const googleLogo = path.resolve(__dirname, 'images/google-logo.svg');
      return /*#__PURE__*/React.createElement("section", {
        id: "google-sso",
        className: loginButtonClass
      }, /*#__PURE__*/React.createElement("div", {
        className: "sso-separator"
      }, "or"), /*#__PURE__*/React.createElement("button", {
        id: "google-sso-btn",
        onClick: this.handleGoogleSso,
        disabled: this.state.inProgress
      }, /*#__PURE__*/React.createElement("div", {
        className: "google-sso-btn-content"
      }, /*#__PURE__*/React.createElement("img", {
        src: googleLogo,
        alt: "Google Logo",
        width: "25px"
      }), /*#__PURE__*/React.createElement("span", {
        className: "sso-text"
      }, 'Sign in with Google'))));
    });

    _defineProperty(this, "renderLogin", (i18n, loginButtonClass) => {
      const makerbotLogo = path.resolve(__dirname, 'images/mb_sign_in_logo.svg');
      return /*#__PURE__*/React.createElement("div", this.props, /*#__PURE__*/React.createElement("div", {
        className: "mbLoginModal"
      }, /*#__PURE__*/React.createElement("div", {
        className: "print-logo"
      }, /*#__PURE__*/React.createElement("img", {
        src: makerbotLogo
      }), /*#__PURE__*/React.createElement("label", null, (null, i18n.t('mb:accountSignIn.sign_in')))), this.getErrorDiv(), this.renderUsernameField(i18n), this.renderPasswordField(i18n), this.renderLinks(i18n), this.renderLoginButtonGroup(i18n, loginButtonClass)));
    });

    _defineProperty(this, "render", () => {
      const i18n = this.getFlux().getService('TranslationService');
      let loginButtonClass = 'button primary loginButton';

      if (this.state.showProxy) {
        return /*#__PURE__*/React.createElement(ProxySettingsModal, {
          onDone: () => this.setState({
            showProxy: false
          }),
          flux: this.getFlux()
        });
      }

      if (this.state.inProgress || !this.validateCredentials()) {
        loginButtonClass += ' disabled1';
      }

      return this.renderLogin(i18n, loginButtonClass);
    });
  }

}

MBAccountLoginModal.displayName = 'MBAccountLoginModal';
MBAccountLoginModal.propTypes = {
  onAuthenticationError: React.PropTypes.func
};
module.exports = MBAccountLoginModal;
