// MB-support-plugin/lib/ui/views/mbLogosPanel.js
'use strict';

const React = require('react');

const Isvg = require('react-inlinesvg');

const path = require('path');

const Tooltip = require('eagle-print/views/util/tooltip');

const util = require('../../util');

const StoreSidePanel = require('./storeSidePanel');

class MBLogosPanel extends React.Component {
  constructor(props) {
    super(props);
    this._handleThingiverseLink = this._handleThingiverseLink.bind(this);
    this._handleMakerBotLink = this._handleMakerBotLink.bind(this);
    this._handleMakerBotStoreLink = this._handleMakerBotStoreLink.bind(this);
    this._handleSupportLink = this._handleSupportLink.bind(this);
    this.i18n = this.props.flux.getService('TranslationService');
    this.state = {
      showMakerBotStorePanel: false
    };
  }

  _handleThingiverseLink() {
    const thingiverseLink = 'https://thingiverse.com/?utm_source=mbp&utm_campaign=tvlogo';
    util.sendPrinterAnalytics(this.props.flux, 'Sidebar Link Clicked Thingiverse');

    require('electron').remote.shell.openExternal(thingiverseLink);
  }

  _handleMakerBotLink() {
    const makerbotLink = 'https://makerbot.com/?utm_source=mbp&utm_campaign=mblogo';
    util.sendPrinterAnalytics(this.props.flux, 'Sidebar Link Clicked Makerbot');

    require('electron').remote.shell.openExternal(makerbotLink);
  }

  _handleMakerBotStoreLink() {
    // Send analytics only from side panel icon click
    this.props && util.sendPrinterAnalytics(this.props.flux, 'Sidebar Link Clicked Store'); // Expand StoreSidePanel Component

    this.setState({
      showMakerBotStorePanel: !this.state.showMakerBotStorePanel
    });
  }

  _handleSupportLink() {
    const makerbotSupportLink = 'https://support.makerbot.com/?utm_source=mbp&utm_campaign=mbsupport';
    util.sendPrinterAnalytics(this.props.flux, 'Sidebar Link Clicked Support');

    require('electron').remote.shell.openExternal(makerbotSupportLink);
  }

  renderIcon(link) {
    const {
      id,
      onClickHandler,
      iconSrcPath,
      contentTxt
    } = link;
    const icon = /*#__PURE__*/React.createElement("div", {
      className: "item logo"
    }, /*#__PURE__*/React.createElement("a", {
      onClick: onClickHandler
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: util.getPathForIsvg(iconSrcPath)
    })));
    return /*#__PURE__*/React.createElement(Tooltip, {
      key: id,
      id: id,
      addTo: icon,
      placement: "right",
      content: contentTxt
    });
  }

  render() {
    const links = [{
      id: 'makerbotTooltip',
      onClickHandler: this._handleMakerBotLink,
      iconSrcPath: path.resolve(__dirname, 'images/nav-makerbot.svg'),
      contentTxt: this.i18n.t('mb:sidePanel.icons.makerbotIcon.context')
    }, {
      id: 'thingiverseTooltip',
      onClickHandler: this._handleThingiverseLink,
      iconSrcPath: path.resolve(__dirname, 'images/nav-thingiverse.svg'),
      contentTxt: this.i18n.t('mb:sidePanel.icons.thingiverseIcon.context')
    }, {
      id: 'makerbotStoreTooltip',
      onClickHandler: this._handleMakerBotStoreLink,
      iconSrcPath: path.resolve(__dirname, 'images/nav-logo-shop.svg'),
      contentTxt: this.i18n.t('mb:sidePanel.icons.storeIcon.context')
    }, {
      id: 'makerbotSupportTooltip',
      onClickHandler: this._handleSupportLink,
      iconSrcPath: path.resolve(__dirname, 'images/help-icon.svg'),
      contentTxt: this.i18n.t('mb:sidePanel.icons.supportIcon.context')
    }];
    return /*#__PURE__*/React.createElement("div", {
      className: "leftPanel"
    }, links.map(link => this.renderIcon(link)), this.state.showMakerBotStorePanel ? /*#__PURE__*/React.createElement(StoreSidePanel, {
      handleIconClick: this._handleMakerBotStoreLink,
      showPanel: this.state.showMakerBotStorePanel,
      i18n: this.i18n
    }) : null);
  }

}

MBLogosPanel.displayName = 'MBLogosPanel';
MBLogosPanel.propTypes = {
  flux: React.PropTypes.object.isRequired
};
module.exports = MBLogosPanel;
