// MB-support-plugin/lib/ui/views/printMaterialBadge.js
"use strict";

var _react = _interopRequireDefault(require("react"));

var _path = _interopRequireDefault(require("path"));

var _flux_component = _interopRequireDefault(require("eagle-print/views/flux_component"));

var _styles = require("./styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PrintMaterialBadge extends _flux_component.default {
  constructor(props) {
    super(props);
  }

  _realRadius() {
    // pull them out to make code formatting nicer
    const {
      r,
      strokeWidth
    } = this.props;
    return r - (strokeWidth + strokeWidth / 2);
  }

  getRemainingMaterial() {
    const dashArray = this._realRadius() * Math.PI * 2;
    const dashOffset = dashArray - dashArray * this.props.materialRemainingPercentage / 100;
    const remainingStyle = Object.assign({}, (0, _styles.roundedCapCircle)(), {
      stroke: this.props.remainingMaterialColor,
      strokeWidth: this.props.strokeWidth,
      strokeDasharray: dashArray,
      strokeDashoffset: dashOffset
    });
    return /*#__PURE__*/_react.default.createElement("circle", {
      style: remainingStyle,
      cx: this.props.r,
      cy: this.props.r,
      r: this._realRadius()
    });
  } // this will stop the needed material display stroke from rotating beyond a full spool (aka 100%)


  materialDisplaySafeguard() {
    return this.props.materialNeededPercentage >= 100 ? 100 : this.props.materialNeededPercentage;
  }

  getNeededMaterial() {
    const dashArray = this._realRadius() * Math.PI * 2;
    const dashOffset = dashArray - dashArray * this.materialDisplaySafeguard() / 100; // const test = dashArray - (dashArray * this.props.materialRemainingPercentage) / 100;

    /* manipulating the transform property with a rudimentary calculation in the interest of time. this keeps the needed 
    display parked on the end of the remaining display while factoring in the diff so it doesn't extend beyond the remaining stroke. 
    TODO: find a smarter way. */

    const neededStyle = Object.assign({}, (0, _styles.roundedCapCircle)(), {
      stroke: this.props.neededMaterialColor,
      strokeWidth: this.props.strokeWidth,
      // transform: `rotate(${test}deg)`,
      strokeDasharray: dashArray,
      strokeDashoffset: dashOffset
    });
    return /*#__PURE__*/_react.default.createElement("circle", {
      style: neededStyle,
      cx: this.props.r,
      cy: this.props.r,
      r: this._realRadius()
    });
  }

  getFilamentError() {
    return this.props.showMaterialError ? /*#__PURE__*/_react.default.createElement("img", {
      className: "filament-error",
      src: _path.default.resolve(__dirname, 'images/error-triangle.svg')
    }) : null;
  }

  getSpoolCount() {
    return /*#__PURE__*/_react.default.createElement("span", {
      className: "spool-count"
    }, this.props.showSpoolCount);
  }

  getBackgroundCircle() {
    const style = {
      strokeWidth: '1px',
      fill: 'none',
      stroke: '#C4C4C4'
    };
    return /*#__PURE__*/_react.default.createElement("circle", {
      style: style,
      cx: this.props.r,
      cy: this.props.r,
      r: this.props.r - 1
    });
  }

  render() {
    const width = this.props.r * 2;
    const height = this.props.r * 2;
    const viewBox = `0 0 ${width} ${height}`;
    return /*#__PURE__*/_react.default.createElement("svg", Object.assign({}, this.props, {
      className: 'CircularProgress',
      width,
      height,
      viewBox
    }), this.getBackgroundCircle(), this.getRemainingMaterial(), this.getNeededMaterial(), this.getFilamentError(), this.getSpoolCount());
  }

}

PrintMaterialBadge.propTypes = Object.assign({}, PrintMaterialBadge.propTypes, {
  printer: _react.default.PropTypes.string.isRequired,
  strokeWidth: _react.default.PropTypes.number,
  r: _react.default.PropTypes.number
});
module.exports = PrintMaterialBadge;
