// MB-support-plugin/lib/ui/views/printButton.js
'use strict';

const React = require('react');

const _ = require('lodash');

const path = require('path');

const analyticUtils = require('eagle-print/util/analyticUtils');

const MenuItemData = require('eagle-print/views/menu/menu_item_data');

const {
  ExtruderTypeEnum
} = require('../../constants');

const util = require('../../util');

const UnsupportedExtruderConfigurationsDialog = require('./unsupportedExtruderConfigurationsDialog');

class PrintButton extends React.Component {
  constructor(props) {
    super(props);
    this._boundOnClick = this.onClick.bind(this);
    this._exportPrint = this.exportPrint.bind(this);
  }

  exportPrint() {
    const remote = require('electron').remote;

    const dialog = remote.dialog;
    const self = this;
    const filters = this.props.flux.getService('ExportService').getSupportedFormats(this.props.printer);
    const defaultPath = localStorage.getItem('lastExportDialogPath');
    const i18n = this.props.flux.getService('TranslationService');
    const spoolMaterials = this.props.flux.store('PrintSettingsStore').getState().materials.spoolMaterials;
    const attachedExtruders = this.props.printer.getAttachedExtruders();
    const hasExtruderMaterialMismatch = this.props.printer.hasExtruderMaterialMismatch();
    const {
      // extruderSetupError pretains to the extruder loadout on
      // the carriage; missing, mismatched, or unsupported
      extruderSetupError,
      perExtruderErrorType,
      supportedConfigs
    } = this.props.printer.checkForUnsupportedExtruderConfigurations() ? this.props.printer.checkForUnsupportedExtruderConfigurations() : {};

    if (extruderSetupError) {
      const props = {
        machine_name: this.props.printer.getName(),
        perExtruderErrorType,
        supportedConfigs,
        i18n
      };
      return this.showExtruderSetupErrorWarning(props);
    } else if (hasExtruderMaterialMismatch) {
      this.showExtruderMaterialMismatchWarning();
    } else if ((spoolMaterials.includes(null) || spoolMaterials.includes(undefined)) && !attachedExtruders.includes(ExtruderTypeEnum.mk14_e)) {
      this.showNonLabsMaterialWarning();
    } else {
      const fileName = dialog.showSaveDialog(remote.getCurrentWindow(), {
        title: self.props.i18n('mb:printButton.export'),
        filters: [filters],
        defaultPath
      });

      if (fileName) {
        self.props.onClick(fileName);

        if (defaultPath) {
          localStorage.setItem('lastExportDialogPath', path.dirname(fileName));
        }
      }
    }
  }

  canPrint() {
    const status = this.props.printer.getStatus();
    return status && status.state && status.state.toLowerCase() !== 'offline' && status.state.toLowerCase() !== 'unauthenticated';
  }

  openContextMenu(event) {
    event.preventDefault();
    event.stopPropagation();
    const menuItem = [new MenuItemData({
      path: 'export',
      label: 'Export',
      action: () => this.exportPrint()
    })];
    this.props.flux.getService('MenuService').renderContextMenu(menuItem, event.clientX, event.clientY);
  }

  onClick() {
    const materials = this.props.flux.store('PrintSettingsStore').getState().materials; // Gotta put this null check in for the grunt tests. Ah well.

    const spoolMaterials = materials ? materials.spoolMaterials : null;
    const attachedExtruders = this.props.printer.getAttachedExtruders();
    const ModalService = this.props.flux.getService('ModalService');

    if ((spoolMaterials.includes(null) || spoolMaterials.includes(undefined)) && !attachedExtruders.includes(ExtruderTypeEnum.mk14_e)) {
      this.showNonLabsMaterialWarning();
    } else {
      this.handlePrintButtonClick();
    }
  }

  handlePrintButtonClick() {
    const printer = this.props.printer;
    const trackingService = this.props.flux.getService('TrackingService');
    const i18n = this.props.flux.getService('TranslationService');

    if (this.canPrint()) {
      const trackingInfo = util.trackingInfoForPrinter(printer);
      trackingInfo.extruder_type = printer.getCurrentExtruderType();
      const printSettings = analyticUtils.getPrintSettingsForAnalyticEvents(this.props.flux, printer);
      trackingService.addEvent(_.merge({
        name: 'Print Started'
      }, trackingInfo, printSettings));
      const {
        // extruderSetupError pretains to the extruder loadout on
        // the carriage; missing, mismatched, or unsupported
        extruderSetupError,
        perExtruderErrorType,
        supportedConfigs
      } = printer.checkForUnsupportedExtruderConfigurations() ? printer.checkForUnsupportedExtruderConfigurations() : {};
      const hasExtruderMaterialMismatch = printer.hasExtruderMaterialMismatch();

      if (extruderSetupError) {
        const props = {
          machine_name: printer.getName(),
          perExtruderErrorType,
          supportedConfigs,
          i18n
        };
        return this.showExtruderSetupErrorWarning(props);
      }

      if (hasExtruderMaterialMismatch) {
        return this.showExtruderMaterialMismatchWarning();
      }

      return this.props.onClick(undefined, true);
    } else {
      return this.exportPrint();
    }
  }

  showExtruderMaterialMismatchWarning() {
    const modalService = this.props.flux.getService('ModalService');
    return modalService.confirm({
      title: this.props.i18n('mb:extruderMaterialErrorModal.title'),
      body: this.props.i18n('mb:extruderMaterialErrorModal.body'),
      confirmText: this.props.i18n('mb:extruderMaterialErrorModal.confirmText'),
      modalClass: 'w400 warningModal',
      hideClose: true,
      backdrop: 'static',
      keyboard: false,
      onConfirm: () => {}
    });
  }

  showExtruderSetupErrorWarning(payload) {
    const modalService = this.props.flux.getService('ModalService');
    return modalService.show( /*#__PURE__*/React.createElement(UnsupportedExtruderConfigurationsDialog, payload), {
      title: this.props.i18n('mb:unsupported_extruder_configurations.title_modal', {
        machine_name: payload.machine_name
      }),
      backdrop: 'static',
      modalClass: 'w400 warningModal'
    });
  }

  showNonLabsMaterialWarning() {
    const modalService = this.props.flux.getService('ModalService');
    return modalService.confirm({
      title: this.props.i18n('mb:unknownMaterialWarningModal.title'),
      body: this.props.i18n('mb:unknownMaterialWarningModal.text'),
      confirmText: this.props.i18n('mb:unknownMaterialWarningModal.acknowledge'),
      modalClass: 'w400 warningModal',
      hideClose: true,
      backdrop: 'static',
      keyboard: false,
      onConfirm: () => {}
    });
  }

  getPrintLabel() {
    return this.canPrint() ? this.props.i18n('printer_list.print') : this.props.i18n('mb:printButton.export');
  }

  getDropdown(disabled) {
    return this.canPrint() ? /*#__PURE__*/React.createElement("div", {
      id: "dropdown",
      className: `printButtonDropdown${disabled ? ' disabled' : ''}`,
      onClick: disabled ? undefined : event => this.openContextMenu(event)
    }, /*#__PURE__*/React.createElement("div", {
      className: "printButtonDropdownLabel"
    }, "...")) : null;
  }

  render() {
    const state = this.props.printer.getStatus().state.toLowerCase();
    const busy = state !== 'idle';
    const validationStore = this.props.flux.store('ValidationStore');
    const trayInvalid = validationStore.getTrayBodiesOutsideBuildEnvelope().size > 0 || validationStore.getRaftsOutsideBuildEnvelope().length > 0 || validationStore.isPurgeTowerOutOfBounds();
    const disabled = trayInvalid || this.props.disabled || busy && state !== 'offline';
    const classes = this.props.printerPopupClasses + (disabled ? ' disabled' : '');
    return /*#__PURE__*/React.createElement("div", {
      className: "printButtonContainer"
    }, /*#__PURE__*/React.createElement("div", {
      id: this.canPrint() ? 'printButton' : 'exportButton',
      className: classes,
      onClick: disabled ? undefined : this._boundOnClick
    }, /*#__PURE__*/React.createElement("div", {
      className: "printButtonLabel"
    }, this.getPrintLabel())), this.getDropdown(disabled));
  }

}

PrintButton.displayName = 'PrintButton';
PrintButton.propTypes = {
  i18n: React.PropTypes.func.isRequired,
  flux: React.PropTypes.object.isRequired,
  onClick: React.PropTypes.func.isRequired,
  printer: React.PropTypes.object.isRequired,
  printerPopupClasses: React.PropTypes.string,
  disabled: React.PropTypes.bool
};
module.exports = PrintButton;
