// MB-support-plugin/lib/ui/views/printProgressText.js
/* eslint-disable react/jsx-key */
'use strict';

const React = require('react');

const Button = require('eagle-print/views/ui/button.js');

const FluxComponent = require('eagle-print/views/flux_component');

const PrinterDetailsError = require('./PrinterDetailsError');

const {
  maybe,
  just,
  nothing
} = require('../../utils/maybe');

class PrintProgressText extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    const flux = this.getFlux();
    this._actions = flux.actions.PrinterDetailsActions;
    this._i18n = flux.getService('TranslationService');
    this._tracking = flux.getService('TrackingService');
  }

  getFormattedTime(time) {
    return maybe(time).flatMap(time => time > 0 ? just(time) : nothing()).map(seconds => {
      const hours = Math.floor(seconds / 3600);
      const minutes = Math.floor((seconds - hours * 3600) / 60);
      return /*#__PURE__*/React.createElement("span", {
        style: {
          fontWeight: 'bold'
        }
      }, `${hours}h ${minutes}m`);
    });
  } // Remove path and .makerbot extension from pathname if they exist


  getFileName(inFile) {
    if (inFile) {
      let returnFile = inFile;
      const backSlashIndex = returnFile.lastIndexOf('/');
      if (backSlashIndex >= 0) returnFile = returnFile.substr(backSlashIndex + 1);
      if (returnFile.lastIndexOf('.makerbot') >= 0) returnFile = returnFile.replace('.makerbot', '');
      return returnFile;
    }
  }

  _estimatedTime() {
    const label = this._i18n.t('mb:printerStatus.description.time_remaining');

    const {
      step
    } = this.props.detailsStatus;
    if (!(step.printing || step.paused)) return null;
    return this.props.currentProcess.apply(cp => {
      if (cp.progress < 100) {
        return this.getFormattedTime(cp.time_remaining).map(time => /*#__PURE__*/React.createElement("div", null, `${label}: `, time)).lift();
      }
    });
  }

  _extruderTemp(extruder, isDual) {
    if (this.props.detailsStatus.has.extruder) {
      const currentTemp = extruder.current_temperature;
      return /*#__PURE__*/React.createElement("div", null, isDual ? this._i18n.t('mb:printerStatus.description.extruder_index', {
        index: extruder.index + 1
      }) : this._i18n.t('mb:printerStatus.description.extruder'), /*#__PURE__*/React.createElement("span", {
        style: {
          fontWeight: 'bold'
        }
      }, ` ${currentTemp + this._targetExtruderTemp(extruder)}`), ' °C');
    }
  }

  _targetExtruderTemp(extruder) {
    const {
      step
    } = this.props.detailsStatus;
    if (step.printing || step.paused || step.zCalFilament || step.startPrint) return null;
    return this.props.currentProcess.apply(() => {
      const targetTemp = extruder.target_temperature;

      if (targetTemp > 0) {
        return ` | ${targetTemp}`;
      } else {
        return null;
      }
    });
  }

  _chamberTemp() {
    if (this.props.detailsStatus.has.chamber) {
      const currentTemp = this.props.printer.getChamberInfo()[0].current_temperature;
      return /*#__PURE__*/React.createElement("div", null, this._i18n.t('mb:printerStatus.description.chamber'), /*#__PURE__*/React.createElement("span", {
        style: {
          fontWeight: 'bold'
        }
      }, ` ${currentTemp + this._targetChamberTemp()}`), ' °C');
    }
  }

  _targetChamberTemp() {
    const {
      step
    } = this.props.detailsStatus;
    if (step.printing || step.paused || step.zCalFilament || step.startPrint) return null;

    if (this.props.detailsStatus.has.chamber) {
      const targetTemp = this.props.printer.getChamberInfo()[0].target_temperature;

      if (targetTemp > 0) {
        return ` | ${targetTemp}`;
      } else {
        return null;
      }
    }
  }

  _platformTemp() {
    if (this.props.detailsStatus.has.platform) {
      const currentTemp = this.props.printer.getPlatformInfo()[0].current_temperature;
      return /*#__PURE__*/React.createElement("div", null, this._i18n.t('mb:printerStatus.description.platform'), /*#__PURE__*/React.createElement("span", {
        style: {
          fontWeight: 'bold'
        }
      }, ` ${currentTemp + this._targetPlatformTemp()}`), " \xB0C");
    }
  }

  _targetPlatformTemp() {
    const {
      step
    } = this.props.detailsStatus;
    if (step.printing || step.paused || step.zCalFilament || step.startPrint) return null;

    if (this.props.detailsStatus.has.platform) {
      const targetTemp = this.props.printer.getPlatformInfo()[0].target_temperature;

      if (targetTemp > 0) {
        return ` | ${targetTemp}`;
      } else {
        return null;
      }
    }
  }

  renderText(button) {
    const {
      step,
      has
    } = this.props.detailsStatus;
    const {
      label,
      description
    } = this.props.progressText(this._i18n, this.props.printer);

    const estimatedTime = this._estimatedTime();

    const isDual = this.props.printer.getExtruderInfo().length > 1;

    const chamberTemp = this._chamberTemp();

    const platformTemp = this._platformTemp();

    const extruderTemp = [];
    this.props.printer.getExtruderInfo().map(extruder => {
      const tempInfo = this._extruderTemp(extruder, isDual);

      extruderTemp.push(tempInfo);
    });
    const fileName = this.props.currentProcess.apply(cp => {
      if ((step.printing || step.completed || step.paused) && has.extruder) {
        return /*#__PURE__*/React.createElement("div", {
          className: "printerStatus"
        }, this.getFileName(cp.filename));
      }
    });
    const progress = this.props.currentProcess.apply(cp => cp.progress != null && step.transferring ? /*#__PURE__*/React.createElement("div", null, `${cp.progress}%`) : null);
    return /*#__PURE__*/React.createElement("div", {
      id: "PrinterDetailsInfoContainer"
    }, /*#__PURE__*/React.createElement("div", {
      className: "printerStatus"
    }, label), fileName, progress, /*#__PURE__*/React.createElement("div", null, description, /*#__PURE__*/React.createElement("div", null, this.renderButton(button))), extruderTemp, chamberTemp, platformTemp, estimatedTime);
  }

  renderError() {
    return /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement(PrinterDetailsError, {
      flux: this.props.flux,
      printer: this.props.printer,
      error: this.props.printer.getLastError(),
      closePanel: () => this.props.onBack()
    }));
  }

  renderButton(button) {
    if (!button) return null;
    const {
      label,
      onClick,
      primary
    } = button(this._i18n, this._actions);
    const buttonType = primary ? 'primary' : 'secondary';
    return /*#__PURE__*/React.createElement(Button, {
      type: buttonType,
      label: label,
      onClick: onClick
    });
  }

  render() {
    // const { step } = this.props.detailsStatus;
    const lastError = this.props.printer.getLastError();

    if (lastError) {
      return this.renderError();
    }

    return this.renderText(this.props.progressButtons);
  }

}

PrintProgressText.displayName = 'PrintProgressText';
PrintProgressText.propTypes = Object.assign({}, PrintProgressText.propsTypes, {
  currentProcess: React.PropTypes.object.isRequired,
  detailsStatus: React.PropTypes.object.isRequired,
  progressText: React.PropTypes.func,
  progressButtons: React.PropTypes.object
});
module.exports = PrintProgressText;
