// MB-support-plugin/lib/ui/views/printProgressBadge.js
'use strict';

var _react = _interopRequireDefault(require("react"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _path = _interopRequireDefault(require("path"));

var _flux_component = _interopRequireDefault(require("eagle-print/views/flux_component"));

var _styles = _interopRequireDefault(require("./styles"));

var _printerDetails = require("../../utils/printerDetails");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class PrintProgressBadge extends _flux_component.default {
  constructor(props, context) {
    super(props, context); // because GrabCAD
    // more specifically, GrabCAD's widget provider doesn't allow
    // us to add useful properties to this component. That's not a
    // big deal when we're in the PrinterDetailsPanel, but it's
    // a problem when we're in the PrinterList. So, we have to add
    // these useful objects in a kind of silly way, instead of just
    // getting them from the props.

    this.detailsStatus = (0, _printerDetails.printerDetailsStatus)(this.props.printer, this.props.overrideState);
    this.currentProcess = (0, _printerDetails.currentPrinterProcess)(this.props.printer, this.props.overrideState);
  }

  componentDidUpdate() {
    // also need to update it since we can't just use props for this
    this.detailsStatus = (0, _printerDetails.printerDetailsStatus)(this.props.printer, this.props.overrideState);
    this.currentProcess = (0, _printerDetails.currentPrinterProcess)(this.props.printer, this.props.overrideState);
  }

  _realRadius() {
    // pull them out to make code formatting nicer
    const {
      r,
      strokeWidth
    } = this.props;
    return r - (strokeWidth + strokeWidth / 2);
  }

  _realPercent() {
    return this.currentProcess.apply(cp => cp.progress);
  }

  _circleClass() {
    const {
      step,
      is,
      has
    } = this.detailsStatus;
    if (step.completed) return 'makerbot-blue-stroke';
    if (step.busy || step.transferring || step.updatingFirmware || step.wifiSetUp) return 'makerbot-light-grey-stroke';
    if (step.printing) return 'makerbot-red-stroke';
    if (step.paused || step.startPrint || step.error || step.clearFilament || is.authenticating || is.unauthenticated || has.error || !has.extruder) return 'makerbot-orange-stroke';
    return '';
  }

  _circleStyle() {
    const {
      step,
      has,
      is
    } = this.detailsStatus;
    if (step.busy || step.transferring || step.updatingFirmware || step.wifiSetUp) return _styles.default.dashedCircle(this._realRadius(), 4);
    if (step.printing) return Object.assign({}, _styles.default.roundedCapCircle(), {
      strokeDasharray: this.currentProcess.apply(() => dashArray(this._realRadius(), this._realPercent()))
    });
    if (step.error || has.error || step.startPrint || step.clearFilament || is.authenticating || is.unauthenticated || !has.extruder) return Object.assign({}, _styles.default.roundedCapCircle(), {
      animation: 'pulse 1.5s linear infinite'
    });
    if (step.paused) return _styles.default.dashedCircle(this._realRadius(), 2);
    if (is.idle || !has.process) return Object.assign({}, _styles.default.roundedCapCircle(), {
      stroke: 'none',
      opacity: 0
    });
    return _styles.default.roundedCapCircle();
  }

  getPrinterIcon() {
    const filePath = _path.default.resolve(__dirname, `images/BotIcon_${this.props.printer.getGender().botType}.svg`);

    if (_fsExtra.default.existsSync(filePath)) return filePath;
    return _path.default.resolve(__dirname, 'images/m_log_placeholder.svg');
  }

  getProgressCircle() {
    return /*#__PURE__*/_react.default.createElement("circle", {
      className: this._circleClass(),
      cx: this.props.r,
      cy: this.props.r,
      r: this._realRadius(),
      strokeWidth: `${this.props.strokeWidth}px`,
      style: this._circleStyle()
    });
  }

  getBackgroundCircle() {
    const {
      has,
      is
    } = this.detailsStatus;
    const style = {
      strokeWidth: '1px',
      fill: 'none',
      stroke: '#8c8c8c'
    };

    if (this.props.r <= 26 && (is.idle || !has.process)) {
      style.stroke = 'none';
      style.opacity = 0;
    }

    return /*#__PURE__*/_react.default.createElement("circle", {
      style: style,
      cx: this.props.r,
      cy: this.props.r,
      r: this.props.r - 1
    });
  }

  _printingText() {
    const {
      step
    } = this.detailsStatus;
    const isSmallIcon = this.props.r <= 15;
    const isMediumIcon = this.props.r <= 26;
    const smallStyle = {
      transform: 'rotate(0deg)'
    };
    const largeStyle = Object.assign({}, smallStyle, {
      fontSize: `${this.props.r * 2 / 3}px`
    });

    if (step.printing && isSmallIcon) {
      return /*#__PURE__*/_react.default.createElement("image", {
        xlinkHref: _path.default.resolve(__dirname, 'images/play.svg'),
        x: this.props.r / 2,
        y: this.props.r / 2,
        height: this.props.r,
        width: this.props.r,
        style: smallStyle
      });
    } // Medium size icons show percent complete but not '%' symbol


    return /*#__PURE__*/_react.default.createElement("text", {
      className: "makerbot-grey-fill",
      style: largeStyle,
      x: this.props.r,
      y: this.props.r,
      dy: ".4em",
      textAnchor: "middle"
    }, /*#__PURE__*/_react.default.createElement("tspan", {
      className: "value"
    }, this._realPercent()), /*#__PURE__*/_react.default.createElement("tspan", {
      className: "percent"
    }, isMediumIcon ? '' : '%'));
  }

  _progressIconProps() {
    const style = {
      transform: 'rotate(0deg)'
    };
    const {
      step,
      has,
      is
    } = this.detailsStatus;

    if (step.completed) {
      return {
        image: 'images/print-done.svg',
        style
      };
    }

    if (step.busy || step.transferring || step.updatingFirmware || step.wifiSetUp) {
      return {
        image: 'images/gears-icon.svg',
        style: {
          animation: 'spinner 30s linear infinite',
          animationDirection: 'reverse',
          transformOrigin: 'center center'
        }
      };
    }

    if (step.error || has.error || step.startPrint || step.clearFilament || is.authenticating || is.unauthenticated || !has.extruder) {
      return {
        image: 'images/alert-icon.svg',
        style
      };
    }

    if (step.paused) {
      return {
        image: 'images/pause-icon.svg',
        style
      };
    }

    if (is.idle || !has.process) {
      return {
        image: this.getPrinterIcon(),
        style: Object.assign({}, style, {
          transformOrigin: 'center',
          transform: `rotate(0deg) scale(2,2)`
        })
      };
    }

    return {
      image: this.getPrinterIcon(),
      style
    };
  }

  getProgressText() {
    if (this.detailsStatus.step.printing) return this._printingText();

    const {
      image,
      style
    } = this._progressIconProps();

    return /*#__PURE__*/_react.default.createElement("image", {
      xlinkHref: _path.default.resolve(__dirname, image),
      x: this.props.r / 2,
      y: this.props.r / 2,
      height: this.props.r,
      width: this.props.r,
      style: style
    });
  }

  render() {
    const width = this.props.r * 2;
    const height = this.props.r * 2;
    const viewBox = `0 0 ${width} ${height}`;
    return /*#__PURE__*/_react.default.createElement("svg", Object.assign({}, this.props, {
      className: 'CircularProgress',
      width,
      height,
      viewBox
    }), this.getBackgroundCircle(), this.getProgressCircle(), this.getProgressText());
  }

}

PrintProgressBadge.defaultProps = {
  r: 50,
  strokeWidth: 1
};
PrintProgressBadge.propTypes = Object.assign({}, PrintProgressBadge.propTypes, {
  printer: _react.default.PropTypes.object.isRequired,
  strokeWidth: _react.default.PropTypes.number,
  r: _react.default.PropTypes.number
});

const dashArray = (r, perc) => {
  const dash = r * Math.PI * 2 * (perc / 100);
  return `${dash}, ${r * Math.PI * 4}`;
};

module.exports = PrintProgressBadge;
