// MB-support-plugin/lib/ui/views/printerDetailsPanelButton.js
'use strict';

const React = require('react');

const styles = require('./styles');

class PrinterDetailsPanelButton extends React.Component {
  constructor(props) {
    super(props);
  }

  _getPendingSpinner() {
    const width = 48;
    const height = 48;
    const icon_d = 32;
    const spinner_d = 22;
    return /*#__PURE__*/React.createElement("svg", {
      width: width,
      height: height,
      viewBox: `0 0 ${width} ${height}`
    }, /*#__PURE__*/React.createElement("circle", {
      style: {
        fill: '#444' // charcoal...

      },
      cx: width / 2,
      cy: height / 2,
      r: icon_d / 2
    }), /*#__PURE__*/React.createElement("circle", {
      stroke: "#fff",
      cx: width / 2,
      cy: height / 2,
      r: spinner_d / 2,
      strokeWidth: `${3}px`,
      style: styles.dashedCircle(spinner_d / 2, 2)
    }));
  }

  render() {
    // if the button is enabled but not in a waiting state(clicked by user, etc) show it as disabled
    // if the button is waiting, show a spinner, otherwise a disabled button with icon
    return /*#__PURE__*/React.createElement("div", {
      className: this.props.enabled && !this.props.waiting ? '' : 'disabled',
      onClick: this.props.buttonClick
    }, /*#__PURE__*/React.createElement("span", {
      className: "isvg loaded"
    }, this.props.waiting ? this._getPendingSpinner() : /*#__PURE__*/React.createElement("img", {
      className: this.props.enabled ? '' : 'disabled',
      src: this.props.buttonIcon
    })), /*#__PURE__*/React.createElement("div", {
      className: "buttonLabel"
    }, this.props.buttonText));
  }

}

PrinterDetailsPanelButton.displayName = 'Printer Details Panel Button';
PrinterDetailsPanelButton.propTypes = {
  buttonText: React.PropTypes.string.isRequired,
  buttonIcon: React.PropTypes.string.isRequired,
  buttonClick: React.PropTypes.func.isRequired,
  enabled: React.PropTypes.bool.isRequired,
  waiting: React.PropTypes.bool.isRequired
};
module.exports = PrinterDetailsPanelButton;
