// MB-support-plugin/lib/ui/views/PrinterDetailsError.js
'use strict';

const Button = require('eagle-print/views/ui/button.js');

const React = require('react');

const _ = require('lodash');

const FluxComponent = require('eagle-print/views/flux_component');

const Isvg = require('react-inlinesvg');

const UrlConst = require('../../constants/urls').url;

const LinkClickUtils = require('../../utils/linkClick');

const BirdWingError = require('../../croissant').BirdWingError;

const Errors = require('../../errors');

class PrinterDetailsError extends FluxComponent {
  constructor(props) {
    super(props);
    this._actions = this.getFlux().actions.PrinterDetailsActions;
  }

  _getErrStringOverrides(err) {
    const i18n = this.props.flux.getService('TranslationService');
    const bwErrors = BirdWingError.Error;

    switch (err.code) {
      case Errors.MBPluginError.GeneralAuthError:
        return {
          title: i18n.t('mb:errors.authenticating.title'),
          message: i18n.t('mb:errors.authenticating.general')
        };

      case Errors.MBPluginError.DisconnectWhileAuthingError:
        return {
          title: i18n.t('mb:errors.authenticating.title'),
          message: i18n.t('mb:errors.authenticating.connection')
        };

      case bwErrors.kToolNotCalibrated:
        return {
          title: i18n.t('mb:errors.calibration_error.title'),
          message: i18n.t('mb:errors.calibration_error.message')
        };

      case bwErrors.kOutOfFilament:
        return this.props.printer.currentProcess.methods.indexOf('load_filament') > -1 ? {
          title: i18n.t('mb:errors.out_of_filament_error.load_filament_title'),
          message: i18n.t('mb:errors.out_of_filament_error.load_filament_message')
        } : {};

      default:
        return {};
    }
  }

  _getErrInfo(err) {
    const errInfo = Errors.getErrInfo(this.props.printer, err); // apply overrides, for things where we don't want the default message
    // from bwcoreutils (and plugin-generated errors which don't have default
    // messages)

    _.extend(errInfo, this._getErrStringOverrides(err));

    return errInfo;
  }

  _renderButtons(err) {
    const i18n = this.props.flux.getService('TranslationService');
    const printer = this.props.printer;
    const bwAction = BirdWingError.Action;

    const errInfo = this._getErrInfo(err);

    switch (errInfo.action) {
      case Errors.MBPluginErrAction.NONE:
        return [];

      case Errors.MBPluginErrAction.RETRY_AUTH:
        return [this._renderButton('tryAgainButton', i18n.t('mb:errors.button_text.tryAgain'), () => {
          printer.acknowledgeError(err.error_id);
          this.props.flux.actions.printer.authenticate(printer);
        })];

      case Errors.MBPluginErrAction.ACKNOWLEDGE_AND_CLOSE:
        return [this._renderButton('okButton', i18n.t('mb:errors.button_text.OK'), () => {
          printer.acknowledgeError(err.error_id);
          this.props.closePanel();
        })];

      case bwAction.ACKNOWLEDGE:
        return [this._renderButton('okButton', i18n.t('mb:errors.button_text.OK'), () => this._actions.acknowledgeError(printer.getId(), err.error_id))];

      case bwAction.CALIBRATE_Z:
        return [this._renderButton('runZCalButton', i18n.t('mb:errors.button_text.runZCal'), () => {
          printer.acknowledgeError(err.error_id);
          printer.invoke('CalibrateZOffset').done();
        })];

      case Errors.MBPluginErrAction.LOAD_FILAMENT:
        return [this._renderButton('loadFilamentButton', i18n.t('mb:errors.button_text.loadFilament'), () => {
          printer.acknowledgeError(err.error_id);
          const lf = 'load_filament';
          printer.currentProcess.methods.indexOf(lf) > -1 ? printer.invoke('ProcessMethod', [lf]).done() : printer.invoke('LoadFilament').done();
        })];
      //case bwAction.CHANGE_FILAMENT:
      //case bwAction.PREHEAT_TIMEOUT:
      //case bwAction.ZPAUSED:
      //case bwAction.UNLOAD_FILAMENT:

      case bwAction.ATTACH_TOOL:
        return [this._renderButton('okButton', i18n.t('mb:errors.button_text.OK'), () => this._actions.attachExtruder(printer.getId()))];
      //case bwAction.ACKNOWLEDGE_OR_CONTINUE:

      default:
        return [this._renderButton('okButton', i18n.t('mb:errors.button_text.OK'), () => {
          printer.acknowledgeError(err.error_id);
        })];
    }
  }

  _renderButton(key, text, action) {
    return /*#__PURE__*/React.createElement("li", {
      key: key
    }, /*#__PURE__*/React.createElement(Button, {
      type: "secondary confirmButton printerStatusButton",
      label: text,
      onClick: action
    }));
  }

  render() {
    const errInfo = this._getErrInfo(this.props.error);

    if (UrlConst.error[errInfo.error]) {
      return /*#__PURE__*/React.createElement("div", {
        id: "PrinterDetailsInfoContainer"
      }, /*#__PURE__*/React.createElement("div", {
        className: "printerStatus"
      }, /*#__PURE__*/React.createElement("a", {
        className: "supportLink",
        onClick: () => {
          LinkClickUtils.clickAction(this.getFlux(), `{$errInfo.error}_link_clicked`, UrlConst.error[errInfo.error]);
        }
      }, errInfo.title, /*#__PURE__*/React.createElement(Isvg, {
        className: "helperLinkIcon",
        src: "icons/helpful-link.svg"
      }))), /*#__PURE__*/React.createElement("div", null, errInfo.message, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("ul", {
        style: {
          listStyleType: 'none',
          margin: 0,
          padding: 0
        }
      }, this._renderButtons(this.props.error)))));
    } else {
      return /*#__PURE__*/React.createElement("div", {
        id: "PrinterDetailsInfoContainer"
      }, /*#__PURE__*/React.createElement("div", {
        className: "printerStatus"
      }, errInfo.title), /*#__PURE__*/React.createElement("div", null, errInfo.message, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("ul", {
        style: {
          listStyleType: 'none',
          margin: 0,
          padding: 0
        }
      }, this._renderButtons(this.props.error)))));
    }
  }

}

PrinterDetailsError.displayName = 'PrinterDetailsError';
PrinterDetailsError.propTypes = {
  flux: React.PropTypes.object.isRequired,
  printer: React.PropTypes.object.isRequired,
  error: React.PropTypes.object.isRequired,
  closePanel: React.PropTypes.func.isRequired
};
module.exports = PrinterDetailsError;
