// MB-support-plugin/lib/ui/views/PrinterDetailsPanel.js
'use strict';

const React = require('react');

const Isvg = require('react-inlinesvg');

const PrinterCardContainer = require('eagle-print/views/printers/printer_card_container');

const FluxComponent = require('eagle-print/views/flux_component');

const CameraFeed = require('./cameraFeed');

const PrinterDetailsPanelButton = require('./printerDetailsPanelButton');

const PrintProgressBadge = require('./printProgressBadge');

const PrintProgressText = require('./printProgressText');

const PrinterUtilityPanel = require('./utilities/PrinterUtilityPanel');

const PrinterFilamentPanel = require('./filament');

const DualFilamentPanel = require('./dualFilament');

const FirmwareUpdateBanner = require('./firmwareUpdateBanner');

const HeaderSelectButton = require('./headerSelectButton');

const {
  PanelStates
} = require('../../constants/printerDetails');

class PrinterDetailsPanel extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this._flux = this.getFlux();
    this._store = this._flux.store('PrinterDetailsStore');
    this._actions = this._flux.actions.PrinterDetailsActions;
    this._i18n = this._flux.getService('TranslationService');
    this._tracking = this._flux.getService('TrackingService');
    this.onChange = this.onChange.bind(this); // STATE:
    // {
    //     detailsStatus: Object,
    //     currentProcess: Object,
    //     currentPanel: String,
    //     panelSwipeDirection: String,
    //     actionButtons: Immutable.Map,
    //     progressText: Object,
    //     progressButtons: Immutable.Map
    // }

    this.state = Object.assign({}, this.state, this._store.getState(this.props.printer.getId()));
  }

  componentDidUpdate(prevProps) {
    const prevPrinterId = prevProps.printer.getId();
    const currentPrinterId = this.props.printer.getId(); // if the printer is different and is unauthenticated, then authenticate it

    if (prevPrinterId !== currentPrinterId && !this.state.detailsStatus.is.authenticated) {
      this.getFlux().actions.printer.authenticate(this.props.printer);
    }
  }

  componentDidMount() {
    super.componentDidMount();

    if (!this.state.detailsStatus.is.authenticated) {
      this.getFlux().actions.printer.authenticate(this.props.printer);
    }

    this._store.addListener('change', this.onChange);
  }

  componentWillUnmount() {
    this._store.removeListener('change', this.onChange);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.shouldShowMainDetailsPanel && nextProps.shouldShowMainDetailsPanel() && this.state.showPanel !== PanelStates.Details) {
      this.onOpenDetails();
    }
  }

  onChange() {
    this.setState(this._store.getState(this.props.printer.getId()));
  }

  onOpenDetails() {
    this._actions.details();
  }

  renderMonitorPill() {
    const flux = this.getFlux();
    return /*#__PURE__*/React.createElement("div", {
      id: "printerDetailsBody"
    }, /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("div", {
      id: "printerDetailsProgressBadge"
    }, /*#__PURE__*/React.createElement(PrintProgressBadge, {
      printer: this.props.printer,
      currentProcess: this.state.currentProcess,
      strokeWidth: 6,
      r: 50
    })), /*#__PURE__*/React.createElement(PrintProgressText, {
      currentProcess: this.state.currentProcess,
      detailsStatus: this.state.detailsStatus,
      progressText: this.state.progressText,
      progressButtons: this.state.progressButtons,
      printer: this.props.printer,
      flux: flux,
      onBack: this.props.onBack
    })));
  }

  onBack() {
    if (!this.state.detailsStatus.Authenticated) {
      this.getFlux().actions.printer.authenticateCancel();
    }

    this.props.onBack();
  }

  renderButtons() {
    // Button:
    // {
    //      label: i18n => String,
    //      icon: String,
    //      order: Int,
    //      enabled: Bool,
    //      shouldShow: Bool,
    //      onClick: () => q.Promise()
    // }
    return this.state.actionButtons.map((v, k) => /*#__PURE__*/React.createElement(PrinterDetailsPanelButton, {
      key: k // Waiting for render refactor before these buttons can see the
      // light of day...
      //waiting: !!this.state.button2Waiting,
      ,
      waiting: v.waiting,
      buttonText: v.label(this._i18n),
      buttonIcon: v.icon,
      enabled: v.enabled,
      buttonClick: v.onClick
    })).toArray();
  }

  renderCard() {
    const info = this.props.printer.getInfo() || {};

    switch (this.state.currentPanel) {
      case PanelStates.Utilities:
        return /*#__PURE__*/React.createElement(PrinterUtilityPanel, {
          key: "printerUtilities",
          printer: this.props.printer,
          onClose: this.props.onClose,
          onBack: () => this.onOpenDetails(),
          onResetSuccess: this.props.onBack,
          flux: this._flux
        });

      case PanelStates.Filament:
        return this.props.printer.isSixthGen() ? /*#__PURE__*/React.createElement(DualFilamentPanel, {
          key: "dualFilament",
          printer: this.props.printer,
          onClose: this.props.onClose,
          onBack: () => this.onOpenDetails(),
          flux: this._flux
        }) : /*#__PURE__*/React.createElement(PrinterFilamentPanel, {
          key: "printerFilament",
          printer: this.props.printer,
          onClose: this.props.onClose,
          onBack: () => this.onOpenDetails(),
          flux: this._flux
        });

      default:
        // Details panel
        return /*#__PURE__*/React.createElement("div", {
          id: "printerDetailsPanel",
          className: "printerPanel",
          key: "printerDetails"
        }, /*#__PURE__*/React.createElement("div", {
          className: "printerPanelHeader"
        }, /*#__PURE__*/React.createElement("div", {
          className: "printerPanelBackButton",
          onClick: () => this.onBack()
        }, /*#__PURE__*/React.createElement(Isvg, {
          src: "icons/back-arrow.svg"
        })), /*#__PURE__*/React.createElement("div", {
          className: "printerPanelHeaderLabel"
        }, /*#__PURE__*/React.createElement("h1", null, this.props.printer.getName()), /*#__PURE__*/React.createElement("h3", null, /*#__PURE__*/React.createElement("span", {
          className: "printer-model"
        }, info.model))), /*#__PURE__*/React.createElement(HeaderSelectButton, {
          flux: this._flux,
          printer: this.props.printer
        }), /*#__PURE__*/React.createElement("div", {
          className: "printerPanelArrowDown",
          onClick: this.props.onClose
        }, /*#__PURE__*/React.createElement(Isvg, {
          src: "icons/back-arrow.svg"
        }))), /*#__PURE__*/React.createElement("div", {
          id: "printerDetailsContents"
        }, /*#__PURE__*/React.createElement("div", {
          style: {
            position: 'relative',
            overflowX: 'hidden',
            overflowY: 'auto'
          }
        }, /*#__PURE__*/React.createElement(CameraFeed, {
          printer: this.props.printer
        }), /*#__PURE__*/React.createElement(FirmwareUpdateBanner, {
          flux: this._flux,
          printer: this.props.printer
        })), this.renderMonitorPill(), /*#__PURE__*/React.createElement("div", {
          id: "printerDetailsButtonBar"
        }, this.renderButtons())));
    }
  }

  render() {
    return (
      /*#__PURE__*/
      //spread combined obj as props
      React.createElement(PrinterCardContainer, Object.assign({}, this.props, {
        swipeDirection: this.state.panelSwipeDirection
      }), this.renderCard())
    );
  }

}

PrinterDetailsPanel.propTypes = Object.assign({}, FluxComponent.propTypes, {
  onClose: React.PropTypes.func.isRequired,
  onBack: React.PropTypes.func.isRequired,
  shouldShowMainDetailsPanel: React.PropTypes.func,
  printer: React.PropTypes.shape({
    acknowledgeLastError: React.PropTypes.func.isRequired,
    getId: React.PropTypes.func.isRequired,
    getInfo: React.PropTypes.func.isRequired,
    getName: React.PropTypes.func.isRequired,
    getStatus: React.PropTypes.func.isRequired,
    invoke: React.PropTypes.func.isRequired
  }).isRequired
});
PrinterDetailsPanel.displayName = 'PrinterDetailsPanel';
module.exports = PrinterDetailsPanel;
