// MB-support-plugin/lib/ui/views/proxySettingsModal.js
'use strict';

const React = require('react');

const DOM = React.DOM;
const ce = React.createElement;

const FluxComponent = require('eagle-print/views/flux_component');

const Dropdown = require('eagle-print/views/ui/input_dropdown');

const crypto = require('crypto');

class ProxySettingsModal extends FluxComponent {
  constructor(props) {
    super(props);
    this.configService = this.getFlux().getService('ConfigService');
    this.i18n = this.getFlux().getService('TranslationService');
    this.trackingService = this.getFlux().getService('TrackingService');
    this.state = {
      protocol: 'http://',
      address: '',
      port: '',
      username: '',
      password: ''
    };
  }

  componentDidMount() {
    // Grab config and set the state.
    const proxyConfig = this.configService.get('proxySettings');

    if (proxyConfig) {
      const password = proxyConfig.password ? this._decryptPassword(proxyConfig.password) : '';
      this.setState({
        protocol: proxyConfig.protocol,
        address: proxyConfig.address,
        port: proxyConfig.port,
        username: proxyConfig.username,
        password
      });
    }
  }

  _encryptPassword(password) {
    const cipher = crypto.createCipher('aes192', 'full speed ahead');
    return cipher.update(password, 'utf8', 'base64') + cipher.final('base64');
  }

  _decryptPassword(encryptedPassword) {
    const decipher = crypto.createDecipher('aes192', 'full speed ahead');
    return decipher.update(encryptedPassword, 'base64', 'utf8') + decipher.final('utf8');
  }

  saveProxySettings() {
    const proxySettings = Object.assign({}, this.state);

    if (proxySettings.password) {
      proxySettings.password = this._encryptPassword(proxySettings.password);
    }

    this.trackingService.addEvent({
      name: 'Proxy Settings Saved'
    });
    this.configService.set('proxySettings', proxySettings);
    this.configService.save();
    this.props.onDone();
  }

  handleProtocolChange(option) {
    this.setState({
      protocol: option.value
    });
  }

  handleAddressChange(e) {
    this.setState({
      address: e.target.value
    });
  }

  handlePortChange(e) {
    this.setState({
      port: e.target.value
    });
  }

  handleUserNameChange(e) {
    this.setState({
      username: e.target.value
    });
  }

  handlePasswordChange(e) {
    this.setState({
      password: e.target.value
    });
  }

  render() {
    return DOM.div({
      className: 'proxy-settings-modal'
    }, DOM.div({
      className: 'form-group'
    }, DOM.div({}, DOM.h3({}, this.i18n.t('proxy_settings.modal_header'))), DOM.div({}, DOM.label({}, this.i18n.t('proxy_settings.protocol')), DOM.br({}), ce(Dropdown, {
      name: 'protocol',
      value: this.state.protocol,
      options: [{
        value: 'http://',
        label: this.i18n.t('proxy_settings.http')
      }, {
        value: 'https://',
        label: this.i18n.t('proxy_settings.https')
      }],
      onChange: this.handleProtocolChange.bind(this)
    })), DOM.div({}, DOM.label({}, this.i18n.t('proxy_settings.address')), DOM.br({}), DOM.input({
      value: this.state.address,
      onChange: this.handleAddressChange.bind(this)
    })), DOM.div({}, DOM.label({}, this.i18n.t('proxy_settings.port')), DOM.br({}), DOM.input({
      value: this.state.port,
      onChange: this.handlePortChange.bind(this)
    })), DOM.div({}, DOM.label({}, this.i18n.t('proxy_settings.user')), DOM.br({}), DOM.input({
      value: this.state.username,
      onChange: this.handleUserNameChange.bind(this)
    })), DOM.div({}, DOM.label({}, this.i18n.t('proxy_settings.password')), DOM.br({}), DOM.input({
      value: this.state.password,
      type: 'password',
      onChange: this.handlePasswordChange.bind(this)
    }))), DOM.div({
      className: 'buttonGroup'
    }, DOM.button({
      className: 'button secondary',
      onClick: () => this.props.onDone()
    }, this.i18n.t('system_preferences.cancel')), DOM.button({
      className: 'button primary',
      onClick: () => this.saveProxySettings()
    }, this.i18n.t('system_preferences.save'))));
  }

}

module.exports = ProxySettingsModal;
