// MB-support-plugin/lib/ui/views/storeSidePanel.js
"use strict";

const React = require('react');

const Isvg = require('react-inlinesvg');

const path = require('path');

const listensToClickOutside = require('react-onclickoutside/decorator'); // const util = require('../../util');


class StoreSidePanel extends React.Component {
  constructor(props) {
    super(props);
    this.renderHeader = this.renderHeader.bind(this);
  } // Required for HOC listensToClickOutside to work properly


  handleClickOutside(_event) {
    this.props.showPanel ? this.props.handleIconClick() : null;
  }

  renderHeader() {
    return /*#__PURE__*/React.createElement("div", {
      className: "storeHeader"
    }, /*#__PURE__*/React.createElement("div", {
      className: "closeButton closePanel",
      onClick: this.props.handleIconClick
    }, /*#__PURE__*/React.createElement(Isvg, {
      src: "icons/x-button.svg"
    })), /*#__PURE__*/React.createElement("div", null, /*#__PURE__*/React.createElement("span", null, this.props.i18n.t('mb:sidePanel.materials.header.title'))));
  }

  renderStoreLink(el) {
    const {
      imgSrc,
      copyText,
      mbLink
    } = el;

    const handleOnClick = () => {
      require('electron').remote.shell.openExternal(mbLink);
    };

    return /*#__PURE__*/React.createElement("div", {
      key: imgSrc,
      className: "storeLink"
    }, /*#__PURE__*/React.createElement("img", {
      src: path.resolve(__dirname, imgSrc)
    }), /*#__PURE__*/React.createElement("button", {
      onClick: handleOnClick
    }, /*#__PURE__*/React.createElement("span", null, copyText)));
  }

  renderStoreLinks() {
    const data = [{
      imgSrc: 'images/method_filament.png',
      copyText: this.props.i18n.t('mb:sidePanel.materials.method.text'),
      mbLink: 'https://store.makerbot.com/3d-printers-materials/method-spools?utm_source=mbprint&utm_medium=display&utm_campaign=methodmaterials'
    }, {
      imgSrc: 'images/rep_filament.png',
      copyText: this.props.i18n.t('mb:sidePanel.materials.replicator.text'),
      mbLink: 'https://store.makerbot.com/3d-printers-materials/replicator-spools?utm_source=mbprint&utm_medium=display&utm_campaign=repmaterials'
    }, {
      imgSrc: 'images/sketch_filament.png',
      copyText: this.props.i18n.t('mb:sidePanel.materials.sketch.text'),
      mbLink: 'https://store.makerbot.com/3d-printers-materials/sketch-spools?utm_source=mbprint&utm_medium=display&utm_campaign=sketchmaterials'
    }];
    return /*#__PURE__*/React.createElement("div", null, data.map(el => this.renderStoreLink(el)));
  }

  render() {
    return /*#__PURE__*/React.createElement("div", {
      ref: "popout",
      className: "item account"
    }, /*#__PURE__*/React.createElement("div", {
      className: "accountPanel"
    }, /*#__PURE__*/React.createElement("div", {
      className: "storePanel"
    }, this.renderHeader(), this.renderStoreLinks())));
  }

}

StoreSidePanel.displayName = 'MBStoreSidePanel';
StoreSidePanel.propTypes = {
  showPanel: React.PropTypes.bool.isRequired,
  handleIconClick: React.PropTypes.func.isRequired,
  i18n: React.PropTypes.object.isRequired
};
module.exports = listensToClickOutside(StoreSidePanel);
