// MB-support-plugin/lib/ui/views/unsupportedExtruderConfigurationsDialog.js
"use strict";

const FluxComponent = require('eagle-print/views/flux_component');

const React = require('react');

class UnsupportedExtruderConfigurationsDialog extends FluxComponent {
  constructor(props, context) {
    super(props, context);
    this.i18n = this.props.i18n;
    this.supportedPairConfigs = this.translateExtruderTypeToPrettier(this.props.supportedConfigs.pairedExtrudersConfigs);
    this.supportedPerExtruderConfigs = this.translateExtruderTypeToPrettier(this.props.supportedConfigs.perExtruderConfigs);
  }

  translateExtruderTypeToPrettier(arr) {
    return arr.map(extruders => extruders.map(extruderId => this.i18n.t(`mb:printSettings.${extruderId}`)));
  }

  convertListToStr(arr, sepStr = 'and') {
    if (arr.length === 1) {
      return arr[0];
    } else if (arr.length === 2) {
      return `${arr[0]} ${sepStr} ${arr[1]}`;
    } else {
      return arr.reduce((res, el, i) => {
        i === arr.length - 1 ? res += `${sepStr} ${el}` : res += `${el}, `;
        return res;
      }, '');
    }
  }

  generateMissingExtruderListItem(index) {
    const msg = this.i18n.t('mb:unsupported_extruder_configurations.missing_extruders', {
      extruderSlot: index + 1,
      supportedExtrudersList: this.convertListToStr(this.supportedPerExtruderConfigs[index], 'or')
    });
    return /*#__PURE__*/React.createElement("li", null, msg);
  }

  generateUnsupportedExtruderListItem() {
    return /*#__PURE__*/React.createElement("li", {
      key: `${this.supportedPairConfigs.join('-')}-mismatched-error`
    }, this.supportedPairConfigs.map((pairConfig, index) => {
      const msg = this.i18n.t('mb:unsupported_extruder_configurations.unsupported_extruders', {
        extruder1: pairConfig[0],
        extruder2: pairConfig[1]
      });
      return index === this.supportedPairConfigs.length - 1 ? /*#__PURE__*/React.createElement("span", null, msg) : /*#__PURE__*/React.createElement("span", null, msg, /*#__PURE__*/React.createElement("br", null));
    }));
  }

  generateMismatchExtruderListItem(index) {
    const supportedExtruders = this.supportedPerExtruderConfigs[index];
    const msg = this.i18n.t('mb:unsupported_extruder_configurations.mismatched_extruders', {
      supportedExtruders: this.convertListToStr(supportedExtruders, 'or'),
      extruderSlot: index + 1
    });
    return /*#__PURE__*/React.createElement("li", null, msg);
  }

  generateListItemsFromErrorType() {
    return this.props.perExtruderErrorType.map((errorType, index) => {
      let result;

      switch (errorType) {
        case 'missing':
          result = this.generateMissingExtruderListItem(index);
          break;

        case 'mismatched':
          result = this.generateMismatchExtruderListItem(index);
          break;

        case 'unsupported':
          result = this.generateUnsupportedExtruderListItem();
          break;

        default:
          result = null;
      }

      return result;
    }).reduce((res, arr) => res.concat(arr), []);
  }

  render() {
    return /*#__PURE__*/React.createElement("ul", null, this.generateListItemsFromErrorType());
  }

}

module.exports = UnsupportedExtruderConfigurationsDialog;
