// MB-support-plugin/lib/ui/views/WaitableButton.js
'use strict';

const React = require('react');

const Button = require('eagle-print/views/ui/button');

const styles = require('./styles');

class WaitableButton extends React.Component {
  constructor(props) {
    super(props);
  }

  _getSpinner() {
    const width = this.props.spinnerDiameter + this.props.strokeWidth;
    const height = width;
    return /*#__PURE__*/React.createElement("div", {
      style: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0
      }
    }, /*#__PURE__*/React.createElement("svg", {
      key: "spinner",
      style: {
        zIndex: 1000
      },
      width: width,
      height: height
    }, /*#__PURE__*/React.createElement("circle", {
      stroke: "#fff",
      cx: width / 2,
      cy: height / 2,
      r: this.props.spinnerDiameter / 2,
      strokeWidth: `${this.props.strokeWidth}px`,
      style: styles.dashedCircle(this.props.spinnerDiameter / 2, 2)
    })));
  }

  render() {
    const props = this.props.buttonProps;
    let overrides = {};

    if (this.props.waiting) {
      overrides = {
        classNames: `${props.className} waiting`,
        label: /*#__PURE__*/React.createElement("div", {
          style: {
            position: 'relative'
          }
        }, props.label, this._getSpinner()),
        disabled: true
      };
    }

    return /*#__PURE__*/React.createElement(Button, Object.assign(props, overrides));
  }

}

WaitableButton.displayName = 'WaitableButton';
WaitableButton.propTypes = {
  buttonProps: React.PropTypes.object.isRequired,
  waiting: React.PropTypes.bool.isRequired,
  spinnerDiameter: React.PropTypes.number.isRequired,
  strokeWidth: React.PropTypes.number.isRequired
};
WaitableButton.defaultProps = {
  spinnerDiameter: 15,
  strokeWidth: 3
};
module.exports = WaitableButton;
