// MB-support-plugin/lib/ui/views/widget_provider.js
'use strict';

const WidgetProvider = require('eagle-print').views.widget.WidgetProvider;

const WidgetDescriptor = require('eagle-print').views.widget.WidgetDescriptor;

const PrinterUtilityPanel = require('./utilities/PrinterUtilityPanel');

const PrinterDetailsPanel = require('./PrinterDetailsPanel');

const PrintButton = require('./printButton');

const PrintEstimates = require('./estimates');

const PrintProgressBadge = require('./printProgressBadge');

const PrintSettingsPanel = require('./printSettings/printSettingsPanel');

const MBAccountSidebarWidget = require('./mb_account_sidebar_widget');

const MBLogosPanel = require('./mbLogosPanel');

const PrinterConnectionIcon = require('./printerConnectionIcon');

class MBWidgetProvider extends WidgetProvider {
  constructor() {
    super();
  }
  /** @override */


  getHandlers() {
    return {
      'printer.details': 'onPrinterDetails',
      'printer.details.utilities': 'onPrinterDetailsUtilities',
      'print.button': 'onPrintButton',
      'print.estimates': 'onPrintEstimates',
      'printer.badge': 'onPrinterBadge',
      'printer.connectionIcon': 'onPrinterConnectionIcon',
      'properties.toolbar.materials': 'onPrintSettings',
      'sidebar.items': 'getSidebarWidget',
      'sidebar.logo': 'onMBLogosPanel'
    };
  }

  onMBLogosPanel(props) {
    return new WidgetDescriptor({
      widgetClass: MBLogosPanel
    });
  }

  getSidebarWidget(props) {
    return new WidgetDescriptor({
      groupItemId: 'cloud_widget',
      groupItemOrder: 100,
      widgetClass: MBAccountSidebarWidget
    });
  }

  onPrintSettings(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrintSettingsPanel
      });
    }
  }

  onPrinterConnectionIcon(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrinterConnectionIcon
      });
    }
  }

  onPrinterBadge(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrintProgressBadge
      });
    }
  }

  onPrinterDetails(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrinterDetailsPanel
      });
    }
  }

  onPrinterDetailsUtilities(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrinterUtilityPanel
      });
    }
  }

  onPrintButton(props) {
    if (props.printer) {
      return new WidgetDescriptor({
        widgetClass: PrintButton
      });
    }
  }

  onPrintEstimates(props) {
    return new WidgetDescriptor({
      widgetClass: PrintEstimates
    });
  }

}

module.exports = MBWidgetProvider;
