// MB-support-plugin/lib/actions/copyBotLogs.js
'use strict';

const R = require('ramda');

const tmp = require('tmp');

const q = require('q');

const fs = require('fs-extra');

const {
  getPrinter
} = require('../utils/printerDetails');

const {
  Actions
} = require('../constants/copyBotLogs');

const {
  getI18n,
  safeKey
} = require('../util');

const zipFilename = 'bot_logs.zip';

const defaultSavePath = printerId => `~/Documents/MakerbotLogs_${safeKey(printerId)}.zip`; // prevent calling zipLogs multiple times in a row


let debounce = [];
module.exports = Object.freeze({
  copyLogs(printerId) {
    const invoke = _invoke(this.flux, printerId);

    if (debounce.includes(printerId)) return;
    debounce = R.concat(printerId);
    this.dispatch(Actions.CopyStart, {
      printerId
    });
    return zipLogs(invoke, zipFilename).then(zipPath => {
      getLogs(invoke, zipPath).then(tmpPath => this.dispatch(Actions.CopySuccess, {
        printerId,
        tmpPath
      })).catch(err => this.dispatch(Actions.CopyFail, {
        printerId,
        err
      })).done();
    }).catch(err => this.dispatch(Actions.CopyFail, {
      printerId,
      err
    })).finally(() => debounce = R.without([printerId], debounce)).done();
  },

  sendToSupport(printerId) {
    const i18n = getI18n(this.flux);
    const tmpPath = getTmpPath(this.flux, printerId);
    const SupportPlugin = this.flux.getPlugin('Support');
    const ModalService = this.flux.getService('ModalService');
    if (!fs.existsSync(tmpPath)) return q().then(() => this.dispatch(Actions.NoFile, {
      printerId
    }));
    const supportModal = SupportPlugin.getModal(this.flux, {
      onMount: component => {
        if (component) filePathToObj(this.flux, tmpPath).then(component.manualAddFile).done();
      }
    }, () => this.dispatch(Actions.SendToSupportSuccess, {
      printerId
    }), err => this.dispatch(Actions.SendToSupportFail, {
      printerId,
      err
    }));
    ModalService.show(supportModal, {
      title: i18n.t('mb:copyBotLogs.send')
    });
  },

  saveLogs(printerId) {
    // have to do it this way because mb-plugin doesn't have electron
    // initalized, so requiring it at the top of the file, like a
    // normal person causes the tests to break
    const {
      dialog
    } = require('electron').remote;

    const i18n = getI18n(this.flux);
    const tmpPath = getTmpPath(this.flux, printerId);

    const dispatch = () => this.dispatch(Actions.Save, {
      printerId
    });

    if (!fs.existsSync(tmpPath)) {
      this.dispatch(Actions.NoFile, {
        printerId,
        tmpPath
      });
    } else {
      dialog.showSaveDialog({
        title: i18n.t('mb:copyBotLogs.save'),
        defaultPath: defaultSavePath(printerId)
      }, saveFile(dispatch, tmpPath));
    }
  },

  timeout(printerId, err) {
    this.dispatch(Actions.Timeout, {
      printerId,
      err
    });
  }

});

const getTmpPath = (flux, printerId) => flux.store('CopyBotLogs').getTmpPath(printerId);

const _invoke = R.curry((flux, printerId, method, params) => getPrinter(flux, printerId).invoke(method, params));

const zipLogs = R.curry((invoke, zipPath) => invoke('ZipLogs', [`/home/${zipPath}`]).then(() => zipPath));
const getLogs = R.curry((invoke, zipPath) => {
  const tmpPath = tmp.tmpNameSync({
    postfix: '.zip'
  });
  return invoke('Get', [zipPath, tmpPath]).then(() => {
    return tmpPath;
  });
});
const saveFile = R.curry((dispatch, tmpPath, path) => tmpPath && path ? fs.copy(tmpPath, path, {}, dispatch) : undefined);

const filePathToObj = (flux, filePath) => flux.getService('FileService').getFileStats(filePath).then(stats => ({
  filePath,
  size: stats['size'],
  error: null
}));
