// MB-support-plugin/lib/actions/networkPreferencesActions.js
'use strict';

const q = require('q');

const _ = require('lodash');

const {
  getPrinter
} = require('../utils/printerDetails');

const {
  Actions,
  NetworkStateEnum,
  InvalidPassword
} = require('../constants/networkPreferences');

const ActionCreators = Object.freeze({
  networkChange(printerId) {
    return updateNetworkState(this.flux, this.dispatch, printerId);
  },

  enableWifi(printerId) {
    const printer = getPrinter(this.flux, printerId);
    this.dispatch(Actions.Waiting);
    return printer.invoke('WifiEnable').then(() => updateNetworkState(this.flux, this.dispatch, printerId)).catch(handleError(this.dispatch)).done();
  },

  disableWifi(printerId) {
    const printer = getPrinter(this.flux, printerId);
    this.dispatch(Actions.Waiting);
    return printer.invoke('WifiDisable').then(() => updateNetworkState(this.flux, this.dispatch, printerId)).catch(handleError(this.dispatch)).done();
  },

  connectToNetwork(printerId, path, pw) {
    const printer = getPrinter(this.flux, printerId);
    const trackEvent = trackingEvent(this.flux, printer);
    const forget = ActionCreators.forgetNetwork.bind(this);
    this.dispatch(Actions.Waiting);
    return connectToWifi(printer, path, pw).then(() => trackEvent('Printer Successfully Connected to Wifi')).then(() => updateNetworkState(this.flux, this.dispatch, printerId)).catch(e => {
      const err = JSON.parse(e.message).error;
      trackEvent('Printer Failed to Connect to Wifi');
      if (err.code === InvalidPassword) return forget(path);
      return handleError(this.dispatch)(err);
    }).done();
  },

  disconnectFromNetwork(printerId, path) {
    const printer = getPrinter(this.flux, printerId);
    const trackEvent = trackingEvent(this.flux, printer);
    return disconnectFromWifi(printer, path).then(() => trackEvent('Printer Successfully Disconnected from Wifi')).then(() => updateNetworkState(this.flux, this.dispatch, printerId)).catch(err => {
      trackEvent('Printer Failed to Disconnect from Wifi');
      return handleError(this.dispatch)(err);
    }).done();
  },

  forgetNetwork(printerId, path) {
    const printer = getPrinter(this.flux, printerId);
    return printer.invoke('WifiForget', [path]).then(() => this.dispatch(Actions.WifiForget)).then(() => updateNetworkState(this.flux, this.dispatch, printerId)).done();
  },

  setStaticIP(printerId, path, ip, netmask, gateway, dnsServers, use_static = true) {
    const printer = getPrinter(this.flux, printerId); // since many of the params for this function are optional in kaiten
    // I have to map over and make sure any that are not explictly set
    // to a real value are converted to `undefined`. Kaiten will handle
    // that appropriately.

    const args = _.map([ip, netmask, gateway, dnsServers, use_static], a => a === null ? undefined : a);

    return printer.invoke('SetStaticIpv4', [path, ...args]).then(() => this.dispatch(Actions.SetStaticIP, {
      dnsServers,
      enable: use_static
    })).catch(handleError(this.dispatch)).done();
  },

  prepopulateStaticIP(printerId, path) {
    //Clear current field values while waiting for update
    this.dispatch(Actions.PrepopulateStaticIPFields, {});

    if (path) {
      const printer = getPrinter(this.flux, printerId);
      printer.invoke('GetStaticIpv4', [path]).then(result => {
        if (result.use_static) {
          this.dispatch(Actions.PrepopulateStaticIPFields, Object.assign({}, result));
        }
      }).done();
    }
  },

  changeStaticIPAddress(ipAddress) {
    this.dispatch(Actions.ChangeStaticIPAddress, {
      ipAddress
    });
  },

  changeNetmask(netmask) {
    this.dispatch(Actions.ChangeNetmask, {
      netmask
    });
  },

  changeGateway(gateway) {
    this.dispatch(Actions.ChangeGateway, {
      gateway
    });
  },

  changeDNSServers(dnsServers) {
    this.dispatch(Actions.ChangeDNSServers, {
      dnsServers: _.unique(dnsServers)
    });
  },

  newDNSItem(item) {
    this.dispatch(Actions.NewDNSItem, {
      item
    });
  },

  selectDNSItem(item) {
    this.dispatch(Actions.SelectDNSItem, {
      item
    });
  },

  selectNetwork(path) {
    this.dispatch(Actions.SelectNetwork, {
      path
    });
  },

  changePasswordAttempt(password) {
    this.dispatch(Actions.ChangePasswordAttempt, {
      password
    });
  },

  togglePasswordVisibility() {
    this.dispatch(Actions.TogglePasswordVisibility);
  },

  clearError() {
    this.dispatch(Actions.ClearError);
  }

});
/*****************************************************************************/

/*                              HELPERS                                      */

/*****************************************************************************/

const handleError = dispatch => error => dispatch(Actions.Error, {
  error
});

const getTracking = flux => flux.getService('TrackingService');

const trackingEvent = (flux, printer) => name => getTracking(flux).addEvent(_.merge({
  name
}, printer.getTrackingInfo()));

const fetchAvailableNetworks = (flux, printerId) => getPrinter(flux, printerId).invoke('WifiScan');

const setAsWifi = action => (flux, dispatch, printerId, networkState) => fetchAvailableNetworks(flux, printerId).then(availableNetworks => dispatch(action, {
  printerId,
  availableNetworks,
  networkState
})).catch(handleError(dispatch)).done();

const setAsEthernetConnected = (flux, dispatch, printerId, networkState) => dispatch(Actions.EthernetConnected, {
  printerId,
  networkState
});

const setAsUnconnected = (flux, dispatch, printerId, networkState) => dispatch(Actions.Unconnected, {
  printerId,
  networkState
}); // Depending on the connection type, different behaviors are needed.
// Instead of having a long if-else tree, I thought a declarative dictionary
// would be easier on the eyes.


const networkStateFlow = Object.freeze({
  [NetworkStateEnum.state.WIFI]: setAsWifi(Actions.WifiConnected),
  [NetworkStateEnum.state.ETHERNET]: setAsEthernetConnected,
  [NetworkStateEnum.wifi.ENABLED]: setAsWifi(Actions.WifiReady),
  [NetworkStateEnum.wifi.DISABLED]: setAsUnconnected
}); // Grab the appropriate state updater for the connection type

const networkStateUpdater = (flux, dispatch, printerId) => networkState => {
  const connFunc = networkStateFlow[networkState.state] || networkStateFlow[networkState.wifi];
  return connFunc(flux, dispatch, printerId, networkState);
};

const updateNetworkState = (flux, dispatch, printerId) => {
  const printer = getPrinter(flux, printerId);
  const updateState = networkStateUpdater(flux, dispatch, printerId);
  if (printer.isAuthenticated()) return printer.invoke('NetworkState').then(updateState).catch(handleError(dispatch)).done();
  return q().then(() => dispatch(Actions.Waiting)).done();
};

const connectToWifi = (printer, path, pw) => _.size(pw) === 0 ? printer.invoke('WifiConnect', [path]) : printer.invoke('WifiConnect', [path, pw]);

const disconnectFromWifi = (printer, path) => printer.invoke('WifiDisconnect', [path]);

module.exports = ActionCreators;
