// MB-support-plugin/lib/actions/printerDetailsActions.js
'use strict';

const q = require('q');

const R = require('ramda');

const React = require('react');

const {
  Dispatch
} = require('../constants/printerDetails');

const {
  getPrinter,
  currentPrinterProcess
} = require('../utils/printerDetails');

const {
  getTracking,
  getLogging,
  getI18n,
  getModal,
  trackingInfoForPrinter,
  getConfig
} = require('../util');

const EETermsAndConditionsDialog = require('../ui/views/eeTermsAndConditionsDialog');

const firmwareUtils = require('../firmwareUtils');

const processMethod = (printer, method) => printer.invoke('ProcessMethod', [method]);

const acknowledgeError = printer => processMethod(printer, 'acknowledge_error');
/**
 * Updates the printer with either the latest version the printer knows about, or with
 * a custom firmware file that the user chooses.
 */


const handleUpdateFW = (printer, flux) => {
  /* including the tracking service in showFWCompatabilityModal() and
  passing as a param blows up the modal service, so doing it here ¯\_(ツ)_/¯ */
  const tracking = getTracking(flux);
  firmwareUtils.updateFirmware(printer, flux);
  tracking.addEvent({
    name: 'firmware_update_initialized'
  });
};

module.exports = {
  done(id) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.AcknowledgeCompleted;
    this.dispatch(Dispatch.ButtonWait, {
      key: 'done',
      action: dispatchMethod
    });
    processMethod(printer, 'acknowledge_completed').then(() => this.dispatch(dispatchMethod, {
      printer
    }));
  },

  resume(id) {
    const printer = getPrinter(this.flux, id);

    if (printer.processMethodExists('suspend')) {
      const dispatchMethod = Dispatch.Suspend;
      this.dispatch(Dispatch.ButtonWait, {
        key: 'pause',
        action: dispatchMethod
      });
      return processMethod(printer, 'suspend').then(() => this.dispatch(dispatchMethod, {
        id,
        printer
      }));
    }

    if (printer.processMethodExists('resume')) {
      const dispatchMethod = Dispatch.Resume; // if there's an error, but the process is resumable,
      // clear the error first, then resume...

      if (printer.getLastError()) printer.acknowledgeLastError();
      this.dispatch(Dispatch.ButtonWait, {
        key: 'resume',
        action: dispatchMethod
      });
      return processMethod(printer, 'resume').then(() => this.dispatch(dispatchMethod, {
        id
      }));
    }
  },

  cancel(id) {
    const modalService = getModal(this.flux);
    const i18n = getI18n(this.flux);
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.Cancel;

    const cancelProcess = () => {
      const currentProcess = currentPrinterProcess(printer).apply(R.identity);
      this.dispatch(Dispatch.ButtonWait, {
        key: 'cancel',
        action: dispatchMethod
      });
      return printer.cancel().then(() => this.dispatch(dispatchMethod, {
        printer,
        currentProcess
      }));
    };

    currentPrinterProcess(printer).apply(cp => {
      if (cp.name === 'PrintProcess') {
        modalService.confirm({
          title: i18n.t('mb:printCancelModal.title'),
          body: i18n.t('mb:printCancelModal.body'),
          cancelText: i18n.t('mb:printCancelModal.cancel'),
          confirmText: i18n.t('mb:printCancelModal.confirm'),
          modalClass: 'w400 warningModal',
          onConfirm: cancelProcess
        });
      } else {
        cancelProcess();
      }
    });
    return q();
  },

  stopFilament(id) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.Cancel;
    this.dispatch(Dispatch.ButtonWait, {
      key: 'stopFilament',
      action: dispatchMethod
    });
    return printer.stopFilament().then(() => this.dispatch(dispatchMethod));
  },

  printAgain(id) {
    const printer = getPrinter(this.flux, id);
    const currentProcess = currentPrinterProcess(printer).check;
    const dispatchMethod = Dispatch.PrintAgain;
    const ackMethod = currentProcess(cp => cp.step === 'failed') ? 'acknowledge_failure' : 'acknowledge_completed';
    this.dispatch(Dispatch.ButtonWait, {
      key: 'printAgain',
      action: dispatchMethod,
      printer
    });
    return processMethod(printer, ackMethod).then(() => printer.invoke('PrintAgain').then(() => this.dispatch(dispatchMethod)));
  },

  acknowledgeError(id, errID) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.AcknowledgeError;
    if (errID) printer.acknowledgeError(errID);else printer.acknowledgeLastError();
    this.dispatch(Dispatch.ButtonWait, {
      key: 'acknowledgeError',
      action: dispatchMethod
    });
    acknowledgeError(printer);
    return q().then(() => this.dispatch(dispatchMethod, {
      id
    }));
  },

  acknowledgeResumeError(id) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.AcknowledgeResumeError;
    this.dispatch(Dispatch.ButtonWait, {
      key: 'acknowledgeResumeError',
      action: dispatchMethod
    });
    return acknowledgeError(printer).then(() => this.dispatch(dispatchMethod));
  },

  acknowledgeFailure(id) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.AcknowledgeFailure;
    this.dispatch(Dispatch.ButtonWait, {
      key: 'acknowledgeFailure',
      action: dispatchMethod
    });
    return processMethod(printer, 'acknowledge_failure').then(() => this.dispatch(dispatchMethod));
  },

  acknowledgeSuccess(id) {
    const printer = getPrinter(this.flux, id);
    const dispatchMethod = Dispatch.AcknowledgeSuccess;
    this.dispatch(Dispatch.ButtonWait, {
      key: 'acknowledgeSuccess',
      action: dispatchMethod
    });
    return processMethod(printer, 'acknowledge_success').then(() => this.dispatch(dispatchMethod));
  },

  acknowledgeBuildPlateCleared(id) {
    const printer = getPrinter(this.flux, id);
    const tracking = getTracking(this.flux);
    const logging = getLogging(this.flux);
    return printer.acknowledgeBuildPlateCleared().then(() => tracking.addEvent(R.merge({
      name: 'Confirm Clear Build Plate'
    }, trackingInfoForPrinter(printer)))).catch(err => {
      if (err.type === 'UnauthenticatedError') logging.warn(err);
    });
  },

  attachExtruder(id) {
    const printer = getPrinter(this.flux, id);
    const lpt = 'load_print_tool';

    if (printer.processMethodExists(lpt)) {
      return processMethod(printer, lpt);
    }

    return printer.invoke('LoadPrintTool', [0]);
  },

  clearFilament(id) {
    const printer = getPrinter(this.flux, id);
    const logging = getLogging(this.flux); // right now ZCal is the only process that uses 'continue_process'
    // to continue to unload filament. if we add any other processes
    // that need to unload filament, we should clean this up

    return processMethod(printer, 'continue_process').catch(err => logging.warn(err));
  },

  zCalFilament(id) {
    const printer = getPrinter(this.flux, id);
    const logging = getLogging(this.flux);
    return processMethod(printer, 'continue_process').catch(err => logging.warn(err));
  },

  extruderChange(id, index, config) {
    return q(this.dispatch(Dispatch.ExtruderChange, {
      id,
      index,
      config
    }));
  },

  networkChange(id, state) {
    return q(this.dispatch(Dispatch.NetworkChange, {
      id,
      state
    }));
  },

  details() {
    return q().delay(0).then(() => this.dispatch(Dispatch.ShowPanelDetails));
  },

  filament() {
    return q(this.dispatch(Dispatch.ShowPanelFilament));
  },

  utilities() {
    return q(this.dispatch(Dispatch.ShowPanelUtilities));
  },

  system(id, info) {
    return this.dispatch(Dispatch.System, {
      id,
      info
    });
  },

  state(id, info) {
    return this.dispatch(Dispatch.State, {
      id,
      info
    });
  },

  status(id, info) {
    return this.dispatch(Dispatch.Status, {
      id,
      info
    });
  },

  // we're not handling this right now
  error() {
    return q();
  },

  buttonWait(key, action) {
    return this.dispatch(Dispatch.ButtonWait, {
      key,
      action
    });
  },

  showEETermsAndConditions() {
    const flux = this.flux;
    const modalService = getModal(flux);
    const configService = getConfig(flux);
    const i18n = getI18n(flux);

    if (!configService.get('eeTermsSeen')) {
      modalService.show( /*#__PURE__*/React.createElement(EETermsAndConditionsDialog, null), {
        title: i18n.t('mb:ee_terms_and_conditions.title'),
        hideClose: true,
        backdrop: 'static',
        keyboard: false
      });
    }
  },

  showFWCompatabilityModal(printer) {
    const flux = this.flux;
    const modalService = getModal(flux);
    const i18n = getI18n(flux);
    modalService.confirm({
      title: i18n.t('mb:firmware.compatability.title'),
      __html: i18n.t('mb:firmware.compatability.text', {
        printerName: printer._printerInfo.name
      }),
      confirmText: i18n.t('mb:firmware.compatability.update_button'),
      cancelText: i18n.t('mb:firmware.compatability.close_button'),
      modalClass: 'w400 warningModal',
      onConfirm: () => handleUpdateFW(printer, flux),
      onClose: modalService.hide()
    });
  }

};
