// MB-support-plugin/lib/actions/printSettingsActions.js
'use strict';

const React = require('react');

const q = require('q');

const Isvg = require('react-inlinesvg');

const R = require('ramda');

const DOM = React.DOM;
const ce = React.createElement;

const AdvancedSettingsModal = require('../ui/views/printSettings/advancedSettingsModal');

const LinkClickUtils = require('../utils/linkClick');

const UrlConst = require('../constants/urls').url;

const {
  Actions
} = require('../constants/print_settings');

const printSettingsActions = {
  toggleProp(key) {
    this.dispatch('MB_TOGGLE_PROP', {
      key
    });
  },

  bulkToggleProp(active) {
    this.dispatch('MB_BULK_TOGGLE_PROPS', {
      active
    });
  },

  resetToDefault(key) {
    this.dispatch('MB_RESET_TO_DEFAULT', {
      key
    });
  },

  resetAllToDefault() {
    this.dispatch('MB_RESET_ALL_TO_DEFAULT');
  },

  removeAllSettings() {
    this.dispatch('MB_REMOVE_ALL_SETTINGS');
  },

  getExtruderTypes(extruderSchema, authenticated) {
    this.dispatch('GET_EXTRUDER_TYPES', {
      extruderSchema,
      authenticated
    });
  },

  getSupportedConfigurations() {
    this.dispatch('GET_SUPPORTED_CONFIGURATIONS');
  },

  loadSchema(payload) {
    /**
     * Requires printer id to avoid unnecessary changes to
     * PrintSettingsStore state. Only update the store given the
     * printer id matches the currently selected printer id
     *
     * payload
     *  id:         printer id
     *  addDelay:   an option to add a delay of ~3 seconds to allow time
     *              for the RFID reader to read the spool tags. (Only
     *              used in makerbot-printer.js' updateSpoolMaterialsFrom-
     *              SystemInfo)
     *
     * -- Ray
     **/
    const currentPrinter = this.flux.store('PrinterStore').getSelectedPrinter(); // For archetype printers, exit out of loadSchema
    // Later note: This doesn't actually work, and moreover, it really shouldn't.
    // I don't see a reason why an archetype printer's schema shouldn't be loaded,
    // Leaving this here for now, in case removing it breaks something,
    // but will have to look into the necessity of this line in the future
    // xoxo Leor
    // Discovered this seems to help when loading old project versions

    if (!currentPrinter) return;

    if (payload && payload.id === currentPrinter.getId()) {
      if (payload.spoolsInfo) {
        return q().delay(0).then(() => this.dispatch('MB_GET_PRINTER_SCHEMA', {
          spoolsInfo: payload.spoolsInfo
        }));
      }

      return q().delay(0).then(() => this.dispatch('MB_GET_PRINTER_SCHEMA', {}));
    }
  },

  launchAdvSettingsModal(flux, extraProps) {
    this.dispatch('MB_PRINT_MODE_MENU_OPENED');
    flux.getService('ModalService').show(ce(AdvancedSettingsModal, R.merge(extraProps, {
      flux
    })), {
      title: DOM.a({
        className: 'supportLink',
        onClick: () => {
          LinkClickUtils.clickAction(flux, 'custom_print_modes_link_clicked', UrlConst.customPrintModes);
        }
      }, extraProps.i18n('mb:printSettings.advancedSettings.title'), ce(Isvg, {
        className: 'helperLinkIcon',
        src: 'icons/helpful-link-white.svg'
      })),
      modalClass: 'narrow-padding advanced-settings-modal-container'
    });
  },

  closeAdvSettingsModal(flux, selectedPreset) {
    flux.getService('ModalService').hide();
    this.dispatch(Actions.CloseModal, {
      selectedPreset
    });
  },

  // create a preset in memory and save it to user_config
  createPreset(name, originalPrintMode) {
    const store = this.flux.store('PrintSettingsStore');
    const deferred = q.defer();
    const errorMsg = R.cond([[R.isNil, R.always('Preset needs a name')], [store.hasPresetWithName, R.always('Preset with that name already exists')], [R.T, R.always('Unknown error')]]);

    if (name && !store.hasPresetWithName(name)) {
      this.dispatch(Actions.CreatePreset, {
        name,
        originalPrintMode
      });
      deferred.resolve();
    } else {
      const error = errorMsg(name);
      this.dispatch(Actions.PresetError, {
        error
      });
      deferred.reject(error);
    }

    return deferred.promise;
  },

  // edit the currently enabled in-memory preset and save it to user_config
  updatePreset(name) {
    this.dispatch(Actions.UpdatePreset, {
      name
    });
  },

  // export the current preset to a file
  writePreset(name) {
    this.dispatch(Actions.WritePreset, {
      name
    });
  },

  destroyPreset(name) {
    this.dispatch(Actions.DestroyPreset, {
      name
    });
  },

  readPresets() {
    this.dispatch(Actions.ReadPresets);
  },

  applyPreset(name) {
    this.dispatch(Actions.ApplyPreset, {
      name
    });
  },

  unapplyPreset() {
    this.dispatch(Actions.UnapplyPreset);
  },

  renamePreset(name) {
    this.dispatch(Actions.RenamePreset, {
      name
    });
  },

  revertChanges() {
    this.dispatch(Actions.RevertChanges);
  },

  updateExtruder() {
    this.dispatch(Actions.UpdateExtruder);
  },

  updateExtruderPair() {
    this.dispatch(Actions.UpdateExtruderPair);
  },

  updateMaterial(index, val) {
    this.dispatch(Actions.UpdateMaterial, {
      index,
      val
    });
  },

  updateMaterials(payload) {
    this.dispatch(Actions.UpdateMaterials, payload);
  },

  updateSupports(supportExtruder, supportType) {
    this.dispatch(Actions.UpdateSupports, {
      supportExtruder,
      supportType
    });
  },

  updatePrintMode(printMode) {
    this.dispatch(Actions.UpdatePrintMode, printMode);
  },

  updatePurgeWallCenter(purgeWallCenter) {
    // cancel builds (aka. slices) when moving purge tower around
    const viewerStore = this.flux.store('ViewerStore');
    const activeTray = viewerStore.getActiveTray();
    this.flux.actions.printer.cancelBuildsForTray(activeTray.getId());
    this.dispatch(Actions.UpdatePurgeWallCenter, purgeWallCenter);
    this.dispatch('PURGE_TOWER_CHANGED');
  },

  updatePurgeWall() {
    this.dispatch('PURGE_TOWER_CHANGED');
  }

};
module.exports = printSettingsActions;
