// MB-support-plugin/lib/stores/analyticsStore.js
'use strict';

const Fluxxor = require('fluxxor');

const JOB_TYPE = require('eagle-print/stores/buildStore').JOB_TYPE;

const R = require('ramda');

const {
  Dispatch
} = require('../constants/printerDetails');

const BotLogsActions = require('../constants/copyBotLogs').Actions;

const PrintSettingsActions = require('../constants/print_settings').Actions;

module.exports = Fluxxor.createStore({
  actions: {
    PREPARE_PRINT: 'onPreparePrint',
    BUILD_FAILED: 'onBuildFailed',
    BUILD_SUCCEEDED: 'onBuildSucceeded',
    EXPORT_TOOLPATH: 'onExportToolpath',
    EXPORT_TOOLPATH_COMPLETE: 'onExportToolpathComplete',
    NEW_PROJECT: 'onNewProject',
    TRAY_CREATED: 'onTrayCreated',
    TRAY_DELETED: 'onTrayDeleted',
    ADD_PRINTER: 'onAddPrinter',
    MB_RESET_TO_DEFAULT: 'onResetPrintSettingToDefault',
    MB_RESET_ALL_TO_DEFAULT: 'onResetAllPrintSettingsToDefault',
    MB_REMOVE_ALL_SETTINGS: 'onRemoveAllPrintSettings',
    MB_CLICK_ACCOUNT_ICON: 'onAccountIconClick',
    MB_PRINT_MODE_MENU_OPENED: 'onPrintModeMenuOpened',
    MB_LOG_IN: 'onLogIn',
    [Dispatch.AcknowledgeCompleted]: 'onAcknowledgePrintCompleted',
    [Dispatch.Suspend]: 'onPrintSuspend',
    [Dispatch.Resume]: 'onPrintResume',
    [Dispatch.Cancel]: 'onPrintCancel',
    [Dispatch.PrintAgain]: 'onPrintAgain',
    [BotLogsActions.CopyStart]: 'onBotLogsCopyStart',
    [BotLogsActions.CopySuccess]: 'onBotLogsCopySuccess',
    [PrintSettingsActions.CloseModal]: 'onCloseModal',
    [PrintSettingsActions.DestroyPreset]: 'onDestroyPreset',
    [PrintSettingsActions.UpdatePreset]: 'onUpdatePreset',
    [PrintSettingsActions.CreatePreset]: 'onCreatePreset',
    [PrintSettingsActions.ReadPresets]: 'onReadPresets',
    [PrintSettingsActions.WritePreset]: 'onWritePreset',
    [PrintSettingsActions.RenamePreset]: 'onRenamePreset',
    [PrintSettingsActions.ToggleProp]: 'onAdvancedToggle'
  },

  getTrackingService() {
    return this.flux.getService('TrackingService');
  },

  getPrinter() {
    return this.flux.store('PrinterStore').getSelectedPrinter();
  },

  onPreparePrint({
    jobType,
    filteredSettings
  }) {
    switch (jobType) {
      case JOB_TYPE.PRINT:
        this.getTrackingService().addEvent(R.merge({
          name: 'Print Slice Started'
        }, filteredSettings));
        break;

      case JOB_TYPE.EXPORT:
        this.getTrackingService().addEvent(R.merge({
          name: 'Export Started'
        }, filteredSettings));
        break;

      default:
        this.getTrackingService().addEvent({
          name: 'Preview Slice Started'
        });
    }
  },

  onBuildFailed({
    jobType,
    filteredSettings,
    err
  }) {
    if (err && err.type === 'cancel') {
      switch (jobType) {
        case JOB_TYPE.PRINT:
          this.getTrackingService().addEvent(R.merge({
            name: 'Print Slice Cancelled Automatic'
          }, filteredSettings));
          break;

        case JOB_TYPE.EXPORT:
          this.getTrackingService().addEvent(R.merge({
            name: 'Export Cancelled Automatic'
          }, filteredSettings));
          break;

        default:
          this.getTrackingService().addEvent({
            name: 'Preview Slice Cancelled Automatic'
          });
      }
    } else {
      switch (jobType) {
        case JOB_TYPE.PRINT:
          this.getTrackingService().addEvent(R.merge({
            name: 'Print Slice Failed',
            error: err
          }, filteredSettings));
          break;

        case JOB_TYPE.EXPORT:
          this.getTrackingService().addEvent(R.merge({
            name: 'Export Failed',
            error: err
          }, filteredSettings));
          break;

        default:
          this.getTrackingService().addEvent({
            name: 'Preview Slice Failed',
            error: err
          });
      }
    }
  },

  onBuildSucceeded({
    jobType,
    filteredSettings
  }) {
    switch (jobType) {
      case JOB_TYPE.PRINT:
        this.getTrackingService().addEvent(R.merge({
          name: 'Print Slice Completed'
        }, filteredSettings));
        break;

      case JOB_TYPE.EXPORT:
        this.getTrackingService().addEvent(R.merge({
          name: 'Export Completed'
        }, filteredSettings));
        break;

      default:
        this.getTrackingService().addEvent({
          name: 'Preview Slice Completed'
        });
    }
  },

  onExportToolpath({
    filteredSettings
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: 'Export Started'
    }, filteredSettings));
  },

  onExportToolpathComplete({
    filteredSettings,
    error
  }) {
    const err = error ? {
      error
    } : {};
    this.getTrackingService().addEvent(R.mergeAll([{
      name: `Export ${error ? 'Failed' : 'Completed'}`
    }, filteredSettings, err]));
  },

  onNewProject() {
    this.getTrackingService().addEvent({
      name: 'New Project Created'
    });
  },

  onTrayCreated() {
    this.getTrackingService().addEvent({
      name: 'New Tray Added'
    });
  },

  onTrayDeleted() {
    this.getTrackingService().addEvent({
      name: 'Tray Deleted'
    });
  },

  onAddPrinter() {
    this.getTrackingService().addEvent({
      name: 'Add Printer From Network Printers Panel'
    });
  },

  onAcknowledgePrintCompleted({
    printer
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: 'Print Done Clicked'
    }, printer.getTrackingInfo()));
  },

  onPrintSuspend({
    printer
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: 'Print Paused'
    }, printer.getTrackingInfo()));
  },

  onPrintResume({
    printer
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: 'Print Resumed'
    }, printer.getTrackingInfo()));
  },

  onPrintCancel({
    printer,
    currentProcess
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: `${currentProcess.name} Cancelled`
    }, printer.getTrackingInfo()));
  },

  onPrintAgain({
    printer
  }) {
    this.getTrackingService().addEvent(R.merge({
      name: 'Print Again'
    }, printer.getTrackingInfo()));
  },

  onBotLogsCopyStart() {
    this.getTrackingService().addEvent({
      name: 'Copy Printer Logs Started'
    });
  },

  onBotLogsCopySuccess() {
    this.getTrackingService().addEvent({
      name: 'Copy Printer Logs Completed'
    });
  },

  onResetPrintSettingToDefault({
    key
  }) {
    this.getTrackingService().addEvent({
      name: 'Reset To Default',
      setting: key
    });
  },

  onResetAllPrintSettingsToDefault() {
    this.getTrackingService().addEvent(R.merge({
      name: 'Reset All Settings to Default'
    }, this.getPrinter().getTrackingInfo()));
  },

  onRemoveAllPrintSettings() {
    this.getTrackingService().addEvent(R.merge({
      name: 'Remove All Settings'
    }, this.getPrinter().getTrackingInfo()));
  },

  onAccountIconClick() {
    this.getTrackingService().addEvent({
      name: 'Sidebar Link Clicked Account'
    });
  },

  onPrintModeMenuOpened() {
    this.getTrackingService().addEvent({
      name: 'Print Mode Menu Opened'
    });
  },

  onLogIn({
    userId
  }) {
    this.getTrackingService().setUserId(userId);
    this.getTrackingService().addEvent({
      name: 'Log In'
    });
  },

  onCloseModal() {
    const {
      printMode,
      currentPreset
    } = this.flux.store('PrintSettingsStore').getState();
    const currentPrintMode = currentPreset || printMode.selected;
    this.getTrackingService().addEvent({
      name: 'Print Mode Selected',
      printMode: currentPrintMode,
      isCustomPrintMode: currentPreset != null
    });
  },

  onDestroyPreset({
    name
  }) {
    this.getTrackingService().addEvent({
      name: 'Print Mode Removed',
      printMode: name
    });
  },

  onUpdatePreset({
    name
  }) {
    this.getTrackingService().addEvent({
      name: 'Print Mode Saved',
      printMode: name
    });
  },

  onCreatePreset({
    name,
    originalPrintMode
  }) {
    this.getTrackingService().addEvent({
      name: 'Print Mode Created',
      printMode: name,
      originalPrintMode
    });
  },

  onReadPresets() {
    this.getTrackingService().addEvent({
      name: 'Print Mode Imported'
    });
  },

  onWritePreset({
    name
  }) {
    this.getTrackingService().addEvent({
      name: 'Print Mode Exported',
      printMode: name
    });
  },

  onRenamePreset({
    name
  }) {
    this.getTrackingService().addEvent({
      name: 'Print Mode Renamed',
      printMode: name
    });
  },

  onAdvancedToggle({
    key
  }) {
    const PrintSettingsStore = this.flux.store('PrintSettingsStore');
    const prevValue = R.propOr(false, key, PrintSettingsStore.getVisibleSettings());
    const eventName = `Setting Visibility ${prevValue ? 'Enabled' : 'Disabled'}`;
    this.getTrackingService().addEvent({
      name: eventName,
      setting: key
    });
  }

});
