// MB-support-plugin/lib/stores/build_estimates_store.js
/**
 * @module BuildEstimateStore
 * @desc
 * The Build Estimate Store:
 * - listens to TOOLPATH_UNZIPPED action & opens the estimates modal here in the store.
 *
 * NOTE : BuildEstimateStore is mb only & is different from eagle-print/stores/buildStore's BUILD_TYPES.ESTIMATE
 *
 *
 * @author Ritika Shandilya
 */
'use strict';

const Fluxxor = require('fluxxor');

const Store = Fluxxor.createStore({
  actions: {
    TOOLPATH_UNZIPPED: 'onToolPathUnzipped',
    IMPORT_TOOLPATH_UNZIPPED: 'onToolPathUnzipped',
    CLEAR_BUILDS: 'onClearBuilds'
  },

  /**
   * @typedef {Object} BuildEstimateStore
   */
  initialize() {
    this._currentEstimate = null;
  },

  getCurrentEstimate() {
    return this._currentEstimate;
  },

  setCurrentEstimate(newEstimate) {
    this._currentEstimate = newEstimate;
  },

  onToolPathUnzipped(payload) {
    const extrusionDistances = payload.unzipped.metaData.extrusion_distances_mm;
    const extruderCount = extrusionDistances.length < 2 || extrusionDistances[1] <= 0 ? 1 : extrusionDistances.length;
    const estimateInfo = {
      time: payload.unzipped.metaData.duration_s,
      mass: payload.unzipped.metaData.extrusion_masses_g,
      distance: payload.unzipped.metaData.extrusion_distances_mm,
      botType: payload.unzipped.metaData.bot_type,
      material: payload.unzipped.metaData.materials,
      extruderCount
    };
    this.setCurrentEstimate({
      estimateInfo,
      task: payload.task
    });
  },

  onClearBuilds() {
    this.setCurrentEstimate(null);
  }

});
module.exports = Store;
