// MB-support-plugin/lib/stores/copyBotLogs.js
'use strict';

const R = require('ramda');

const {
  createStore
} = require('fluxxor');

const {
  Actions,
  CHANGE_EVENT
} = require('../constants/copyBotLogs');

const {
  getLogging,
  safeKey,
  getSafeKeyProp,
  setSafeKeyProp,
  removeSafeKeyProp,
  appendSafeKey,
  removeSafeKey
} = require('../util'); // A map whose keys are a printer's id (string) and a path to a temp file on
// the user's computer (string) after the transfer from the bot.


let logsCache = {}; // A list of printer id's that are currently copying bot logs to the user's
// computer

let isCopyingList = []; // A map whose keys are a printer's id (string) and any error that occurred
// while zipping, transferring, saving or sending the bot logs

let errors = {};

const isCopying = printerId => isCopyingList.includes(safeKey(printerId));

module.exports = createStore({
  actions: {
    [Actions.CopyStart]: 'onCopyStart',
    [Actions.CopySuccess]: 'onCopySuccess',
    [Actions.CopyFail]: 'onCopyFail',
    [Actions.SendToSupportSuccess]: 'onSendToSupportSuccess',
    [Actions.SendToSupportFail]: 'onSendToSupportFail',
    [Actions.Save]: 'onSave',
    [Actions.NoFile]: 'onNoFile',
    [Actions.Timeout]: 'onTimeout'
  },

  getTmpPath(printerId) {
    return getSafeKeyProp(printerId, logsCache);
  },

  getState(printerId) {
    return {
      printerId,
      tmpPath: getSafeKeyProp(printerId, logsCache),
      isCopying: isCopying(printerId),
      error: getSafeKeyProp(printerId, errors)
    };
  },

  onCopyStart({
    printerId
  }) {
    isCopyingList = R.uniq(appendSafeKey(printerId, isCopyingList));
    errors = removeSafeKeyProp(printerId, errors);
    this.emit(CHANGE_EVENT);
  },

  onCopySuccess({
    printerId,
    tmpPath
  }) {
    logsCache = setSafeKeyProp(printerId, tmpPath, logsCache);
    isCopyingList = removeSafeKey(printerId, isCopyingList);
    errors = removeSafeKeyProp(printerId, errors);
    this.emit(CHANGE_EVENT);
  },

  onCopyFail({
    printerId,
    err
  }) {
    const logger = getLogging(this.flux);
    errors = setSafeKeyProp(printerId, err, errors);
    isCopyingList = removeSafeKey(printerId, isCopyingList);
    logger.error(err);
    this.emit(CHANGE_EVENT);
  },

  onSendToSupportSuccess({
    printerId
  }) {
    errors = removeSafeKeyProp(printerId, errors);
    this.emit(CHANGE_EVENT);
  },

  onSendToSupportFail({
    printerId,
    err
  }) {
    const logger = getLogging(this.flux);
    errors = setSafeKeyProp(printerId, err, errors);
    logger.error(err);
    this.emit(CHANGE_EVENT);
  },

  onSave({
    printerId
  }) {
    errors = removeSafeKeyProp(printerId, errors);
    this.emit(CHANGE_EVENT);
  },

  onNoFile({
    printerId,
    tmpPath
  }) {
    const logger = getLogging(this.flux);
    const err = new Error('The temporary file no longer exists. Try again.');
    logsCache = removeSafeKeyProp(printerId, logsCache);
    isCopyingList = removeSafeKey(printerId, isCopyingList);
    errors = setSafeKeyProp(printerId, err, errors);
    logger.error(err, tmpPath);
    this.emit(CHANGE_EVENT);
  },

  onTimeout({
    printerId,
    err
  }) {
    const logger = getLogging(this.flux); // If we're not currently copying for this printer, then just
    // move along. No reason to record the error here.

    if (!isCopying(printerId)) return;
    logsCache = removeSafeKeyProp(printerId, logsCache);
    isCopyingList = removeSafeKey(printerId, isCopyingList);
    errors = setSafeKeyProp(printerId, err, errors);
    logger.error(err);
    this.emit(CHANGE_EVENT);
  }

});
