// MB-support-plugin/lib/stores/networkPreferences.js
'use strict';

const Fluxxor = require('fluxxor');

const _ = require('lodash');

const {
  maybe,
  just,
  nothing,
  isNothing,
  getProp,
  getPropLift
} = require('../utils/maybe');

const {
  getPrinter
} = require('../utils/printerDetails');

const {
  PrinterConnTypeEnum
} = require('../constants');

const {
  Actions,
  CHANGE_EVENT,
  NetworkStateEnum
} = require('../constants/networkPreferences'); // Network State


let networkState = nothing();
let availableNetworks = [];
let selectedNetwork = nothing();
let prevSelectedNetwork = nothing(); // Messaging

let currentError = nothing();
let statusMsg = nothing();
let currentMessage = nothing(); // Progress indication:

let waiting = false; // WIFI Password

let passwordAttempt = '';
let showPassword = false; // Static IP Edit containers:
// When the user enables static IP, we need the real values from the printer,
// but changes don't take effect until the user hits 'apply', thus we need
// to have containers to edit the values without changing the preceived state
// of the printer.
// NOTE: There is currently NO VALIDATION as per the spec, that may change...

let editStaticIPAddress = null;
let editNetmask = null;
let editGateway = null;
let editDNSServers = null;
let newDNSItem = '';
let selectedDNSItem = nothing();
const Store = Fluxxor.createStore({
  actions: {
    [Actions.Waiting]: 'onWaiting',
    [Actions.Connecting]: 'onConnecting',
    [Actions.WifiConnected]: 'onWifiConnected',
    [Actions.EthernetConnected]: 'onEthernetConnected',
    [Actions.WifiReady]: 'onWifiReady',
    [Actions.Unconnected]: 'onUnconnected',
    [Actions.SelectNetwork]: 'onSelectNetwork',
    [Actions.ChangePasswordAttempt]: 'onChangePasswordAttempt',
    [Actions.TogglePasswordVisibility]: 'onTogglePasswordVisibility',
    [Actions.WifiForget]: 'onWifiForget',
    [Actions.PrepopulateStaticIPFields]: 'onPrepopulateStaticIPFields',
    [Actions.ChangeStaticIPAddress]: 'onChangeStaticIPAddress',
    [Actions.ChangeNetmask]: 'onChangeNetmask',
    [Actions.ChangeGateway]: 'onChangeGateway',
    [Actions.ChangeDNSServers]: 'onChangeDNSServers',
    [Actions.SelectDNSItem]: 'onSelectDNSItem',
    [Actions.NewDNSItem]: 'onNewDNSItem',
    [Actions.SetStaticIP]: 'onSetStaticIP',
    [Actions.Error]: 'onError',
    [Actions.ClearError]: 'onClearError'
  },

  getState() {
    const networkProp = prop => getPropLift(prop, networkState);

    const getDNS = getDNSServers(selectedDNSItem.lift());
    const currentPath = getPropLift('path', getNetworkByName(networkProp('name')));
    const offline = networkProp('state') === NetworkStateEnum.state.OFFLINE;
    const connectedByEthernet = networkProp('state') === NetworkStateEnum.state.ETHERNET; // we need two different "can edit" flags, one for wifi, and the other
    // for static IP. They both share some common attributes, so I broke
    // that out into a single flag. That's why it's the "base".

    const canEditBase = printer => printer.getConnectionType() === PrinterConnTypeEnum.USB && !isNothing(networkState) && !waiting;

    const canEditWifi = printer => canEditBase(printer) && !connectedByEthernet; // offline but with network selected, or connected by ethernet


    const canEditStaticIP = printer => canEditBase(printer) && (offline && !isNothing(selectedNetwork) || connectedByEthernet); // we'll be enabling some buttons depending on whether this stuff
    // has been edited by the user


    const staticIPEdited = !!(editStaticIPAddress || editNetmask || editGateway || editDNSServers);
    const staticIPAddress = getPropLift('ip', networkState) || getPropLift('ip', selectedNetwork);
    const netmask = getPropLift('netmask', networkState) || getPropLift('netmask', selectedNetwork);
    const gateway = getPropLift('gateway', networkState) || getPropLift('gateway', selectedNetwork);
    const dnsServers = getDNS(editDNSServers, getPropLift('dns', selectedNetwork), getPropLift('dns', networkState));
    return {
      networkState,
      availableNetworks,
      selectedNetwork,
      currentPath,
      currentError,
      statusMsg,
      currentMessage,
      prevSelectedNetwork,
      enableWifi: networkProp('wifi') === 'enabled',
      waiting,
      passwordAttempt,
      showPassword,
      canEditWifi,
      canEditStaticIP,
      staticIPEdited,
      enableStaticIP: networkProp('static'),
      dnsServers,
      newDNSItem,
      staticIPAddress,
      netmask,
      gateway,
      editStaticIPAddress,
      editNetmask,
      editGateway
    };
  },

  onWaiting() {
    waiting = true;
    return this.emit(CHANGE_EVENT);
  },

  onWifiConnected(payload) {
    networkState = maybe(payload.networkState);
    selectedNetwork = nothing();
    availableNetworks = payload.availableNetworks || [];
    statusMsg = just('mb:network.status.connected');
    currentError = nothing();
    waiting = false;

    this._resetStaticIPEdits();

    return this.emit(CHANGE_EVENT);
  },

  onEthernetConnected(payload) {
    const connType = getPrinter(this.flux, payload.printerId).getConnectionType();
    prevSelectedNetwork = selectedNetwork;
    selectedNetwork = nothing();
    networkState = maybe(payload.networkState);
    availableNetworks = payload.availableNetworks || [];
    statusMsg = just('mb:network.status.connectedByEthernet');
    waiting = false;
    currentMessage = connType === PrinterConnTypeEnum.USB ? just('mb:network.messages.connectedByEthernetAndUsb') : just('mb:network.messages.connectedByEthernet');

    this._resetStaticIPEdits();

    return this.emit(CHANGE_EVENT);
  },

  onWifiReady(payload) {
    prevSelectedNetwork = selectedNetwork;
    networkState = maybe(payload.networkState);
    availableNetworks = payload.availableNetworks || [];
    statusMsg = just('mb:network.status.notConnected');
    currentError = nothing();
    currentMessage = nothing();
    passwordAttempt = '';
    waiting = false;
    return this.emit(CHANGE_EVENT);
  },

  onUnconnected(payload) {
    prevSelectedNetwork = selectedNetwork;
    availableNetworks = payload.availableNetworks || [];
    selectedNetwork = getNetworkByName(getProp('name', networkState));
    networkState = maybe(payload.networkState);
    statusMsg = just('mb:network.status.unconnected');
    currentError = nothing();
    currentMessage = just('mb:network.messages.unconnected');
    waiting = false;
    return this.emit(CHANGE_EVENT);
  },

  onSelectNetwork({
    path
  }) {
    prevSelectedNetwork = selectedNetwork;
    selectedNetwork = getNetworkByPath(path);
    return this.emit(CHANGE_EVENT);
  },

  onChangePasswordAttempt({
    password
  }) {
    passwordAttempt = password;
    return this.emit(CHANGE_EVENT);
  },

  onTogglePasswordVisibility() {
    showPassword = !showPassword;
    return this.emit(CHANGE_EVENT);
  },

  onWifiForget() {
    prevSelectedNetwork = selectedNetwork.map(network => network.value.password = 'required');
    selectedNetwork = prevSelectedNetwork;
    return this.emit(CHANGE_EVENT);
  },

  onSetStaticIP({
    dnsServers,
    enable
  }) {
    editDNSServers = dnsServers;
    newDNSItem = null;
    return this.emit(CHANGE_EVENT);
  },

  onPrepopulateStaticIPFields(result) {
    editStaticIPAddress = result.ip;
    editNetmask = result.netmask;
    editGateway = result.gateway;
    editDNSServers = result.dns;
    return this.emit(CHANGE_EVENT);
  },

  onChangeStaticIPAddress({
    ipAddress
  }) {
    editStaticIPAddress = ipAddress;
    return this.emit(CHANGE_EVENT);
  },

  onChangeNetmask({
    netmask
  }) {
    editNetmask = netmask;
    return this.emit(CHANGE_EVENT);
  },

  onChangeGateway({
    gateway
  }) {
    editGateway = gateway;
    return this.emit(CHANGE_EVENT);
  },

  onNewDNSItem({
    item
  }) {
    newDNSItem = item;
    return this.emit(CHANGE_EVENT);
  },

  onChangeDNSServers({
    dnsServers
  }) {
    editDNSServers = dnsServers.filter(s => _.size(s) > 0);
    newDNSItem = '';
    return this.emit(CHANGE_EVENT);
  },

  onSelectDNSItem({
    item
  }) {
    selectedDNSItem = maybe(item);
    return this.emit(CHANGE_EVENT);
  },

  onError({
    error
  }) {
    // ridiculous error handling, because errors here can come in a variety
    // of shapes and sizes...
    const parsedErr = parseJSON(error.message || error);
    const err = errorMsg(parsedErr) || parsedErr;
    waiting = false;
    currentError = just(err);
    console.error(error.message);
    this.emit(CHANGE_EVENT);
  },

  onClearError() {
    currentError = nothing();
    this.emit(CHANGE_EVENT);
  },

  _resetStaticIPEdits() {
    editStaticIPAddress = null;
    editNetmask = null;
    editGateway = null;
    editDNSServers = null;
  }

});

const getNetworkByPath = path => availableNetworks.reduce((acc, v) => v.path === path ? just(v) : acc, nothing());

const getNetworkByName = name => availableNetworks.reduce((acc, v) => v.name === name ? just(v) : acc, nothing());

const getDNSServers = selectedItem => (...serverLists) => {
  if (serverLists.length === 0) return [];
  const recurse = getDNSServers(selectedItem);
  const [head, ...tail] = serverLists;
  if (head != null) return head.map(item => ({
    label: item,
    value: item,
    selected: selectedItem === item
  }));
  return recurse(...tail);
};

const parseJSON = msg => {
  try {
    return JSON.parse(msg);
  } catch (e) {
    return msg;
  }
};

const errorMsg = _.property('error.message');

module.exports = Store;
