// MB-support-plugin/lib/stores/printerDetailsStore.js
/* eslint-disable no-unreachable */
'use strict';

const Fluxxor = require('fluxxor');

const _ = require('lodash');

const Immutable = require('immutable');

const {
  SWIPE_LEFT,
  SWIPE_RIGHT
} = require('eagle-print/views/ui/card_container');

const {
  either
} = require('../utils/maybe');

const {
  Dispatch,
  PanelStates,
  ActionButtons,
  ProgressButtons
} = require('../constants/printerDetails');

const {
  getPrinter,
  printerStatus,
  currentPrinterProcess,
  printerDetailsStatus
} = require('../utils/printerDetails'); // NOTE: Since printers are already stored in PrinterStore, there's no reason
// to be redundant here. Most of the state coming out of the printers is
// stored there. For the most part, we're just handing MB specific events
// and returning formatted data for the UI.


let currentPrinterID = '';
let currentPanel = PanelStates.Details;
let panelSwipeDirection = SWIPE_LEFT; // which action buttons are in a "waiting" (spinner) state

let waitingButtons = Immutable.Map();
const CHANGE = 'change';
module.exports = Fluxxor.createStore({
  actions: {
    [Dispatch.AcknowledgeCompleted]: 'onAcknowledgeCompleted',
    [Dispatch.AcknowledgeError]: 'onAcknowledgeError',
    [Dispatch.AcknowledgeResumeError]: 'onAcknowledgeResumeError',
    [Dispatch.AcknowledgeFailure]: 'onAcknowledgeFailure',
    [Dispatch.AcknowledgeSuccess]: 'onAcknowledgeSuccess',
    [Dispatch.Suspend]: 'onSuspend',
    [Dispatch.Resume]: 'onResume',
    [Dispatch.Cancel]: 'onCancel',
    [Dispatch.StopFilament]: 'onStopFilament',
    [Dispatch.PrintAgain]: 'onPrintAgain',
    [Dispatch.ShowPanelDetails]: 'onShowPanelDetails',
    [Dispatch.ShowPanelFilament]: 'onShowPanelFilament',
    [Dispatch.ShowPanelUtilities]: 'onShowPanelUtilities',
    [Dispatch.System]: 'onSystem',
    [Dispatch.State]: 'onState',
    [Dispatch.Status]: 'onStatus',
    [Dispatch.ButtonWait]: 'onButtonWait',
    [Dispatch.ExtruderChange]: 'onExtruderChange'
  },

  getState(printerID) {
    currentPrinterID = printerID;
    const actions = this.flux.actions.PrinterDetailsActions;
    const printer = getPrinter(this.flux, printerID);
    const pStatus = printerStatus(printer);
    const detailsStatus = printerDetailsStatus(printer);
    const currentProcess = currentPrinterProcess(printer);
    const progressButtons = currentProcess.apply(cp => ProgressButtons(printer.getId()).get(cp.step));
    const extruders = printer.getExtruderInfo(); // if there is a current process, use the i18n from that,
    // otherwise use the state

    const progressText = either(pStatus.apply(labelStatus), currentProcess.apply(labelProcess)).map(extruderDetached(extruders)).lift();

    const waitList = _.flatten(waitingButtons.reduce((acc, v) => acc.concat(v), []));

    const actionButtons = formatButtons(printerID, detailsStatus, actions, waitList);
    return {
      detailsStatus,
      currentProcess,
      currentPanel,
      panelSwipeDirection,
      actionButtons,
      progressText,
      progressButtons
    };
  },

  // take a particular set of buttons out of their waiting state
  _unwaitButton(type) {
    waitingButtons = waitingButtons.set(type, []);
  },

  onAcknowledgeCompleted(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onAcknowledgeError(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onAcknowledgeResumeError(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onAcknowledgeFailure(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onAcknowledgeSuccess(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onSuspend(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onResume(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onCancel(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onStopFilament(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onPrintAgain(payload, type) {
    this._unwaitButton(type);

    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onShowPanelDetails(payload, type) {
    this._unwaitButton(type);

    currentPanel = PanelStates.Details;
    panelSwipeDirection = SWIPE_RIGHT;
    return this.emit(CHANGE);
  },

  onShowPanelFilament(payload, type) {
    this._unwaitButton(type);

    currentPanel = PanelStates.Filament;
    panelSwipeDirection = SWIPE_LEFT;
    return this.emit(CHANGE);
  },

  onShowPanelUtilities(payload, type) {
    this._unwaitButton(type);

    currentPanel = PanelStates.Utilities;
    panelSwipeDirection = SWIPE_LEFT;
    return this.emit(CHANGE);
  },

  onSystem(payload) {
    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onState(payload) {
    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onStatus(payload) {
    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  // set a particular button into its "waiting" (spinner) state
  onButtonWait(payload) {
    const waiting = waitingButtons.get(payload.action, []);
    waiting.push(payload.key);
    waitingButtons = waitingButtons.set(payload.action, waiting);
    this.emit(CHANGE);
  },

  onExtruderChange(payload) {
    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  },

  onNetworkChange(payload) {
    return payload.id === currentPrinterID ? this.emit(CHANGE) : null;
  }

});

const formatButtons = (printerID, status, actions, waitList) => ActionButtons.map((v, k) => _.merge({}, v, {
  icon: v.icon(status),
  enabled: v.enabled(status),
  shouldShow: v.shouldShow(status),
  onClick: v.onClick(actions, printerID),
  waiting: waitList.includes(k)
})).filter(b => b.shouldShow || b.default) // include defaults for now
.groupBy(b => b.order).map(g => g.size > 1 ? g.filter(b => !b.default) : g) // remove defaults
.flatten().sortBy(b => b.order); // helper to set default values for i18n


const i18nOpts = val => ({
  defaultValue: val
});

const processI18nKey = cp => `mb:printerStatus.${cp.name}.${cp.step.toLowerCase()}`;

const labelProcess = cp => i18n => ({
  label: i18n.t(`${processI18nKey(cp)}.label`, i18nOpts(_.startCase(cp.step))),
  description: i18n.t(`${processI18nKey(cp)}.description`, i18nOpts(''))
});

const stateI18nKey = state => `mb:printerStatus.state.${state.toLowerCase()}`;

const labelStatus = status => (i18n, printer) => {
  // Not the best way to handle this fire_e copy change for auth'ing -- Ray
  const isSixthGen = printer && printer.isSixthGen();
  let i18nStrDescription;

  if (['authenticating', 'confirm'].includes(status.state.toLowerCase()) && isSixthGen) {
    i18nStrDescription = `${stateI18nKey(status.state)}.fire_e` + `.description`;
  } else {
    i18nStrDescription = `${stateI18nKey(status.state)}.description`;
  }

  return {
    label: i18n.t(`${stateI18nKey(status.state)}.label`, i18nOpts(_.startCase(status.state))),
    description: i18n.t(i18nStrDescription, i18nOpts(''))
  };
}; // basically, for the most part, checking status and process is enough
// to get all the possible text states for the panel, except for if
// the extruder has been detached. So, this function will transform
// the progressText in just such an event, otherwise it just passes it
// on through.


const extruderDetached = extruders => progressText => {
  const extrudersPresent = extruders.map(extruder => extruder.tool_present);

  switch (extrudersPresent.length) {
    case 1:
      if (!extrudersPresent[0]) {
        return i18n => ({
          label: i18n.t('mb:errors.extruder_detached.title'),
          description: i18n.t('mb:errors.extruder_detached.message')
        });
      } else {
        return progressText;
      }

      break;

    case 2:
      if (extrudersPresent[0] && extrudersPresent[1]) {
        return progressText;
      } else if (extrudersPresent[0] != extrudersPresent[1]) {
        //logical equivalent to XOR
        const slotMissingExtruder = extrudersPresent[1] ? 1 : 2;
        return i18n => ({
          label: i18n.t('mb:errors.extruder_detached.title'),
          description: i18n.t('mb:errors.extruder_detached.missing_one_message', {
            extruderNumber: slotMissingExtruder
          })
        });
      } else {
        return i18n => ({
          label: i18n.t('mb:errors.extruder_detached.title'),
          description: i18n.t('mb:errors.extruder_detached.missing_both_message')
        });
      }

      break;

    default:
      return progressText;
  }
};
