// MB-support-plugin/lib/stores/printSettingsStore.js
/**
 * @module PrintSettingsStore
 * @desc
 * The Print Settings Store attempt to simplify the tedious business of adding overrides to print settings. Using the
 * store is fairly straight forward, however the way it's put together under the hood may be a little confusing at
 * first glance. Fortunately, I'm here to explain.
 *
 * @example
 * class myComponent extends FluxComponent {
 *   constructor(props) {
 *      super(props);
 *      this._store = this.getFlux().store('PrintSettingsStore');
 *      this.state = this._store.getState();
 *   }
 *
 *   componentDidMount() {
 *      this._store.on("change", () => this.setState(this._store.getState()));
 *   }
 *
 *   render() {
 *      // render only active settings and their values
 *      return (
 *          <ul>
 *              {
 *                  R.pipe(
 *                      R.pickBy(propOr(false, 'active')),
 *                      R.mapObjIndexed((v, k) => <li>{v.name}: {v.value})
 *                  )(this.state.settings)
 *              }
 *          </ul>
 *      );
 *   }
 * }
 *
 * @author Jon Laing & Wenson Tsai
 */

/**
 * NOTE!
 * Every setting that Slice Config/Miracle Grue slices with
 * is currently validated when saving and loading from project.json.
 * When a new setting is added,
 * we MUST UPDATE mb-plugin/mb-project-validator-provider.js
 * adding property to the schema, or else projects with these schemas
 * will not be able to load!
 */
'use strict';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const Fluxxor = require('fluxxor');

const R = require('ramda');

const _ = require('lodash');

const fs = require('fs-extra');

const q = require('q');

const {
  decodeKey
} = require('@makerbot/settingsui');

const CHANGE_EVENT = 'change';
const ONUPDATE_EXTRUDERS_COMPLETED_EVENT = 'ONUPDATE_EXTRUDERS_COMPLETED_EVENT';

const makerbot = require('../makerbot');

const L = require('../utils/printSettings/lenses');

const reducers = require('../utils/printSettings/reducers');

const format = require('../utils/printSettings/format');

const Constants = require('../constants/print_settings');

const printStoreUtils = require('../utils/printSettings/printSettingsStoreUtils');

const {
  Actions
} = Constants;

const {
  ExtruderTypeEnum,
  ExtruderToDefaultMaterial
} = require('../constants');

const {
  fileSafeString,
  getI18n
} = require('../util');

const {
  just,
  nothing
} = require('../utils/maybe');

let sanitizePreset, presetToJSON;
makerbot.getVersion().then(R.trim).then(v => {
  sanitizePreset = reducers.sanitizePreset(v);
  presetToJSON = reducers.presetToJSON(v);
}).done();
/**
 * Dictionary of all the settings mappings for the Print Settings Panel.
 * We use this so we don't have to keep passing strings around and then
 * type them wrong, causing them to fail silently.
 * @constant {Object<string, string[]>}
 * @private
 */

const SettingsKeys = Constants.SettingsKeys;
const PresetKeyBlacklist = Constants.PresetKeyBlacklist; /////////// STORED VALUES ///////////////

/**
 * This holds all the possible override settings that comes
 * from the Slice Config. This is NOT for holding values. The
 * user should not have any ability to change this. The only
 * time this should change is if the printer changed. We keep this
 * separate from the user-bound data to make sure the schema of possible
 * overrides stays clean. We do this in hopes of avoiding bugs caused
 * by too many confusing tree traversals and inline mutations.
 * @see Store.getState()
 * @private
 * @type {Object<string, any>}
 */

let availableSettings = {};
/**
 * This indicates whether a particular override has been
 * activated by the user. This becomes more powerful when
 * it is combined with `availableOverrides` and `values`.
 * @see Store.getState()
 * @example
 * R.prop('infill', visibleSettings) == true;
 * @type {Object<string, boolean>}
 * @private
 */

let visibleSettings = {};
/**
 * This holds that values of the active settings.
 * @see Store.getState()
 * @example
 * R.prop('layer_height', values); // 0.2
 * @type {Object<string, any>}
 * @private
 */

let values = {};
let supportedExtruders = []; // Lists supported extruders per extruder-index
// ie. [[mk13, mk13_impla, mk13_experimental]]  ** 5th Gen Printers **
// ie. [[mk14, mk14_hot], [mk14_s, mk14_hot_s]] ** 6th Gen Printers **

let supportedExtrudersPerIndex = []; // Lists supported extruder pair configurations
// ie. [[mk13], [mk13_impla], [mk13_experimental]]  ** 5th Gen Printers **
// ie. [[mk14, mk14_s], [mk14_hot, mk14_hot_s]]     ** 6th Gen Printers **

let supportedExtruderPairConfigs = [];
let selectedMaterials = [];
let supportedMaterials = [];
let materialTranslations = {};
let spoolMaterials = [];
let printMode = null;
let supportedPrintModes = [];
/**
 * This array will hold the x,y coordinates for the purgeWallCenter.
 * For 4th & 5th gen printers, purgeWallCenter will be an empty array [].
 *
 * For 6th gen printers, the purgeWallCenter will either be initial_value
 * -or- the value return by the viewer
 *
 * -- Ray
 **/

let purgeWallCenter = [];
/**
 * Flag to determine if the settings have been loaded from user_config.json
 * @type {boolean}
 * @private
 */

let settingsLoaded = false;
let firstTrayLoaded = false;
let supportedConfigurations = [];
const PRESET_EXT = Constants.PRESET_EXT; // User presets are held here in memory, copied from the user config

let presets = []; // the current preset the user has selected. It will not be set if the user
// hasn't set anything

let currentPreset = null;
let presetError = nothing(); //////////// END STORED VALUES ///////////

module.exports = Fluxxor.createStore({
  actions: {
    [Actions.ToggleProp]: 'onAdvancedToggle',
    [Actions.BulkToggleProp]: 'onBulkAdvancedToggle',
    [Actions.ResetToDefault]: 'resetToDefault',
    [Actions.GetPrinterSchema]: 'onGetPrinterSchema',
    [Actions.UpdateProperty]: 'onUpdateProperty',
    [Actions.SelectPrinter]: 'onSelectPrinter',
    [Actions.GetExtruderTypes]: 'getExtruderTypes',
    [Actions.GetSupportedConfigurations]: 'getSupportedConfigurations',
    [Actions.InitalizeValues]: 'initializeValues',
    [Actions.AuthenticationWaiting]: 'onSelectPrinter',
    [Actions.ResetAllToDefault]: 'onResetAllToDefault',
    [Actions.RemoveAllSettings]: 'onRemoveAllSettings',
    [Actions.CompareSettings]: 'compareSettings',
    [Actions.CreatePreset]: 'onCreatePreset',
    [Actions.WritePreset]: 'onWritePreset',
    [Actions.UpdatePreset]: 'onUpdatePreset',
    [Actions.DestroyPreset]: 'onDestroyPreset',
    [Actions.ReadPresets]: 'onReadPresets',
    [Actions.ApplyPreset]: 'onApplyPreset',
    [Actions.UnapplyPreset]: 'onUnapplyPreset',
    [Actions.RenamePreset]: 'onRenamePreset',
    [Actions.PresetError]: 'onPresetError',
    [Actions.CloseModal]: 'onCloseModal',
    [Actions.RevertChanges]: 'onRevertChanges',
    [Actions.UpdateExtruder]: 'onUpdateExtruder',
    [Actions.UpdateExtruderPair]: 'onUpdateExtruderPair',
    [Actions.UpdateMaterial]: 'onUpdateMaterial',
    [Actions.UpdateMaterials]: 'onUpdateMaterials',
    [Actions.UpdateSupports]: 'onUpdateSupports',
    [Actions.UpdatePrintMode]: 'onUpdatePrintMode',
    [Actions.UpdatePurgeWallCenter]: 'onUpdatePurgeWallCenter'
  },

  /**
   * @typedef {Object} PrintSettingsStoreState
   * @property {Object<String, Object>} settings - All of the settings formatted for consumption by the view
   */

  /**
   * This returns the current state fully formatted for the UI
   *
   * @example
   * // inside a React component
   * this.setState(this.getFlux().store('PrintSettingsStore').getState());
   *
   * // getting all the settings for a particular tab
   * let infillSettings = R.filter(v => v.groups.includes('infill', this.state.settings);
   *
   * // getting all active settings
   * let visibleSettings = R.filter(R.propOr(false, 'visible'), this.state.settings);
   *
   * // getting the value for a particular override
   * let val = R.path(['layer_height', 'value'], this.state.settings);
   *
   * @returns {PrintSettingsStoreState}
   */
  getState() {
    const settings = reducers.state(availableSettings, visibleSettings, values, Constants.GroupDict, Constants.Options, Constants.Format, Constants.Sanitization, Constants.SettingsTypes);
    return {
      // extruders lives outside of settings now
      // in contrast to attached_extruder0, which used to live in settings
      extruders: {
        supported: supportedExtruders,
        supportedExtrudersPerIndex,
        supportedExtruderPairConfigs
      },
      materials: {
        selected: selectedMaterials,
        supported: supportedMaterials,
        translations: materialTranslations,
        spoolMaterials
      },
      printMode: {
        selected: printMode,
        supported: supportedPrintModes
      },
      settings,
      presets,
      currentPreset,
      presetError,
      purgeWallCenter
    };
  },

  hasExtruderMaterialMismatch(supportedMaterialsFromFirmware) {
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();
    const isLiveBot = printer && printer.isConnected() && printer.isAuthenticated();
    let hasMismatch = false;

    if (isLiveBot && printer.isSixthGen()) {
      const attachedExtruders = printer.getAttachedExtruders(); // Only the newer versions of Method/Method X firmware give us supported materials
      // If we don't get that info, use the hardcoded values from _getSupportedMaterialsByExtruder()

      const supportedMaterialsByExtruder = supportedMaterialsFromFirmware.length > 0 ? supportedMaterialsFromFirmware : this._getSupportedMaterialsByExtruder(attachedExtruders, printer);

      if (spoolMaterials.length > 0 && supportedMaterialsByExtruder.length > 0) {
        spoolMaterials.map((spoolMaterial, index) => {
          if (spoolMaterial && supportedMaterialsByExtruder[index] && !supportedMaterialsByExtruder[index].includes(spoolMaterial) && attachedExtruders[index] !== ExtruderTypeEnum.mk14_e) {
            hasMismatch = true;
          }
        });
      }
    }

    return hasMismatch;
  },

  setSupportVisibility() {
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();

    if (printer && printer.isSixthGen()) {
      visibleSettings.doSupport = false;
    } else {
      visibleSettings.supportType = true;
    }
  },

  getVisibleSettings() {
    return visibleSettings;
  },

  getOverrides() {
    return _.mapKeys(_.omit(values, PresetKeyBlacklist), (val, key) => {
      return decodeKey(key);
    });
  },

  hasOverridesForPrintMode(currentPrintMode) {
    const currentPrintModeChanged = currentPrintMode === printMode;
    const hasOverride = R.pipe(R.omit(PresetKeyBlacklist), R.values, R.length, R.lt(0))(values);
    return !currentPreset && currentPrintModeChanged && hasOverride;
  },

  hasOverridesForPreset(preset) {
    return preset && currentPreset === preset && !R.equals(reducers.getPresetByName(preset, presets).settings, R.omit(PresetKeyBlacklist, values));
  },

  hasPresetWithName(name) {
    return reducers.getPresetByName(name, presets) != null;
  },

  /**
   * Fires when an advanced setting is toggled in the Advanced Settings Panel,
   * or when we update a setting in that panel.
   * It manipulates visibleSettings and then emits a change event.
   * @param {object} payload - the payload of the dispatched event. Should have `key` in it.
   * If we call it from onUpdateProperty(), it should have `visible` in it, set to true,
   * so that newly-changed settings are always made visible.
   * @private
   */
  onAdvancedToggle(payload) {
    const setting = payload.key; // If we know what value we want to set visibility to, use that. If not, just do a straightforward toggle

    const toggledOn = payload.visible !== undefined ? payload.visible : !R.propOr(false, setting, visibleSettings);
    visibleSettings = R.assoc(setting, toggledOn, visibleSettings); // So, product decided things should work in a certain way, but I think
    // they're going to regret that decision in a month or two, so I'm
    // leaving this here, commented out, so it's nice and easy to revert
    // to the intuitive way it used to work. -- xoxo, Jon
    // if(!toggledOn) {
    //     printStoreUtils.updateProjectStoreProp(setting, undefined);
    //     values = R.dissoc(setting, values);
    // }
    // emitting up here so the interface can proceed while we do some background stuff

    this.emit(CHANGE_EVENT);

    const schemaInitialValue = this._loadSchemaInitialValue();

    if (schemaInitialValue) {
      schemaInitialValue.then(() => {
        // Save Toggled (Active) settings to user_config
        // (APP-WIDE settings).
        printStoreUtils.saveToUserConfig('activeAdvSettings', visibleSettings, this.flux);
        printStoreUtils.saveToUserConfig('printSettings', values, this.flux);
        printStoreUtils.updateProjectStoreProp(SettingsKeys.Materials, selectedMaterials, true, this.flux);
        printStoreUtils.updateProjectStoreProp(SettingsKeys.PrintModes, values.printMode, true, this.flux);
      }).catch(console.error).done();
    }
  },

  /**
   * Resets setting back to Slice Config evaluated `initial_value` for setting.
   * This requires calling Slice Config with print_mode and overrides object with the
   * setting which we want `initial_value` (default) for removed from overrides.
   * getSchema will always override value as `initial_value` if there is one set.
   * @param {object} payload - the payload of the dispatched event. Should have `key` in it.
   * @private
   */
  resetToDefault(payload) {
    const setting = payload.key;
    values = printStoreUtils.persistValues(setting, undefined, false, values, this.flux);
    printStoreUtils.clearBuilds(this.flux);
    return this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)).catch(console.error);
  },

  /**
   * Fires when a property on a print setting changes
   * @param {object} payload - the payload of the dispatched event. Should have `path` and `value` in it.
   * @private
   */
  onUpdateProperty(payload) {
    const key = format.keyFromPath(payload.path);

    if (key === SettingsKeys.Supports) {
      // Reset to default with 'doPurgeWall' when switching between support types
      // Applies mainly to the sixth gen printers
      values = printStoreUtils.persistValues(SettingsKeys.PurgeWall, undefined, false, values, this.flux);
    }

    if (R.has(key, visibleSettings)) {
      if (key === SettingsKeys.PrintModes) ({
        values,
        currentPreset
      } = printStoreUtils.resetAllToDefault(this.flux, values));

      if (R.has(key, availableSettings) || key === SettingsKeys.PrintModes) {
        values = printStoreUtils.persistValues(key, payload.value, false, values, this.flux);
        this.emit(CHANGE_EVENT);

        this._loadSchemaInitialValue().done();
      }
    } else {
      // Avoid passing non print settings related properties; info.extruder, info.meshColor, etc..
      if (R.has(key, availableSettings) || key === SettingsKeys.PrintModes) {
        values = printStoreUtils.persistValues(key, payload.value, false, values, this.flux); // toggle the visibilty on print setting and run through
        // the _loadSchemaInitalValue routine

        this.onAdvancedToggle({
          key,
          visible: true
        });
      }
    }
  },

  onResetAllToDefault() {
    ({
      values,
      currentPreset
    } = printStoreUtils.resetAllToDefault(this.flux, values));
    this.emit('initialize_values');
  },

  onRemoveAllSettings() {
    visibleSettings = printStoreUtils.removeAllSettings(this.flux);
    this.setSupportVisibility();
    this.emit(CHANGE_EVENT);
  },

  onRevertChanges() {
    if (currentPreset) {
      // have to set currentPreset to null for onApplyPreset to work properly
      const presetBuf = currentPreset;
      currentPreset = null;
      this.onApplyPreset({
        name: presetBuf
      });
    } else {
      const printModeBuf = R.view(L.printMode, values);
      ({
        values,
        currentPreset
      } = printStoreUtils.resetAllToDefault(this.flux, values));
      values = printStoreUtils.persistValues(SettingsKeys.PrintModes, printModeBuf, false, values, this.flux);
      this.emit(CHANGE_EVENT);

      this._loadSchemaInitialValue().done();
    }
  },

  onBulkAdvancedToggle(payload) {
    visibleSettings = R.merge(visibleSettings, payload.active);
    values = reducers.cleanValues(values, visibleSettings);
    this.emit(CHANGE_EVENT);
  },

  onGetPrinterSchema(payload) {
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();
    const extruders = printer && printer.getAttachedExtruders();
    const {
      resetToDefaultExtruders
    } = payload;
    const isPrinterConnectedAndAuthed = printer && printer.isConnected() && printer.isAuthenticated(); // Ignore Extruder Args if offline/archetype printers and get default extruders

    const ignoreExtruderArgs = resetToDefaultExtruders && !isPrinterConnectedAndAuthed; // For archetype printers, exit out of loading schema

    if (!printer) return; // if payload has spoolsInfo, then it means spools have been changed
    // on a live printer

    if (payload.spoolsInfo) {
      this._persistValues(SettingsKeys.ChamberTemp, null, true); // this._persistValues(SettingsKeys.BuildPlaneTemp, null, true);

    }

    const ConfigService = this.flux.getService('ConfigService');
    const activeTray = this.flux.store('ProjectStore').getProject().getActiveTray();
    const tray = activeTray.getBuildInfo(printer);
    if (R.isEmpty(presets)) presets = ConfigService.get('presets') || [];
    return this.updatePrinterSpoolInfo(printer, payload) // Passing `resetToDefaultsExtruders` from 'onSelectPrinter'
    // to avoid passing extruders as args to sliceconfig
    // BUT we don't ignore extruder selections if it's a live printer
    .then(() => makerbot.getSchema(tray, printer, ignoreExtruderArgs, this.flux)).then(schema => {
      // In rare cases, if getSchema gets called too quickly, it
      // will return empty string as a result. Let's exit before
      // things error out.
      // -- Ray
      if (_.isEmpty(schema)) return; // Passing `resetToDefaultExtruders` from 'onSelectPrinter'
      // to force update `extruders` to the schema's attached extruders

      this._parseSchema(schema.getFullSchema(), resetToDefaultExtruders); // emit early so the UI can render while we do the rest of the updating


      this.emit(CHANGE_EVENT);
      printStoreUtils.updateProjectStoreProp(SettingsKeys.Extruders, extruders, true, this.flux);
      printStoreUtils.updateProjectStoreProp(SettingsKeys.Materials, selectedMaterials, true, this.flux);
      printStoreUtils.updateProjectStoreProp(SettingsKeys.PrintModes, values.printMode, true, this.flux);
    }).catch(err => console.error(err));
  },

  updatePrinterSpoolInfo(printer, payload) {
    /**
     * update spoolsMaterials if...
     *   payload contains spoolsInfo sent from makerbot-printer.js
     *   on spool material changes or on reconnect
     *   -or-
     *   invoke getSpoolInfo from printer when switching between printers
     **/
    const parseSpoolsInfo = spoolsInfo => {
      return spoolsInfo ? spoolsInfo.map((spoolInfo, bayIndex) => {
        const {
          material_name
        } = spoolInfo;
        const attachedExtruders = printer.getAttachedExtruders(); // If we have information from the spool, show that

        if (!R.isNil(material_name)) {
          return material_name;
        } else if (bayIndex === 0 && printer.isConnected() && printer.isAuthenticated()) {
          // Live bots with no spool info (for the first bay only)
          return null;
        } else {
          return ExtruderToDefaultMaterial[attachedExtruders[bayIndex]];
        }
      }) : []; // 4th & 5th Gen Printers + Sketch Printers, set to empty
    };

    if (payload && payload.spoolsInfo) {
      return q().then(() => {
        spoolMaterials = parseSpoolsInfo(payload.spoolsInfo);
      });
    }

    return printer.getSpoolInfo().then(spoolsInfo => {
      spoolMaterials = parseSpoolsInfo(spoolsInfo);
    });
  },

  /**
   * Clears all values when the printer is changed, and sets the panel to prepare to close.
   * @private
   */
  onSelectPrinter() {
    this.waitFor(['PrinterStore'], () => {
      selectedMaterials = [];
      spoolMaterials = [];
      purgeWallCenter = [];
      const settingsWithBotSpecificValues = [SettingsKeys.ChamberTemp, SettingsKeys.Supports // SettingsKeys.BuildPlaneTemp,
      ];
      settingsWithBotSpecificValues.map(setting => {
        this._persistValues(setting, null, true);
      });
      /**
       * *** HACK ***
       * Pass a flag down to makerbot.js and _parseSchema to
       * avoid using tray.bodyProperties' extruder.value when
       * switching between different 6th gen printers; ie.
       * Method => Method X.
       *
       * It will reset extruders back to defaults instead of
       * persisting as sticky extruders. This hack will maintain
       * all 5th gen's sticky extruder selection.
       *
       * -- Ray
       **/

      const printer = this.flux.store('PrinterStore').getSelectedPrinter();
      const resetToDefaultExtruders = printer && printer.isSixthGen();
      const isPrinterConnectedAndAuthed = printer && printer.isConnected() && printer.isAuthenticated();

      if (resetToDefaultExtruders && !isPrinterConnectedAndAuthed) {
        const botType = printer.getGender().botType;
        let defaultExtruders = [];

        if (botType === 'lava_f') {
          defaultExtruders = ['mk14_hot', 'mk14_hot_s'];
        } else {
          defaultExtruders = ['mk14', 'mk14_s'];
        } // Update attachedExtruders to archetype_printer


        printer.setAttachedExtruders(defaultExtruders); // Update the side-effects

        printStoreUtils.updateProjectStoreProp(SettingsKeys.Extruders, defaultExtruders, true, this.flux); // Reset materials & support settings

        this.onUpdateExtruderPair();
      }

      this.onGetPrinterSchema({
        resetToDefaultExtruders
      });
      this.setSupportVisibility();
      this.emit('close_panel');
    });
  },

  /**
   * Loads the project settings into this store and sets the active settings.
   * @param {object} filteredSettings - Project Settings that have been filtered for things we do not want.
   * @param {Immutable Map} projectSettings - Settings from the project.
   */
  loadProjectSettings(filteredSettings, projectSettings) {
    const ConfigService = this.flux.getService('ConfigService');
    let extruders;
    values = R.mapObjIndexed((v, k) => projectSettings.getProperty(k).getValue(), filteredSettings);
    const isActive = R.pipe(R.mapObjIndexed(() => true), R.omit([SettingsKeys.Extruders, SettingsKeys.Materials, SettingsKeys.PrintModes]));
    visibleSettings = isActive(values);
    ConfigService.delete('printSettings');
    ConfigService.delete('activeAdvSettings');
    ConfigService.delete('extruders');
    ConfigService.delete('materials');
    R.mapObjIndexed((v, k) => {
      if (k === 'extruders') {
        extruders = v;
      }

      if (k === 'materials') {
        selectedMaterials = v;
      }

      if (k === 'printMode') {
        printMode = v;
      }

      values = printStoreUtils.persistValues(k, v, false, values, this.flux);
    }, values);
    printStoreUtils.saveToUserConfig('activeAdvSettings', visibleSettings, this.flux);
    printStoreUtils.saveToUserConfig('printSettings', values, this.flux);
    printStoreUtils.saveToUserConfig('extruders', extruders, this.flux);
    printStoreUtils.saveToUserConfig('materials', selectedMaterials, this.flux);

    this._loadSchemaInitialValue();

    this.emit(CHANGE_EVENT);
  },

  /**
   * Compares this stores settings vs project settings and fires modal if they differ.
   * @param  {object} payload - Contains the project object, dispached from fileActions.
   */
  compareSettings(payload) {
    const i18n = this.flux.getService('TranslationService');
    const ModalService = this.flux.getService('ModalService');
    const project = payload.project;
    const activeTray = project.getActiveTray();
    const trayBodies = activeTray.getTrayBodies();
    const bodyProperties = trayBodies.map(body => body.getOccurrence().getProperties()).toArray();
    const projectSettings = bodyProperties[0]; // Remove all references to 'defaultSupportMaterial' from occurrences in the project

    const projectSettingsJsonKeys = Object.keys(projectSettings.toJSON());
    projectSettingsJsonKeys.includes('defaultSupportMaterial') && printStoreUtils.updateProjectStoreProp('defaultSupportMaterial', // undefined value will remove the property from occurrence
    undefined, false, this.flux);
    const filterSettings = R.pipe(R.pickBy((v, k) => R.contains(k, R.values(SettingsKeys))), R.mapObjIndexed(R.prop('value')));
    const filteredSettings = filterSettings(projectSettings.toJSON()); // Ok so we do not have values/custom settings when we open the app and
    // do not have a model on the build plate. So if that is the case then
    // lets automatically load project settings.
    // if (valuesJSON.hasOwnProperty('applyToAll') &&
    //     Object.keys(valuesJSON).length == 1 ) {

    if (R.keys(values).length === 0) {
      this.loadProjectSettings(filteredSettings, projectSettings);
    } else if (!R.equals(values, filteredSettings)) {
      ModalService.confirm({
        body: i18n.t('mb:projectSettingsModal.body'),
        cancelText: i18n.t('mb:projectSettingsModal.projectSettingsText'),
        confirmText: i18n.t('mb:projectSettingsModal.currentSettingsText'),
        modalClass: 'w400 warningModal',
        hideClose: true,
        backdrop: 'static',
        keyboard: false,
        onCancel: () => {
          // Load project settings into this store. Project Store is already updated.
          this.loadProjectSettings(filteredSettings, projectSettings);
        },
        onConfirm: () => {
          // Load project then overwrite settings since the project
          // settings were already loaded.
          //Does not exist in values so lets not use it in project.
          R.mapObjIndexed((v, k) => R.has(k, values) ? printStoreUtils.updateProjectStoreProp(k, R.prop(k, values), true, this.flux) : printStoreUtils.updateProjectStoreProp(k, undefined, true, this.flux), filteredSettings);
        }
      });
    }
  },

  /**
   * Fires when a property on a print setting changes
   * @param {object} payload - the payload of the dispatched event. Should have `path` and `value` in it.
   * @private
   */
  initializeValues() {
    const ConfigService = this.flux.getService('ConfigService');
    const userConfigPrintSettings = ConfigService.get('printSettings');
    const userConfigActiveAdvSettings = ConfigService.get('activeAdvSettings');
    const activeTray = this.flux.store('ProjectStore').getProject().getActiveTray();
    const trayBodiesExist = activeTray.toJSON().trayBodies.size > 0;

    if (!settingsLoaded) {
      // Load 'activeAdvSettings' from user_config.json if app just loaded, and settings exist
      if (userConfigActiveAdvSettings) {
        // If app just started, use user_config 'activeAdvSettings'
        // set values (overrides) to 'printSettings'
        visibleSettings = userConfigActiveAdvSettings;
      } // If this is the first time we are seeing a tray with bodies, we need
      // to load up user_config.json's saved 'printSettings'


      if (!firstTrayLoaded && userConfigPrintSettings) {
        // If app just started, use user_config 'printSettings'
        // set values (overrides) to 'printSettings'
        values = userConfigPrintSettings; // Leaving this here to remove any previous references in
        // user_config's print settings

        values = R.dissoc('defaultSupportMaterial', values);
      }
    } // moving this up so the ui can chug along while we update the project store


    this.emit('initialize_values'); // Toggle to false until next app start.

    if (!settingsLoaded) settingsLoaded = true;
    if (!firstTrayLoaded) firstTrayLoaded = true; // Re-organizing this code to read from saved user config once at store init time,
    // saving it in store's values and visibleSettings objects.
    // However, we need to update the tray body properties every time or else we miss
    // some cases and the overrides go missing.
    // For example: when a model is added and deleted, we need to reapply overrides.
    // Not the most efficient way of doing things, and maybe someone else has
    // better ideas, but, definitely applying overrides only the first time will not work.

    if (userConfigPrintSettings && trayBodiesExist) {
      R.mapObjIndexed((v, k) => printStoreUtils.updateProjectStoreProp(k, v, true, this.flux), values);
    }
  },

  /**
   * We have this separate function for processing extruder types, partly because the offline
   * bots have some odd behavior in this context, but mainly because we need to grab this to
   * appropriately process Print Modes.
   * @param {object} data - the available configurations as well as the default
   * @private
   */
  getExtruderTypes(payload) {
    const normalized = this._extruder(payload);

    printStoreUtils.updateProjectStoreProp(SettingsKeys.Extruder, normalized, true, this.flux);
    this.emit(CHANGE_EVENT);
  },

  /**
   * Cache the supported configs once from sliceconfig on application load
   **/
  getSupportedConfigurations() {
    if (!supportedConfigurations.length) {
      supportedConfigurations = makerbot.getSupportedConfigurations();
    }
  },

  onCreatePreset({
    name
  }) {
    if (this.hasPresetWithName(name)) return;
    const preset = reducers.valuesToPreset(name, values);
    presets = R.append(preset, presets);
    currentPreset = name;
    presetError = nothing();

    this._savePresets();

    this.emit(CHANGE_EVENT);
  },

  onUpdatePreset() {
    if (!currentPreset) return;
    const preset = reducers.valuesToPreset(currentPreset, values);
    presets = reducers.updatePresetByName(currentPreset, preset, presets);

    this._savePresets();

    this.emit(CHANGE_EVENT);
  },

  onWritePreset() {
    if (!currentPreset) return;

    const {
      dialog
    } = require('electron').remote;

    const i18n = getI18n(this.flux);
    const messageStore = this.flux.store('MessageStore');
    const preset = reducers.getPresetByName(currentPreset, presets); // really not sure how this would happen, but I'd rather not debug it later if it does

    if (!preset) {
      console.error("Couldn't find preset:", currentPreset);
      return;
    }

    dialog.showSaveDialog({
      title: i18n.t('mb:printSettings.prests.saveDialog'),
      defaultPath: `${R.isEmpty(fileSafeString(currentPreset)) ? fileSafeString('New Custom Preset') : fileSafeString(currentPreset)}.${PRESET_EXT}`,
      filters: [{
        name: 'Makerbot Print Presets',
        extensions: [PRESET_EXT]
      }]
    }, filename => {
      if (filename) fs.writeJson(filename, presetToJSON(preset), err => {
        if (err) throw err;
        messageStore.onMessage({
          type: 'information',
          message: i18n.t('mb:printSettings.presets.exportSuccess')
        });
      });
      this.emit(CHANGE_EVENT);
    });
  },

  onDestroyPreset() {
    if (!currentPreset) return;
    presets = reducers.removePresetByName(currentPreset, presets);
    ({
      values,
      currentPreset
    } = printStoreUtils.resetAllToDefault(this.flux, values));

    this._savePresets(presets);

    this.emit(CHANGE_EVENT);
  },

  onReadPresets() {
    const {
      dialog
    } = require('electron').remote;

    const i18n = getI18n(this.flux);
    const messageStore = this.flux.store('MessageStore');

    const readJSON = path => fs.readJson(path, (err, obj) => {
      if (err) {
        messageStore.onMessage({
          type: 'error',
          message: i18n.t('mb:printSettings.presets.importFailure')
        });
        return;
      }

      const preset = sanitizePreset(obj, presets);
      presets = R.append(preset, presets);
      messageStore.onMessage({
        type: 'information',
        message: i18n.t('mb:printSettings.presets.importSuccess')
      });

      this._savePresets();

      this.emit(CHANGE_EVENT);
    });

    dialog.showOpenDialog({
      title: i18n.t('mb:printSettings.presets.import'),
      properties: ['openFile', 'multiSelections'],
      filters: [{
        name: 'Makerbot Print Presets',
        extensions: [PRESET_EXT]
      }]
    }, R.pipe(R.defaultTo([]), R.sortBy(R.prop('name')), R.when(R.compose(R.not, R.isEmpty), R.map(readJSON))));
  },

  onApplyPreset({
    name
  }) {
    if (currentPreset && currentPreset === name) return;
    const preset = reducers.getPresetByName(name, presets);
    if (!preset) console.error('No preset by that name');
    const filteredSettings = {};
    Object.keys(preset.settings).map(setting => {
      /**
       * Preset settings need to be filtered based on the
       * availableSettings, which gets updated upon loadSchema,
       * depending on what extruder is attached. (This is mainly for
       * filtering out EE settings on non-EE-attached printer)
       **/
      if (Object.keys(availableSettings).includes(setting)) {
        filteredSettings[setting] = preset.settings[setting];
      } // If the setting we're looking at is a print mode,
      // we check if it's supported by the current printer


      if (setting === Constants.SettingsKeys.PrintModes && supportedPrintModes.includes(preset.settings[setting])) {
        printMode = preset.settings[setting];
        filteredSettings[setting] = preset.settings[setting];
      }
    });
    const projectStoreSettings = R.merge(R.mapObjIndexed((v, k) => R.contains(k, PresetKeyBlacklist) ? v : undefined, values), filteredSettings); // taking this async so as not to block with these slow operations

    q().then(() => {
      /**
       * 'printStoreUtils.resetAllToDefault' handles the values in projectSettings
       * '_loadSchemaInitialValue' will handle the
       * printSettingsStore's state
       **/
      ({
        values,
        currentPreset
      } = printStoreUtils.resetAllToDefault(this.flux, values));

      this._loadSchemaInitialValue();
    }).then(() => {
      values = R.merge(filteredSettings, R.pick(PresetKeyBlacklist, values));
      currentPreset = name;
      this.emit(CHANGE_EVENT);
    }).then(() => {
      printStoreUtils.saveToUserConfig('printSettings', values, this.flux);
      R.mapObjIndexed((v, k) => printStoreUtils.updateProjectStoreProp(k, v, false, this.flux), projectStoreSettings);
    }).catch(console.error).done();
  },

  onRenamePreset({
    name
  }) {
    this.onDestroyPreset();
    this.onCreatePreset({
      name
    });
  },

  onUnapplyPreset() {
    if (!currentPreset) return;
    ({
      values,
      currentPreset
    } = printStoreUtils.resetAllToDefault(this.flux, values));
    this.emit(CHANGE_EVENT);
  },

  onPresetError({
    error
  }) {
    presetError = just(error);
    this.emit(CHANGE_EVENT);
  },

  onCloseModal() {
    presetError = nothing();
    this.emit(CHANGE_EVENT);
  },

  resetMaterialSpecificSettings(selectedMaterials) {
    //Since these settings are material-specific, they need to be
    //reset each time the material changes. (This includes when the
    //extruder changes)
    purgeWallCenter = [];

    this._persistValues('materials', selectedMaterials, true);

    this._persistValues(SettingsKeys.ChamberTemp, null, true);

    this._persistValues(SettingsKeys.Supports, null, true); // this._persistValues(SettingsKeys.BuildPlaneTemp, null, true);

  },

  onUpdateExtruder() {
    // A change in extruders implies a change in materials
    selectedMaterials = [];
    this.resetMaterialSpecificSettings(selectedMaterials);

    this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)) // emit event for diagnostics purposes
    .then(() => this.emit(ONUPDATE_EXTRUDERS_COMPLETED_EVENT)).done();
  },

  onUpdateExtruderPair() {
    // A change in extruders implies a change in materials
    selectedMaterials = [];
    /**
     * Since we're essentially allowing sliceocnfig get_schema to return
     * default materials, but ATM sliceconfig has an issue where it doesn't
     * return proper initial values for certain settings.
     * Instead, we'll update the `selectedMaterials` based on default
     * material per extruder type.
     *
     * -- Ray
     **/

    selectedMaterials = this._getDefaultMaterialsBasedOnExtruders();
    this.resetMaterialSpecificSettings(selectedMaterials);

    this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)) // emit event for diagnostics purposes
    .then(() => this.emit(ONUPDATE_EXTRUDERS_COMPLETED_EVENT)).done();
  },

  onUpdateMaterial({
    index,
    val
  }) {
    if (index > 1 || index < 0) return;
    selectedMaterials[index] = val;
    this.resetMaterialSpecificSettings(selectedMaterials);

    this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)).done();
  },

  onUpdateMaterials({
    id,
    spoolsInfo
  }) {
    const currentPrinter = this.flux.store('PrinterStore').getSelectedPrinter();
    if (!(id === currentPrinter.getId())) return;
    q(this.updatePrinterSpoolInfo(currentPrinter, {
      spoolsInfo
    })).then(() => {
      selectedMaterials = spoolMaterials;
      this.resetMaterialSpecificSettings(selectedMaterials);
    }).then(() => {
      this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)).done();
    }).done();
  },

  onUpdateSupports({
    supportExtruder,
    supportType
  }) {
    this._persistValues(SettingsKeys.Supports, supportType);

    this.emit(CHANGE_EVENT);
    printStoreUtils.clearBuilds(this.flux);
  },

  onUpdatePrintMode(newPrintMode) {
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();
    printMode = newPrintMode;

    this._persistValues(SettingsKeys.PrintModes, printMode, true);

    this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)).done();
  },

  onUpdatePurgeWallCenter(newPurgeWallcenter) {
    purgeWallCenter = newPurgeWallcenter;

    this._loadSchemaInitialValue().then(() => this.emit(CHANGE_EVENT)).done();
  },

  /************************************************************************/

  /*                          UTILITY METHODS                             */

  /************************************************************************/

  /**
   * TODO: Potential for refactor, maybe not manage 3 independently...
   *
   * Saves values across 3 objects that need to be in sync:
   * (1) tray.bodyProperties through Project Store - necessary for slicing
   * (2) values object from current store - representation of overrides
   * (3) user_config.json - app-wide saved settings under `printSettings`
   * @param {string} setting - setting key
   * @param {string} value - value to persist on key
   * @private
   */
  _persistValues(setting, value, skipUndoStack = false) {
    // Save to tray.bodyProperties for slicing
    printStoreUtils.updateProjectStoreProp(setting, value, skipUndoStack, this.flux);
    if (value != null) values = R.assoc(setting, value, values);else values = R.dissoc(setting, values); // Save to user_config as APP-WIDE settings.

    printStoreUtils.saveToUserConfig('printSettings', values, this.flux);
  },

  /**
   * Calls sliceConfig get_schema with overrides and print_mode
   * to get new initial_value for a setting, saves to override object,
   * and sets as new `initial_value` within availableSettings object.
   * @param {string} setting - the advanced setting key
   * @returns {string} - new evaluated initial_value
   * @private
   */
  _loadSchemaInitialValue() {
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();

    if (printer) {
      const activeTray = this.flux.store('ProjectStore').getProject().getActiveTray();
      const tray = activeTray.getBuildInfo(printer);
      return makerbot.getSchema(tray, printer, null, this.flux).then(schema => {
        // In rare cases, if getSchema gets called too quickly, it
        // will return empty string as a result. Let's exit before
        // things error out.
        // -- Ray
        if (_.isEmpty(schema)) return;
        const fullSchema = schema.getFullSchema();

        this._parseSchema(fullSchema); // emitting again because the setting might have changed its
        // initial value, so we want the UI to reflect this change.
        // There may be a slight delay, but as long as the rest of the
        // UI changes quickly, this slight lag here shouldn't be an
        // issue for the user.


        this.emit(CHANGE_EVENT);
      });
    }
  },

  /**
   * This is a helper function to parse the JSON that comes from
   * Slice Config into a useable format for availableSettings.
   * @private
   * @param {object} schema - the object that comes from Slice Config
   */
  _parseSchema(schema, resetToDefaultExtruders = false) {
    if (!schema || !schema.user_settings) return;
    const {
      user_settings,
      extruders: schemaExtruders,
      materials: schemaMaterials,
      materials_translations: schemaMaterialTranslations,
      print_modes: schemaPrintModes
    } = schema;
    if (!user_settings || !schemaExtruders) return;
    const printer = this.flux.store('PrinterStore').getSelectedPrinter();
    const currentBotType = printer.getGender().botType;
    const filteredSettings = printStoreUtils.filterIncompatibleOverrides(user_settings, currentBotType); // list of the settings no longer available (if any) from the last time we parsed the schema

    const schemaKeyDiff = R.pipe(R.keys, R.without(R.keys(filteredSettings)), R.without([SettingsKeys.PrintModes, SettingsKeys.Extruders, SettingsKeys.Materials]))(availableSettings);
    availableSettings = reducers.availableSettings(filteredSettings); // if a setting is no longer available remove it

    R.map(k => values = printStoreUtils.persistValues(k, undefined, false, values, this.flux), schemaKeyDiff);
    supportedExtruders = schemaExtruders.supported;
    ({
      supportedExtruderPairConfigs,
      supportedExtrudersPerIndex
    } = this.generateSupportedExtrudersInfo({
      supportedConfigurations,
      printer
    }));
    supportedPrintModes = schemaPrintModes.supported;

    const supportedMaterialsByExtruder = this._getSupportedMaterialsByExtruder(printer.getAttachedExtruders(), printer, schemaMaterials, supportedConfigurations);

    supportedMaterials = supportedMaterialsByExtruder;
    materialTranslations = schemaMaterialTranslations; // Assign materials to selectedMaterials from either spoolMaterials or through schema's default materials

    selectedMaterials = this._getSelectedMaterials({
      spoolMaterials,
      selectedMaterials,
      supportedMaterialsByExtruder,
      extruders: printer.getAttachedExtruders(),
      printerIsConnected: printer.isConnected(),
      printMode,
      isSketchPrinter: printer.isSketchPrinter(),
      isFifthGen: printer.isFifthGen()
    }); // If a newly selected printer doesn't support the currently selected print mode, reset to default
    // Otherwise, persist the previously-selected print mode.

    if (printMode === null || !supportedPrintModes.includes(printMode)) {
      printMode = schemaPrintModes.default;
      values = R.assoc('printMode', printMode, values);
    }
    /**
     * NOTE:
     * Regardless if a `supporType` selection triggers an initial value of false
     * with `doPurgeWall`, `purgeWallCenter` will report an initial value. No longer
     * we can use `user_settings.purgeWallCenter` to reset `purgeWallCenter` back to an
     * empty array. ::eelroll::
     * -- Ray
     **/


    if (user_settings.doPurgeWall && user_settings.doPurgeWall.initial_value || values.doPurgeWall) {
      if (_.isEmpty(purgeWallCenter)) {
        purgeWallCenter = user_settings.purgeWallCenter && user_settings.purgeWallCenter.initial_value || [];
      }
    } else {
      purgeWallCenter = [];
    }
  },

  _getSelectedMaterials(payload) {
    const {
      spoolMaterials,
      supportedMaterialsByExtruder,
      extruders,
      printerIsConnected,
      printMode,
      isSketchPrinter,
      isFifthGen
    } = payload;
    let selectedMaterials = payload.selectedMaterials;
    const hasLabsExtruder = extruders[0] === ExtruderTypeEnum.mk14_e;

    if (printerIsConnected && !isFifthGen && spoolMaterials !== null && !hasLabsExtruder) {
      // Live bot with smart spool capabilities (so, live 6th gen bots)
      // but without a LABS Extruder.
      // This includes 6th gen bots with empty spool bays.
      // xoxo Leor
      spoolMaterials.forEach((material, bayIndex) => {
        if (material && selectedMaterials[bayIndex] !== material) {
          selectedMaterials[bayIndex] = material;
        }
      });
    } else {
      // Any archetype bot, or a live bot without smart spool capabilities
      // (so, all 5th gens), or any 6th gen with an experimental LABS extruder
      const shouldMaterialChange = selectedMaterials.length === 0 || selectedMaterials[0] === undefined || !supportedMaterialsByExtruder[0].includes(selectedMaterials[0] || hasLabsExtruder);

      if (shouldMaterialChange) {
        if (extruders.length > 1) {
          selectedMaterials[0] = supportedMaterialsByExtruder[0][0];
          selectedMaterials[1] = supportedMaterialsByExtruder[1][0];
        } else if (isSketchPrinter) {
          // Default to [ PLA ]
          selectedMaterials[0] = supportedMaterialsByExtruder[0][1];
        } else {
          selectedMaterials = supportedMaterialsByExtruder[0];
        }
      }
    }

    return selectedMaterials;
  },

  getDefaultMaterials(supportedConfigurations, options) {
    // returns nested array of materials per extruder by way
    // of slice config's supportedConfigurations' e.g. [['abs'],['sr30']]
    const {
      printer,
      printMode,
      extruders
    } = options;

    try {
      return supportedConfigurations.filter(config => config.machine_id === printer.getGender().botType).filter(config => printMode ? config.print_mode === printMode : true).filter(config => _.isEqual(config.extruder_ids, extruders)).map(config => config.material_ids).reduce((accArr, materials) => {
        materials.map((material, index) => {
          !accArr[index] ? accArr[index] = [material] : accArr[index].push(material);
        });
        return accArr;
      }, []).map(bayMaterials => _.uniq(bayMaterials));
    } catch (err) {
      console.error(err);
    }
  },

  generateSupportedExtrudersInfo(payload) {
    const {
      printer,
      supportedConfigurations
    } = payload;
    const {
      botType
    } = printer.getGender();
    const filteredResults = supportedConfigurations.filter(config => config.machine_id === botType) // Don't want to filter by printMode or else we'll limit
    // the extruders lists. Extruder dropdown should dictate
    // supported print modes...
    .map(config => config.extruder_ids);
    const supportedExtruderPairConfigs = filteredResults.reduce((accArr, extruderIds) => {
      // if [mk14, mk14_s] not contain in [[mk14_hot, mk14_hots], [...]]
      !R.contains(extruderIds, accArr) // Update R.contains (Deprecated) to R.includes
      ? accArr.push(extruderIds) : null;
      return accArr;
    }, []);
    const supportedExtrudersPerIndex = filteredResults.reduce((accArr, extruderIds) => {
      extruderIds.map((extruderId, index) => {
        // ie. if accArr == [undefined, undefined]
        if (!accArr[index]) {
          accArr[index] = [extruderId];
        } else {
          // ie. if mk13_impla not in [[mk12, mk13, ...]]
          // Update R.contains (Deprecated) to R.includes
          // w/ Ramda ^0.26
          if (!R.contains(extruderId, accArr[index]) && extruderId !== ExtruderTypeEnum.mk13_experimental) {
            accArr[index].push(extruderId);
          }
        }
      });
      return accArr;
    }, []);
    return {
      supportedExtruderPairConfigs,
      supportedExtrudersPerIndex
    };
  },

  _getSupportedMaterialsByExtruder(extruders, printer) {
    const currentBotType = printer.getGender().botType; //materials supported by each extruder on the bot

    const materials = extruders.map((extruder, index) => {
      return supportedConfigurations.reduce(function (materialsForThisExtruder, config) {
        if (config.machine_id === currentBotType && _.isEqual(config.extruder_ids, extruders) && !materialsForThisExtruder.includes(config.material_ids[index])) {
          materialsForThisExtruder.push(config.material_ids[index]);
        }

        return materialsForThisExtruder;
      }, []);
    });
    return materials;
  },

  _savePresets() {
    printStoreUtils.saveToUserConfig('presets', presets, this.flux);
  },

  _getDefaultMaterialsBasedOnExtruders() {
    const enumExtruderToDefaultMaterials = _objectSpread({}, ExtruderToDefaultMaterial);

    const printer = this.flux.store('PrinterStore').getSelectedPrinter();
    const extruders = printer && printer.getAttachedExtruders();
    return extruders.map(extruder => enumExtruderToDefaultMaterials[extruder]);
  }

});
